# Miscellaneous scripts

Scripts that don't quite fit into formal analysis directories. Largely related to formatting/generating supplementary tables

If its your first time running code from this project, run `renv::restore()` in your R console to install the required dependencies

## clean_event_coordinates.R

Taking the PAPA quantification GTF as input, this script collapses unique intervals to semi-colon separated strings for each le_id.

Outputs a TSV with the le_id, start and end coordinates. Intervals are paired according to their index in the start and end strings i.e. to reconstruct the interval, pair the 1st element in the start column with the 1st element in the end column.

## get_non_cryptic_apas_table.R

Returns a summary of regulated, non-cryptic APAs across in vitro experiments, where regulated = padj < 0.05 and non-cryptic = not in a cryptic APA-containing gene.

Requires as input:

- DEXSeq results table combined across all experiments (produced by `../preprocessing/scripts/clean_papa_tbls.R`)
- Collapsed coordinates TSV (produced by `clean_event_coordinates.R`)

## merge_iclip_summary.R

Combine binarised cryptic event iCLIP binding summaries for different event types and terminii (+/- 500bp) into a single table. Output df contains columns le_id | gene_name | event_type | binding_start | binding_end

Requires as input:

- Per-event binding summaries split by event type and interval generated by `../motifs/scripts/merge_iclip_summary.R`

## combine_supp_data.R

Script to generate supplemental data file (XLSX and individual supplementary TSVs). Requires as input:

- table of dataset characteristics (provided with repo)
- table mapping cell types/model system to the experiments for which they were used (provided with repo)
- table containing set of primer sequences used for different analyses (mainly 3'RACE, provided with repo)
- outputs of `clean_event_coordinates.R`, `get_non_cryptic_apas_table.R` and `merge_iclip_summary.R`
- NYGC metadata summary statistics split by the analysis 'disease group' (produced by `../postmortem/scripts/nygc_metadata_summary.R`)
- cryptic ALE SJ quantification and selectivity classification (produced by `../postmortem/scripts/clean_ale_junction_counts.R`)
- Riboseq differential expression results for cryptic APAs (produced by `../riboseq/plot_riboseq.R`)
- BED files of representative last exon coordinates for each APA category (produced by `../motifs/notebooks/define_iclip_regions_<ale|d3utr>.ipynb` - both scripts used)
- Summary dataframe of cryptic event annotation and expression across in-vitro datasets (produced by `../preprocessing/scripts/manual_validation_summary.R`)
- List of target genes for ELK1 & ELK4 in HeLa ChIP-seq data (produced by `../tf_activity/scripts/write_lists_hela_ko.R`)