{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from define_helpers import select_rep_site, _define_cryptic_status, select_rep_prox_site, select_rep_five_end, construct_bed_name\n",
    "import os\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# approach\n",
    "\n",
    "### Input last exons - selecting representative PAS/last exon:\n",
    "\n",
    "le_ids can have multiple predictions, need to select a single one.\n",
    "Strategy:\n",
    "(starting with novel last exons used for quantification) \n",
    "1. Prefer events that match atlas\n",
    "2. if multiple, select min atlas distance\n",
    "3. If matches, s\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## To-dos:\n",
    "- fix updating to atlas for -ve strand matching pas\n",
    "- run selection for all novel 3'UTR extensions\n",
    "- Implement selection for annotated proximal sites of 3'UTR extensions (i.e. most distal annotated site)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "\n",
    "cryptics_df = pd.read_csv(\"../../preprocessing/processed/cryptics_summary_all_events.tsv\", sep=\"\\t\")\n",
    "\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
      "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
      "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
      "       'le_number', 'le_id'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074713_S59.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_3.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24475256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24473773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-B_S2.334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24474852</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_3.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36889</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074729_S12.25635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36890</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36891</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.26168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36892</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234583</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_0_2.21889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36893</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234594</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.24449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36894 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24475256  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24473773  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234583  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234594  | 13234826  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 36,894 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(quant_uniq_le.columns)\n",
    "quant_uniq_le.subset(lambda df: ~df.transcript_id.str.startswith(\"ENST\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['experiment_name', 'binID', 'groupID', 'featureID', 'exonBaseMean',\n",
       "       'dispersion', 'stat', 'pvalue', 'padj', 'UsageCoefficient_base',\n",
       "       'UsageCoefficient_contrast', 'log2fold_contrast_base', 'le_id',\n",
       "       'gene.qvalue', 'contrast_name', 'gene_id', 'gene_name', 'event_type',\n",
       "       'annot_status', 'transcript_id', 'chromosome', 'strand', 'start', 'end',\n",
       "       'mean_PPAU_base', 'mean_PPAU_treatment', 'delta_PPAU_treatment_control',\n",
       "       'simple_event_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
       "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
       "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
       "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
       "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
       "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Event type - bleedthrough - number of IDs - 21\n",
      "Event type - bleedthrough,distal_3utr_extension - number of IDs - 3\n",
      "Event type - bleedthrough,spliced - number of IDs - 12\n",
      "Event type - distal_3utr_extension - number of IDs - 104\n",
      "Event type - spliced - number of IDs - 119\n",
      "Number of cryptic IDs failing manual validation - 34\n"
     ]
    }
   ],
   "source": [
    "# summary df of cryptic events following manual validation of bleedthrough events\n",
    "cryptics_df = pd.read_csv(\"../../preprocessing/processed/cryptics_summary_all_events_bleedthrough_manual_validation.tsv\", sep=\"\\t\")\n",
    "\n",
    "# load in IDs to remove entirely from analysis (remove as cryptic, but make sure nto included in overall background)\n",
    "with open(\"../../preprocessing/processed/cryptics_manual_validation_fail_le_ids.txt\", \"r\") as infile:\n",
    "    manual_validation_fail_ids = [line.rstrip(\"\\n\") for line in infile]\n",
    "\n",
    "# get a dict of set of le_ids in each case\n",
    "event_type_le_ids = cryptics_df.groupby(\"simple_event_type\")[\"le_id\"].agg(set).to_dict()\n",
    "event_type_gene_names = cryptics_df.groupby(\"simple_event_type\")[\"gene_name\"].agg(set).to_dict()\n",
    "\n",
    "for event_type, ids in event_type_le_ids.items():\n",
    "    print(f\"Event type - {event_type} - number of IDs - {len(ids)}\")\n",
    "\n",
    "print(f\"Number of cryptic IDs failing manual validation - {len(manual_validation_fail_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_name</th>\n",
       "      <th>binID</th>\n",
       "      <th>groupID</th>\n",
       "      <th>featureID</th>\n",
       "      <th>exonBaseMean</th>\n",
       "      <th>dispersion</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>UsageCoefficient_base</th>\n",
       "      <th>...</th>\n",
       "      <th>annot_status</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mean_PPAU_base</th>\n",
       "      <th>mean_PPAU_treatment</th>\n",
       "      <th>delta_PPAU_treatment_control</th>\n",
       "      <th>simple_event_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000021645.20:E001</td>\n",
       "      <td>ENSG00000021645.20</td>\n",
       "      <td>E001</td>\n",
       "      <td>32.627088</td>\n",
       "      <td>0.044220</td>\n",
       "      <td>90.962787</td>\n",
       "      <td>1.463997e-21</td>\n",
       "      <td>2.250563e-19</td>\n",
       "      <td>1.452069</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.TDP-4.18197.2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>7.829867e+07</td>\n",
       "      <td>7.830620e+07</td>\n",
       "      <td>0.001914</td>\n",
       "      <td>0.033848</td>\n",
       "      <td>0.031934</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000021645.20:E002</td>\n",
       "      <td>ENSG00000021645.20</td>\n",
       "      <td>E002</td>\n",
       "      <td>622.521901</td>\n",
       "      <td>0.022117</td>\n",
       "      <td>135.739619</td>\n",
       "      <td>2.274890e-31</td>\n",
       "      <td>5.061630e-29</td>\n",
       "      <td>12.675156</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000557594.5,ENST00000428277.6,ENST000002...</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>7.986104e+39</td>\n",
       "      <td>7.986248e+39</td>\n",
       "      <td>0.998086</td>\n",
       "      <td>0.934776</td>\n",
       "      <td>-0.063310</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000021645.20:E002</td>\n",
       "      <td>ENSG00000021645.20</td>\n",
       "      <td>E002</td>\n",
       "      <td>622.521901</td>\n",
       "      <td>0.022117</td>\n",
       "      <td>135.739619</td>\n",
       "      <td>2.274890e-31</td>\n",
       "      <td>5.061630e-29</td>\n",
       "      <td>12.675156</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000557594.5,ENST00000428277.6,ENST000002...</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>7.986104e+39</td>\n",
       "      <td>7.986248e+39</td>\n",
       "      <td>0.998086</td>\n",
       "      <td>0.934776</td>\n",
       "      <td>-0.063310</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000048649.14:E001</td>\n",
       "      <td>ENSG00000048649.14</td>\n",
       "      <td>E001</td>\n",
       "      <td>163.262346</td>\n",
       "      <td>0.037855</td>\n",
       "      <td>88.319669</td>\n",
       "      <td>5.568698e-21</td>\n",
       "      <td>8.407740e-19</td>\n",
       "      <td>4.515264</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.11065.7,PAPA.TDP43-G_S...</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>7.781315e+71</td>\n",
       "      <td>7.781379e+71</td>\n",
       "      <td>0.090158</td>\n",
       "      <td>0.462779</td>\n",
       "      <td>0.372622</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000048649.14:E001</td>\n",
       "      <td>ENSG00000048649.14</td>\n",
       "      <td>E001</td>\n",
       "      <td>163.262346</td>\n",
       "      <td>0.037855</td>\n",
       "      <td>88.319669</td>\n",
       "      <td>5.568698e-21</td>\n",
       "      <td>8.407740e-19</td>\n",
       "      <td>4.515264</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.11065.7,PAPA.TDP43-G_S...</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>7.781315e+71</td>\n",
       "      <td>7.781379e+71</td>\n",
       "      <td>0.090158</td>\n",
       "      <td>0.462779</td>\n",
       "      <td>0.372622</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209479</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000280789.2:E002</td>\n",
       "      <td>ENSG00000280789.2</td>\n",
       "      <td>E002</td>\n",
       "      <td>99.077638</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>0.006587</td>\n",
       "      <td>9.353145e-01</td>\n",
       "      <td>9.785953e-01</td>\n",
       "      <td>5.982423</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_4.15153.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2.981955e+07</td>\n",
       "      <td>2.982252e+07</td>\n",
       "      <td>0.181220</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.152113</td>\n",
       "      <td>distal_3utr_extension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209480</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000280832.2:E001</td>\n",
       "      <td>ENSG00000280832.2</td>\n",
       "      <td>E001</td>\n",
       "      <td>75.778902</td>\n",
       "      <td>0.099478</td>\n",
       "      <td>0.075875</td>\n",
       "      <td>7.829679e-01</td>\n",
       "      <td>9.189656e-01</td>\n",
       "      <td>6.014731</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000626810.1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>1.263502e+08</td>\n",
       "      <td>1.263518e+08</td>\n",
       "      <td>0.133144</td>\n",
       "      <td>0.147606</td>\n",
       "      <td>0.014462</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209481</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000280832.2:E002</td>\n",
       "      <td>ENSG00000280832.2</td>\n",
       "      <td>E002</td>\n",
       "      <td>402.882205</td>\n",
       "      <td>0.085732</td>\n",
       "      <td>0.087526</td>\n",
       "      <td>7.673463e-01</td>\n",
       "      <td>9.134343e-01</td>\n",
       "      <td>12.145436</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000629441.2</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>1.263409e+08</td>\n",
       "      <td>1.263423e+08</td>\n",
       "      <td>0.866856</td>\n",
       "      <td>0.852394</td>\n",
       "      <td>-0.014462</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209482</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000282508.2:E001</td>\n",
       "      <td>ENSG00000282508.2</td>\n",
       "      <td>E001</td>\n",
       "      <td>43.093715</td>\n",
       "      <td>0.060733</td>\n",
       "      <td>0.017535</td>\n",
       "      <td>8.946533e-01</td>\n",
       "      <td>9.665471e-01</td>\n",
       "      <td>4.668928</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000633109.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>2.112510e+05</td>\n",
       "      <td>2.117150e+05</td>\n",
       "      <td>0.678888</td>\n",
       "      <td>0.582547</td>\n",
       "      <td>-0.096341</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209483</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000282508.2:E002</td>\n",
       "      <td>ENSG00000282508.2</td>\n",
       "      <td>E002</td>\n",
       "      <td>27.050964</td>\n",
       "      <td>0.069219</td>\n",
       "      <td>0.014790</td>\n",
       "      <td>9.032054e-01</td>\n",
       "      <td>9.693684e-01</td>\n",
       "      <td>3.817323</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000631988.1,ENST00000631797.1,ENST000006...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>2.014082e+29</td>\n",
       "      <td>2.016622e+29</td>\n",
       "      <td>0.149081</td>\n",
       "      <td>0.131692</td>\n",
       "      <td>-0.017389</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>209085 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                experiment_name                    binID             groupID  \\\n",
       "0             brown_i3_cortical  ENSG00000021645.20:E001  ENSG00000021645.20   \n",
       "1             brown_i3_cortical  ENSG00000021645.20:E002  ENSG00000021645.20   \n",
       "2             brown_i3_cortical  ENSG00000021645.20:E002  ENSG00000021645.20   \n",
       "3             brown_i3_cortical  ENSG00000048649.14:E001  ENSG00000048649.14   \n",
       "4             brown_i3_cortical  ENSG00000048649.14:E001  ENSG00000048649.14   \n",
       "...                         ...                      ...                 ...   \n",
       "209479  zanovello_skndz_curve_1   ENSG00000280789.2:E002   ENSG00000280789.2   \n",
       "209480  zanovello_skndz_curve_1   ENSG00000280832.2:E001   ENSG00000280832.2   \n",
       "209481  zanovello_skndz_curve_1   ENSG00000280832.2:E002   ENSG00000280832.2   \n",
       "209482  zanovello_skndz_curve_1   ENSG00000282508.2:E001   ENSG00000282508.2   \n",
       "209483  zanovello_skndz_curve_1   ENSG00000282508.2:E002   ENSG00000282508.2   \n",
       "\n",
       "       featureID  exonBaseMean  dispersion        stat        pvalue  \\\n",
       "0           E001     32.627088    0.044220   90.962787  1.463997e-21   \n",
       "1           E002    622.521901    0.022117  135.739619  2.274890e-31   \n",
       "2           E002    622.521901    0.022117  135.739619  2.274890e-31   \n",
       "3           E001    163.262346    0.037855   88.319669  5.568698e-21   \n",
       "4           E001    163.262346    0.037855   88.319669  5.568698e-21   \n",
       "...          ...           ...         ...         ...           ...   \n",
       "209479      E002     99.077638   12.000000    0.006587  9.353145e-01   \n",
       "209480      E001     75.778902    0.099478    0.075875  7.829679e-01   \n",
       "209481      E002    402.882205    0.085732    0.087526  7.673463e-01   \n",
       "209482      E001     43.093715    0.060733    0.017535  8.946533e-01   \n",
       "209483      E002     27.050964    0.069219    0.014790  9.032054e-01   \n",
       "\n",
       "                padj  UsageCoefficient_base  ...  annot_status  \\\n",
       "0       2.250563e-19               1.452069  ...         novel   \n",
       "1       5.061630e-29              12.675156  ...     annotated   \n",
       "2       5.061630e-29              12.675156  ...     annotated   \n",
       "3       8.407740e-19               4.515264  ...         novel   \n",
       "4       8.407740e-19               4.515264  ...     annotated   \n",
       "...              ...                    ...  ...           ...   \n",
       "209479  9.785953e-01               5.982423  ...         novel   \n",
       "209480  9.189656e-01               6.014731  ...     annotated   \n",
       "209481  9.134343e-01              12.145436  ...     annotated   \n",
       "209482  9.665471e-01               4.668928  ...     annotated   \n",
       "209483  9.693684e-01               3.817323  ...     annotated   \n",
       "\n",
       "                                            transcript_id chromosome  strand  \\\n",
       "0                                      PAPA.TDP-4.18197.2      chr14       +   \n",
       "1       ENST00000557594.5,ENST00000428277.6,ENST000002...      chr14       +   \n",
       "2       ENST00000557594.5,ENST00000428277.6,ENST000002...      chr14       +   \n",
       "3       PAPA.doxconc_DOX_0075_2.11065.7,PAPA.TDP43-G_S...      chr11       -   \n",
       "4       PAPA.doxconc_DOX_0075_2.11065.7,PAPA.TDP43-G_S...      chr11       -   \n",
       "...                                                   ...        ...     ...   \n",
       "209479                   PAPA.chx_tdp_CTRL_ctrl_4.15153.2      chr16       +   \n",
       "209480                                  ENST00000626810.1      chr11       -   \n",
       "209481                                  ENST00000629441.2      chr11       -   \n",
       "209482                                  ENST00000633109.1      chr19       -   \n",
       "209483  ENST00000631988.1,ENST00000631797.1,ENST000006...      chr19       -   \n",
       "\n",
       "               start           end mean_PPAU_base mean_PPAU_treatment  \\\n",
       "0       7.829867e+07  7.830620e+07       0.001914            0.033848   \n",
       "1       7.986104e+39  7.986248e+39       0.998086            0.934776   \n",
       "2       7.986104e+39  7.986248e+39       0.998086            0.934776   \n",
       "3       7.781315e+71  7.781379e+71       0.090158            0.462779   \n",
       "4       7.781315e+71  7.781379e+71       0.090158            0.462779   \n",
       "...              ...           ...            ...                 ...   \n",
       "209479  2.981955e+07  2.982252e+07       0.181220            0.333333   \n",
       "209480  1.263502e+08  1.263518e+08       0.133144            0.147606   \n",
       "209481  1.263409e+08  1.263423e+08       0.866856            0.852394   \n",
       "209482  2.112510e+05  2.117150e+05       0.678888            0.582547   \n",
       "209483  2.014082e+29  2.016622e+29       0.149081            0.131692   \n",
       "\n",
       "       delta_PPAU_treatment_control      simple_event_type  \n",
       "0                          0.031934                spliced  \n",
       "1                         -0.063310                spliced  \n",
       "2                         -0.063310                spliced  \n",
       "3                          0.372622                spliced  \n",
       "4                          0.372622                spliced  \n",
       "...                             ...                    ...  \n",
       "209479                     0.152113  distal_3utr_extension  \n",
       "209480                     0.014462                spliced  \n",
       "209481                    -0.014462                spliced  \n",
       "209482                    -0.096341                spliced  \n",
       "209483                    -0.017389                spliced  \n",
       "\n",
       "[209085 rows x 28 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter out failed IDs from dexseq df (so not included as background)\n",
    "dexseq_df = dexseq_df[~dexseq_df[\"le_id\"].isin(manual_validation_fail_ids)]\n",
    "dexseq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3369"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel 3'UTR ectensions\n",
    "le_id_d3utr = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "len(le_id_d3utr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic 3'UTR extensions - 104\n",
      "Number of cryptic 3'UTR extension-containing genes - 104\n"
     ]
    }
   ],
   "source": [
    "cryp_le_ids = event_type_le_ids[\"distal_3utr_extension\"]\n",
    "cryp_gene_names = event_type_gene_names[\"distal_3utr_extension\"]\n",
    "\n",
    "print(f\"Number of cryptic 3'UTR extensions - {len(cryp_le_ids)}\")\n",
    "print(f\"Number of cryptic 3'UTR extension-containing genes - {len(cryp_gene_names)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add le_id used in downstream analysis\n",
    "novel_le = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of novel 3'UTR extenstion intervals before selecting representative LEs - 15810\n",
      "Number of novel 3'UTR extension intervals after selecting representative LEs - 3398\n",
      "{'atlas_1_pred': 1549, 'atlas_max_datasets': 534, 'atlas_max_datasets_shortest': 344, 'motif_1_min': 889, 'motif_shortest_min': 53}\n"
     ]
    }
   ],
   "source": [
    "# Select representative PAS for distal 3'UTR extension events (novel)\n",
    "novel_le_rep_d3utr, rep_choices_d3utr = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_d3utr))), id_col=\"le_id_quant\")\n",
    "\n",
    "\n",
    "print(f\"Number of novel 3'UTR extenstion intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_d3utr))))}\")\n",
    "print(f\"Number of novel 3'UTR extension intervals after selecting representative LEs - {len(novel_le_rep_d3utr)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_d3utr.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check if any IDs dropped\n",
    "le_id_d3utr - set(novel_le_rep_d3utr.le_id_quant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       0  1\n",
      "6     ENSG00000001461.17  6\n",
      "85    ENSG00000023902.14  3\n",
      "98    ENSG00000031698.13  3\n",
      "102   ENSG00000033122.21  5\n",
      "143    ENSG00000053372.5  3\n",
      "...                  ... ..\n",
      "1137  ENSG00000129824.16  2\n",
      "1798   ENSG00000154620.6  2\n",
      "2091  ENSG00000165246.15  3\n",
      "60    ENSG00000012817.16  3\n",
      "3081   ENSG00000230663.1  3\n",
      "\n",
      "[3398 rows x 2 columns]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6       ENSG00000001461.17_5\n",
       "85      ENSG00000023902.14_2\n",
       "98      ENSG00000031698.13_2\n",
       "102     ENSG00000033122.21_4\n",
       "143      ENSG00000053372.5_2\n",
       "                ...         \n",
       "1137    ENSG00000129824.16_1\n",
       "1798     ENSG00000154620.6_1\n",
       "2091    ENSG00000165246.15_2\n",
       "60      ENSG00000012817.16_2\n",
       "3081     ENSG00000230663.1_2\n",
       "Name: 0, Length: 3398, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to select representative proximal site for each distal site, since quantification approach lumps together all annotated proximal sites (and compares them to extension)\n",
    "# since le_ids are annotated sequentially, know that immediately preceding le_number is the partner le_id\n",
    "d3utr_le_id_split = novel_le_rep_d3utr.le_id_quant.str.split(\"_\", regex=False, expand=True)\n",
    "print(d3utr_le_id_split)\n",
    "d3utr_le_id_split[1] = d3utr_le_id_split[1].astype(int).subtract(1)\n",
    "\n",
    "# reconstruct le_id (corresponding to proximal le_id)\n",
    "d3utr_prox_le_id = d3utr_le_id_split[0].str.cat(d3utr_le_id_split[1].astype(str), sep=\"_\")\n",
    "d3utr_prox_le_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150159404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150159497</td>\n",
       "      <td>150160065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150159404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>133.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12702</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12703</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705416</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12704</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12705</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705424</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12706</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>17552799</td>\n",
       "      <td>17553954</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>17553954</td>\n",
       "      <td>17554041</td>\n",
       "      <td>ENST00000458667.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000230663.1_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12707 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 150158818 | 150159404 | ...   |\n",
       "| chr1         | .          | exon         | 150159497 | 150160065 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 19705416  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19705419  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19705424  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 17552799  | 17553954  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 12,707 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset for annotated last exons with 3'end proximal to 3'UTR extensions\n",
    "quant_uniq_le_d3utr_prox = quant_uniq_le.subset(lambda df: df.le_id.isin(set(d3utr_prox_le_id.values)))\n",
    "quant_uniq_le_d3utr_prox\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000126214.22_3',\n",
       " 'ENSG00000160439.16_4',\n",
       " 'ENSG00000165156.15_3',\n",
       " 'ENSG00000176222.9_3',\n",
       " 'ENSG00000196268.12_4',\n",
       " 'ENSG00000213923.13_4'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check if any cryptic genes dropped\n",
    "# [le_id.split(\"_\")] prevents trying to unpack each character in string\n",
    "# list(le_id_d3utr)\n",
    "\n",
    "missing_ids_prox = set(\"_\".join([i, str(int(n) - 1)]) for le_id in le_id_d3utr for i,n in [le_id.split(\"_\")]) - set(quant_uniq_le_d3utr_prox.le_id)\n",
    "# go back to original ids\n",
    "missing_ids = set(\"_\".join([i, str(int(n) + 1)]) for le_id in missing_ids_prox for i,n in [le_id.split(\"_\")])\n",
    "# set(quant_uniq_le_d3utr_prox.le_id)\n",
    "missing_ids\n",
    "# set(len(le_id.split(\"_\")) for le_id in le_id_d3utr)\n",
    "\n",
    "# - set(quant_uniq_le_d3utr_prox.le_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [le_id, gene_name, simple_event_type, n_exper_cryptic, experiment_name, annot_status, chromosome, start, end, strand]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# are these cryptic?\n",
    "print(cryptics_df[cryptics_df.le_id.isin(missing_ids)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150159497</td>\n",
       "      <td>150160065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>109237373</td>\n",
       "      <td>109237730</td>\n",
       "      <td>ENST00000468588.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000031698.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>70121779</td>\n",
       "      <td>70121994</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000033122.21</td>\n",
       "      <td>LRRC7</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>70107826</td>\n",
       "      <td>70121779</td>\n",
       "      <td>ENST00000415775.2</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000033122.21</td>\n",
       "      <td>LRRC7</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000033122.21_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4387</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>13704435</td>\n",
       "      <td>13705224</td>\n",
       "      <td>ENST00000284856.4</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000154620.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4388</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845650</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000165246.15</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>14830519,14830519,14830519,14830519</td>\n",
       "      <td>14840412,14840412,14840412,14840412</td>\n",
       "      <td>ENST00000339174.9,ENST00000382868.5,ENST000003...</td>\n",
       "      <td>NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000165246.15</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000165246.15_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4389</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>17552799</td>\n",
       "      <td>17553954</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>17553954</td>\n",
       "      <td>17554041</td>\n",
       "      <td>ENST00000458667.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000230663.1_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4392 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 150159497 | 150160065 | ...   |\n",
       "| chr1         | .          | exon         | 109237730 | 109238182 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 14840412  | 14845650  | ...   |\n",
       "| chrY         | .          | exon         | 19703864  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19703864  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 17552799  | 17553954  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 4,392 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select representative proximal site for partner annotated ALEs (most distal of annotated PAS)\n",
    "sel_le_d3utr_prox = select_rep_prox_site(quant_uniq_le_d3utr_prox, id_col=\"le_id\")\n",
    "sel_le_d3utr_prox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000126214.22_2',\n",
       " 'ENSG00000160439.16_3',\n",
       " 'ENSG00000165156.15_2',\n",
       " 'ENSG00000176222.9_2',\n",
       " 'ENSG00000196268.12_3',\n",
       " 'ENSG00000213923.13_3'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check no further cryptics/le_ids dropped\n",
    "set(\"_\".join([i, str(int(n) - 1)]) for le_id in le_id_d3utr for i,n in [le_id.split(\"_\")]) - set(sel_le_d3utr_prox.le_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same 6 IDs. Don't think losing 6 events from background is going to make a dramatic difference...\n",
    "# remove missing IDs from novel distal DF, so don't have inbalanced numbers of proximal/distal events\n",
    "novel_le_rep_d3utr = novel_le_rep_d3utr.subset(lambda df: ~df.le_id_quant.isin(missing_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene counts for the number of SH-SY5Y datasets in which genes are expressed\n",
      "experiment_name\n",
      "3    2582\n",
      "1    1068\n",
      "2     704\n",
      "Name: count, dtype: int64\n",
      "Number of ns ALE-containing genes expressed in all SH-SY5Y datasets - 1577\n",
      "Number of ns ALE isoforms of genes expressed in all SH-SY5Y datasets - 3624\n",
      "Of the genes containing ALE isoforms, how many datasets are the individual isoforms expressed in?\n",
      "experiment_name\n",
      "3    3304\n",
      "1     185\n",
      "2     135\n",
      "Name: count, dtype: int64\n",
      "Number of ns ALE isoforms expressed in all SH-SY5Y datasets - 3304\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Alternative definition of background (more conservative)\n",
    "# not significant in any SH-SY5Y dataset, but expressed/evaluated in at least one of them (so know is multi-ALE in at least one dataset)\n",
    "# NB: extra-check that expressed/evaluated in all 3? Then confident that is a multi-ALE gene robustly expressed/detected\n",
    "shsy5y_datasets = [\"zanovello_shsy5y_curve_0075\", \"zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl\", \"brown_shsy5y\"]\n",
    "dexseq_df_shsy5y = dexseq_df.loc[dexseq_df.experiment_name.isin(shsy5y_datasets), :]\n",
    "\n",
    "# first get an experiment count for each event\n",
    "gene_n_shsy5y_expressed = (dexseq_df_shsy5y\n",
    ".groupby(\"groupID\")\n",
    "[\"experiment_name\"]\n",
    ".nunique()\n",
    ")\n",
    "\n",
    "print(\"Gene counts for the number of SH-SY5Y datasets in which genes are expressed\")\n",
    "print(gene_n_shsy5y_expressed.value_counts())\n",
    "\n",
    "# get genes expressed in all SH-SY5Y datasets but with no sig change in any of them \n",
    "ns_gene_ids_shsy5y = (dexseq_df_shsy5y.loc[dexseq_df_shsy5y.groupID.isin(set(gene_n_shsy5y_expressed[gene_n_shsy5y_expressed == len(shsy5y_datasets)].index)), :]\n",
    ".assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "               .groupby(\"groupID\")\n",
    "               [\"reg_status\"]\n",
    "               .sum()\n",
    "               .loc[lambda x: x == 0]\n",
    ")\n",
    "\n",
    "# set of\n",
    "ns_le_ids_shsy5y = set(dexseq_df_shsy5y.loc[dexseq_df_shsy5y[\"groupID\"].isin(ns_gene_ids_shsy5y.index), \"le_id\"])\n",
    "\n",
    "\n",
    "print(f\"Number of ns ALE-containing genes expressed in all SH-SY5Y datasets - {len(ns_gene_ids_shsy5y.index)}\")\n",
    "print(f\"Number of ns ALE isoforms of genes expressed in all SH-SY5Y datasets - {len(set(ns_le_ids_shsy5y))}\")\n",
    "\n",
    "# le_ids in genes with no differential usage in any dataset that are also expressed/assessed in all datasets\n",
    "ns_le_ids_shsy5y_all = set(dexseq_df_shsy5y[dexseq_df_shsy5y.le_id.isin(ns_le_ids_shsy5y)]\n",
    ".groupby(\"le_id\")\n",
    "[\"experiment_name\"]\n",
    ".nunique()\n",
    ".loc[lambda x: x == 3]\n",
    ".index\n",
    ")\n",
    "\n",
    "print(\"Of the genes containing ALE isoforms, how many datasets are the individual isoforms expressed in?\")\n",
    "print(dexseq_df_shsy5y[dexseq_df_shsy5y.le_id.isin(ns_le_ids_shsy5y)]\n",
    ".groupby(\"le_id\")\n",
    "[\"experiment_name\"]\n",
    ".nunique()\n",
    ".value_counts()\n",
    ")\n",
    "\n",
    "print(f\"Number of ns ALE isoforms expressed in all SH-SY5Y datasets - {len(ns_le_ids_shsy5y_all)}\")\n",
    "# (dexseq_df.assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "#                .groupby(\"groupID\")\n",
    "#                [\"reg_status\"]\n",
    "#                .sum()\n",
    "#                \n",
    "#              )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now want to combine the two with minimal info required for making maps\n",
    "# 3'end coordinates, gene name, le_id, site type, regulation status\n",
    "\n",
    "# define site_type columns\n",
    "sel_le_d3utr_prox.site_type = \"proximal\"\n",
    "novel_le_rep_d3utr.site_type = \"distal\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1636"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# le_ids of novel extensions with no change in any experiment\n",
    "ns_le_ids_d3utr = set(dexseq_df.loc[(dexseq_df[\"le_id\"].isin(ns_le_ids_shsy5y_all.intersection(le_id_d3utr))), \"le_id\"].drop_duplicates())\n",
    "\n",
    "\n",
    "# now get le_ids of partner proximal les\n",
    "ns_le_ids_d3utr_prtnr = set([\"_\".join([le_id.split(\"_\")[0], str(int(le_id.split(\"_\")[1]) - 1)]) for le_id in ns_le_ids_d3utr])\n",
    "print(len(ns_le_ids_d3utr) == len(ns_le_ids_d3utr_prtnr))\n",
    "ns_le_ids_d3utr_comb = ns_le_ids_d3utr.union(ns_le_ids_d3utr_prtnr)\n",
    "len(ns_le_ids_d3utr_comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure proximal partner of cryptic les are labelled\n",
    "cryp_le_ids_prtnr = set([\"_\".join([le_id.split(\"_\")[0], str(int(le_id.split(\"_\")[1]) - 1)]) for le_id in list(cryp_le_ids)])\n",
    "cryp_le_ids_comb = cryp_le_ids.union(cryp_le_ids_prtnr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids_comb, ns_le_ids_d3utr_comb, id_col=\"le_id\"))\n",
    "novel_le_rep_d3utr = novel_le_rep_d3utr.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids_comb, ns_le_ids_d3utr_comb, id_col=\"le_id_quant\"))\n",
    "\n",
    "\n",
    "\n",
    "# sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))\n",
    "# sel_novel_le_d3utr = sel_novel_le_d3utr.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id_ref.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cryptic_status\n",
      "NULL          2486\n",
      "background     802\n",
      "cryptic        104\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of duplicated IDs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          20\n",
       "background     9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(novel_le_rep_d3utr.cryptic_status.value_counts())\n",
    "print(\"\\nNumber of duplicated IDs\")\n",
    "# how many IDs are duplicated with distinct regions?\n",
    "novel_d3utr_dup_ids = set(novel_le_rep_d3utr.drop_duplicate_positions().as_df().le_id_quant.value_counts().loc[lambda x: x > 1].index)\n",
    "# subset is for le_ids that have more than one entry. Then collapse to a single entry for each le_id and count the number of unique IDs\n",
    "novel_le_rep_d3utr[novel_le_rep_d3utr.le_id_quant.isin(novel_d3utr_dup_ids)].drop_duplicate_positions().as_df().drop_duplicates(subset=\"le_id_quant\").cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          3231\n",
       "background    1012\n",
       "cryptic        149\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of events in selected proximal df - should be seeing 104 & 802\n",
    "sel_le_d3utr_prox.cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          24\n",
       "background     6\n",
       "cryptic        4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many events are duplicated when selecting a representative proximal PAS?\n",
    "sel_d3utr_prox_dup_ids = set(sel_le_d3utr_prox.drop_duplicate_positions().le_id.value_counts().loc[lambda x: x > 1].index)\n",
    "sel_le_d3utr_prox[sel_le_d3utr_prox.le_id.isin(sel_d3utr_prox_dup_ids)].drop_duplicate_positions().as_df().drop_duplicates(subset=\"le_id\").cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proximal PAS regions with distinct 5'ends \n",
      "34\n",
      "Number of proximal PAS regions with distinct 3'ends\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of proximal PAS regions with distinct 5'ends \")\n",
    "print(sel_le_d3utr_prox.subset(lambda df: df.le_id.isin(sel_d3utr_prox_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.Start.nunique() if (df.Strand == \"+\").all() else df.End.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )\n",
    "\n",
    "print(\"Number of proximal PAS regions with distinct 3'ends\")\n",
    "print(sel_le_d3utr_prox.subset(lambda df: df.le_id.isin(sel_d3utr_prox_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.End.nunique() if (df.Strand == \"+\").all() else df.Start.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of distal PAS regions with distinct 5'ends \n",
      "0\n",
      "Number of distal PAS regions with distinct 3'ends\n",
      "28\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of distal PAS regions with distinct 5'ends \")\n",
    "print(novel_le_rep_d3utr.subset(lambda df: df.le_id_quant.isin(novel_d3utr_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.Start.nunique() if (df.Strand == \"+\").all() else df.End.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )\n",
    "\n",
    "print(\"Number of distal PAS regions with distinct 3'ends\")\n",
    "print(novel_le_rep_d3utr.subset(lambda df: df.le_id_quant.isin(novel_d3utr_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.End.nunique() if (df.Strand == \"+\").all() else df.Start.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Proximal regions - all have different 5'ends/exon starts\n",
    "Distal regions - all have distinct 3'ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name_ref</th>\n",
       "      <th>Name</th>\n",
       "      <th>cryptic_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70439100</td>\n",
       "      <td>70445530</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116761.12_2</td>\n",
       "      <td>CTH</td>\n",
       "      <td>1:70445533:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70439100</td>\n",
       "      <td>70445536</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116761.12_2</td>\n",
       "      <td>CTH</td>\n",
       "      <td>1:70445533:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33494217</td>\n",
       "      <td>33501611</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000121903.15_4</td>\n",
       "      <td>ZSCAN20</td>\n",
       "      <td>1:33501609:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33494217</td>\n",
       "      <td>33501607</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000121903.15_4</td>\n",
       "      <td>ZSCAN20</td>\n",
       "      <td>1:33501609:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44779122</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142937.12_2</td>\n",
       "      <td>RPS8</td>\n",
       "      <td>1:44779125:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44779128</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142937.12_2</td>\n",
       "      <td>RPS8</td>\n",
       "      <td>1:44779125:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42775208</td>\n",
       "      <td>42781977</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164008.17_2</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>1:42781980:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42775208</td>\n",
       "      <td>42781983</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164008.17_2</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>1:42781980:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173606034</td>\n",
       "      <td>173610215</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000183831.7_3</td>\n",
       "      <td>ANKRD45</td>\n",
       "      <td>1:173606033:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173606030</td>\n",
       "      <td>173610215</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000183831.7_3</td>\n",
       "      <td>ANKRD45</td>\n",
       "      <td>1:173606033:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr2</td>\n",
       "      <td>130346355</td>\n",
       "      <td>130351066</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000136718.10_4</td>\n",
       "      <td>IMP4</td>\n",
       "      <td>2:130351065:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>130346355</td>\n",
       "      <td>130351064</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000136718.10_4</td>\n",
       "      <td>IMP4</td>\n",
       "      <td>2:130351065:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr4</td>\n",
       "      <td>182883977</td>\n",
       "      <td>182891477</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000129187.15_2</td>\n",
       "      <td>DCTD</td>\n",
       "      <td>4:182883985:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>182883991</td>\n",
       "      <td>182891477</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000129187.15_2</td>\n",
       "      <td>DCTD</td>\n",
       "      <td>4:182883985:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>139329344</td>\n",
       "      <td>139332519</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000015479.20_2</td>\n",
       "      <td>MATR3</td>\n",
       "      <td>5:139332518:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr5</td>\n",
       "      <td>139329344</td>\n",
       "      <td>139332517</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000015479.20_2</td>\n",
       "      <td>MATR3</td>\n",
       "      <td>5:139332518:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr5</td>\n",
       "      <td>168262048</td>\n",
       "      <td>168264155</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000145934.17_3</td>\n",
       "      <td>TENM2</td>\n",
       "      <td>5:168264156:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr5</td>\n",
       "      <td>168262048</td>\n",
       "      <td>168264157</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000145934.17_3</td>\n",
       "      <td>TENM2</td>\n",
       "      <td>5:168264156:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr5</td>\n",
       "      <td>175916678</td>\n",
       "      <td>175918172</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000051596.10_5</td>\n",
       "      <td>THOC3</td>\n",
       "      <td>5:175916678:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr5</td>\n",
       "      <td>175916676</td>\n",
       "      <td>175918172</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000051596.10_5</td>\n",
       "      <td>THOC3</td>\n",
       "      <td>5:175916678:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr6</td>\n",
       "      <td>111597837</td>\n",
       "      <td>111602289</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000231889.8_6</td>\n",
       "      <td>TRAF3IP2-AS1</td>\n",
       "      <td>6:111602292:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr6</td>\n",
       "      <td>111597837</td>\n",
       "      <td>111602295</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000231889.8_6</td>\n",
       "      <td>TRAF3IP2-AS1</td>\n",
       "      <td>6:111602292:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr7</td>\n",
       "      <td>100092240</td>\n",
       "      <td>100093133</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000166508.18_3</td>\n",
       "      <td>MCM7</td>\n",
       "      <td>7:100092237:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr7</td>\n",
       "      <td>100092232</td>\n",
       "      <td>100093133</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000166508.18_3</td>\n",
       "      <td>MCM7</td>\n",
       "      <td>7:100092237:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr8</td>\n",
       "      <td>28750656</td>\n",
       "      <td>28756560</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000012232.10_2</td>\n",
       "      <td>EXTL3</td>\n",
       "      <td>8:28756561:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr8</td>\n",
       "      <td>28750656</td>\n",
       "      <td>28756562</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000012232.10_2</td>\n",
       "      <td>EXTL3</td>\n",
       "      <td>8:28756561:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr9</td>\n",
       "      <td>130232989</td>\n",
       "      <td>130241144</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000107130.10_2</td>\n",
       "      <td>NCS1</td>\n",
       "      <td>9:130241143:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr9</td>\n",
       "      <td>130232989</td>\n",
       "      <td>130241142</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000107130.10_2</td>\n",
       "      <td>NCS1</td>\n",
       "      <td>9:130241143:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr9</td>\n",
       "      <td>34680905</td>\n",
       "      <td>34682352</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230074.4_2</td>\n",
       "      <td>ENSG00000230074</td>\n",
       "      <td>9:34682351:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr9</td>\n",
       "      <td>34680905</td>\n",
       "      <td>34682350</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230074.4_2</td>\n",
       "      <td>ENSG00000230074</td>\n",
       "      <td>9:34682351:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr11</td>\n",
       "      <td>36229159</td>\n",
       "      <td>36232257</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179241.13_2</td>\n",
       "      <td>LDLRAD3</td>\n",
       "      <td>11:36232255:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr11</td>\n",
       "      <td>36229159</td>\n",
       "      <td>36232253</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179241.13_2</td>\n",
       "      <td>LDLRAD3</td>\n",
       "      <td>11:36232255:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr12</td>\n",
       "      <td>106885903</td>\n",
       "      <td>106889321</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000111785.21_3</td>\n",
       "      <td>RIC8B</td>\n",
       "      <td>12:106889314:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr12</td>\n",
       "      <td>106885903</td>\n",
       "      <td>106889307</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000111785.21_3</td>\n",
       "      <td>RIC8B</td>\n",
       "      <td>12:106889314:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr13</td>\n",
       "      <td>102678226</td>\n",
       "      <td>102679165</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134900.12_3</td>\n",
       "      <td>TPP2</td>\n",
       "      <td>13:102679169:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr13</td>\n",
       "      <td>102678226</td>\n",
       "      <td>102679173</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134900.12_3</td>\n",
       "      <td>TPP2</td>\n",
       "      <td>13:102679169:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54538062</td>\n",
       "      <td>54545302</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100532.13_3</td>\n",
       "      <td>CGRRF1</td>\n",
       "      <td>14:54545301:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54538062</td>\n",
       "      <td>54545300</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100532.13_3</td>\n",
       "      <td>CGRRF1</td>\n",
       "      <td>14:54545301:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr15</td>\n",
       "      <td>73769469</td>\n",
       "      <td>73770615</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260469.3_2</td>\n",
       "      <td>INSYN1-AS1</td>\n",
       "      <td>15:73770613:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr15</td>\n",
       "      <td>73769469</td>\n",
       "      <td>73770611</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260469.3_2</td>\n",
       "      <td>INSYN1-AS1</td>\n",
       "      <td>15:73770613:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr15</td>\n",
       "      <td>76347900</td>\n",
       "      <td>76348736</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000140386.13_5</td>\n",
       "      <td>SCAPER</td>\n",
       "      <td>15:76347904:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr15</td>\n",
       "      <td>76347906</td>\n",
       "      <td>76348736</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000140386.13_5</td>\n",
       "      <td>SCAPER</td>\n",
       "      <td>15:76347904:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr17</td>\n",
       "      <td>34637176</td>\n",
       "      <td>34639316</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181291.8_2</td>\n",
       "      <td>TMEM132E</td>\n",
       "      <td>17:34639317:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr17</td>\n",
       "      <td>34637176</td>\n",
       "      <td>34639318</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181291.8_2</td>\n",
       "      <td>TMEM132E</td>\n",
       "      <td>17:34639317:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr17</td>\n",
       "      <td>43763584</td>\n",
       "      <td>43769814</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000108861.9_2</td>\n",
       "      <td>DUSP3</td>\n",
       "      <td>17:43763586:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr17</td>\n",
       "      <td>43763586</td>\n",
       "      <td>43769814</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000108861.9_2</td>\n",
       "      <td>DUSP3</td>\n",
       "      <td>17:43763586:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr21</td>\n",
       "      <td>42296163</td>\n",
       "      <td>42304387</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160179.19_3</td>\n",
       "      <td>ABCG1</td>\n",
       "      <td>21:42304386:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr21</td>\n",
       "      <td>42296163</td>\n",
       "      <td>42304385</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160179.19_3</td>\n",
       "      <td>ABCG1</td>\n",
       "      <td>21:42304386:+</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chrX</td>\n",
       "      <td>71169253</td>\n",
       "      <td>71175311</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196338.15_3</td>\n",
       "      <td>NLGN3</td>\n",
       "      <td>X:71175306:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrX</td>\n",
       "      <td>71169253</td>\n",
       "      <td>71175301</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196338.15_3</td>\n",
       "      <td>NLGN3</td>\n",
       "      <td>X:71175306:+</td>\n",
       "      <td>background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127449</td>\n",
       "      <td>130129628</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3</td>\n",
       "      <td>AIFM1</td>\n",
       "      <td>X:130127455:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127459</td>\n",
       "      <td>130129628</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3</td>\n",
       "      <td>AIFM1</td>\n",
       "      <td>X:130127455:-</td>\n",
       "      <td>NULL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 70439100  | 70445530  | +            | ...   |\n",
       "| chr1         | 70439100  | 70445536  | +            | ...   |\n",
       "| chr1         | 33494217  | 33501611  | +            | ...   |\n",
       "| chr1         | 33494217  | 33501607  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 71169253  | 71175311  | +            | ...   |\n",
       "| chrX         | 71169253  | 71175301  | +            | ...   |\n",
       "| chrX         | 130127449 | 130129628 | -            | ...   |\n",
       "| chrX         | 130127459 | 130129628 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 52 rows and 8 columns from 16 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: le_id, gene_name_ref, Name, cryptic_status"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le_rep_d3utr.subset(lambda df: df.le_id.isin(novel_d3utr_dup_ids)).drop_duplicate_positions()[[\"le_id\", \"gene_name_ref\", \"Name\", \"cryptic_status\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proximal PAS regions with distinct 5'ends \n",
      "0\n",
      "Number of distal PAS regions with distinct 3'ends\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# All duplicates are equidistant to the atlas site. For the sake of simplicity and consistency arbitrarily select the most distal site for each le_id\n",
    "# Distal - select most distal/3' PAS for each le_id\n",
    "# Proximal - like ALEs, select the most proximal/3' site (i.e. shortest annotated exon 5'end)\n",
    "\n",
    "sel_le_d3utr_prox = select_rep_five_end(sel_le_d3utr_prox, \"le_id\")\n",
    "novel_le_rep_d3utr = select_rep_prox_site(novel_le_rep_d3utr, \"le_id_quant\")\n",
    "\n",
    "\n",
    "print(\"Number of proximal PAS regions with distinct 5'ends \")\n",
    "print(sel_le_d3utr_prox.subset(lambda df: df.le_id.isin(sel_d3utr_prox_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.Start.nunique() if (df.Strand == \"+\").all() else df.End.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )\n",
    "\n",
    "\n",
    "print(\"Number of distal PAS regions with distinct 3'ends\")\n",
    "print(novel_le_rep_d3utr.subset(lambda df: df.le_id_quant.isin(novel_d3utr_dup_ids))\n",
    " .drop_duplicate_positions()\n",
    " .as_df()\n",
    " .groupby([\"le_id\", \"Strand\"])\n",
    " .apply(lambda df: df.End.nunique() if (df.Strand == \"+\").all() else df.Start.nunique()).sort_values()\n",
    " .loc[lambda x: x > 1].index.nunique()\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct 'name' string with minimal annotation information\n",
    "# le_id|gene_name|site_type|cryptic_status\n",
    "\n",
    "sel_le_d3utr_prox = (sel_le_d3utr_prox\n",
    "                     .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"ref_gene_name\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"gene_name_common\",\n",
    "                          lambda df: pd.Series(np.where(df[\"gene_name\"].isna(),\n",
    "                                                        df[\"ref_gene_name\"],\n",
    "                                                        df[\"gene_name\"])))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_common\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    " )\n",
    "\n",
    "\n",
    "novel_le_rep_d3utr = (novel_le_rep_d3utr\n",
    "                      .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"Name\",\n",
    "         lambda df: df[\"le_id_quant\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# quick double check that all cryptic events are retained\n",
    "print(cryp_le_ids_prtnr - set(sel_le_d3utr_prox.subset(lambda df: df.cryptic_status.eq(\"cryptic\")).le_id))\n",
    "cryp_le_ids - set(novel_le_rep_d3utr.subset(lambda df: df.cryptic_status.eq(\"cryptic\")).le_id_quant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2040\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6220805</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000158286.13_3|RNF207|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6221220</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000158286.13_4|RNF207|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11296049</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000120942.14_2|UBIAD1|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11299544</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000120942.14_3|UBIAD1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1789</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149923199</td>\n",
       "      <td>149930081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197021.9_4|EOLA2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1790</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154505228</td>\n",
       "      <td>154506906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000071889.17_3|FAM3A|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154506158</td>\n",
       "      <td>154506906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000071889.17_2|FAM3A|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1792</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155491666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.12_5|TMLHE|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1793</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489010</td>\n",
       "      <td>155491666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.12_4|TMLHE|proximal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1794 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 6219235   | 6220805   | .          | +            | ...   |\n",
       "| chr1         | 6219235   | 6221220   | .          | +            | ...   |\n",
       "| chr1         | 11294872  | 11296049  | .          | +            | ...   |\n",
       "| chr1         | 11294872  | 11299544  | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 154505228 | 154506906 | .          | -            | ...   |\n",
       "| chrX         | 154506158 | 154506906 | .          | -            | ...   |\n",
       "| chrX         | 155478419 | 155491666 | .          | -            | ...   |\n",
       "| chrX         | 155489010 | 155491666 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 1,794 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate output bed file of proximal and distal PAS\n",
    "utr3_out_bed = pr.concat([sel_le_d3utr_prox[[\"Score\", \"Name\"]],\n",
    "                          novel_le_rep_d3utr[[\"Score\", \"Name\"]]]\n",
    "                          )\n",
    "\n",
    "print(len(utr3_out_bed))\n",
    "# drop duplicate entries\n",
    "utr3_out_bed = utr3_out_bed.drop_duplicate_positions().sort()\n",
    "utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>event_type</th>\n",
       "      <th>reg_status</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>proximal</td>\n",
       "      <td>background</td>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>proximal</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>distal</td>\n",
       "      <td>background</td>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>distal</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  event_type  reg_status  count\n",
       "0   proximal  background    793\n",
       "1   proximal     cryptic    104\n",
       "2     distal  background    793\n",
       "3     distal     cryptic    104"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get event counts for cryptic/bg for each region set\n",
    "# event_counts = {}\n",
    "# for nm, gr in {\"spliced_all_background\": spliced_out_bed,\n",
    "#                \"spliced_shsy5y_background\": spliced_out_bed_shsy5y_all,\n",
    "#                \"bleedthrough_all_background\": bleed_out_bed,\n",
    "#                \"bleed_out_bed_shsy5y_all\": bleed_out_bed_shsy5y_all}.items():\n",
    "    \n",
    "#     event_counts[nm] = gr.Name.str.split(\"|\", expand=True)[3].rename(\"reg_status\").value_counts()\n",
    "\n",
    "event_counts = {region: utr3_out_bed.subset(lambda df: df.Name.str.contains(region, regex=False)).Name.str.split(\"|\", expand=True)[3].rename(\"reg_status\").value_counts() \n",
    "                for region in [\"proximal\", \"distal\"]}\n",
    "\n",
    "\n",
    "\n",
    "event_counts_df = pd.concat(event_counts, names=[\"event_type\"]).reset_index()\n",
    "event_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"../processed/iclip_regions\"):\n",
    "    os.makedirs(\"../processed/iclip_regions\")\n",
    "    \n",
    "utr3_out_bed.to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.last_exons.bed\")\n",
    "# proximal 3'UTRs\n",
    "utr3_out_bed.subset(lambda df: df.Name.str.contains(\"proximal\", regex=False)).to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.last_exons.proximal.bed\")\n",
    "# distal 3'UTRs\n",
    "utr3_out_bed.subset(lambda df: df.Name.str.contains(\"distal\", regex=False)).to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.last_exons.distal.bed\")\n",
    "\n",
    "# also make one with just polYA sites\n",
    "utr3_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.pas.bed\")\n",
    "# just proximal\n",
    "utr3_out_bed.three_end().subset(lambda df: df.Name.str.contains(\"proximal\", regex=False)).to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.pas.proximal.bed\")\n",
    "# just distal\n",
    "utr3_out_bed.three_end().subset(lambda df: df.Name.str.contains(\"distal\", regex=False)).to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.pas.distal.bed\")\n",
    "\n",
    "event_counts_df.to_csv(\"../processed/iclip_regions/2023-12-03_papa_d3utr.event_counts.tsv\", sep=\"\\t\", header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+------------+--------------+-------+\n",
      "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
      "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
      "|--------------+-----------+-----------+------------+--------------+-------|\n",
      "| chr1         | 6219235   | 6221220   | .          | +            | ...   |\n",
      "| chr1         | 11294872  | 11299544  | .          | +            | ...   |\n",
      "| chr1         | 15939295  | 15940477  | .          | +            | ...   |\n",
      "| chr1         | 19258680  | 19264037  | .          | +            | ...   |\n",
      "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
      "| chrX         | 123600573 | 123601338 | .          | -            | ...   |\n",
      "| chrX         | 149923199 | 149930081 | .          | -            | ...   |\n",
      "| chrX         | 154505228 | 154506906 | .          | -            | ...   |\n",
      "| chrX         | 155478419 | 155491666 | .          | -            | ...   |\n",
      "+--------------+-----------+-----------+------------+--------------+-------+\n",
      "Stranded PyRanges object has 897 rows and 6 columns from 23 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "1 hidden columns: Name\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6220805</td>\n",
       "      <td>6221220</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000158286.13_4|RNF207|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11296049</td>\n",
       "      <td>11299544</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000120942.14_3|UBIAD1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19260128</td>\n",
       "      <td>19264037</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000053372.5_3|MRTO4|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24472976</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_6|NIPAL3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935</th>\n",
       "      <td>chrX</td>\n",
       "      <td>110673482</td>\n",
       "      <td>110673855</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000101938.15_2|CHRDL1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>936</th>\n",
       "      <td>chrX</td>\n",
       "      <td>117894639</td>\n",
       "      <td>117897812</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000003096.14_2|KLHL13|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>chrX</td>\n",
       "      <td>119584579</td>\n",
       "      <td>119588336</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000186416.18_2|NKRF|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>938</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154505228</td>\n",
       "      <td>154506158</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000071889.17_3|FAM3A|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155489010</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.12_5|TMLHE|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>940 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 6220805   | 6221220   | .          | +            | ...   |\n",
       "| chr1         | 11296049  | 11299544  | .          | +            | ...   |\n",
       "| chr1         | 15940456  | 15940477  | .          | +            | ...   |\n",
       "| chr1         | 19260128  | 19264037  | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 117894639 | 117897812 | .          | -            | ...   |\n",
       "| chrX         | 119584579 | 119588336 | .          | -            | ...   |\n",
       "| chrX         | 154505228 | 154506158 | .          | -            | ...   |\n",
       "| chrX         | 155478419 | 155489010 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 940 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# also make a BED with distal bins\n",
    "utr3_out_bed_prox = utr3_out_bed.subset(lambda df: df.Name.str.contains(\"proximal\", regex=False))\n",
    "utr3_out_bed_dist = utr3_out_bed.subset(lambda df: df.Name.str.contains(\"distal\", regex=False))\n",
    "\n",
    "print(utr3_out_bed_dist)\n",
    "# uniq region of distal 3'UTR\n",
    "utr3_out_bed_dist_bin = utr3_out_bed_dist.subtract(utr3_out_bed_prox, strandedness=\"same\")\n",
    "utr3_out_bed_dist_bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "utr3_out_bed_dist_bin.to_bed(\"../processed/iclip_regions/2023-12-03_papa_cryptic_d3utr.background_shsy5y.last_exon_bins.distal.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
