{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# approach\n",
    "\n",
    "### Input last exons - selecting representative PAS/last exon:\n",
    "\n",
    "le_ids can have multiple predictions, need to select a single one.\n",
    "Strategy:\n",
    "(starting with novel last exons used for quantification) \n",
    "1. Prefer events that match atlas\n",
    "2. if multiple, select min atlas distance\n",
    "3. If matches, s\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## To-dos:\n",
    "- fix updating to atlas for -ve strand matching pas\n",
    "- run selection for all novel 3'UTR extensions\n",
    "- Implement selection for annotated proximal sites of 3'UTR extensions (i.e. most distal annotated site)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
      "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
      "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
      "       'le_number', 'le_id'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074713_S59.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_3.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24475256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24473773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-B_S2.334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24474852</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_3.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36889</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074729_S12.25635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36890</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36891</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.26168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36892</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234583</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_0_2.21889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36893</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234594</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.24449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36894 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24475256  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24473773  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234583  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234594  | 13234826  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 36,894 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(quant_uniq_le.columns)\n",
    "quant_uniq_le.subset(lambda df: ~df.transcript_id.str.startswith(\"ENST\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['experiment_name', 'binID', 'groupID', 'featureID', 'exonBaseMean',\n",
       "       'dispersion', 'stat', 'pvalue', 'padj', 'UsageCoefficient_base',\n",
       "       'UsageCoefficient_contrast', 'log2fold_contrast_base', 'le_id',\n",
       "       'gene.qvalue', 'contrast_name', 'gene_id', 'gene_name', 'event_type',\n",
       "       'annot_status', 'transcript_id', 'chromosome', 'strand', 'start', 'end',\n",
       "       'mean_PPAU_base', 'mean_PPAU_treatment', 'delta_PPAU_treatment_control',\n",
       "       'simple_event_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
       "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
       "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
       "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
       "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
       "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3369"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel 3'UTR ectensions\n",
    "le_id_d3utr = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "len(le_id_d3utr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def n_uniq_coords(df):\n",
    "\n",
    "    if df[\"Strand\"].unique()[0] == \"+\":\n",
    "        n_coords = df[\"End\"].nunique()\n",
    "    \n",
    "    else:\n",
    "        n_coords = df[\"Start\"].nunique()\n",
    "\n",
    "    return n_coords\n",
    "\n",
    "\n",
    "def select_rep_site(df):\n",
    "# First select sites of le_id that match atlas site (if applicable)\n",
    "    if df[\"atlas_filter\"].astype(int).sum() > 0:\n",
    "\n",
    "        out_df = df[df[\"atlas_filter\"].eq(\"1\")]\n",
    "\n",
    "        # 2. Within these, see if multiple different 3'ends predicted\n",
    "        ### TODO: use Name (atlas ID) instead of 3' coordinate to match\n",
    "        n_atlas = out_df[\"Name\"].nunique()\n",
    "\n",
    "        if n_atlas > 1:\n",
    "                # multiple different atlas sites predicted\n",
    "                # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                end3_c = out_df.groupby(\"Name\")[\"experiment_id\"].nunique()\n",
    "                # have a series with Name (PolyASite ID) as index labels, count of experimnts as values\n",
    "                # get the IDs with largest number of unqiue experiments (as as series)\n",
    "                end3_c = end3_c[end3_c == end3_c.max()]\n",
    "\n",
    "                if len(end3_c) > 1:\n",
    "                    # multiple 3'ends/ atlas sites predicted in same number of datasets\n",
    "                    # raise Exception(f\"{out_df['le_id_quant'].drop_duplicates().iloc[0]} - le_id has multiple predicted atlas sites\")\n",
    "                    # prefer the shortest extension in this case (conservative, defines smaller window in)\n",
    "\n",
    "                    # calculate length\n",
    "                    out_df = out_df[out_df[\"Name\"].isin(end3_c.index)]\n",
    "                    out_df[\"reg_len\"] = out_df.End - out_df.Start\n",
    "\n",
    "                    # select shortest extension\n",
    "                    out_df = out_df[out_df.reg_len == out_df.reg_len.min()]\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"]).drop(columns=\"reg_len\")\n",
    "\n",
    "                    return out_df, \"atlas_max_datasets_shortest\"\n",
    "                                    \n",
    "                else:\n",
    "                    # 1 pas remaining\n",
    "                    end3_coord = end3_c.index[0]\n",
    "\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df[out_df[\"Name\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "                    decision_str = \"atlas_max_datasets\"\n",
    "\n",
    "\n",
    "        else:\n",
    "            # only single atlas site predicted - just select a single row\n",
    "            end3_coord = out_df[\"Name\"].drop_duplicates().index[0]\n",
    "            # Select a single row for le_id corresponding to selected PAS\n",
    "            out_df = out_df[out_df[\"Name\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "            decision_str = \"atlas_1_pred\"\n",
    "        \n",
    "\n",
    "        return out_df, decision_str\n",
    "\n",
    "    else:\n",
    "        # TODO: implement filtering for events called with motifs only\n",
    "        out_df = df[df[\"atlas_filter\"].eq(\"0\")]\n",
    "\n",
    "        # select predicted site with motif that is closest to expected distance (21nt upstream)\n",
    "        # -74_TATAAA,-14_TATAAA\n",
    "        # [-74_TATAAA, -14_TATAAA]\n",
    "        # [-74, -14]\n",
    "        # absolute distance from expected position (21nt upstream)\n",
    "        # select min deviating motif for each predicted PAS\n",
    "        pas_m = out_df[\"pas_motifs\"].apply(lambda x: min([abs(21 - abs(int(i.split(\"_\")[0]))) for i in x.split(\",\")]))\n",
    "\n",
    "        # select predicted pas with minimum deviation from 21nt\n",
    "        pas_m = pas_m[pas_m == pas_m.min()]\n",
    "\n",
    "        if len(pas_m) > 1:\n",
    "            # multiple predicted PAS with motif same distance from expected\n",
    "\n",
    "            # check if exactly the same coordinate predicted for these events\n",
    "\n",
    "            # (if true, just sort alphabetically by experiment id and pick first)\n",
    "            # otherwise, select shortest\n",
    "            n_pas = n_uniq_coords(out_df)\n",
    "\n",
    "            if n_pas > 1:\n",
    "                # select shortest extension\n",
    "                # calculate length\n",
    "                out_df = out_df.loc[pas_m.index, :]\n",
    "\n",
    "                out_df[\"reg_len\"] = out_df.End - out_df.Start\n",
    "\n",
    "                # select shortest extension\n",
    "                out_df = out_df[out_df.reg_len == out_df.reg_len.min()]\n",
    "                # Select a single row for le_id corresponding to selected PAS\n",
    "                out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"]).drop(columns=\"reg_len\")\n",
    "\n",
    "            else:\n",
    "                # just a single coord predicted in multiple datasets - \n",
    "                out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "            return out_df, \"motif_shortest_min\"\n",
    "        \n",
    "        else:\n",
    "            out_df = out_df.loc[pas_m.index, :]\n",
    "            \n",
    "\n",
    "            decision_str = \"motif_1_min\"\n",
    "\n",
    "\n",
    "        return out_df, decision_str\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'atlas_1_pred': 41, 'atlas_max_datasets': 20, 'atlas_max_datasets_shortest': 8, 'motif_1_min': 29, 'motif_shortest_min': 2}\n"
     ]
    }
   ],
   "source": [
    "#### SELECTING REPRESENTATIVE PAS\n",
    "# LEs have to be quantified\n",
    "novel_le_df = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"])).as_df()\n",
    "\n",
    "\n",
    "i = 0\n",
    "\n",
    "sel_dfs = []\n",
    "sel_dict = {\"atlas_1_pred\": [],\n",
    "            \"atlas_max_datasets\": [],\n",
    "            \"atlas_max_datasets_shortest\": [],\n",
    "            \"motif_1_min\": [],\n",
    "            \"motif_shortest_min\": []}\n",
    "\n",
    "le_grpd = iter(novel_le_df[novel_le_df.le_id.isin(le_id_d3utr)].groupby(\"le_id_quant\"))\n",
    "\n",
    "\n",
    "\n",
    "while i < 100:\n",
    "        le_id, df = next(le_grpd)\n",
    "        # print(le_id)\n",
    "        out_df, decision_str = select_rep_site(df)\n",
    "        sel_dfs.append(out_df)\n",
    "        sel_dict[decision_str].append(le_id)\n",
    "        i +=1  \n",
    "\n",
    "\n",
    "\n",
    "sel_novel_le_df = pd.concat(sel_dfs)               \n",
    "    \n",
    "print({dec: len(ids) for dec, ids in sel_dict.items()})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>le_id_quant</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>experiment_count</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33375</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>X:7897884:-</td>\n",
       "      <td>7897883</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.26964.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33376</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>X:7897884:-</td>\n",
       "      <td>7897883</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26054.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33377</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898079</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.NT_19074717_S21.25514.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33378</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898128</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.26703.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33379</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898132</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.25616.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33380</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898136</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.TDP43_ctrl_1.25803.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33381</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898241</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.NT_19074721_S50.25909.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                le_id_quant  \\\n",
       "33375  ENSG00000006757.12_2   \n",
       "33376  ENSG00000006757.12_2   \n",
       "33377  ENSG00000006757.12_2   \n",
       "33378  ENSG00000006757.12_2   \n",
       "33379  ENSG00000006757.12_2   \n",
       "33380  ENSG00000006757.12_2   \n",
       "33381  ENSG00000006757.12_2   \n",
       "\n",
       "                                           experiment_id         Name  \\\n",
       "33375  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...  X:7897884:-   \n",
       "33376  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...  X:7897884:-   \n",
       "33377                               seddighi_i3_cortical  X:7898247:-   \n",
       "33378  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...  X:7898247:-   \n",
       "33379                               seddighi_i3_cortical  X:7898247:-   \n",
       "33380  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...  X:7898247:-   \n",
       "33381                               seddighi_i3_cortical  X:7898247:-   \n",
       "\n",
       "         Start      End experiment_count                    transcript_id  \\\n",
       "33375  7897883  7900817                2         PAPA.ctrl_ctrl_2.26964.1   \n",
       "33376  7897883  7900817                2        PAPA.TDP43_ctrl_4.26054.2   \n",
       "33377  7898079  7900817                2     PAPA.NT_19074717_S21.25514.1   \n",
       "33378  7898128  7900817                2         PAPA.ctrl_ctrl_4.26703.2   \n",
       "33379  7898132  7900817                2  PAPA.TDP43_19065403_S23.25616.1   \n",
       "33380  7898136  7900817                2        PAPA.TDP43_ctrl_1.25803.1   \n",
       "33381  7898241  7900817                2     PAPA.NT_19074721_S50.25909.1   \n",
       "\n",
       "      Strand  \n",
       "33375      -  \n",
       "33376      -  \n",
       "33377      -  \n",
       "33378      -  \n",
       "33379      -  \n",
       "33380      -  \n",
       "33381      -  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le_df[novel_le_df[\"le_id_quant\"].isin(sel_dict[\"atlas_multiple_sites\"])][[\"le_id_quant\", \"experiment_id\", \"Name\", \"Start\", \"End\", \"experiment_count\", \"transcript_id\", \"Strand\"]].loc[lambda df: df.le_id_quant == ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>le_id_quant</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>experiment_count</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13794</th>\n",
       "      <td>ENSG00000001629.10_3</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>7:92401774:+</td>\n",
       "      <td>92398211</td>\n",
       "      <td>92401760</td>\n",
       "      <td>6</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10802.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14443</th>\n",
       "      <td>ENSG00000005108.17_3</td>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>7:11370437:-</td>\n",
       "      <td>11370452</td>\n",
       "      <td>11375878</td>\n",
       "      <td>7</td>\n",
       "      <td>PAPA.doxconc_NT_0_2.7361.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14450</th>\n",
       "      <td>ENSG00000006468.14_4</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>7:13890964:-</td>\n",
       "      <td>13890946</td>\n",
       "      <td>13896087</td>\n",
       "      <td>3</td>\n",
       "      <td>PAPA.TDP-3.9905.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33381</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898241</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.NT_19074721_S50.25909.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13230</th>\n",
       "      <td>ENSG00000008018.9_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>6:170534538:-</td>\n",
       "      <td>170534537</td>\n",
       "      <td>170535405</td>\n",
       "      <td>3</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.9994.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32454</th>\n",
       "      <td>ENSG00000015475.19_2</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>22:17733561:-</td>\n",
       "      <td>17733560</td>\n",
       "      <td>17735591</td>\n",
       "      <td>4</td>\n",
       "      <td>PAPA.CTRL-5.21657.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21869</th>\n",
       "      <td>ENSG00000018236.15_4</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>12:41073374:+</td>\n",
       "      <td>41069958</td>\n",
       "      <td>41073383</td>\n",
       "      <td>6</td>\n",
       "      <td>PAPA.TDP43-F_S6.12521.4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>ENSG00000031698.13_3</td>\n",
       "      <td>zanovello_shsy5y_chx_kd_only</td>\n",
       "      <td>1:109238791:+</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238792</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_4.997.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                le_id_quant  \\\n",
       "13794  ENSG00000001629.10_3   \n",
       "14443  ENSG00000005108.17_3   \n",
       "14450  ENSG00000006468.14_4   \n",
       "33381  ENSG00000006757.12_2   \n",
       "13230   ENSG00000008018.9_2   \n",
       "32454  ENSG00000015475.19_2   \n",
       "21869  ENSG00000018236.15_4   \n",
       "764    ENSG00000031698.13_3   \n",
       "\n",
       "                                           experiment_id           Name  \\\n",
       "13794  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...   7:92401774:+   \n",
       "14443                        zanovello_shsy5y_curve_0075   7:11370437:-   \n",
       "14450                               humphrey_i3_cortical   7:13890964:-   \n",
       "33381                               seddighi_i3_cortical    X:7898247:-   \n",
       "13230                               seddighi_i3_cortical  6:170534538:-   \n",
       "32454                               humphrey_i3_cortical  22:17733561:-   \n",
       "21869                                  brown_i3_cortical  12:41073374:+   \n",
       "764                         zanovello_shsy5y_chx_kd_only  1:109238791:+   \n",
       "\n",
       "           Start        End experiment_count                   transcript_id  \\\n",
       "13794   92398211   92401760                6       PAPA.TDP43_ctrl_2.10802.2   \n",
       "14443   11370452   11375878                7      PAPA.doxconc_NT_0_2.7361.1   \n",
       "14450   13890946   13896087                3               PAPA.TDP-3.9905.1   \n",
       "33381    7898241    7900817                2    PAPA.NT_19074721_S50.25909.1   \n",
       "13230  170534537  170535405                3  PAPA.TDP43_19065409_S29.9994.1   \n",
       "32454   17733560   17735591                4             PAPA.CTRL-5.21657.1   \n",
       "21869   41069958   41073383                6         PAPA.TDP43-F_S6.12521.4   \n",
       "764    109237730  109238792                2  PAPA.chx_tdp_CTRL_ctrl_4.997.3   \n",
       "\n",
       "      Strand  \n",
       "13794      +  \n",
       "14443      -  \n",
       "14450      -  \n",
       "33381      -  \n",
       "13230      -  \n",
       "32454      -  \n",
       "21869      +  \n",
       "764        +  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_novel_le_df[sel_novel_le_df[\"le_id_quant\"].isin(sel_dict[\"atlas_max_datasets_shortest\"])][[\"le_id_quant\", \"experiment_id\", \"Name\", \"Start\", \"End\", \"experiment_count\", \"transcript_id\", \"Strand\"]]\n",
    "#.loc[lambda df: df.le_id_quant == \"ENSG00000006757.12_2\", :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
      "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
      "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
      "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
      "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
      "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU',\n",
      "       'le_id_quant'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>atlas_filter</th>\n",
       "      <th>Name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14815</th>\n",
       "      <td>92198195</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92197498:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14816</th>\n",
       "      <td>92198602</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14817</th>\n",
       "      <td>92199253</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14818</th>\n",
       "      <td>92199257</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Start       End atlas_filter          Name                 le_id  \\\n",
       "14815  92198195  92200804            0  7:92197498:-  ENSG00000001631.17_3   \n",
       "14816  92198602  92200804            0  7:92199243:-  ENSG00000001631.17_3   \n",
       "14817  92199253  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "14818  92199257  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "\n",
       "      gene_name_ref  \n",
       "14815         KRIT1  \n",
       "14816         KRIT1  \n",
       "14817         KRIT1  \n",
       "14818         KRIT1  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"].columns)\n",
    "novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"][[\"Start\", \"End\", \"atlas_filter\", \"Name\", \"le_id\", \"gene_name_ref\"]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
