{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# approach\n",
    "\n",
    "### Input last exons - selecting representative PAS/last exon:\n",
    "\n",
    "le_ids can have multiple predictions, need to select a single one.\n",
    "Strategy:\n",
    "(starting with novel last exons used for quantification) \n",
    "1. Prefer events that match atlas\n",
    "2. if multiple, select min atlas distance\n",
    "3. If matches, s\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## To-dos:\n",
    "- fix updating to atlas for -ve strand matching pas\n",
    "- run selection for all novel 3'UTR extensions\n",
    "- Implement selection for annotated proximal sites of 3'UTR extensions (i.e. most distal annotated site)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
      "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
      "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
      "       'le_number', 'le_id'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074713_S59.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_3.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24475256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24473773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-B_S2.334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24474852</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_3.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36889</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074729_S12.25635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36890</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36891</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.26168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36892</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234583</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_0_2.21889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36893</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234594</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.24449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36894 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24475256  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24473773  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234583  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234594  | 13234826  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 36,894 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(quant_uniq_le.columns)\n",
    "quant_uniq_le.subset(lambda df: ~df.transcript_id.str.startswith(\"ENST\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['experiment_name', 'binID', 'groupID', 'featureID', 'exonBaseMean',\n",
       "       'dispersion', 'stat', 'pvalue', 'padj', 'UsageCoefficient_base',\n",
       "       'UsageCoefficient_contrast', 'log2fold_contrast_base', 'le_id',\n",
       "       'gene.qvalue', 'contrast_name', 'gene_id', 'gene_name', 'event_type',\n",
       "       'annot_status', 'transcript_id', 'chromosome', 'strand', 'start', 'end',\n",
       "       'mean_PPAU_base', 'mean_PPAU_treatment', 'delta_PPAU_treatment_control',\n",
       "       'simple_event_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
       "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
       "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
       "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
       "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
       "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3369"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel 3'UTR ectensions\n",
    "le_id_d3utr = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "len(le_id_d3utr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def n_uniq_coords(df):\n",
    "\n",
    "    if df[\"Strand\"].unique()[0] == \"+\":\n",
    "        n_coords = df[\"End\"].nunique()\n",
    "    \n",
    "    else:\n",
    "        n_coords = df[\"Start\"].nunique()\n",
    "\n",
    "    return n_coords\n",
    "\n",
    "\n",
    "def select_rep_site(df):\n",
    "# First select sites of le_id that match atlas site (if applicable)\n",
    "    if df[\"atlas_filter\"].astype(int).sum() > 0:\n",
    "\n",
    "        out_df = df[df[\"atlas_filter\"].eq(\"1\")]\n",
    "\n",
    "        # 2. Within these, see if multiple different 3'ends predicted\n",
    "        ### TODO: use Name (atlas ID) instead of 3' coordinate to match\n",
    "        n_atlas = out_df[\"Name\"].nunique()\n",
    "\n",
    "        if n_atlas > 1:\n",
    "                # multiple different atlas sites predicted\n",
    "                # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                end3_c = out_df.groupby(\"Name\")[\"experiment_id\"].nunique()\n",
    "                # have a series with Name (PolyASite ID) as index labels, count of experimnts as values\n",
    "                # get the IDs with largest number of unqiue experiments (as as series)\n",
    "                end3_c = end3_c[end3_c == end3_c.max()]\n",
    "\n",
    "                if len(end3_c) > 1:\n",
    "                    # multiple 3'ends/ atlas sites predicted in same number of datasets\n",
    "                    # raise Exception(f\"{out_df['le_id_quant'].drop_duplicates().iloc[0]} - le_id has multiple predicted atlas sites\")\n",
    "                    # prefer the shortest extension in this case (conservative, defines smaller window in)\n",
    "\n",
    "                    # calculate length\n",
    "                    out_df = out_df[out_df[\"Name\"].isin(end3_c.index)]\n",
    "                    out_df[\"reg_len\"] = out_df.End - out_df.Start\n",
    "\n",
    "                    # select shortest extension\n",
    "                    out_df = out_df[out_df.reg_len == out_df.reg_len.min()]\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"]).drop(columns=\"reg_len\")\n",
    "\n",
    "                    return out_df, \"atlas_max_datasets_shortest\"\n",
    "                                    \n",
    "                else:\n",
    "                    # 1 pas remaining\n",
    "                    end3_coord = end3_c.index[0]\n",
    "\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df[out_df[\"Name\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "                    decision_str = \"atlas_max_datasets\"\n",
    "\n",
    "\n",
    "        else:\n",
    "            # only single atlas site predicted - just select a single row\n",
    "            end3_coord = out_df[\"Name\"].drop_duplicates().index[0]\n",
    "            # Select a single row for le_id corresponding to selected PAS\n",
    "            out_df = out_df[out_df[\"Name\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "            decision_str = \"atlas_1_pred\"\n",
    "        \n",
    "\n",
    "        return out_df, decision_str\n",
    "\n",
    "    else:\n",
    "        # TODO: implement filtering for events called with motifs only\n",
    "        out_df = df[df[\"atlas_filter\"].eq(\"0\")]\n",
    "\n",
    "        # select predicted site with motif that is closest to expected distance (21nt upstream)\n",
    "        # -74_TATAAA,-14_TATAAA\n",
    "        # [-74_TATAAA, -14_TATAAA]\n",
    "        # [-74, -14]\n",
    "        # absolute distance from expected position (21nt upstream)\n",
    "        # select min deviating motif for each predicted PAS\n",
    "        pas_m = out_df[\"pas_motifs\"].apply(lambda x: min([abs(21 - abs(int(i.split(\"_\")[0]))) for i in x.split(\",\")]))\n",
    "\n",
    "        # select predicted pas with minimum deviation from 21nt\n",
    "        pas_m = pas_m[pas_m == pas_m.min()]\n",
    "\n",
    "        if len(pas_m) > 1:\n",
    "            # multiple predicted PAS with motif same distance from expected\n",
    "\n",
    "            # check if exactly the same coordinate predicted for these events\n",
    "\n",
    "            # (if true, just sort alphabetically by experiment id and pick first)\n",
    "            # otherwise, select shortest\n",
    "            n_pas = n_uniq_coords(out_df)\n",
    "\n",
    "            if n_pas > 1:\n",
    "                # select shortest extension\n",
    "                # calculate length\n",
    "                out_df = out_df.loc[pas_m.index, :]\n",
    "\n",
    "                out_df[\"reg_len\"] = out_df.End - out_df.Start\n",
    "\n",
    "                # select shortest extension\n",
    "                out_df = out_df[out_df.reg_len == out_df.reg_len.min()]\n",
    "                # Select a single row for le_id corresponding to selected PAS\n",
    "                out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"]).drop(columns=\"reg_len\")\n",
    "\n",
    "            else:\n",
    "                # just a single coord predicted in multiple datasets - \n",
    "                out_df = out_df.sort_values(by=\"experiment_id\").drop_duplicates(subset=[\"Start\", \"End\", \"Strand\"])\n",
    "\n",
    "            return out_df, \"motif_shortest_min\"\n",
    "        \n",
    "        else:\n",
    "            out_df = out_df.loc[pas_m.index, :]\n",
    "            \n",
    "\n",
    "            decision_str = \"motif_1_min\"\n",
    "\n",
    "\n",
    "        return out_df, decision_str\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'atlas_1_pred': 41, 'atlas_max_datasets': 20, 'atlas_max_datasets_shortest': 8, 'motif_1_min': 29, 'motif_shortest_min': 2}\n"
     ]
    }
   ],
   "source": [
    "#### SELECTING REPRESENTATIVE PAS\n",
    "# LEs have to be quantified\n",
    "novel_le_df = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"])).as_df()\n",
    "\n",
    "\n",
    "i = 0\n",
    "\n",
    "sel_dfs = []\n",
    "sel_dict = {\"atlas_1_pred\": [],\n",
    "            \"atlas_max_datasets\": [],\n",
    "            \"atlas_max_datasets_shortest\": [],\n",
    "            \"motif_1_min\": [],\n",
    "            \"motif_shortest_min\": []}\n",
    "\n",
    "le_grpd = iter(novel_le_df[novel_le_df.le_id.isin(le_id_d3utr)].groupby(\"le_id_quant\"))\n",
    "\n",
    "\n",
    "\n",
    "while i < 100:\n",
    "        le_id, df = next(le_grpd)\n",
    "        # print(le_id)\n",
    "        out_df, decision_str = select_rep_site(df)\n",
    "        sel_dfs.append(out_df)\n",
    "        sel_dict[decision_str].append(le_id)\n",
    "        i +=1  \n",
    "\n",
    "\n",
    "\n",
    "sel_novel_le_df = pd.concat(sel_dfs)               \n",
    "    \n",
    "print({dec: len(ids) for dec, ids in sel_dict.items()})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# novel_le_df[novel_le_df[\"le_id_quant\"].isin(sel_dict[\"atlas_multiple_sites\"])][[\"le_id_quant\", \"experiment_id\", \"Name\", \"Start\", \"End\", \"experiment_count\", \"transcript_id\", \"Strand\"]].loc[lambda df: df.le_id_quant == ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>le_id_quant</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>experiment_count</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13794</th>\n",
       "      <td>ENSG00000001629.10_3</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>7:92401774:+</td>\n",
       "      <td>92398211</td>\n",
       "      <td>92401760</td>\n",
       "      <td>6</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10802.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14443</th>\n",
       "      <td>ENSG00000005108.17_3</td>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>7:11370437:-</td>\n",
       "      <td>11370452</td>\n",
       "      <td>11375878</td>\n",
       "      <td>7</td>\n",
       "      <td>PAPA.doxconc_NT_0_2.7361.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14450</th>\n",
       "      <td>ENSG00000006468.14_4</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>7:13890964:-</td>\n",
       "      <td>13890946</td>\n",
       "      <td>13896087</td>\n",
       "      <td>3</td>\n",
       "      <td>PAPA.TDP-3.9905.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33381</th>\n",
       "      <td>ENSG00000006757.12_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>X:7898247:-</td>\n",
       "      <td>7898241</td>\n",
       "      <td>7900817</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.NT_19074721_S50.25909.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13230</th>\n",
       "      <td>ENSG00000008018.9_2</td>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>6:170534538:-</td>\n",
       "      <td>170534537</td>\n",
       "      <td>170535405</td>\n",
       "      <td>3</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.9994.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32454</th>\n",
       "      <td>ENSG00000015475.19_2</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>22:17733561:-</td>\n",
       "      <td>17733560</td>\n",
       "      <td>17735591</td>\n",
       "      <td>4</td>\n",
       "      <td>PAPA.CTRL-5.21657.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21869</th>\n",
       "      <td>ENSG00000018236.15_4</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>12:41073374:+</td>\n",
       "      <td>41069958</td>\n",
       "      <td>41073383</td>\n",
       "      <td>6</td>\n",
       "      <td>PAPA.TDP43-F_S6.12521.4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>ENSG00000031698.13_3</td>\n",
       "      <td>zanovello_shsy5y_chx_kd_only</td>\n",
       "      <td>1:109238791:+</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238792</td>\n",
       "      <td>2</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_4.997.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                le_id_quant  \\\n",
       "13794  ENSG00000001629.10_3   \n",
       "14443  ENSG00000005108.17_3   \n",
       "14450  ENSG00000006468.14_4   \n",
       "33381  ENSG00000006757.12_2   \n",
       "13230   ENSG00000008018.9_2   \n",
       "32454  ENSG00000015475.19_2   \n",
       "21869  ENSG00000018236.15_4   \n",
       "764    ENSG00000031698.13_3   \n",
       "\n",
       "                                           experiment_id           Name  \\\n",
       "13794  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...   7:92401774:+   \n",
       "14443                        zanovello_shsy5y_curve_0075   7:11370437:-   \n",
       "14450                               humphrey_i3_cortical   7:13890964:-   \n",
       "33381                               seddighi_i3_cortical    X:7898247:-   \n",
       "13230                               seddighi_i3_cortical  6:170534538:-   \n",
       "32454                               humphrey_i3_cortical  22:17733561:-   \n",
       "21869                                  brown_i3_cortical  12:41073374:+   \n",
       "764                         zanovello_shsy5y_chx_kd_only  1:109238791:+   \n",
       "\n",
       "           Start        End experiment_count                   transcript_id  \\\n",
       "13794   92398211   92401760                6       PAPA.TDP43_ctrl_2.10802.2   \n",
       "14443   11370452   11375878                7      PAPA.doxconc_NT_0_2.7361.1   \n",
       "14450   13890946   13896087                3               PAPA.TDP-3.9905.1   \n",
       "33381    7898241    7900817                2    PAPA.NT_19074721_S50.25909.1   \n",
       "13230  170534537  170535405                3  PAPA.TDP43_19065409_S29.9994.1   \n",
       "32454   17733560   17735591                4             PAPA.CTRL-5.21657.1   \n",
       "21869   41069958   41073383                6         PAPA.TDP43-F_S6.12521.4   \n",
       "764    109237730  109238792                2  PAPA.chx_tdp_CTRL_ctrl_4.997.3   \n",
       "\n",
       "      Strand  \n",
       "13794      +  \n",
       "14443      -  \n",
       "14450      -  \n",
       "33381      -  \n",
       "13230      -  \n",
       "32454      -  \n",
       "21869      +  \n",
       "764        +  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_novel_le_df[sel_novel_le_df[\"le_id_quant\"].isin(sel_dict[\"atlas_max_datasets_shortest\"])][[\"le_id_quant\", \"experiment_id\", \"Name\", \"Start\", \"End\", \"experiment_count\", \"transcript_id\", \"Strand\"]]\n",
    "#.loc[lambda df: df.le_id_quant == \"ENSG00000006757.12_2\", :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
      "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
      "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
      "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
      "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
      "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU',\n",
      "       'le_id_quant'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>atlas_filter</th>\n",
       "      <th>Name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14815</th>\n",
       "      <td>92198195</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92197498:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14816</th>\n",
       "      <td>92198602</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14817</th>\n",
       "      <td>92199253</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14818</th>\n",
       "      <td>92199257</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Start       End atlas_filter          Name                 le_id  \\\n",
       "14815  92198195  92200804            0  7:92197498:-  ENSG00000001631.17_3   \n",
       "14816  92198602  92200804            0  7:92199243:-  ENSG00000001631.17_3   \n",
       "14817  92199253  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "14818  92199257  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "\n",
       "      gene_name_ref  \n",
       "14815         KRIT1  \n",
       "14816         KRIT1  \n",
       "14817         KRIT1  \n",
       "14818         KRIT1  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"].columns)\n",
    "novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"][[\"Start\", \"End\", \"atlas_filter\", \"Name\", \"le_id\", \"gene_name_ref\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'atlas_1_pred': 1511, 'atlas_max_datasets': 437, 'atlas_max_datasets_shortest': 315, 'motif_1_min': 849, 'motif_shortest_min': 56}\n"
     ]
    }
   ],
   "source": [
    "# now for all 3'utr extensions\n",
    "sel_dfs_d3utr = []\n",
    "sel_dict_d3utr = {\"atlas_1_pred\": [],\n",
    "                  \"atlas_max_datasets\": [],\n",
    "                  \"atlas_max_datasets_shortest\": [],\n",
    "                  \"motif_1_min\": [],\n",
    "                  \"motif_shortest_min\": []}\n",
    "\n",
    "\n",
    "for le_id, df in novel_le_df[novel_le_df.le_id.isin(le_id_d3utr)].groupby(\"le_id_quant\"):\n",
    "        out_df, decision_str = select_rep_site(df)\n",
    "        sel_dfs_d3utr.append(out_df)\n",
    "        sel_dict_d3utr[decision_str].append(le_id)\n",
    "\n",
    "\n",
    "sel_novel_le_df_d3utr = pd.concat(sel_dfs_d3utr)  \n",
    "sel_novel_le_d3utr = pr.PyRanges(sel_novel_le_df_d3utr)             \n",
    "    \n",
    "print({dec: len(ids) for dec, ids in sel_dict_d3utr.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        0  1\n",
      "33668  ENSG00000000003.15  2\n",
      "33669  ENSG00000000003.15  2\n",
      "31647  ENSG00000000419.14  2\n",
      "113    ENSG00000001461.17  6\n",
      "13794  ENSG00000001629.10  3\n",
      "...                   ... ..\n",
      "31482   ENSG00000285230.3  2\n",
      "33348   ENSG00000285756.2  4\n",
      "2663    ENSG00000286019.1  2\n",
      "29844   ENSG00000286104.1  4\n",
      "941     ENSG00000286219.2  2\n",
      "\n",
      "[2196 rows x 2 columns]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "33668    ENSG00000000003.15_1\n",
       "33669    ENSG00000000003.15_1\n",
       "31647    ENSG00000000419.14_1\n",
       "113      ENSG00000001461.17_5\n",
       "13794    ENSG00000001629.10_2\n",
       "                 ...         \n",
       "31482     ENSG00000285230.3_1\n",
       "33348     ENSG00000285756.2_3\n",
       "2663      ENSG00000286019.1_1\n",
       "29844     ENSG00000286104.1_3\n",
       "941       ENSG00000286219.2_1\n",
       "Name: 0, Length: 2196, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to select representative proximal site for each distal site, since quantification approach lumps together all annotated proximal sites (and compares them to extension)\n",
    "# since le_ids are annotated sequentially, know that immediately preceding le_number is the partner le_id\n",
    "d3utr_le_id_split = sel_novel_le_df_d3utr.le_id.str.split(\"_\", regex=False, expand=True)\n",
    "print(d3utr_le_id_split)\n",
    "d3utr_le_id_split[1] = d3utr_le_id_split[1].astype(int).subtract(1)\n",
    "\n",
    "# reconstruct le_id (corresponding to proximal le_id)\n",
    "d3utr_prox_le_id = d3utr_le_id_split[0].str.cat(d3utr_le_id_split[1].astype(str), sep=\"_\")\n",
    "d3utr_prox_le_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>109237373</td>\n",
       "      <td>109237730</td>\n",
       "      <td>ENST00000468588.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000031698.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238169</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>109237373</td>\n",
       "      <td>109237730</td>\n",
       "      <td>ENST00000468588.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000031698.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>210674287</td>\n",
       "      <td>210676290</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>210623670,210623670,210623670,210623670,210623...</td>\n",
       "      <td>210674287,210674287,210674287,210674287,210674...</td>\n",
       "      <td>ENST00000261458.8,ENST00000367009.2,ENST000003...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000054392.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6126</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>3853009</td>\n",
       "      <td>3854927</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>3854927</td>\n",
       "      <td>3860789</td>\n",
       "      <td>ENST00000381108.3</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000285756.2_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6127</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>3854348</td>\n",
       "      <td>3854927</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>3854927</td>\n",
       "      <td>3866921</td>\n",
       "      <td>ENST00000456563.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000285756.2_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6128</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>2866792</td>\n",
       "      <td>2867268</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>2865245,2865245</td>\n",
       "      <td>2866792,2866792</td>\n",
       "      <td>ENST00000250784.13,ENST00000477725.1</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000129824.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6129</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>2866792</td>\n",
       "      <td>2866956</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>2865245,2865245</td>\n",
       "      <td>2866792,2866792</td>\n",
       "      <td>ENST00000250784.13,ENST00000477725.1</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000129824.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6130</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>13704435</td>\n",
       "      <td>13705224</td>\n",
       "      <td>ENST00000284856.4</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000154620.6_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6131 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 109237730 | 109238182 | ...   |\n",
       "| chr1         | .          | exon         | 109237730 | 109238169 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 153587931 | 153588394 | ...   |\n",
       "| chrX         | .          | exon         | 153587931 | 153588394 | ...   |\n",
       "| chrX         | .          | exon         | 3853009   | 3854927   | ...   |\n",
       "| chrX         | .          | exon         | 3854348   | 3854927   | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6,131 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset for annotated last exons with 3'end proximal to 3'UTR extensions\n",
    "quant_uniq_le_d3utr_prox = quant_uniq_le.subset(lambda df: df.le_id.isin(set(d3utr_prox_le_id.values)))\n",
    "quant_uniq_le_d3utr_prox\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since proximal_le_id collapses any annotated last exon into a single isoform, will need to select a representative proximal polyA site\n",
    "# will arbitrarily select the most distal annotated site (in case of multiple)\n",
    "\n",
    "def select_rep_prox_site(df: pd.DataFrame):\n",
    "    '''_summary_\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : _type_\n",
    "        _description_\n",
    "    '''\n",
    "\n",
    "    if (df.Strand == \"+\").all():\n",
    "        # select the row with most distal 3'end (largest End coordinate)\n",
    "        return df[df[\"End\"] == df[\"End\"].max()]\n",
    "    \n",
    "    elif (df.Strand == \"-\").all():\n",
    "        # select the row with most distal 3'end (smallest Start coordinate)\n",
    "        return df[df[\"Start\"] == df[\"Start\"].min()]\n",
    "    \n",
    "    else:\n",
    "        raise ValueError(\"Strand column values must be all '+' or '-'\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>109237373</td>\n",
       "      <td>109237730</td>\n",
       "      <td>ENST00000468588.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000031698.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>210674287</td>\n",
       "      <td>210676296</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>210623670,210623670,210623670,210623670,210623...</td>\n",
       "      <td>210674287,210674287,210674287,210674287,210674...</td>\n",
       "      <td>ENST00000261458.8,ENST00000367009.2,ENST000003...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000054392.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>210674287</td>\n",
       "      <td>210676296</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>210623670,210623670,210623670,210623670,210623...</td>\n",
       "      <td>210674287,210674287,210674287,210674287,210674...</td>\n",
       "      <td>ENST00000261458.8,ENST00000367009.2,ENST000003...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000054392.13</td>\n",
       "      <td>HHAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000054392.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2077</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>153587595</td>\n",
       "      <td>153588394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.22033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000262919.8</td>\n",
       "      <td>CCNQ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000262919.8_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2078</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>153587595</td>\n",
       "      <td>153588394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.22033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000262919.8</td>\n",
       "      <td>CCNQ,CCNQ,CCNQ,CCNQ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000262919.8_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2079</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>3853009</td>\n",
       "      <td>3854927</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>3854927</td>\n",
       "      <td>3860789</td>\n",
       "      <td>ENST00000381108.3</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000285756.2</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000285756.2_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2080</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>2866792</td>\n",
       "      <td>2867268</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>2865245,2865245</td>\n",
       "      <td>2866792,2866792</td>\n",
       "      <td>ENST00000250784.13,ENST00000477725.1</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000129824.16</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000129824.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>13704435</td>\n",
       "      <td>13705224</td>\n",
       "      <td>ENST00000284856.4</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000154620.6_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2082 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 109237730 | 109238182 | ...   |\n",
       "| chr1         | .          | exon         | 210674287 | 210676296 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 46497726  | 46501335  | ...   |\n",
       "| chrX         | .          | exon         | 153587595 | 153588394 | ...   |\n",
       "| chrX         | .          | exon         | 153587595 | 153588394 | ...   |\n",
       "| chrX         | .          | exon         | 3853009   | 3854927   | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2,082 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_le_d3utr_prox = quant_uniq_le_d3utr_prox.apply(lambda df: df.groupby(\"le_id\").apply(select_rep_prox_site).reset_index(drop=True))\n",
    "sel_le_d3utr_prox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic 3'UTR extension-containing genes - 66\n",
      "Number of ns 3'UTR extension-containing genes - 811\n"
     ]
    }
   ],
   "source": [
    "# now want to combine the two with minimal info required for making maps\n",
    "# 3'end coordinates, gene name, le_id, site type, regulation status\n",
    "\n",
    "# define site_type columns\n",
    "sel_le_d3utr_prox.site_type = \"proximal\"\n",
    "sel_novel_le_d3utr.site_type = \"distal\"\n",
    "\n",
    "# define cryptic le_ids\n",
    "# define cryptic gene_ids\n",
    "dexseq_df_cryp = dexseq_df[(dexseq_df[\"padj\"].lt(0.05)) & (dexseq_df[\"mean_PPAU_base\"].le(0.1)) & (dexseq_df[\"delta_PPAU_treatment_control\"].ge(0.1))]\n",
    "cryp_le_ids = set(dexseq_df_cryp.le_id)\n",
    "cryp_gene_names = set(dexseq_df_cryp.gene_name)\n",
    "\n",
    "# background - ns in any dataset (to do this, get list of sig in any dataset)\n",
    "# get gene IDs with no regulated ALEs in any dataset\n",
    "# returns pd.Series\n",
    "ns_gene_ids = (dexseq_df.assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "               .groupby(\"groupID\")\n",
    "               [\"reg_status\"]\n",
    "               .sum()\n",
    "               .loc[lambda x: x == 0]\n",
    "             )\n",
    "\n",
    "# if one isoform of pair is called as cryptic, label both as cryptic\n",
    "cryp_gene_ids_d3utr = (pd.concat([sel_le_d3utr_prox.as_df()[[\"le_id\", \"gene_id\"]].drop_duplicates(),\n",
    "           sel_novel_le_d3utr.as_df()[[\"le_id\", \"gene_id_ref\"]].rename(columns={\"gene_id_ref\": \"gene_id\"}).drop_duplicates()],\n",
    "           ignore_index=True)\n",
    "           .assign(cryptic_status=lambda df: np.where(df[\"le_id\"].isin(cryp_le_ids), 1, 0))\n",
    "           # for each gene_name, if any isoforms (prox/distal are called as cryptic, annotated as cryptic)\n",
    "           .groupby(\"gene_id\")\n",
    "           [\"cryptic_status\"].sum()\n",
    "           .loc[lambda x: x > 0]\n",
    ")\n",
    "\n",
    "ns_gene_ids_d3utr = (pd.concat([sel_le_d3utr_prox.as_df()[[\"le_id\", \"gene_id\"]].drop_duplicates(),\n",
    "           sel_novel_le_d3utr.as_df()[[\"le_id\", \"gene_id_ref\"]].rename(columns={\"gene_id_ref\": \"gene_id\"}).drop_duplicates()],\n",
    "           ignore_index=True)\n",
    "           .assign(cryptic_status=lambda df: np.where(df[\"gene_id\"].isin(ns_gene_ids.index), 1, 0))\n",
    "           # for each gene_name, if any isoforms (prox/distal are called as cryptic, annotated as cryptic)\n",
    "           .groupby(\"gene_id\")\n",
    "           [\"cryptic_status\"].sum()\n",
    "           .loc[lambda x: x > 0]\n",
    ")\n",
    "\n",
    "\n",
    "print(f\"Number of cryptic 3'UTR extension-containing genes - {len(cryp_gene_ids_d3utr)}\")\n",
    "print(f\"Number of ns 3'UTR extension-containing genes - {len(ns_gene_ids_d3utr)}\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# sel_le_d3utr_prox\n",
    "# sel_novel_le_d3utr\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _define_cryptic_status(df, cryptic_ids: iter, ns_ids: iter, id_col: str) -> pd.Series:\n",
    "    '''_summary_\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : _type_\n",
    "        _description_\n",
    "    cryptic_ids : iter\n",
    "        _description_\n",
    "    ns_ids : iter\n",
    "        _description_\n",
    "    id_col : str\n",
    "        _description_\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.Series\n",
    "        _description_\n",
    "    '''\n",
    "\n",
    "    assert id_col in df.columns\n",
    "\n",
    "    choices = [\"cryptic\", \"background\"]\n",
    "    conds = [df[id_col].isin(cryptic_ids), # in cryptic list\n",
    "             ~df[id_col].isin(cryptic_ids) & df[id_col].isin(ns_ids) # not in cryptic list & in ns list\n",
    "             ]\n",
    "    \n",
    "    return pd.Series(np.select(conds, choices, default=\"NULL\"), index=df.index)\n",
    "\n",
    "\n",
    "sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_gene_ids_d3utr.index, ns_gene_ids_d3utr.index, id_col=\"gene_id\"))\n",
    "sel_novel_le_d3utr = sel_novel_le_d3utr.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_gene_ids_d3utr.index, ns_gene_ids_d3utr.index, id_col=\"gene_id_ref\"))\n",
    "\n",
    "\n",
    "\n",
    "# sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))\n",
    "# sel_novel_le_d3utr = sel_novel_le_d3utr.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id_ref.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct 'name' string with minimal annotation information\n",
    "# le_id|gene_name|site_type|cryptic_status\n",
    "\n",
    "sel_le_d3utr_prox = (sel_le_d3utr_prox\n",
    "                     .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"ref_gene_name\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"gene_name_common\",\n",
    "                          lambda df: pd.Series(np.where(df[\"gene_name\"].isna(),\n",
    "                                                        df[\"ref_gene_name\"],\n",
    "                                                        df[\"gene_name\"])))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_common\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    " )\n",
    "\n",
    "\n",
    "sel_novel_le_d3utr = (sel_novel_le_d3utr\n",
    "                      .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"Name\",\n",
    "         lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2163\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_5|NIPAL3|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_6|NIPAL3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28496533</td>\n",
       "      <td>28500572</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000204138.13_3|PHACTR4|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28497271</td>\n",
       "      <td>28500364</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000204138.13_2|PHACTR4|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28499419</td>\n",
       "      <td>28500364</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000204138.13_2|PHACTR4|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155276210</td>\n",
       "      <td>155278064</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155962.13_1|CLIC2|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2866792</td>\n",
       "      <td>2867268</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000129824.16_1|RPS4Y1|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1991</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2866792</td>\n",
       "      <td>2867772</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000129824.16_2|RPS4Y1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1992</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_1|TMSB4Y|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_2|TMSB4Y|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1994 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 24468985  | 24472976  | .          | +            | ...   |\n",
       "| chr1         | 24468985  | 24476667  | .          | +            | ...   |\n",
       "| chr1         | 28496533  | 28500572  | .          | +            | ...   |\n",
       "| chr1         | 28497271  | 28500364  | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrY         | 2866792   | 2867268   | .          | +            | ...   |\n",
       "| chrY         | 2866792   | 2867772   | .          | +            | ...   |\n",
       "| chrY         | 13705224  | 13706024  | .          | +            | ...   |\n",
       "| chrY         | 13705224  | 13706471  | .          | +            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 1,994 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate output bed file of proximal and distal PAS\n",
    "utr3_out_bed = pr.concat([sel_le_d3utr_prox[[\"Score\", \"Name\"]],\n",
    "                          sel_novel_le_d3utr[[\"Score\", \"Name\"]]]\n",
    "                          )\n",
    "\n",
    "print(len(utr3_out_bed))\n",
    "# drop duplicate entries\n",
    "utr3_out_bed = utr3_out_bed.apply(lambda df: df.drop_duplicates()).sort()\n",
    "\n",
    "utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"../processed/iclip_regions\"):\n",
    "    os.makedirs(\"../processed/iclip_regions\")\n",
    "    \n",
    "utr3_out_bed.to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_d3utr.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "utr3_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_d3utr.pas.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
