{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from define_helpers import select_rep_site, _define_cryptic_status, select_rep_prox_site\n",
    "from collections import Counter\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "cryptics_df = pd.read_csv(\"../../preprocessing/processed/cryptics_summary_all_events.tsv\", sep=\"\\t\")\n",
    "\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "simple_event_type\n",
       "spliced                  130806\n",
       "bleedthrough              41453\n",
       "distal_3utr_extension     37225\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel bleedthroughs or ALEs\n",
    "# le_id_spliced = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "# le_id_bleed = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "dexseq_df.simple_event_type.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "le_id_spliced = dexseq_df.loc[dexseq_df.simple_event_type == \"spliced\", \"le_id\"]\n",
    "le_id_bleed = dexseq_df.loc[dexseq_df.simple_event_type == \"bleedthrough\", \"le_id\"]\n",
    "le_id_d3utr = dexseq_df.loc[dexseq_df.simple_event_type == \"distal_3utr_extension\", \"le_id\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of spliced ALE events - 7978\n",
      "Number of bleedthrough ALE events - 3599\n",
      "Number of 3'UTR-ALE events - 3369\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of spliced ALE events - {len(set(le_id_spliced))}\")\n",
    "print(f\"Number of bleedthrough ALE events - {len(set(le_id_bleed))}\")\n",
    "print(f\"Number of 3'UTR-ALE events - {len(set(le_id_d3utr))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "simple_event_type\n",
       "spliced                               119\n",
       "distal_3utr_extension                 104\n",
       "bleedthrough                           55\n",
       "bleedthrough,spliced                   12\n",
       "bleedthrough,distal_3utr_extension      3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cryptics_df.simple_event_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "annot_status\n",
       "novel              204\n",
       "annotated           70\n",
       "annotated,novel     19\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cryptics_df.annot_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "annot_status     simple_event_type                 \n",
       "annotated        bleedthrough                           18\n",
       "                 bleedthrough,spliced                    4\n",
       "                 spliced                                48\n",
       "annotated,novel  bleedthrough,spliced                    6\n",
       "                 spliced                                13\n",
       "novel            bleedthrough                           37\n",
       "                 bleedthrough,distal_3utr_extension      3\n",
       "                 bleedthrough,spliced                    2\n",
       "                 distal_3utr_extension                 104\n",
       "                 spliced                                58\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cryptics_df[[\"annot_status\",\"simple_event_type\"]].value_counts(sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a set of novel cryptic IDs so can track where they go\n",
    "le_id_spliced_novel_cryp = set(cryptics_df[(cryptics_df[\"annot_status\"] == \"novel\") & (cryptics_df[\"simple_event_type\"] == \"spliced\")].le_id)\n",
    "le_id_bleed_novel_cryp = set(cryptics_df[(cryptics_df[\"annot_status\"] == \"novel\") & (cryptics_df[\"simple_event_type\"] == \"bleedthrough\")].le_id)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of spliced ALE events (after removing those with a novel extended 3'UTR) - 4854\n"
     ]
    }
   ],
   "source": [
    "# good idea to ensure don't consider spliced LEs that have a 3'UTR extension\n",
    "# UPDATE now need to select representative proximal site for each distal site, since quantification approach lumps together all annotated proximal sites (and compares them to extension) \n",
    "# since le_ids are annotated sequentially, know that immediately succeeding le_number is the partner le_id\n",
    "le_id_spliced_spl = le_id_spliced.str.split(\"_\", regex=False, expand=True)\n",
    "le_id_spliced_spl[1] = le_id_spliced_spl[1].astype(int).add(1)\n",
    "# reconstruct le_id (corresponding to theoretical distal 3'UTR extension le_id)\n",
    "le_id_spliced_d3utr = le_id_spliced_spl[0].str.cat(le_id_spliced_spl[1].astype(str), sep=\"_\")\n",
    "\n",
    "# now have theoretical distal 3'UTR le_ids, intersect with actual ids - any overlapping will be removed from downstream analysis\n",
    "spliced_d3utr_olap = le_id_spliced_d3utr.isin(le_id_d3utr.values)\n",
    "# print(f\"Number of spliced last exons with 3'UTR extension - {spliced_d3utr_olap.sum()}\")\n",
    "\n",
    "# now use mask to remove these ALEs\n",
    "le_id_spliced = le_id_spliced[~spliced_d3utr_olap]\n",
    "print(f\"Number of spliced ALE events (after removing those with a novel extended 3'UTR) - {len(set(le_id_spliced))}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add le_id used in downstream analysis\n",
    "novel_le = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(novel_le.subset(lambda df: df.le_id_quant.isin(le_id_spliced_novel_cryp)).le_id_quant))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'  Chromosome     Start       End Strand          Name\\n0       chr8  79616821  79617073      +  8:79617071:+\\n1       chr8  79616821  79617049      +  8:79617071:+\\n2       chr8  79616821  79617073      +  8:79617071:+\\n3       chr8  79616821  79617073      +  8:79617071:+\\n4       chr8  79616821  79617071      +  8:79617071:+\\n5       chr8  79616821  79617073      +  8:79617071:+\\n6       chr8  79616821  79623859      +  8:79623859:+'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.subset(lambda df: df.gene_name_ref.str.contains(\"STMN2\", regex=False))[[\"Name\"]].as_df().to_string()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of novel spliced intervals before selecting representative LEs - 6209\n",
      "Number of novel spliced intervals after selecting representative LEs - 1818\n",
      "{'atlas_1_pred': 571, 'atlas_max_datasets': 143, 'atlas_max_datasets_shortest': 98, 'motif_1_min': 321, 'motif_shortest_min': 160}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Select representative PAS for bleedthroughs and spliced events (novel)\n",
    "novel_le_rep_spliced, rep_choices_spliced = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_spliced))), id_col=\"le_id_quant\")\n",
    "\n",
    "\n",
    "print(f\"Number of novel spliced intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_spliced))))}\")\n",
    "print(f\"Number of novel spliced intervals after selecting representative LEs - {len(novel_le_rep_spliced)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_spliced.items()})\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# see if any missing ids\n",
    "missing_cryp_spl = le_id_spliced_novel_cryp - set(novel_le_rep_spliced.subset(lambda df: df.le_id_quant.isin(le_id_spliced_novel_cryp)).le_id_quant)\n",
    "len(missing_cryp_spl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bleedthrough intervals before selecting representative LEs - 5952\n",
      "Number of bleedthrough intervals after selecting representative LEs - 1714\n",
      "{'atlas_1_pred': 590, 'atlas_max_datasets': 152, 'atlas_max_datasets_shortest': 105, 'motif_1_min': 350, 'motif_shortest_min': 114}\n"
     ]
    }
   ],
   "source": [
    "# Select representative PAS for bleedthroughs and spliced events\n",
    "novel_le_rep_bleed, rep_choices_bleed = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_bleed))), id_col=\"le_id_quant\")\n",
    "\n",
    "print(f\"Number of bleedthrough intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_bleed))))}\")\n",
    "print(f\"Number of bleedthrough intervals after selecting representative LEs - {len(novel_le_rep_bleed)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_bleed.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# see if any missing ids\n",
    "missing_cryp_bleed = le_id_bleed_novel_cryp - set(novel_le_rep_bleed.subset(lambda df: df.le_id_quant.isin(le_id_bleed_novel_cryp)).le_id_quant)\n",
    "len(missing_cryp_bleed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic ALE-containing genes - 283\n",
      "Number of cryptic ALE isoforms - 283\n",
      "Number of ns ALE-containing genes - 2937\n",
      "Number of ns ALE isoforms - 6365\n"
     ]
    }
   ],
   "source": [
    "# now want to generate an ID with minimal info required for making maps\n",
    "# 3'end coordinates/last exon, gene name, le_id, site type, regulation status\n",
    "# site_type - spliced/bleedthrough\n",
    "# regulation_status - cryptic/background\n",
    "\n",
    "\n",
    "# define cryptic le_ids\n",
    "# define cryptic gene_ids\n",
    "cryp_le_ids = set(cryptics_df.le_id)\n",
    "cryp_gene_names = set(cryptics_df.gene_name)\n",
    "\n",
    "\n",
    "# define cryptic & background le_ids\n",
    "# background - ns in any dataset (to do this, get list of sig in any dataset)\n",
    "# get gene IDs with no regulated ALEs in any dataset\n",
    "# returns pd.Series (index = groupIDs)\n",
    "ns_gene_ids = (dexseq_df.assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "               .groupby(\"groupID\")\n",
    "               [\"reg_status\"]\n",
    "               .sum()\n",
    "               .loc[lambda x: x == 0]\n",
    "             )\n",
    "\n",
    "# set of le_ids that belong to genes with no significant ALEs (is this too conservative?)\n",
    "ns_le_ids = set(dexseq_df.loc[dexseq_df[\"groupID\"].isin(ns_gene_ids.index), \"le_id\"])\n",
    "\n",
    "\n",
    "print(f\"Number of cryptic ALE-containing genes - {len(cryp_gene_names)}\")\n",
    "print(f\"Number of cryptic ALE isoforms - {len(cryp_gene_names)}\")\n",
    "\n",
    "print(f\"Number of ns ALE-containing genes - {len(set(ns_gene_ids.index))}\")\n",
    "print(f\"Number of ns ALE isoforms - {len(set(ns_le_ids))}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          1139\n",
       "background     508\n",
       "cryptic         67\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for novel bleedthroughs\n",
    "novel_le_rep_bleed = novel_le_rep_bleed.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id_quant\"))\n",
    "novel_le_rep_bleed.cryptic_status.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "background    404\n",
       "cryptic        75\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for novel spliced events\n",
    "novel_le_rep_spliced = novel_le_rep_spliced.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id_quant\"))\n",
    "novel_le_rep_spliced.as_df().drop_duplicates(subset=\"le_id_quant\").cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct 'name' string with minimal annotation information\n",
    "# le_id|gene_name|site_type|cryptic_status\n",
    "\n",
    "novel_le_rep_spliced = (novel_le_rep_spliced\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"spliced\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id_quant\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    ")\n",
    "\n",
    "novel_le_rep_bleed = (novel_le_rep_bleed\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"bleedthrough\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id_quant\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117069.15_2</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61824444</td>\n",
       "      <td>61825501</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132849.22_1</td>\n",
       "      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245464258</td>\n",
       "      <td>245471621</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162849.16_2</td>\n",
       "      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112540178</td>\n",
       "      <td>112542090</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000007341.19_4</td>\n",
       "      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243612605</td>\n",
       "      <td>243613034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000117020.19_2</td>\n",
       "      <td>ENSG00000117020.19_2|AKT3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr21</td>\n",
       "      <td>45792478</td>\n",
       "      <td>45794699</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000183570.17_3</td>\n",
       "      <td>ENSG00000183570.17_3|PCBP3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>chr22</td>\n",
       "      <td>37464524</td>\n",
       "      <td>37465718</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100060.18_2</td>\n",
       "      <td>ENSG00000100060.18_2|MFNG|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>chrX</td>\n",
       "      <td>102721363</td>\n",
       "      <td>102724864</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198908.12_1</td>\n",
       "      <td>ENSG00000198908.12_1|BHLHB9|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98679426</td>\n",
       "      <td>98679978</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000281566.3_1</td>\n",
       "      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18_1</td>\n",
       "      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 76871267  | 76871821  | +            | ...   |\n",
       "| chr1         | 61824444  | 61825501  | +            | ...   |\n",
       "| chr1         | 245464258 | 245471621 | +            | ...   |\n",
       "| chr1         | 112540178 | 112542090 | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr22        | 37464524  | 37465718  | -            | ...   |\n",
       "| chrX         | 102721363 | 102724864 | +            | ...   |\n",
       "| chrX         | 98679426  | 98679978  | +            | ...   |\n",
       "| chrX         | 17835910  | 17837395  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 112 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: le_id, Name"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le_rep_spliced.subset(lambda df: df.cryptic_status == \"cryptic\")[[\"le_id\",\"Name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAPA.CTRL-2.2002.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PAPA.ctrl_ctrl_4.2339.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.1123.3</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "      <td>ENSG00000116128.12_2|BCL9|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.1061.3</td>\n",
       "      <td>ENSG00000116830.12_2</td>\n",
       "      <td>ENSG00000116830.12_2|TTF2|bleedthrough|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>PAPA.TDP-2.25089.1</td>\n",
       "      <td>ENSG00000182518.14_1</td>\n",
       "      <td>ENSG00000182518.14_1|FAM104B|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>PAPA.doxconc_NT_0_3.19568.4</td>\n",
       "      <td>ENSG00000196459.15_1</td>\n",
       "      <td>ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>PAPA.Cont-D_S4.20482.2</td>\n",
       "      <td>ENSG00000197021.9_2</td>\n",
       "      <td>ENSG00000197021.9_2|EOLA2|bleedthrough|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>575 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      transcript_id                 le_id  \\\n",
       "0                 PAPA.TDP-1.2120.4  ENSG00000082497.12_1   \n",
       "1                PAPA.CTRL-2.2002.3  ENSG00000082497.12_1   \n",
       "2           PAPA.ctrl_ctrl_4.2339.3  ENSG00000082497.12_1   \n",
       "3    PAPA.doxconc_DOX_0075_2.1123.3  ENSG00000116128.12_2   \n",
       "4    PAPA.doxconc_DOX_0075_1.1061.3  ENSG00000116830.12_2   \n",
       "..                              ...                   ...   \n",
       "570         PAPA.TDP43-G_S7.20350.2  ENSG00000180182.11_3   \n",
       "571              PAPA.TDP-2.25089.1  ENSG00000182518.14_1   \n",
       "572     PAPA.doxconc_NT_0_3.19568.4  ENSG00000196459.15_1   \n",
       "573          PAPA.Cont-D_S4.20482.2   ENSG00000197021.9_2   \n",
       "574      PAPA.TDP43_ctrl_2.26896.10   ENSG00000213468.7_1   \n",
       "\n",
       "                                                  Name  \n",
       "0    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "1    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "2    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "3       ENSG00000116128.12_2|BCL9|bleedthrough|cryptic  \n",
       "4    ENSG00000116830.12_2|TTF2|bleedthrough|background  \n",
       "..                                                 ...  \n",
       "570    ENSG00000180182.11_3|MED14|bleedthrough|cryptic  \n",
       "571  ENSG00000182518.14_1|FAM104B|bleedthrough|back...  \n",
       "572  ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...  \n",
       "573  ENSG00000197021.9_2|EOLA2|bleedthrough|background  \n",
       "574     ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic  \n",
       "\n",
       "[575 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for spliced events, want to report the last exon coordinates (also split into 5'end & 3'end)\n",
    "# for this purpose, current coordiantes (from input last exons) is sufficient.\n",
    "# but for bleedthrough events, this will contain the complete last exon. So need to return to quant last exons to get the unique regions. \n",
    "\n",
    "# novel_le_rep_bleed - extract transcript_id (i.e. represents selected isoform) & downstream le_id (le_id_quant)\n",
    "novel_rep_bleed_tx2le = novel_le_rep_bleed.as_df()[[\"transcript_id\", \"le_id_quant\", \"Name\"]].rename(columns={\"le_id_quant\": \"le_id\"})\n",
    "novel_rep_bleed_tx2le\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atlas_max_datasets\n"
     ]
    }
   ],
   "source": [
    "for k,v in rep_choices_bleed.items():\n",
    "    if \"ENSG00000082497.12_1\" in v:\n",
    "        print(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210241557</td>\n",
       "      <td>210245767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246847</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246813</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.CTRL-2.2002.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.2339.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147622531</td>\n",
       "      <td>147623356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2|BCL9|bleedthrough|cryptic</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.1123.3</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX</td>\n",
       "      <td>55142867</td>\n",
       "      <td>55146180</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182518.14_1|FAM104B|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-2.25089.1</td>\n",
       "      <td>ENSG00000182518.14_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149931286</td>\n",
       "      <td>149932588</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197021.9_2|EOLA2|bleedthrough|background</td>\n",
       "      <td>PAPA.Cont-D_S4.20482.2</td>\n",
       "      <td>ENSG00000197021.9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825221</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX</td>\n",
       "      <td>13715131</td>\n",
       "      <td>13716003</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...</td>\n",
       "      <td>PAPA.doxconc_NT_0_3.19568.4</td>\n",
       "      <td>ENSG00000196459.15_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>649 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 210241557 | 210245767 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246847 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246813 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246804 | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 55142867  | 55146180  | .          | -            | ...   |\n",
       "| chrX         | 149931286 | 149932588 | .          | -            | ...   |\n",
       "| chrX         | 131823777 | 131825221 | .          | -            | ...   |\n",
       "| chrX         | 13715131  | 13716003  | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 649 rows and 8 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, transcript_id, le_id"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to representative bleedthrough txs, joining in Name information\n",
    "# 5'coord = 1st cooridnate of intron (i.e. where spliced occurs)\n",
    "# 3'coord = final coordinate of last exon\n",
    "novel_le_rep_bleed_quant = quant_uniq_le.apply(lambda df: df.merge(novel_rep_bleed_tx2le, on=[\"transcript_id\", \"le_id\"], how=\"inner\"))\n",
    "\n",
    "novel_le_rep_bleed_quant[[\"Score\", \"Name\", \"transcript_id\", \"le_id\"]].drop_duplicate_positions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Repeat for annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of annotated bleedthrough last exons (assessed by DEXSeq) - 2288\n",
      "Number of annotated spliced last exons (assessed by DEXSeq) - 3561\n"
     ]
    }
   ],
   "source": [
    "# get le_ids in which all transcripts are annotated\n",
    "annotated_le_ids = dexseq_df.groupby(\"le_id\").filter(lambda df: (df.annot_status == \"annotated\").all()).le_id\n",
    "le_id_bleed_annot = set(le_id_bleed).intersection(set(annotated_le_ids))\n",
    "le_id_spliced_annot = set(le_id_spliced).intersection(set(annotated_le_ids))\n",
    "print(f\"Number of annotated bleedthrough last exons (assessed by DEXSeq) - {len(le_id_bleed_annot)}\")\n",
    "print(f\"Number of annotated spliced last exons (assessed by DEXSeq) - {len(le_id_spliced_annot)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset quantification GTF for spliecd last exons \n",
    "ref_le_spliced = quant_uniq_le.subset(lambda df: df.le_id.isin(le_id_spliced_annot))\n",
    "# select representative isoform for each last exon ID (most distal 3'end)\n",
    "ref_le_spliced = select_rep_prox_site(ref_le_spliced, \"le_id\")\n",
    "\n",
    "# subset quantification GTF for bleedthrough last exons \n",
    "ref_le_bleed = quant_uniq_le.subset(lambda df: df.le_id.isin(le_id_bleed_annot))\n",
    "# select representative isoform for each last exon ID (most distal 3'end)\n",
    "ref_le_bleed = select_rep_prox_site(ref_le_bleed, \"le_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          2275\n",
       "background    2176\n",
       "cryptic         39\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for annotatesd spliced last exons\n",
    "ref_le_spliced = ref_le_spliced.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id\"))\n",
    "ref_le_spliced.cryptic_status.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "background    1382\n",
       "NULL          1261\n",
       "cryptic         24\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for annotatesd bleedthrough last exons\n",
    "ref_le_bleed = ref_le_bleed.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id\"))\n",
    "ref_le_bleed.cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109214546</td>\n",
       "      <td>109215153</td>\n",
       "      <td>+</td>\n",
       "      <td>SARS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69873466</td>\n",
       "      <td>69875004</td>\n",
       "      <td>+</td>\n",
       "      <td>LRRC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6634019</td>\n",
       "      <td>6635586</td>\n",
       "      <td>+</td>\n",
       "      <td>THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7779625</td>\n",
       "      <td>7781432</td>\n",
       "      <td>+</td>\n",
       "      <td>VAMP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243305011</td>\n",
       "      <td>243306548</td>\n",
       "      <td>+</td>\n",
       "      <td>SDCCAG8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4485</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12859278</td>\n",
       "      <td>12860839</td>\n",
       "      <td>+</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4486</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4487</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4488</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4489</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13249882</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4490 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | ref_gene_name   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+-----------------|\n",
       "| chr1         | 109214546 | 109215153 | +            | SARS1           |\n",
       "| chr1         | 69873466  | 69875004  | +            | LRRC7           |\n",
       "| chr1         | 6634019   | 6635586   | +            | THAP3           |\n",
       "| chr1         | 7779625   | 7781432   | +            | VAMP3           |\n",
       "| ...          | ...       | ...       | ...          | ...             |\n",
       "| chrY         | 18932316  | 18932841  | -            | TTTY14          |\n",
       "| chrY         | 18872500  | 18872834  | -            | TTTY14          |\n",
       "| chrY         | 18872500  | 18872834  | -            | TTTY14          |\n",
       "| chrY         | 13248378  | 13249882  | -            | UTY             |\n",
       "+--------------+-----------+-----------+--------------+-----------------+\n",
       "Stranded PyRanges object has 4,490 rows and 5 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_le_spliced[[\"ref_gene_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign Name field with minimal info\n",
    "ref_le_spliced = (ref_le_spliced\n",
    "  .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"spliced\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")\n",
    "\n",
    "ref_le_bleed = (ref_le_bleed\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    ".assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"bleedthrough\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "event_type\n",
      "last_exon_spliced                           1190\n",
      "first_exon_spliced                           525\n",
      "internal_exon_spliced                        247\n",
      "internal_exon_extension                      106\n",
      "first_exon_spliced,internal_exon_spliced     106\n",
      "first_exon_extension                          41\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "      <th>cryptic_status</th>\n",
       "      <th>gene_name_ref</th>\n",
       "      <th>site_type</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>39753286</td>\n",
       "      <td>39756769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000084072.17_2</td>\n",
       "      <td>background</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000084072.17_2|PPIE|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>39763688</td>\n",
       "      <td>39763914</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000084072.17_3</td>\n",
       "      <td>background</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000084072.17_3|PPIE|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>170736047</td>\n",
       "      <td>170739421</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116132.12</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000116132.12</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116132.12_2</td>\n",
       "      <td>background</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116132.12_2|PRRX1|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>110232597</td>\n",
       "      <td>110249124</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116396.15</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>...</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000116396.15</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116396.15_2</td>\n",
       "      <td>background</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116396.15_2|KCNC4|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>32855923</td>\n",
       "      <td>32858875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116497.18</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000116497.18</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000116497.18_3</td>\n",
       "      <td>background</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116497.18_3|S100PBP|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>72274254</td>\n",
       "      <td>72274435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198034.11</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000198034.11</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000198034.11_1</td>\n",
       "      <td>background</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000198034.11_1|RPS4X|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1021</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130981589</td>\n",
       "      <td>130981856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000228659.1</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000228659.1</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000228659.1_3</td>\n",
       "      <td>background</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000228659.1_3|LINC01201|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>149527590</td>\n",
       "      <td>149528005</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7_1</td>\n",
       "      <td>background</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7_1|EOLA1-DT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1023</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000176728.10_3</td>\n",
       "      <td>background</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000176728.10_3|TTTY14|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000176728.10_3</td>\n",
       "      <td>background</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000176728.10_3|TTTY14|spliced|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1025 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +21   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 39753286  | 39756769  | ...   |\n",
       "| chr1         | .          | exon         | 39763688  | 39763914  | ...   |\n",
       "| chr1         | .          | exon         | 170736047 | 170739421 | ...   |\n",
       "| chr1         | .          | exon         | 110232597 | 110249124 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | .          | exon         | 72274254  | 72274435  | ...   |\n",
       "| chrX         | .          | exon         | 130981589 | 130981856 | ...   |\n",
       "| chrX         | .          | exon         | 149527590 | 149528005 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,025 rows and 26 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "21 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 14 more.)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since last_exon_spliced annotations includes most distal last exon of a gene, let's compare these ALEs to non-terminal ALEs (i.e. there is a competing splicign event)\n",
    "print(ref_le_spliced.event_type.value_counts())\n",
    "\n",
    "ref_le_spliced = ref_le_spliced.subset(lambda df: df[\"event_type\"].ne(\"last_exon_spliced\"))\n",
    "ref_le_spliced\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61824444</td>\n",
       "      <td>61825501</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245464258</td>\n",
       "      <td>245471621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616614</td>\n",
       "      <td>1619210</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112540178</td>\n",
       "      <td>112542090</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>chr22</td>\n",
       "      <td>37464524</td>\n",
       "      <td>37465718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100060.18_2|MFNG|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>chrX</td>\n",
       "      <td>102721363</td>\n",
       "      <td>102724864</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198908.12_1|BHLHB9|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98679426</td>\n",
       "      <td>98679978</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 76871267  | 76871821  | .          | +            | ...   |\n",
       "| chr1         | 61824444  | 61825501  | .          | +            | ...   |\n",
       "| chr1         | 245464258 | 245471621 | .          | +            | ...   |\n",
       "| chr1         | 1616614   | 1619210   | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 102721363 | 102724864 | .          | +            | ...   |\n",
       "| chrX         | 98679426  | 98679978  | .          | +            | ...   |\n",
       "| chrX         | 17835910  | 17837395  | .          | -            | ...   |\n",
       "| chrX         | 108267641 | 108269152 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 133 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined output BEDs\n",
    "spliced_out_bed = pr.concat([novel_le_rep_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions(),\n",
    "                             ref_le_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions()])\n",
    "\n",
    "bleed_out_bed = pr.concat([novel_le_rep_bleed_quant[[\"Score\", \"Name\"]].drop_duplicate_positions(),\n",
    "                             ref_le_bleed[[\"Score\", \"Name\"]].drop_duplicate_positions()])\n",
    "\n",
    "spliced_out_bed.subset(lambda df: df.Name.str.contains(\"cryptic$\"))\n",
    "\n",
    "# novel_le_rep_bleed_quant[[\"Score\", \"Name\"]].drop_duplicate_positions()\n",
    "# novel_le_rep_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions()\n",
    "\n",
    "\n",
    "# generate output bed file of proximal and distal PAS\n",
    "# utr3_out_bed = pr.concat([sel_le_d3utr_prox[[\"Score\", \"Name\"]],\n",
    "#                           sel_novel_le_d3utr[[\"Score\", \"Name\"]]]\n",
    "#                           )\n",
    "\n",
    "# print(len(utr3_out_bed))\n",
    "# # drop duplicate entries\n",
    "# utr3_out_bed = utr3_out_bed.apply(lambda df: df.drop_duplicates()).sort()\n",
    "\n",
    "# utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617073</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617049</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   |     Start |       End | Score      | Strand       | +1    |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr8         |  79616821 |  79617073 | .          | +            | ...   |\n",
       "| chr8         |  79616821 |  79617049 | .          | +            | ...   |\n",
       "| chr8         |  79616821 |  79617071 | .          | +            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spliced_out_bed.subset(lambda df: df.Name.str.contains(\"STMN2\", regex=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({None: 526, 'atlas_1_pred': 62, 'atlas_max_datasets': 32})"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter([k if le_id.split(\"|\")[0] in v else None for le_id in spliced_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index for k,v in rep_choices_spliced.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147622531</td>\n",
       "      <td>147623356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2|BCL9|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>156139102</td>\n",
       "      <td>156140091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160789.24_11|LMNA|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149834134</td>\n",
       "      <td>149835093</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000270882.3_2|H4C14|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45013791</td>\n",
       "      <td>45013889</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000126088.14_1|UROD|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chr21</td>\n",
       "      <td>42849225</td>\n",
       "      <td>42850040</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000160193.12_1|WDR4|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chrX</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23_3|PCDH11X|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>chrX</td>\n",
       "      <td>107087378</td>\n",
       "      <td>107088435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000089682.17_2|RBM41|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825221</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 147622531 | 147623356 | .          | +            | ...   |\n",
       "| chr1         | 21453372  | 21457150  | .          | +            | ...   |\n",
       "| chr1         | 156139102 | 156140091 | .          | +            | ...   |\n",
       "| chr1         | 149834134 | 149835093 | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 91882906  | 91891321  | .          | +            | ...   |\n",
       "| chrX         | 107087378 | 107088435 | .          | -            | ...   |\n",
       "| chrX         | 40653641  | 40654363  | .          | -            | ...   |\n",
       "| chrX         | 131823777 | 131825221 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 92 rows and 6 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bleed_out_bed.subset(lambda df: df.Name.str.contains(\"cryptic$\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({None: 728,\n",
       "         'atlas_1_pred': 23,\n",
       "         'motif_shortest_min': 13,\n",
       "         'atlas_max_datasets': 9,\n",
       "         'motif_1_min': 2})"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter([k if le_id.split(\"|\")[0] in v else None for le_id in bleed_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index for k,v in rep_choices_spliced.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"../processed/iclip_regions\"):\n",
    "    os.makedirs(\"../processed/iclip_regions\")\n",
    "\n",
    "#### TEMPORARY - work out why duplicate regions/multiple PAS coords for some last exons. To prevent duplciats slect longest per interval\n",
    "spliced_out_bed = select_rep_prox_site(spliced_out_bed, \"Name\")\n",
    "bleed_out_bed = select_rep_prox_site(bleed_out_bed, \"Name\")\n",
    "\n",
    "spliced_out_bed.to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "spliced_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.pas.bed\")\n",
    "# also first coordinate of exon (i.e. interval ending at this coordinate = splice site end, or substract 1 (strand-aware) to get final nucleotide of splice site)\n",
    "spliced_out_bed.five_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.le_start.bed\")\n",
    "\n",
    "bleed_out_bed.to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "bleed_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.pas.bed\")\n",
    "# also first coordinate of exon (i.e. 1st nucleotide of intron\n",
    "bleed_out_bed.five_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.le_start.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
