{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from define_helpers import select_rep_site, _define_cryptic_status, select_rep_prox_site\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "simple_event_type\n",
       "spliced                  130806\n",
       "bleedthrough              41453\n",
       "distal_3utr_extension     37225\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel bleedthroughs or ALEs\n",
    "# le_id_spliced = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "# le_id_bleed = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "dexseq_df.simple_event_type.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "le_id_spliced = dexseq_df.loc[dexseq_df.simple_event_type == \"spliced\", \"le_id\"]\n",
    "le_id_bleed = dexseq_df.loc[dexseq_df.simple_event_type == \"bleedthrough\", \"le_id\"]\n",
    "le_id_d3utr = dexseq_df.loc[dexseq_df.simple_event_type == \"distal_3utr_extension\", \"le_id\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_name</th>\n",
       "      <th>binID</th>\n",
       "      <th>groupID</th>\n",
       "      <th>featureID</th>\n",
       "      <th>exonBaseMean</th>\n",
       "      <th>dispersion</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>UsageCoefficient_base</th>\n",
       "      <th>...</th>\n",
       "      <th>annot_status</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mean_PPAU_base</th>\n",
       "      <th>mean_PPAU_treatment</th>\n",
       "      <th>delta_PPAU_treatment_control</th>\n",
       "      <th>simple_event_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>685.073107</td>\n",
       "      <td>0.013534</td>\n",
       "      <td>127.151527</td>\n",
       "      <td>1.721175e-29</td>\n",
       "      <td>3.384311e-27</td>\n",
       "      <td>10.215189</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.065536</td>\n",
       "      <td>0.226421</td>\n",
       "      <td>0.160885</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>1635.679342</td>\n",
       "      <td>0.013184</td>\n",
       "      <td>130.260073</td>\n",
       "      <td>3.594567e-30</td>\n",
       "      <td>7.236205e-28</td>\n",
       "      <td>17.008705</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.832218</td>\n",
       "      <td>0.649472</td>\n",
       "      <td>-0.182746</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16839</th>\n",
       "      <td>brown_shsy5y</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>646.630524</td>\n",
       "      <td>0.070550</td>\n",
       "      <td>0.725662</td>\n",
       "      <td>3.942930e-01</td>\n",
       "      <td>8.831562e-01</td>\n",
       "      <td>13.366541</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.181492</td>\n",
       "      <td>0.147692</td>\n",
       "      <td>-0.033800</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16840</th>\n",
       "      <td>brown_shsy5y</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>1583.366570</td>\n",
       "      <td>0.069263</td>\n",
       "      <td>0.738940</td>\n",
       "      <td>3.900008e-01</td>\n",
       "      <td>8.809464e-01</td>\n",
       "      <td>16.186991</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.789384</td>\n",
       "      <td>0.806265</td>\n",
       "      <td>0.016880</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25677</th>\n",
       "      <td>brown_skndz</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>123.423529</td>\n",
       "      <td>0.990241</td>\n",
       "      <td>0.184085</td>\n",
       "      <td>6.678856e-01</td>\n",
       "      <td>8.576725e-01</td>\n",
       "      <td>7.870387</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.091220</td>\n",
       "      <td>0.082138</td>\n",
       "      <td>-0.009082</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25678</th>\n",
       "      <td>brown_skndz</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>1382.957785</td>\n",
       "      <td>0.017284</td>\n",
       "      <td>3.819419</td>\n",
       "      <td>5.066181e-02</td>\n",
       "      <td>2.041506e-01</td>\n",
       "      <td>18.595389</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.459231</td>\n",
       "      <td>0.371105</td>\n",
       "      <td>-0.088125</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25679</th>\n",
       "      <td>brown_skndz</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>734.474029</td>\n",
       "      <td>0.013301</td>\n",
       "      <td>5.624554</td>\n",
       "      <td>1.771058e-02</td>\n",
       "      <td>1.009837e-01</td>\n",
       "      <td>14.791017</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.436124</td>\n",
       "      <td>0.489640</td>\n",
       "      <td>0.053516</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34381</th>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>31.461331</td>\n",
       "      <td>0.988492</td>\n",
       "      <td>3.663168</td>\n",
       "      <td>5.562777e-02</td>\n",
       "      <td>1.940285e-01</td>\n",
       "      <td>4.173276</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.019720</td>\n",
       "      <td>0.072751</td>\n",
       "      <td>0.053031</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34382</th>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>232.592245</td>\n",
       "      <td>0.020956</td>\n",
       "      <td>177.904729</td>\n",
       "      <td>1.389688e-40</td>\n",
       "      <td>7.563611e-38</td>\n",
       "      <td>6.775476</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.062796</td>\n",
       "      <td>0.271946</td>\n",
       "      <td>0.209150</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34383</th>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>469.902739</td>\n",
       "      <td>0.024516</td>\n",
       "      <td>174.525798</td>\n",
       "      <td>7.599079e-40</td>\n",
       "      <td>3.877430e-37</td>\n",
       "      <td>9.311451</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.908924</td>\n",
       "      <td>0.596359</td>\n",
       "      <td>-0.312565</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47455</th>\n",
       "      <td>klim_i3_motor</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>144.835749</td>\n",
       "      <td>0.019029</td>\n",
       "      <td>3.054360</td>\n",
       "      <td>8.052074e-02</td>\n",
       "      <td>5.581589e-01</td>\n",
       "      <td>7.224955</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.066292</td>\n",
       "      <td>0.080668</td>\n",
       "      <td>0.014376</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47456</th>\n",
       "      <td>klim_i3_motor</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>611.086934</td>\n",
       "      <td>0.018797</td>\n",
       "      <td>3.084681</td>\n",
       "      <td>7.903278e-02</td>\n",
       "      <td>5.548139e-01</td>\n",
       "      <td>10.885618</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.847995</td>\n",
       "      <td>0.836036</td>\n",
       "      <td>-0.011959</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55826</th>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>1278.572167</td>\n",
       "      <td>0.014028</td>\n",
       "      <td>318.124993</td>\n",
       "      <td>3.709556e-71</td>\n",
       "      <td>4.794186e-69</td>\n",
       "      <td>16.295375</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.084287</td>\n",
       "      <td>0.306735</td>\n",
       "      <td>0.222448</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55827</th>\n",
       "      <td>seddighi_i3_cortical</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>3394.001584</td>\n",
       "      <td>0.013990</td>\n",
       "      <td>318.966062</td>\n",
       "      <td>2.432856e-71</td>\n",
       "      <td>3.191833e-69</td>\n",
       "      <td>24.386196</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.863467</td>\n",
       "      <td>0.668574</td>\n",
       "      <td>-0.194894</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67510</th>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>105.534632</td>\n",
       "      <td>0.005586</td>\n",
       "      <td>0.008014</td>\n",
       "      <td>9.286665e-01</td>\n",
       "      <td>9.841491e-01</td>\n",
       "      <td>7.657459</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.057043</td>\n",
       "      <td>0.074931</td>\n",
       "      <td>0.017888</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67511</th>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>988.462295</td>\n",
       "      <td>0.003196</td>\n",
       "      <td>442.611427</td>\n",
       "      <td>2.924845e-98</td>\n",
       "      <td>1.189716e-95</td>\n",
       "      <td>14.900440</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.137892</td>\n",
       "      <td>0.394443</td>\n",
       "      <td>0.256551</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67512</th>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>951.766052</td>\n",
       "      <td>0.002682</td>\n",
       "      <td>492.797690</td>\n",
       "      <td>3.508085e-109</td>\n",
       "      <td>1.664781e-106</td>\n",
       "      <td>19.629855</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.788094</td>\n",
       "      <td>0.522136</td>\n",
       "      <td>-0.265958</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79770</th>\n",
       "      <td>zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>71.399225</td>\n",
       "      <td>0.119240</td>\n",
       "      <td>1.893448</td>\n",
       "      <td>1.688136e-01</td>\n",
       "      <td>4.246955e-01</td>\n",
       "      <td>5.322357</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.054078</td>\n",
       "      <td>0.086728</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79771</th>\n",
       "      <td>zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>594.115006</td>\n",
       "      <td>0.012225</td>\n",
       "      <td>17.124480</td>\n",
       "      <td>3.500816e-05</td>\n",
       "      <td>5.101975e-04</td>\n",
       "      <td>11.991790</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.222966</td>\n",
       "      <td>0.312574</td>\n",
       "      <td>0.089608</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79772</th>\n",
       "      <td>zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>753.314063</td>\n",
       "      <td>0.009719</td>\n",
       "      <td>27.125775</td>\n",
       "      <td>1.906391e-07</td>\n",
       "      <td>4.859659e-06</td>\n",
       "      <td>13.950307</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.680860</td>\n",
       "      <td>0.559108</td>\n",
       "      <td>-0.121752</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93191</th>\n",
       "      <td>zanovello_shsy5y_curve_00125</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>76.428022</td>\n",
       "      <td>0.073518</td>\n",
       "      <td>0.012217</td>\n",
       "      <td>9.119869e-01</td>\n",
       "      <td>9.997851e-01</td>\n",
       "      <td>5.786969</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.082612</td>\n",
       "      <td>0.076246</td>\n",
       "      <td>-0.006366</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93192</th>\n",
       "      <td>zanovello_shsy5y_curve_00125</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>532.677509</td>\n",
       "      <td>0.006241</td>\n",
       "      <td>1.125615</td>\n",
       "      <td>2.887127e-01</td>\n",
       "      <td>9.671906e-01</td>\n",
       "      <td>11.893295</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.245697</td>\n",
       "      <td>0.263737</td>\n",
       "      <td>0.018040</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93193</th>\n",
       "      <td>zanovello_shsy5y_curve_00125</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>750.349881</td>\n",
       "      <td>0.007609</td>\n",
       "      <td>0.567313</td>\n",
       "      <td>4.513287e-01</td>\n",
       "      <td>9.730105e-01</td>\n",
       "      <td>13.462503</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.638713</td>\n",
       "      <td>0.618608</td>\n",
       "      <td>-0.020105</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99663</th>\n",
       "      <td>zanovello_shsy5y_curve_00187</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>76.565383</td>\n",
       "      <td>0.096247</td>\n",
       "      <td>0.009507</td>\n",
       "      <td>9.223256e-01</td>\n",
       "      <td>9.830233e-01</td>\n",
       "      <td>1.893856</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.082906</td>\n",
       "      <td>0.081571</td>\n",
       "      <td>-0.001334</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99664</th>\n",
       "      <td>zanovello_shsy5y_curve_00187</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>563.779051</td>\n",
       "      <td>0.006498</td>\n",
       "      <td>11.575263</td>\n",
       "      <td>6.683498e-04</td>\n",
       "      <td>1.776802e-02</td>\n",
       "      <td>2.696905</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.245501</td>\n",
       "      <td>0.316147</td>\n",
       "      <td>0.070646</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99665</th>\n",
       "      <td>zanovello_shsy5y_curve_00187</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>694.466108</td>\n",
       "      <td>0.003179</td>\n",
       "      <td>17.797581</td>\n",
       "      <td>2.456955e-05</td>\n",
       "      <td>1.282166e-03</td>\n",
       "      <td>2.869866</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.638621</td>\n",
       "      <td>0.570185</td>\n",
       "      <td>-0.068436</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109899</th>\n",
       "      <td>zanovello_shsy5y_curve_0021</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>86.659312</td>\n",
       "      <td>0.096907</td>\n",
       "      <td>0.350523</td>\n",
       "      <td>5.538175e-01</td>\n",
       "      <td>8.928920e-01</td>\n",
       "      <td>5.995127</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.082769</td>\n",
       "      <td>0.100345</td>\n",
       "      <td>0.017576</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109900</th>\n",
       "      <td>zanovello_shsy5y_curve_0021</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>544.482004</td>\n",
       "      <td>0.004321</td>\n",
       "      <td>7.352049</td>\n",
       "      <td>6.698637e-03</td>\n",
       "      <td>1.075579e-01</td>\n",
       "      <td>11.929101</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.245516</td>\n",
       "      <td>0.285854</td>\n",
       "      <td>0.040337</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109901</th>\n",
       "      <td>zanovello_shsy5y_curve_0021</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>710.106273</td>\n",
       "      <td>0.005473</td>\n",
       "      <td>8.239053</td>\n",
       "      <td>4.099852e-03</td>\n",
       "      <td>7.804047e-02</td>\n",
       "      <td>13.461381</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.638751</td>\n",
       "      <td>0.566147</td>\n",
       "      <td>-0.072605</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119377</th>\n",
       "      <td>zanovello_shsy5y_curve_0025</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>76.458193</td>\n",
       "      <td>0.073991</td>\n",
       "      <td>0.034458</td>\n",
       "      <td>8.527360e-01</td>\n",
       "      <td>9.607204e-01</td>\n",
       "      <td>6.210104</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.082709</td>\n",
       "      <td>0.077357</td>\n",
       "      <td>-0.005352</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119378</th>\n",
       "      <td>zanovello_shsy5y_curve_0025</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>590.881307</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>18.551999</td>\n",
       "      <td>1.653317e-05</td>\n",
       "      <td>5.932833e-04</td>\n",
       "      <td>12.293834</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.245536</td>\n",
       "      <td>0.319812</td>\n",
       "      <td>0.074276</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119379</th>\n",
       "      <td>zanovello_shsy5y_curve_0025</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>711.602055</td>\n",
       "      <td>0.004601</td>\n",
       "      <td>16.252765</td>\n",
       "      <td>5.542865e-05</td>\n",
       "      <td>1.677186e-03</td>\n",
       "      <td>13.828617</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.638778</td>\n",
       "      <td>0.556055</td>\n",
       "      <td>-0.082723</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133660</th>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>63.625651</td>\n",
       "      <td>0.139196</td>\n",
       "      <td>0.456293</td>\n",
       "      <td>4.993615e-01</td>\n",
       "      <td>6.984629e-01</td>\n",
       "      <td>5.499118</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.082579</td>\n",
       "      <td>0.064042</td>\n",
       "      <td>-0.018536</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133661</th>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>506.344592</td>\n",
       "      <td>0.009260</td>\n",
       "      <td>2.732345</td>\n",
       "      <td>9.833479e-02</td>\n",
       "      <td>2.556704e-01</td>\n",
       "      <td>11.220149</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.245655</td>\n",
       "      <td>0.282123</td>\n",
       "      <td>0.036467</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133662</th>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>689.473804</td>\n",
       "      <td>0.004574</td>\n",
       "      <td>2.313277</td>\n",
       "      <td>1.282733e-01</td>\n",
       "      <td>3.071809e-01</td>\n",
       "      <td>12.736122</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.638786</td>\n",
       "      <td>0.609473</td>\n",
       "      <td>-0.029314</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142647</th>\n",
       "      <td>zanovello_skndz_curve_002</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>96.358193</td>\n",
       "      <td>0.045133</td>\n",
       "      <td>1.567616</td>\n",
       "      <td>2.105536e-01</td>\n",
       "      <td>8.742224e-01</td>\n",
       "      <td>6.780247</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.073347</td>\n",
       "      <td>0.098691</td>\n",
       "      <td>0.025344</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142648</th>\n",
       "      <td>zanovello_skndz_curve_002</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>410.408858</td>\n",
       "      <td>0.019788</td>\n",
       "      <td>0.956813</td>\n",
       "      <td>3.279912e-01</td>\n",
       "      <td>8.986341e-01</td>\n",
       "      <td>12.824541</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173664</td>\n",
       "      <td>0.154850</td>\n",
       "      <td>-0.018814</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142649</th>\n",
       "      <td>zanovello_skndz_curve_002</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>765.422724</td>\n",
       "      <td>0.026747</td>\n",
       "      <td>0.090350</td>\n",
       "      <td>7.637325e-01</td>\n",
       "      <td>9.894349e-01</td>\n",
       "      <td>15.333824</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.696851</td>\n",
       "      <td>0.704889</td>\n",
       "      <td>0.008037</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149994</th>\n",
       "      <td>zanovello_skndz_curve_003</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>86.112958</td>\n",
       "      <td>0.031029</td>\n",
       "      <td>0.087214</td>\n",
       "      <td>7.677498e-01</td>\n",
       "      <td>9.789854e-01</td>\n",
       "      <td>6.559546</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.072946</td>\n",
       "      <td>0.079372</td>\n",
       "      <td>0.006427</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149995</th>\n",
       "      <td>zanovello_skndz_curve_003</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>469.110459</td>\n",
       "      <td>0.003460</td>\n",
       "      <td>5.118592</td>\n",
       "      <td>2.367081e-02</td>\n",
       "      <td>3.003310e-01</td>\n",
       "      <td>12.464778</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173598</td>\n",
       "      <td>0.201720</td>\n",
       "      <td>0.028122</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149996</th>\n",
       "      <td>zanovello_skndz_curve_003</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>751.717935</td>\n",
       "      <td>0.005942</td>\n",
       "      <td>3.537481</td>\n",
       "      <td>5.999650e-02</td>\n",
       "      <td>4.767092e-01</td>\n",
       "      <td>14.942506</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.697270</td>\n",
       "      <td>0.661589</td>\n",
       "      <td>-0.035681</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159813</th>\n",
       "      <td>zanovello_skndz_curve_004</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>107.453431</td>\n",
       "      <td>0.038148</td>\n",
       "      <td>1.846493</td>\n",
       "      <td>1.741918e-01</td>\n",
       "      <td>6.484853e-01</td>\n",
       "      <td>6.730290</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.073234</td>\n",
       "      <td>0.103665</td>\n",
       "      <td>0.030432</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159814</th>\n",
       "      <td>zanovello_skndz_curve_004</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>519.592776</td>\n",
       "      <td>0.006799</td>\n",
       "      <td>4.929704</td>\n",
       "      <td>2.639883e-02</td>\n",
       "      <td>2.564286e-01</td>\n",
       "      <td>12.880722</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173714</td>\n",
       "      <td>0.213019</td>\n",
       "      <td>0.039305</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159815</th>\n",
       "      <td>zanovello_skndz_curve_004</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>798.917656</td>\n",
       "      <td>0.008124</td>\n",
       "      <td>7.433330</td>\n",
       "      <td>6.402674e-03</td>\n",
       "      <td>9.740151e-02</td>\n",
       "      <td>15.525733</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.696949</td>\n",
       "      <td>0.639043</td>\n",
       "      <td>-0.057906</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169488</th>\n",
       "      <td>zanovello_skndz_curve_005</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>93.361612</td>\n",
       "      <td>0.050017</td>\n",
       "      <td>0.122989</td>\n",
       "      <td>7.258149e-01</td>\n",
       "      <td>9.269410e-01</td>\n",
       "      <td>7.049336</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.073487</td>\n",
       "      <td>0.088465</td>\n",
       "      <td>0.014978</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169489</th>\n",
       "      <td>zanovello_skndz_curve_005</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>516.867119</td>\n",
       "      <td>0.003791</td>\n",
       "      <td>15.790468</td>\n",
       "      <td>7.075809e-05</td>\n",
       "      <td>2.018021e-03</td>\n",
       "      <td>13.204374</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173613</td>\n",
       "      <td>0.233301</td>\n",
       "      <td>0.059688</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169490</th>\n",
       "      <td>zanovello_skndz_curve_005</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>755.946263</td>\n",
       "      <td>0.009815</td>\n",
       "      <td>7.818465</td>\n",
       "      <td>5.171510e-03</td>\n",
       "      <td>6.303054e-02</td>\n",
       "      <td>15.768428</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.696798</td>\n",
       "      <td>0.648997</td>\n",
       "      <td>-0.047801</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179760</th>\n",
       "      <td>zanovello_skndz_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>100.664169</td>\n",
       "      <td>0.051007</td>\n",
       "      <td>0.294538</td>\n",
       "      <td>5.873270e-01</td>\n",
       "      <td>8.620445e-01</td>\n",
       "      <td>7.064735</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.073287</td>\n",
       "      <td>0.092230</td>\n",
       "      <td>0.018943</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>zanovello_skndz_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>544.545181</td>\n",
       "      <td>0.008389</td>\n",
       "      <td>8.642271</td>\n",
       "      <td>3.284518e-03</td>\n",
       "      <td>3.744826e-02</td>\n",
       "      <td>13.578006</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173718</td>\n",
       "      <td>0.232457</td>\n",
       "      <td>0.058739</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>zanovello_skndz_curve_0075</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>796.666715</td>\n",
       "      <td>0.006041</td>\n",
       "      <td>13.072656</td>\n",
       "      <td>2.996380e-04</td>\n",
       "      <td>5.357255e-03</td>\n",
       "      <td>16.373214</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.696864</td>\n",
       "      <td>0.640334</td>\n",
       "      <td>-0.056530</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190549</th>\n",
       "      <td>zanovello_skndz_curve_01</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>92.209662</td>\n",
       "      <td>0.048289</td>\n",
       "      <td>0.023722</td>\n",
       "      <td>8.775946e-01</td>\n",
       "      <td>9.533776e-01</td>\n",
       "      <td>7.165208</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.073339</td>\n",
       "      <td>0.082073</td>\n",
       "      <td>0.008734</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190550</th>\n",
       "      <td>zanovello_skndz_curve_01</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>532.413758</td>\n",
       "      <td>0.012118</td>\n",
       "      <td>6.576006</td>\n",
       "      <td>1.033626e-02</td>\n",
       "      <td>8.707122e-02</td>\n",
       "      <td>13.599112</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173627</td>\n",
       "      <td>0.232477</td>\n",
       "      <td>0.058850</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190551</th>\n",
       "      <td>zanovello_skndz_curve_01</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>779.008582</td>\n",
       "      <td>0.010487</td>\n",
       "      <td>7.234780</td>\n",
       "      <td>7.150459e-03</td>\n",
       "      <td>6.674859e-02</td>\n",
       "      <td>16.310471</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.696912</td>\n",
       "      <td>0.639784</td>\n",
       "      <td>-0.057128</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200009</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000127511.10:E001</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E001</td>\n",
       "      <td>117.260338</td>\n",
       "      <td>0.038149</td>\n",
       "      <td>1.943215</td>\n",
       "      <td>1.633199e-01</td>\n",
       "      <td>4.311100e-01</td>\n",
       "      <td>6.893364</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685574e+07</td>\n",
       "      <td>0.072946</td>\n",
       "      <td>0.111551</td>\n",
       "      <td>0.038605</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200010</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000127511.10:E002</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E002</td>\n",
       "      <td>613.040549</td>\n",
       "      <td>0.009959</td>\n",
       "      <td>15.518159</td>\n",
       "      <td>8.171646e-05</td>\n",
       "      <td>1.260793e-03</td>\n",
       "      <td>13.163862</td>\n",
       "      <td>...</td>\n",
       "      <td>novel</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.685414e+07</td>\n",
       "      <td>1.685742e+07</td>\n",
       "      <td>0.173565</td>\n",
       "      <td>0.262698</td>\n",
       "      <td>0.089133</td>\n",
       "      <td>bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200011</th>\n",
       "      <td>zanovello_skndz_curve_1</td>\n",
       "      <td>ENSG00000127511.10:E003</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>E003</td>\n",
       "      <td>797.952543</td>\n",
       "      <td>0.009552</td>\n",
       "      <td>21.364045</td>\n",
       "      <td>3.798279e-06</td>\n",
       "      <td>8.142211e-05</td>\n",
       "      <td>15.810042</td>\n",
       "      <td>...</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000601141.5,ENST00000595541.1,ENST000005...</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>1.687850e+47</td>\n",
       "      <td>1.687906e+47</td>\n",
       "      <td>0.697314</td>\n",
       "      <td>0.600909</td>\n",
       "      <td>-0.096406</td>\n",
       "      <td>spliced</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          experiment_name  \\\n",
       "40                                      brown_i3_cortical   \n",
       "41                                      brown_i3_cortical   \n",
       "16839                                        brown_shsy5y   \n",
       "16840                                        brown_shsy5y   \n",
       "25677                                         brown_skndz   \n",
       "25678                                         brown_skndz   \n",
       "25679                                         brown_skndz   \n",
       "34381                                humphrey_i3_cortical   \n",
       "34382                                humphrey_i3_cortical   \n",
       "34383                                humphrey_i3_cortical   \n",
       "47455                                       klim_i3_motor   \n",
       "47456                                       klim_i3_motor   \n",
       "55826                                seddighi_i3_cortical   \n",
       "55827                                seddighi_i3_cortical   \n",
       "67510   zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...   \n",
       "67511   zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...   \n",
       "67512   zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...   \n",
       "79770              zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl   \n",
       "79771              zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl   \n",
       "79772              zanovello_shsy5y_chx_kd_ctl_vs_ctl_ctl   \n",
       "93191                        zanovello_shsy5y_curve_00125   \n",
       "93192                        zanovello_shsy5y_curve_00125   \n",
       "93193                        zanovello_shsy5y_curve_00125   \n",
       "99663                        zanovello_shsy5y_curve_00187   \n",
       "99664                        zanovello_shsy5y_curve_00187   \n",
       "99665                        zanovello_shsy5y_curve_00187   \n",
       "109899                        zanovello_shsy5y_curve_0021   \n",
       "109900                        zanovello_shsy5y_curve_0021   \n",
       "109901                        zanovello_shsy5y_curve_0021   \n",
       "119377                        zanovello_shsy5y_curve_0025   \n",
       "119378                        zanovello_shsy5y_curve_0025   \n",
       "119379                        zanovello_shsy5y_curve_0025   \n",
       "133660                        zanovello_shsy5y_curve_0075   \n",
       "133661                        zanovello_shsy5y_curve_0075   \n",
       "133662                        zanovello_shsy5y_curve_0075   \n",
       "142647                          zanovello_skndz_curve_002   \n",
       "142648                          zanovello_skndz_curve_002   \n",
       "142649                          zanovello_skndz_curve_002   \n",
       "149994                          zanovello_skndz_curve_003   \n",
       "149995                          zanovello_skndz_curve_003   \n",
       "149996                          zanovello_skndz_curve_003   \n",
       "159813                          zanovello_skndz_curve_004   \n",
       "159814                          zanovello_skndz_curve_004   \n",
       "159815                          zanovello_skndz_curve_004   \n",
       "169488                          zanovello_skndz_curve_005   \n",
       "169489                          zanovello_skndz_curve_005   \n",
       "169490                          zanovello_skndz_curve_005   \n",
       "179760                         zanovello_skndz_curve_0075   \n",
       "179761                         zanovello_skndz_curve_0075   \n",
       "179762                         zanovello_skndz_curve_0075   \n",
       "190549                           zanovello_skndz_curve_01   \n",
       "190550                           zanovello_skndz_curve_01   \n",
       "190551                           zanovello_skndz_curve_01   \n",
       "200009                            zanovello_skndz_curve_1   \n",
       "200010                            zanovello_skndz_curve_1   \n",
       "200011                            zanovello_skndz_curve_1   \n",
       "\n",
       "                          binID             groupID featureID  exonBaseMean  \\\n",
       "40      ENSG00000127511.10:E001  ENSG00000127511.10      E001    685.073107   \n",
       "41      ENSG00000127511.10:E002  ENSG00000127511.10      E002   1635.679342   \n",
       "16839   ENSG00000127511.10:E001  ENSG00000127511.10      E001    646.630524   \n",
       "16840   ENSG00000127511.10:E002  ENSG00000127511.10      E002   1583.366570   \n",
       "25677   ENSG00000127511.10:E001  ENSG00000127511.10      E001    123.423529   \n",
       "25678   ENSG00000127511.10:E002  ENSG00000127511.10      E002   1382.957785   \n",
       "25679   ENSG00000127511.10:E003  ENSG00000127511.10      E003    734.474029   \n",
       "34381   ENSG00000127511.10:E001  ENSG00000127511.10      E001     31.461331   \n",
       "34382   ENSG00000127511.10:E002  ENSG00000127511.10      E002    232.592245   \n",
       "34383   ENSG00000127511.10:E003  ENSG00000127511.10      E003    469.902739   \n",
       "47455   ENSG00000127511.10:E001  ENSG00000127511.10      E001    144.835749   \n",
       "47456   ENSG00000127511.10:E002  ENSG00000127511.10      E002    611.086934   \n",
       "55826   ENSG00000127511.10:E001  ENSG00000127511.10      E001   1278.572167   \n",
       "55827   ENSG00000127511.10:E002  ENSG00000127511.10      E002   3394.001584   \n",
       "67510   ENSG00000127511.10:E001  ENSG00000127511.10      E001    105.534632   \n",
       "67511   ENSG00000127511.10:E002  ENSG00000127511.10      E002    988.462295   \n",
       "67512   ENSG00000127511.10:E003  ENSG00000127511.10      E003    951.766052   \n",
       "79770   ENSG00000127511.10:E001  ENSG00000127511.10      E001     71.399225   \n",
       "79771   ENSG00000127511.10:E002  ENSG00000127511.10      E002    594.115006   \n",
       "79772   ENSG00000127511.10:E003  ENSG00000127511.10      E003    753.314063   \n",
       "93191   ENSG00000127511.10:E001  ENSG00000127511.10      E001     76.428022   \n",
       "93192   ENSG00000127511.10:E002  ENSG00000127511.10      E002    532.677509   \n",
       "93193   ENSG00000127511.10:E003  ENSG00000127511.10      E003    750.349881   \n",
       "99663   ENSG00000127511.10:E001  ENSG00000127511.10      E001     76.565383   \n",
       "99664   ENSG00000127511.10:E002  ENSG00000127511.10      E002    563.779051   \n",
       "99665   ENSG00000127511.10:E003  ENSG00000127511.10      E003    694.466108   \n",
       "109899  ENSG00000127511.10:E001  ENSG00000127511.10      E001     86.659312   \n",
       "109900  ENSG00000127511.10:E002  ENSG00000127511.10      E002    544.482004   \n",
       "109901  ENSG00000127511.10:E003  ENSG00000127511.10      E003    710.106273   \n",
       "119377  ENSG00000127511.10:E001  ENSG00000127511.10      E001     76.458193   \n",
       "119378  ENSG00000127511.10:E002  ENSG00000127511.10      E002    590.881307   \n",
       "119379  ENSG00000127511.10:E003  ENSG00000127511.10      E003    711.602055   \n",
       "133660  ENSG00000127511.10:E001  ENSG00000127511.10      E001     63.625651   \n",
       "133661  ENSG00000127511.10:E002  ENSG00000127511.10      E002    506.344592   \n",
       "133662  ENSG00000127511.10:E003  ENSG00000127511.10      E003    689.473804   \n",
       "142647  ENSG00000127511.10:E001  ENSG00000127511.10      E001     96.358193   \n",
       "142648  ENSG00000127511.10:E002  ENSG00000127511.10      E002    410.408858   \n",
       "142649  ENSG00000127511.10:E003  ENSG00000127511.10      E003    765.422724   \n",
       "149994  ENSG00000127511.10:E001  ENSG00000127511.10      E001     86.112958   \n",
       "149995  ENSG00000127511.10:E002  ENSG00000127511.10      E002    469.110459   \n",
       "149996  ENSG00000127511.10:E003  ENSG00000127511.10      E003    751.717935   \n",
       "159813  ENSG00000127511.10:E001  ENSG00000127511.10      E001    107.453431   \n",
       "159814  ENSG00000127511.10:E002  ENSG00000127511.10      E002    519.592776   \n",
       "159815  ENSG00000127511.10:E003  ENSG00000127511.10      E003    798.917656   \n",
       "169488  ENSG00000127511.10:E001  ENSG00000127511.10      E001     93.361612   \n",
       "169489  ENSG00000127511.10:E002  ENSG00000127511.10      E002    516.867119   \n",
       "169490  ENSG00000127511.10:E003  ENSG00000127511.10      E003    755.946263   \n",
       "179760  ENSG00000127511.10:E001  ENSG00000127511.10      E001    100.664169   \n",
       "179761  ENSG00000127511.10:E002  ENSG00000127511.10      E002    544.545181   \n",
       "179762  ENSG00000127511.10:E003  ENSG00000127511.10      E003    796.666715   \n",
       "190549  ENSG00000127511.10:E001  ENSG00000127511.10      E001     92.209662   \n",
       "190550  ENSG00000127511.10:E002  ENSG00000127511.10      E002    532.413758   \n",
       "190551  ENSG00000127511.10:E003  ENSG00000127511.10      E003    779.008582   \n",
       "200009  ENSG00000127511.10:E001  ENSG00000127511.10      E001    117.260338   \n",
       "200010  ENSG00000127511.10:E002  ENSG00000127511.10      E002    613.040549   \n",
       "200011  ENSG00000127511.10:E003  ENSG00000127511.10      E003    797.952543   \n",
       "\n",
       "        dispersion        stat         pvalue           padj  \\\n",
       "40        0.013534  127.151527   1.721175e-29   3.384311e-27   \n",
       "41        0.013184  130.260073   3.594567e-30   7.236205e-28   \n",
       "16839     0.070550    0.725662   3.942930e-01   8.831562e-01   \n",
       "16840     0.069263    0.738940   3.900008e-01   8.809464e-01   \n",
       "25677     0.990241    0.184085   6.678856e-01   8.576725e-01   \n",
       "25678     0.017284    3.819419   5.066181e-02   2.041506e-01   \n",
       "25679     0.013301    5.624554   1.771058e-02   1.009837e-01   \n",
       "34381     0.988492    3.663168   5.562777e-02   1.940285e-01   \n",
       "34382     0.020956  177.904729   1.389688e-40   7.563611e-38   \n",
       "34383     0.024516  174.525798   7.599079e-40   3.877430e-37   \n",
       "47455     0.019029    3.054360   8.052074e-02   5.581589e-01   \n",
       "47456     0.018797    3.084681   7.903278e-02   5.548139e-01   \n",
       "55826     0.014028  318.124993   3.709556e-71   4.794186e-69   \n",
       "55827     0.013990  318.966062   2.432856e-71   3.191833e-69   \n",
       "67510     0.005586    0.008014   9.286665e-01   9.841491e-01   \n",
       "67511     0.003196  442.611427   2.924845e-98   1.189716e-95   \n",
       "67512     0.002682  492.797690  3.508085e-109  1.664781e-106   \n",
       "79770     0.119240    1.893448   1.688136e-01   4.246955e-01   \n",
       "79771     0.012225   17.124480   3.500816e-05   5.101975e-04   \n",
       "79772     0.009719   27.125775   1.906391e-07   4.859659e-06   \n",
       "93191     0.073518    0.012217   9.119869e-01   9.997851e-01   \n",
       "93192     0.006241    1.125615   2.887127e-01   9.671906e-01   \n",
       "93193     0.007609    0.567313   4.513287e-01   9.730105e-01   \n",
       "99663     0.096247    0.009507   9.223256e-01   9.830233e-01   \n",
       "99664     0.006498   11.575263   6.683498e-04   1.776802e-02   \n",
       "99665     0.003179   17.797581   2.456955e-05   1.282166e-03   \n",
       "109899    0.096907    0.350523   5.538175e-01   8.928920e-01   \n",
       "109900    0.004321    7.352049   6.698637e-03   1.075579e-01   \n",
       "109901    0.005473    8.239053   4.099852e-03   7.804047e-02   \n",
       "119377    0.073991    0.034458   8.527360e-01   9.607204e-01   \n",
       "119378    0.004888   18.551999   1.653317e-05   5.932833e-04   \n",
       "119379    0.004601   16.252765   5.542865e-05   1.677186e-03   \n",
       "133660    0.139196    0.456293   4.993615e-01   6.984629e-01   \n",
       "133661    0.009260    2.732345   9.833479e-02   2.556704e-01   \n",
       "133662    0.004574    2.313277   1.282733e-01   3.071809e-01   \n",
       "142647    0.045133    1.567616   2.105536e-01   8.742224e-01   \n",
       "142648    0.019788    0.956813   3.279912e-01   8.986341e-01   \n",
       "142649    0.026747    0.090350   7.637325e-01   9.894349e-01   \n",
       "149994    0.031029    0.087214   7.677498e-01   9.789854e-01   \n",
       "149995    0.003460    5.118592   2.367081e-02   3.003310e-01   \n",
       "149996    0.005942    3.537481   5.999650e-02   4.767092e-01   \n",
       "159813    0.038148    1.846493   1.741918e-01   6.484853e-01   \n",
       "159814    0.006799    4.929704   2.639883e-02   2.564286e-01   \n",
       "159815    0.008124    7.433330   6.402674e-03   9.740151e-02   \n",
       "169488    0.050017    0.122989   7.258149e-01   9.269410e-01   \n",
       "169489    0.003791   15.790468   7.075809e-05   2.018021e-03   \n",
       "169490    0.009815    7.818465   5.171510e-03   6.303054e-02   \n",
       "179760    0.051007    0.294538   5.873270e-01   8.620445e-01   \n",
       "179761    0.008389    8.642271   3.284518e-03   3.744826e-02   \n",
       "179762    0.006041   13.072656   2.996380e-04   5.357255e-03   \n",
       "190549    0.048289    0.023722   8.775946e-01   9.533776e-01   \n",
       "190550    0.012118    6.576006   1.033626e-02   8.707122e-02   \n",
       "190551    0.010487    7.234780   7.150459e-03   6.674859e-02   \n",
       "200009    0.038149    1.943215   1.633199e-01   4.311100e-01   \n",
       "200010    0.009959   15.518159   8.171646e-05   1.260793e-03   \n",
       "200011    0.009552   21.364045   3.798279e-06   8.142211e-05   \n",
       "\n",
       "        UsageCoefficient_base  ...  annot_status  \\\n",
       "40                  10.215189  ...         novel   \n",
       "41                  17.008705  ...     annotated   \n",
       "16839               13.366541  ...         novel   \n",
       "16840               16.186991  ...     annotated   \n",
       "25677                7.870387  ...     annotated   \n",
       "25678               18.595389  ...         novel   \n",
       "25679               14.791017  ...     annotated   \n",
       "34381                4.173276  ...     annotated   \n",
       "34382                6.775476  ...         novel   \n",
       "34383                9.311451  ...     annotated   \n",
       "47455                7.224955  ...         novel   \n",
       "47456               10.885618  ...     annotated   \n",
       "55826               16.295375  ...         novel   \n",
       "55827               24.386196  ...     annotated   \n",
       "67510                7.657459  ...     annotated   \n",
       "67511               14.900440  ...         novel   \n",
       "67512               19.629855  ...     annotated   \n",
       "79770                5.322357  ...     annotated   \n",
       "79771               11.991790  ...         novel   \n",
       "79772               13.950307  ...     annotated   \n",
       "93191                5.786969  ...     annotated   \n",
       "93192               11.893295  ...         novel   \n",
       "93193               13.462503  ...     annotated   \n",
       "99663                1.893856  ...     annotated   \n",
       "99664                2.696905  ...         novel   \n",
       "99665                2.869866  ...     annotated   \n",
       "109899               5.995127  ...     annotated   \n",
       "109900              11.929101  ...         novel   \n",
       "109901              13.461381  ...     annotated   \n",
       "119377               6.210104  ...     annotated   \n",
       "119378              12.293834  ...         novel   \n",
       "119379              13.828617  ...     annotated   \n",
       "133660               5.499118  ...     annotated   \n",
       "133661              11.220149  ...         novel   \n",
       "133662              12.736122  ...     annotated   \n",
       "142647               6.780247  ...     annotated   \n",
       "142648              12.824541  ...         novel   \n",
       "142649              15.333824  ...     annotated   \n",
       "149994               6.559546  ...     annotated   \n",
       "149995              12.464778  ...         novel   \n",
       "149996              14.942506  ...     annotated   \n",
       "159813               6.730290  ...     annotated   \n",
       "159814              12.880722  ...         novel   \n",
       "159815              15.525733  ...     annotated   \n",
       "169488               7.049336  ...     annotated   \n",
       "169489              13.204374  ...         novel   \n",
       "169490              15.768428  ...     annotated   \n",
       "179760               7.064735  ...     annotated   \n",
       "179761              13.578006  ...         novel   \n",
       "179762              16.373214  ...     annotated   \n",
       "190549               7.165208  ...     annotated   \n",
       "190550              13.599112  ...         novel   \n",
       "190551              16.310471  ...     annotated   \n",
       "200009               6.893364  ...     annotated   \n",
       "200010              13.163862  ...         novel   \n",
       "200011              15.810042  ...     annotated   \n",
       "\n",
       "                                            transcript_id chromosome  strand  \\\n",
       "40                               PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "41      ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "16839                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "16840   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "25677                                   ENST00000596802.5      chr19       +   \n",
       "25678                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "25679   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "34381                                   ENST00000596802.5      chr19       +   \n",
       "34382                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "34383   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "47455                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "47456   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "55826                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "55827   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "67510                                   ENST00000596802.5      chr19       +   \n",
       "67511                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "67512   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "79770                                   ENST00000596802.5      chr19       +   \n",
       "79771                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "79772   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "93191                                   ENST00000596802.5      chr19       +   \n",
       "93192                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "93193   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "99663                                   ENST00000596802.5      chr19       +   \n",
       "99664                            PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "99665   ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "109899                                  ENST00000596802.5      chr19       +   \n",
       "109900                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "109901  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "119377                                  ENST00000596802.5      chr19       +   \n",
       "119378                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "119379  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "133660                                  ENST00000596802.5      chr19       +   \n",
       "133661                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "133662  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "142647                                  ENST00000596802.5      chr19       +   \n",
       "142648                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "142649  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "149994                                  ENST00000596802.5      chr19       +   \n",
       "149995                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "149996  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "159813                                  ENST00000596802.5      chr19       +   \n",
       "159814                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "159815  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "169488                                  ENST00000596802.5      chr19       +   \n",
       "169489                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "169490  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "179760                                  ENST00000596802.5      chr19       +   \n",
       "179761                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "179762  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "190549                                  ENST00000596802.5      chr19       +   \n",
       "190550                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "190551  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "200009                                  ENST00000596802.5      chr19       +   \n",
       "200010                           PAPA.ctrl_ctrl_2.24194.4      chr19       +   \n",
       "200011  ENST00000601141.5,ENST00000595541.1,ENST000005...      chr19       +   \n",
       "\n",
       "               start           end mean_PPAU_base mean_PPAU_treatment  \\\n",
       "40      1.685414e+07  1.685742e+07       0.065536            0.226421   \n",
       "41      1.687850e+47  1.687906e+47       0.832218            0.649472   \n",
       "16839   1.685414e+07  1.685742e+07       0.181492            0.147692   \n",
       "16840   1.687850e+47  1.687906e+47       0.789384            0.806265   \n",
       "25677   1.685414e+07  1.685574e+07       0.091220            0.082138   \n",
       "25678   1.685414e+07  1.685742e+07       0.459231            0.371105   \n",
       "25679   1.687850e+47  1.687906e+47       0.436124            0.489640   \n",
       "34381   1.685414e+07  1.685574e+07       0.019720            0.072751   \n",
       "34382   1.685414e+07  1.685742e+07       0.062796            0.271946   \n",
       "34383   1.687850e+47  1.687906e+47       0.908924            0.596359   \n",
       "47455   1.685414e+07  1.685742e+07       0.066292            0.080668   \n",
       "47456   1.687850e+47  1.687906e+47       0.847995            0.836036   \n",
       "55826   1.685414e+07  1.685742e+07       0.084287            0.306735   \n",
       "55827   1.687850e+47  1.687906e+47       0.863467            0.668574   \n",
       "67510   1.685414e+07  1.685574e+07       0.057043            0.074931   \n",
       "67511   1.685414e+07  1.685742e+07       0.137892            0.394443   \n",
       "67512   1.687850e+47  1.687906e+47       0.788094            0.522136   \n",
       "79770   1.685414e+07  1.685574e+07       0.054078            0.086728   \n",
       "79771   1.685414e+07  1.685742e+07       0.222966            0.312574   \n",
       "79772   1.687850e+47  1.687906e+47       0.680860            0.559108   \n",
       "93191   1.685414e+07  1.685574e+07       0.082612            0.076246   \n",
       "93192   1.685414e+07  1.685742e+07       0.245697            0.263737   \n",
       "93193   1.687850e+47  1.687906e+47       0.638713            0.618608   \n",
       "99663   1.685414e+07  1.685574e+07       0.082906            0.081571   \n",
       "99664   1.685414e+07  1.685742e+07       0.245501            0.316147   \n",
       "99665   1.687850e+47  1.687906e+47       0.638621            0.570185   \n",
       "109899  1.685414e+07  1.685574e+07       0.082769            0.100345   \n",
       "109900  1.685414e+07  1.685742e+07       0.245516            0.285854   \n",
       "109901  1.687850e+47  1.687906e+47       0.638751            0.566147   \n",
       "119377  1.685414e+07  1.685574e+07       0.082709            0.077357   \n",
       "119378  1.685414e+07  1.685742e+07       0.245536            0.319812   \n",
       "119379  1.687850e+47  1.687906e+47       0.638778            0.556055   \n",
       "133660  1.685414e+07  1.685574e+07       0.082579            0.064042   \n",
       "133661  1.685414e+07  1.685742e+07       0.245655            0.282123   \n",
       "133662  1.687850e+47  1.687906e+47       0.638786            0.609473   \n",
       "142647  1.685414e+07  1.685574e+07       0.073347            0.098691   \n",
       "142648  1.685414e+07  1.685742e+07       0.173664            0.154850   \n",
       "142649  1.687850e+47  1.687906e+47       0.696851            0.704889   \n",
       "149994  1.685414e+07  1.685574e+07       0.072946            0.079372   \n",
       "149995  1.685414e+07  1.685742e+07       0.173598            0.201720   \n",
       "149996  1.687850e+47  1.687906e+47       0.697270            0.661589   \n",
       "159813  1.685414e+07  1.685574e+07       0.073234            0.103665   \n",
       "159814  1.685414e+07  1.685742e+07       0.173714            0.213019   \n",
       "159815  1.687850e+47  1.687906e+47       0.696949            0.639043   \n",
       "169488  1.685414e+07  1.685574e+07       0.073487            0.088465   \n",
       "169489  1.685414e+07  1.685742e+07       0.173613            0.233301   \n",
       "169490  1.687850e+47  1.687906e+47       0.696798            0.648997   \n",
       "179760  1.685414e+07  1.685574e+07       0.073287            0.092230   \n",
       "179761  1.685414e+07  1.685742e+07       0.173718            0.232457   \n",
       "179762  1.687850e+47  1.687906e+47       0.696864            0.640334   \n",
       "190549  1.685414e+07  1.685574e+07       0.073339            0.082073   \n",
       "190550  1.685414e+07  1.685742e+07       0.173627            0.232477   \n",
       "190551  1.687850e+47  1.687906e+47       0.696912            0.639784   \n",
       "200009  1.685414e+07  1.685574e+07       0.072946            0.111551   \n",
       "200010  1.685414e+07  1.685742e+07       0.173565            0.262698   \n",
       "200011  1.687850e+47  1.687906e+47       0.697314            0.600909   \n",
       "\n",
       "       delta_PPAU_treatment_control simple_event_type  \n",
       "40                         0.160885      bleedthrough  \n",
       "41                        -0.182746           spliced  \n",
       "16839                     -0.033800      bleedthrough  \n",
       "16840                      0.016880           spliced  \n",
       "25677                     -0.009082      bleedthrough  \n",
       "25678                     -0.088125      bleedthrough  \n",
       "25679                      0.053516           spliced  \n",
       "34381                      0.053031      bleedthrough  \n",
       "34382                      0.209150      bleedthrough  \n",
       "34383                     -0.312565           spliced  \n",
       "47455                      0.014376      bleedthrough  \n",
       "47456                     -0.011959           spliced  \n",
       "55826                      0.222448      bleedthrough  \n",
       "55827                     -0.194894           spliced  \n",
       "67510                      0.017888      bleedthrough  \n",
       "67511                      0.256551      bleedthrough  \n",
       "67512                     -0.265958           spliced  \n",
       "79770                      0.032651      bleedthrough  \n",
       "79771                      0.089608      bleedthrough  \n",
       "79772                     -0.121752           spliced  \n",
       "93191                     -0.006366      bleedthrough  \n",
       "93192                      0.018040      bleedthrough  \n",
       "93193                     -0.020105           spliced  \n",
       "99663                     -0.001334      bleedthrough  \n",
       "99664                      0.070646      bleedthrough  \n",
       "99665                     -0.068436           spliced  \n",
       "109899                     0.017576      bleedthrough  \n",
       "109900                     0.040337      bleedthrough  \n",
       "109901                    -0.072605           spliced  \n",
       "119377                    -0.005352      bleedthrough  \n",
       "119378                     0.074276      bleedthrough  \n",
       "119379                    -0.082723           spliced  \n",
       "133660                    -0.018536      bleedthrough  \n",
       "133661                     0.036467      bleedthrough  \n",
       "133662                    -0.029314           spliced  \n",
       "142647                     0.025344      bleedthrough  \n",
       "142648                    -0.018814      bleedthrough  \n",
       "142649                     0.008037           spliced  \n",
       "149994                     0.006427      bleedthrough  \n",
       "149995                     0.028122      bleedthrough  \n",
       "149996                    -0.035681           spliced  \n",
       "159813                     0.030432      bleedthrough  \n",
       "159814                     0.039305      bleedthrough  \n",
       "159815                    -0.057906           spliced  \n",
       "169488                     0.014978      bleedthrough  \n",
       "169489                     0.059688      bleedthrough  \n",
       "169490                    -0.047801           spliced  \n",
       "179760                     0.018943      bleedthrough  \n",
       "179761                     0.058739      bleedthrough  \n",
       "179762                    -0.056530           spliced  \n",
       "190549                     0.008734      bleedthrough  \n",
       "190550                     0.058850      bleedthrough  \n",
       "190551                    -0.057128           spliced  \n",
       "200009                     0.038605      bleedthrough  \n",
       "200010                     0.089133      bleedthrough  \n",
       "200011                    -0.096406           spliced  \n",
       "\n",
       "[56 rows x 28 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df[dexseq_df[\"gene_name\"] == \"SIN3B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of spliced last exons with 3'UTR extension - 58895\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0         ENSG00000021645.20_1\n",
       "1         ENSG00000021645.20_4\n",
       "2         ENSG00000021645.20_4\n",
       "3         ENSG00000048649.14_1\n",
       "4         ENSG00000048649.14_1\n",
       "                  ...         \n",
       "209475     ENSG00000278311.5_5\n",
       "209480     ENSG00000280832.2_1\n",
       "209481     ENSG00000280832.2_2\n",
       "209482     ENSG00000282508.2_4\n",
       "209483     ENSG00000282508.2_6\n",
       "Name: le_id, Length: 71911, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# good idea to ensure don't consider spliced LEs that have a 3'UTR extension\n",
    "# UPDATE now need to select representative proximal site for each distal site, since quantification approach lumps together all annotated proximal sites (and compares them to extension) \n",
    "# since le_ids are annotated sequentially, know that immediately succeeding le_number is the partner le_id\n",
    "le_id_spliced_spl = le_id_spliced.str.split(\"_\", regex=False, expand=True)\n",
    "le_id_spliced_spl[1] = le_id_spliced_spl[1].astype(int).add(1)\n",
    "# reconstruct le_id (corresponding to theoretical distal 3'UTR extension le_id)\n",
    "le_id_spliced_d3utr = le_id_spliced_spl[0].str.cat(le_id_spliced_spl[1].astype(str), sep=\"_\")\n",
    "\n",
    "# now have theoretical distal 3'UTR le_ids, intersect with actual ids - any overlapping will be removed from downstream analysis\n",
    "spliced_d3utr_olap = le_id_spliced_d3utr.isin(le_id_d3utr.values)\n",
    "print(f\"Number of spliced last exons with 3'UTR extension - {spliced_d3utr_olap.sum()}\")\n",
    "\n",
    "# now use mask to remove these ALEs\n",
    "le_id_spliced = le_id_spliced[~spliced_d3utr_olap]\n",
    "le_id_spliced\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of spliced intervals before selecting representative LEs - 6209\n",
      "Number of spliced intervals after selecting representative LEs - 949\n",
      "{'atlas_1_pred': 571, 'atlas_max_datasets': 143, 'atlas_max_datasets_shortest': 98, 'motif_1_min': 321, 'motif_shortest_min': 160}\n"
     ]
    }
   ],
   "source": [
    "# add le_id used in downstream analysis\n",
    "novel_le = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"]))\n",
    "\n",
    "\n",
    "# Select representative PAS for bleedthroughs and spliced events (novel)\n",
    "novel_le_rep_spliced, rep_choices_spliced = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_spliced))), id_col=\"le_id_quant\")\n",
    "\n",
    "\n",
    "print(f\"Number of spliced intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_spliced))))}\")\n",
    "print(f\"Number of spliced intervals after selecting representative LEs - {len(novel_le_rep_spliced)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_spliced.items()})\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bleedthrough intervals before selecting representative LEs - 5952\n",
      "Number of spliced intervals after selecting representative LEs - 931\n",
      "{'atlas_1_pred': 590, 'atlas_max_datasets': 152, 'atlas_max_datasets_shortest': 105, 'motif_1_min': 350, 'motif_shortest_min': 114}\n"
     ]
    }
   ],
   "source": [
    "# Select representative PAS for bleedthroughs and spliced events\n",
    "novel_le_rep_bleed, rep_choices_bleed = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_bleed))), id_col=\"le_id_quant\")\n",
    "\n",
    "print(f\"Number of bleedthrough intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_bleed))))}\")\n",
    "print(f\"Number of spliced intervals after selecting representative LEs - {len(novel_le_rep_bleed)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_bleed.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic ALE-containing genes - 283\n",
      "Number of cryptic ALE isoforms - 283\n",
      "Number of ns ALE-containing genes - 2937\n",
      "Number of ns ALE isoforms - 6365\n"
     ]
    }
   ],
   "source": [
    "# now want to generate an ID with minimal info required for making maps\n",
    "# 3'end coordinates/last exon, gene name, le_id, site type, regulation status\n",
    "# site_type - spliced/bleedthrough\n",
    "# regulation_status - cryptic/background\n",
    "\n",
    "\n",
    "# define cryptic le_ids\n",
    "# define cryptic gene_ids\n",
    "dexseq_df_cryp = dexseq_df[(dexseq_df[\"padj\"].lt(0.05)) & (dexseq_df[\"mean_PPAU_base\"].le(0.1)) & (dexseq_df[\"delta_PPAU_treatment_control\"].ge(0.1))]\n",
    "cryp_le_ids = set(dexseq_df_cryp.le_id)\n",
    "cryp_gene_names = set(dexseq_df_cryp.gene_name)\n",
    "\n",
    "\n",
    "# define cryptic & background le_ids\n",
    "# background - ns in any dataset (to do this, get list of sig in any dataset)\n",
    "# get gene IDs with no regulated ALEs in any dataset\n",
    "# returns pd.Series\n",
    "ns_gene_ids = (dexseq_df.assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "               .groupby(\"groupID\")\n",
    "               [\"reg_status\"]\n",
    "               .sum()\n",
    "               .loc[lambda x: x == 0]\n",
    "             )\n",
    "\n",
    "# set of le_ids that belong to genes with no significant ALEs (is this too conservative?)\n",
    "ns_le_ids = set(dexseq_df.loc[dexseq_df[\"groupID\"].isin(ns_gene_ids.index), \"le_id\"])\n",
    "\n",
    "\n",
    "print(f\"Number of cryptic ALE-containing genes - {len(cryp_gene_names)}\")\n",
    "print(f\"Number of cryptic ALE isoforms - {len(cryp_gene_names)}\")\n",
    "\n",
    "print(f\"Number of ns ALE-containing genes - {len(ns_gene_ids)}\")\n",
    "print(f\"Number of ns ALE isoforms - {len(ns_le_ids)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          619\n",
       "background    285\n",
       "cryptic        27\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for novel bleedthroughs\n",
    "novel_le_rep_bleed = novel_le_rep_bleed.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id_quant\"))\n",
    "novel_le_rep_bleed.cryptic_status.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          604\n",
       "background    306\n",
       "cryptic        39\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for novel spliced events\n",
    "novel_le_rep_spliced = novel_le_rep_spliced.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id_quant\"))\n",
    "novel_le_rep_spliced.cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct 'name' string with minimal annotation information\n",
    "# le_id|gene_name|site_type|cryptic_status\n",
    "\n",
    "novel_le_rep_spliced = (novel_le_rep_spliced\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"spliced\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id_quant\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    ")\n",
    "\n",
    "novel_le_rep_bleed = (novel_le_rep_bleed\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"bleedthrough\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id_quant\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAPA.CTRL-2.2002.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PAPA.ctrl_ctrl_4.2339.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PAPA.TDP43-G_S7.362.1</td>\n",
       "      <td>ENSG00000117682.17_3</td>\n",
       "      <td>ENSG00000117682.17_3|DHDDS|bleedthrough|backgr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAPA.doxconc_NT_0_3.354.1</td>\n",
       "      <td>ENSG00000117682.17_6</td>\n",
       "      <td>ENSG00000117682.17_6|DHDDS|bleedthrough|backgr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>PAPA.TDP-2.25089.1</td>\n",
       "      <td>ENSG00000182518.14_1</td>\n",
       "      <td>ENSG00000182518.14_1|FAM104B|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>PAPA.doxconc_NT_0_3.19568.4</td>\n",
       "      <td>ENSG00000196459.15_1</td>\n",
       "      <td>ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>PAPA.Cont-D_S4.20482.2</td>\n",
       "      <td>ENSG00000197021.9_2</td>\n",
       "      <td>ENSG00000197021.9_2|EOLA2|bleedthrough|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>312 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   transcript_id                 le_id  \\\n",
       "0              PAPA.TDP-1.2120.4  ENSG00000082497.12_1   \n",
       "1             PAPA.CTRL-2.2002.3  ENSG00000082497.12_1   \n",
       "2        PAPA.ctrl_ctrl_4.2339.3  ENSG00000082497.12_1   \n",
       "3          PAPA.TDP43-G_S7.362.1  ENSG00000117682.17_3   \n",
       "4      PAPA.doxconc_NT_0_3.354.1  ENSG00000117682.17_6   \n",
       "..                           ...                   ...   \n",
       "307      PAPA.TDP43-G_S7.20350.2  ENSG00000180182.11_3   \n",
       "308           PAPA.TDP-2.25089.1  ENSG00000182518.14_1   \n",
       "309  PAPA.doxconc_NT_0_3.19568.4  ENSG00000196459.15_1   \n",
       "310       PAPA.Cont-D_S4.20482.2   ENSG00000197021.9_2   \n",
       "311   PAPA.TDP43_ctrl_2.26896.10   ENSG00000213468.7_1   \n",
       "\n",
       "                                                  Name  \n",
       "0    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "1    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "2    ENSG00000082497.12_1|SERTAD4|bleedthrough|back...  \n",
       "3    ENSG00000117682.17_3|DHDDS|bleedthrough|backgr...  \n",
       "4    ENSG00000117682.17_6|DHDDS|bleedthrough|backgr...  \n",
       "..                                                 ...  \n",
       "307    ENSG00000180182.11_3|MED14|bleedthrough|cryptic  \n",
       "308  ENSG00000182518.14_1|FAM104B|bleedthrough|back...  \n",
       "309  ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...  \n",
       "310  ENSG00000197021.9_2|EOLA2|bleedthrough|background  \n",
       "311     ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic  \n",
       "\n",
       "[312 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for spliced events, want to report the last exon coordinates (also split into 5'end & 3'end)\n",
    "# for this purpose, current coordiantes (from input last exons) is sufficient.\n",
    "# but for bleedthrough events, this will contain the complete last exon. So need to return to quant last exons to get the unique regions. \n",
    "\n",
    "# novel_le_rep_bleed - extract transcript_id (i.e. represents selected isoform) & downstream le_id (le_id_quant)\n",
    "novel_rep_bleed_tx2le = novel_le_rep_bleed.as_df()[[\"transcript_id\", \"le_id_quant\", \"Name\"]].rename(columns={\"le_id_quant\": \"le_id\"})\n",
    "novel_rep_bleed_tx2le\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atlas_max_datasets\n"
     ]
    }
   ],
   "source": [
    "for k,v in rep_choices_bleed.items():\n",
    "    if \"ENSG00000082497.12_1\" in v:\n",
    "        print(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210241557</td>\n",
       "      <td>210245767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246847</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-1.2120.4</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246813</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.CTRL-2.2002.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210245927</td>\n",
       "      <td>210246804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000082497.12_1|SERTAD4|bleedthrough|back...</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.2339.3</td>\n",
       "      <td>ENSG00000082497.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26438284</td>\n",
       "      <td>26439856</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117682.17_3|DHDDS|bleedthrough|backgr...</td>\n",
       "      <td>PAPA.TDP43-G_S7.362.1</td>\n",
       "      <td>ENSG00000117682.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>chrX</td>\n",
       "      <td>55142867</td>\n",
       "      <td>55146180</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182518.14_1|FAM104B|bleedthrough|back...</td>\n",
       "      <td>PAPA.TDP-2.25089.1</td>\n",
       "      <td>ENSG00000182518.14_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149931286</td>\n",
       "      <td>149932588</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197021.9_2|EOLA2|bleedthrough|background</td>\n",
       "      <td>PAPA.Cont-D_S4.20482.2</td>\n",
       "      <td>ENSG00000197021.9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825221</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>chrX</td>\n",
       "      <td>13715131</td>\n",
       "      <td>13716003</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000196459.15_1|TRAPPC2|bleedthrough|back...</td>\n",
       "      <td>PAPA.doxconc_NT_0_3.19568.4</td>\n",
       "      <td>ENSG00000196459.15_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 210241557 | 210245767 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246847 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246813 | .          | +            | ...   |\n",
       "| chr1         | 210245927 | 210246804 | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrX         | 55142867  | 55146180  | .          | -            | ...   |\n",
       "| chrX         | 149931286 | 149932588 | .          | -            | ...   |\n",
       "| chrX         | 131823777 | 131825221 | .          | -            | ...   |\n",
       "| chrX         | 13715131  | 13716003  | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 361 rows and 8 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, transcript_id, le_id"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to representative bleedthrough txs, joining in Name information\n",
    "# 5'coord = 1st cooridnate of intron (i.e. where spliced occurs)\n",
    "# 3'coord = final coordinate of last exon\n",
    "novel_le_rep_bleed_quant = quant_uniq_le.apply(lambda df: df.merge(novel_rep_bleed_tx2le, on=[\"transcript_id\", \"le_id\"], how=\"inner\"))\n",
    "\n",
    "novel_le_rep_bleed_quant[[\"Score\", \"Name\", \"transcript_id\", \"le_id\"]].drop_duplicate_positions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Repeat for annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of annotated bleedthrough last exons (assessed by DEXSeq) - 2288\n",
      "Number of annotated spliced last exons (assessed by DEXSeq) - 3561\n"
     ]
    }
   ],
   "source": [
    "# get le_ids in which all transcripts are annotated\n",
    "annotated_le_ids = dexseq_df.groupby(\"le_id\").filter(lambda df: (df.annot_status == \"annotated\").all()).le_id\n",
    "le_id_bleed_annot = set(le_id_bleed).intersection(set(annotated_le_ids))\n",
    "le_id_spliced_annot = set(le_id_spliced).intersection(set(annotated_le_ids))\n",
    "print(f\"Number of annotated bleedthrough last exons (assessed by DEXSeq) - {len(le_id_bleed_annot)}\")\n",
    "print(f\"Number of annotated spliced last exons (assessed by DEXSeq) - {len(le_id_spliced_annot)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset quantification GTF for spliecd last exons \n",
    "ref_le_spliced = quant_uniq_le.subset(lambda df: df.le_id.isin(le_id_spliced_annot))\n",
    "# select representative isoform for each last exon ID (most distal 3'end)\n",
    "ref_le_spliced = select_rep_prox_site(ref_le_spliced, \"le_id\")\n",
    "\n",
    "# subset quantification GTF for bleedthrough last exons \n",
    "ref_le_bleed = quant_uniq_le.subset(lambda df: df.le_id.isin(le_id_bleed_annot))\n",
    "# select representative isoform for each last exon ID (most distal 3'end)\n",
    "ref_le_bleed = select_rep_prox_site(ref_le_bleed, \"le_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          2275\n",
       "background    2176\n",
       "cryptic         39\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for annotatesd spliced last exons\n",
    "ref_le_spliced = ref_le_spliced.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id\"))\n",
    "ref_le_spliced.cryptic_status.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "background    1382\n",
       "NULL          1261\n",
       "cryptic         24\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign cryptic status for annotatesd bleedthrough last exons\n",
    "ref_le_bleed = ref_le_bleed.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids, ns_le_ids, id_col=\"le_id\"))\n",
    "ref_le_bleed.cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109214546</td>\n",
       "      <td>109215153</td>\n",
       "      <td>+</td>\n",
       "      <td>SARS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69873466</td>\n",
       "      <td>69875004</td>\n",
       "      <td>+</td>\n",
       "      <td>LRRC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6634019</td>\n",
       "      <td>6635586</td>\n",
       "      <td>+</td>\n",
       "      <td>THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7779625</td>\n",
       "      <td>7781432</td>\n",
       "      <td>+</td>\n",
       "      <td>VAMP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243305011</td>\n",
       "      <td>243306548</td>\n",
       "      <td>+</td>\n",
       "      <td>SDCCAG8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4485</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12859278</td>\n",
       "      <td>12860839</td>\n",
       "      <td>+</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4486</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4487</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4488</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4489</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13249882</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4490 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | ref_gene_name   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+-----------------|\n",
       "| chr1         | 109214546 | 109215153 | +            | SARS1           |\n",
       "| chr1         | 69873466  | 69875004  | +            | LRRC7           |\n",
       "| chr1         | 6634019   | 6635586   | +            | THAP3           |\n",
       "| chr1         | 7779625   | 7781432   | +            | VAMP3           |\n",
       "| ...          | ...       | ...       | ...          | ...             |\n",
       "| chrY         | 18932316  | 18932841  | -            | TTTY14          |\n",
       "| chrY         | 18872500  | 18872834  | -            | TTTY14          |\n",
       "| chrY         | 18872500  | 18872834  | -            | TTTY14          |\n",
       "| chrY         | 13248378  | 13249882  | -            | UTY             |\n",
       "+--------------+-----------+-----------+--------------+-----------------+\n",
       "Stranded PyRanges object has 4,490 rows and 5 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_le_spliced[[\"ref_gene_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign Name field with minimal info\n",
    "ref_le_spliced = (ref_le_spliced\n",
    "  .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    "         .assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"spliced\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")\n",
    "\n",
    "ref_le_bleed = (ref_le_bleed\n",
    " .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    ".assign(\"site_type\",\n",
    "                 lambda df: pd.Series([\"bleedthrough\"]*len(df.index), index=df.index))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "event_type\n",
      "last_exon_spliced                           1190\n",
      "first_exon_spliced                           525\n",
      "internal_exon_spliced                        247\n",
      "internal_exon_extension                      106\n",
      "first_exon_spliced,internal_exon_spliced     106\n",
      "first_exon_extension                          41\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "      <th>cryptic_status</th>\n",
       "      <th>gene_name_ref</th>\n",
       "      <th>site_type</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>39753286</td>\n",
       "      <td>39756769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000084072.17_2</td>\n",
       "      <td>background</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000084072.17_2|PPIE|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>39763688</td>\n",
       "      <td>39763914</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000084072.17</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000084072.17_3</td>\n",
       "      <td>background</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000084072.17_3|PPIE|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>170736047</td>\n",
       "      <td>170739421</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116132.12</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000116132.12</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116132.12_2</td>\n",
       "      <td>background</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116132.12_2|PRRX1|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>110232597</td>\n",
       "      <td>110249124</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116396.15</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>...</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000116396.15</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116396.15_2</td>\n",
       "      <td>background</td>\n",
       "      <td>KCNC4</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116396.15_2|KCNC4|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>32855923</td>\n",
       "      <td>32858875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000116497.18</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000116497.18</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000116497.18_3</td>\n",
       "      <td>background</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000116497.18_3|S100PBP|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>72274254</td>\n",
       "      <td>72274435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198034.11</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000198034.11</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000198034.11_1</td>\n",
       "      <td>background</td>\n",
       "      <td>RPS4X</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000198034.11_1|RPS4X|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1021</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130981589</td>\n",
       "      <td>130981856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000228659.1</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000228659.1</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000228659.1_3</td>\n",
       "      <td>background</td>\n",
       "      <td>LINC01201</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000228659.1_3|LINC01201|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>149527590</td>\n",
       "      <td>149528005</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7_1</td>\n",
       "      <td>background</td>\n",
       "      <td>EOLA1-DT</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000241489.8,ENSG00000241769.7_1|EOLA1-DT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1023</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000176728.10_3</td>\n",
       "      <td>background</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000176728.10_3|TTTY14|spliced|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>...</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000176728.10</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000176728.10_3</td>\n",
       "      <td>background</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>spliced</td>\n",
       "      <td>ENSG00000176728.10_3|TTTY14|spliced|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1025 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +21   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 39753286  | 39756769  | ...   |\n",
       "| chr1         | .          | exon         | 39763688  | 39763914  | ...   |\n",
       "| chr1         | .          | exon         | 170736047 | 170739421 | ...   |\n",
       "| chr1         | .          | exon         | 110232597 | 110249124 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | .          | exon         | 72274254  | 72274435  | ...   |\n",
       "| chrX         | .          | exon         | 130981589 | 130981856 | ...   |\n",
       "| chrX         | .          | exon         | 149527590 | 149528005 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,025 rows and 26 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "21 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 14 more.)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since last_exon_spliced annotations includes most distal last exon of a gene, let's compare these ALEs to non-terminal ALEs (i.e. there is a competing splicign event)\n",
    "print(ref_le_spliced.event_type.value_counts())\n",
    "\n",
    "ref_le_spliced = ref_le_spliced.subset(lambda df: df[\"event_type\"].ne(\"last_exon_spliced\"))\n",
    "ref_le_spliced\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245464258</td>\n",
       "      <td>245471621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616614</td>\n",
       "      <td>1619210</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112540178</td>\n",
       "      <td>112542090</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93783620</td>\n",
       "      <td>93784191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000137936.18_1|BCAR3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33591806</td>\n",
       "      <td>33592489</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000121904.19_3|CSMD2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20775449</td>\n",
       "      <td>20776596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000127483.19_1|HP1BP3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>163304210</td>\n",
       "      <td>163305313</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000232995.7_2|ENSG00000232995|spliced|cr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr2</td>\n",
       "      <td>135060280</td>\n",
       "      <td>135063187</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000115839.19_2|RAB3GAP1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr2</td>\n",
       "      <td>205776245</td>\n",
       "      <td>205779714</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000118257.17_5|NRP2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr2</td>\n",
       "      <td>3457794</td>\n",
       "      <td>3458181</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000171853.16_4|TRAPPC12|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>29480346</td>\n",
       "      <td>29480787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000171094.18_1|ALK|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>3129918</td>\n",
       "      <td>3130679</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000072756.17_1|TRNT1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr3</td>\n",
       "      <td>69304561</td>\n",
       "      <td>69310517</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000114541.15_3|FRMD4B|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr3</td>\n",
       "      <td>52731023</td>\n",
       "      <td>52732846</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000114904.13_3|NEK4|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr3</td>\n",
       "      <td>186103665</td>\n",
       "      <td>186104572</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244405.8_3|ETV5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr4</td>\n",
       "      <td>4560393</td>\n",
       "      <td>4562914</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000247708.8_2|STX18-AS1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr4</td>\n",
       "      <td>53141117</td>\n",
       "      <td>53143787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184178.17_1|SCFD2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr5</td>\n",
       "      <td>171271150</td>\n",
       "      <td>171271619</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000204764.14_9|RANBP17|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr6</td>\n",
       "      <td>99411735</td>\n",
       "      <td>99412318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000132424.17_1|PNISR|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444214</td>\n",
       "      <td>43446033</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000002746.15_3|HECW1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr7</td>\n",
       "      <td>154587649</td>\n",
       "      <td>154588631</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000130226.17_3|DPP6|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr7</td>\n",
       "      <td>151408559</td>\n",
       "      <td>151409053</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000187260.16_1|WDR86|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617073</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617049</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr8</td>\n",
       "      <td>13195167</td>\n",
       "      <td>13214777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000164741.15_3|DLC1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93662001</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr9</td>\n",
       "      <td>95110511</td>\n",
       "      <td>95111345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000158169.14_3|FANCC|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>132754243</td>\n",
       "      <td>132757132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000068383.19_2|INPP5A|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr10</td>\n",
       "      <td>19818296</td>\n",
       "      <td>19819344</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000120594.17_1|PLXDC2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr10</td>\n",
       "      <td>132352881</td>\n",
       "      <td>132353396</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000148814.18_3|LRRC27|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr10</td>\n",
       "      <td>48452047</td>\n",
       "      <td>48452797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000128805.15_4|ARHGAP22|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr10</td>\n",
       "      <td>1636901</td>\n",
       "      <td>1645966</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185736.16_1|ADARB2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr10</td>\n",
       "      <td>49102171</td>\n",
       "      <td>49103860</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000165633.13_1|VSTM4|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr11</td>\n",
       "      <td>66646026</td>\n",
       "      <td>66646468</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991449</td>\n",
       "      <td>128992046</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991448</td>\n",
       "      <td>128992046</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr11</td>\n",
       "      <td>7018036</td>\n",
       "      <td>7019151</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000149050.10_1|ZNF214|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr12</td>\n",
       "      <td>39571999</td>\n",
       "      <td>39573109</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000173208.4_1|ABCD2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr12</td>\n",
       "      <td>123516660</td>\n",
       "      <td>123518494</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188026.13_1|RILPL1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr12</td>\n",
       "      <td>6844792</td>\n",
       "      <td>6846724</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000111665.12_4|CDCA3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr13</td>\n",
       "      <td>97338646</td>\n",
       "      <td>97341757</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000139793.20_1|MBNL2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr14</td>\n",
       "      <td>31433709</td>\n",
       "      <td>31438140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000203546.7,ENSG00000129480.13_2|ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr14</td>\n",
       "      <td>31433728</td>\n",
       "      <td>31438140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000203546.7,ENSG00000129480.13_2|ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr15</td>\n",
       "      <td>52774457</td>\n",
       "      <td>52777690</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000169856.9_1|ONECUT1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr15</td>\n",
       "      <td>75727496</td>\n",
       "      <td>75732152</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000246877.2_2|DNM1P35|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30413586</td>\n",
       "      <td>30414186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179965.12_1|ZNF771|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30413586</td>\n",
       "      <td>30414223</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179965.12_1|ZNF771|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr17</td>\n",
       "      <td>5232731</td>\n",
       "      <td>5236002</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000261879.8_2|ZNF594-DT|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr17</td>\n",
       "      <td>20236859</td>\n",
       "      <td>20239427</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr17</td>\n",
       "      <td>80436132</td>\n",
       "      <td>80438086</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000173818.17_5|ENDOV|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr17</td>\n",
       "      <td>38544577</td>\n",
       "      <td>38545459</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277363.5_2|SRCIN1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr18</td>\n",
       "      <td>4145795</td>\n",
       "      <td>4147667</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000170579.17_1|DLGAP1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr19</td>\n",
       "      <td>52719723</td>\n",
       "      <td>52721267</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213020.10_1|ZNF611|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr20</td>\n",
       "      <td>5817484</td>\n",
       "      <td>5821165</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000171984.15_2|SHLD1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr20</td>\n",
       "      <td>44708787</td>\n",
       "      <td>44709470</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244558.6_1|KCNK15-AS1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr21</td>\n",
       "      <td>45792478</td>\n",
       "      <td>45794699</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000183570.17_3|PCBP3|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98679426</td>\n",
       "      <td>98679978</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 76871267  | 76871821  | .          | +            | ...   |\n",
       "| chr1         | 245464258 | 245471621 | .          | +            | ...   |\n",
       "| chr1         | 1616614   | 1619210   | .          | +            | ...   |\n",
       "| chr1         | 112540178 | 112542090 | .          | -            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chr20        | 5817484   | 5821165   | .          | +            | ...   |\n",
       "| chr20        | 44708787  | 44709470  | .          | -            | ...   |\n",
       "| chr21        | 45792478  | 45794699  | .          | +            | ...   |\n",
       "| chrX         | 108267641 | 108269152 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 60 rows and 6 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined output BEDs\n",
    "spliced_out_bed = pr.concat([novel_le_rep_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions(),\n",
    "                             ref_le_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions()])\n",
    "\n",
    "bleed_out_bed = pr.concat([novel_le_rep_bleed_quant[[\"Score\", \"Name\"]].drop_duplicate_positions(),\n",
    "                             ref_le_bleed[[\"Score\", \"Name\"]].drop_duplicate_positions()])\n",
    "\n",
    "spliced_out_bed.subset(lambda df: df.Name.str.contains(\"cryptic$\"))\n",
    "\n",
    "# novel_le_rep_bleed_quant[[\"Score\", \"Name\"]].drop_duplicate_positions()\n",
    "# novel_le_rep_spliced[[\"Score\", \"Name\"]].drop_duplicate_positions()\n",
    "\n",
    "\n",
    "# generate output bed file of proximal and distal PAS\n",
    "# utr3_out_bed = pr.concat([sel_le_d3utr_prox[[\"Score\", \"Name\"]],\n",
    "#                           sel_novel_le_d3utr[[\"Score\", \"Name\"]]]\n",
    "#                           )\n",
    "\n",
    "# print(len(utr3_out_bed))\n",
    "# # drop duplicate entries\n",
    "# utr3_out_bed = utr3_out_bed.apply(lambda df: df.drop_duplicates()).sort()\n",
    "\n",
    "# utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45013791</td>\n",
       "      <td>45013889</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000126088.14_1|UROD|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616614</td>\n",
       "      <td>1619210</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000197530.13_1|MIB2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112540178</td>\n",
       "      <td>112540778</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000007341.19_4|ST7L|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112540872</td>\n",
       "      <td>112541950</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000007341.19_4|ST7L|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20775449</td>\n",
       "      <td>20776596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000127483.19_1|HP1BP3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>225651535</td>\n",
       "      <td>225658565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000144460.13_2|NYAP2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>216451064</td>\n",
       "      <td>216451601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000138375.13_1|SMARCAL1|bleedthrough|cry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631010</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11_1|RFTN2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr3</td>\n",
       "      <td>52414587</td>\n",
       "      <td>52416161</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000010318.22_2|PHF7|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr3</td>\n",
       "      <td>113425851</td>\n",
       "      <td>113426123</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206530.11_1|CFAP44|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr3</td>\n",
       "      <td>183298426</td>\n",
       "      <td>183300004</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000053524.13_2|MCF2L2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr4</td>\n",
       "      <td>3430955</td>\n",
       "      <td>3432381</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159788.20_8|RGS12|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr4</td>\n",
       "      <td>83458429</td>\n",
       "      <td>83459144</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000163319.11_1|MRPS18C|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>121826264</td>\n",
       "      <td>121828145</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000138686.10_4|BBS7|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>171271150</td>\n",
       "      <td>171272579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000204764.14_10|RANBP17|bleedthrough|cry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>99411735</td>\n",
       "      <td>99412318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000132424.17_1|PNISR|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444214</td>\n",
       "      <td>43444217</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000002746.15_3|HECW1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43445570</td>\n",
       "      <td>43446033</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000002746.15_3|HECW1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr7</td>\n",
       "      <td>29400828</td>\n",
       "      <td>29401281</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000106069.24_2|CHN2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135159776</td>\n",
       "      <td>135166871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000122783.17_4|CYREN|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135159785</td>\n",
       "      <td>135166871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000122783.17_4|CYREN|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135159792</td>\n",
       "      <td>135166871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000122783.17_4|CYREN|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135159794</td>\n",
       "      <td>135166871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000122783.17_4|CYREN|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135159813</td>\n",
       "      <td>135166871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000122783.17_4|CYREN|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000158604.15_1|TMED4|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr9</td>\n",
       "      <td>128515639</td>\n",
       "      <td>128516262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000119392.16_1|GLE1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137834772</td>\n",
       "      <td>137836136</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181090.21_6|EHMT1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr10</td>\n",
       "      <td>11927500</td>\n",
       "      <td>11929864</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000151461.20_1|UPF2|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr10</td>\n",
       "      <td>119581222</td>\n",
       "      <td>119581921</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000151923.18_1|TIAL1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121959917</td>\n",
       "      <td>121960357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107672.15_3|NSMCE4A|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr10</td>\n",
       "      <td>77793543</td>\n",
       "      <td>77794007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000151208.17_3|DLG5|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr11</td>\n",
       "      <td>47326807</td>\n",
       "      <td>47327186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000110514.19_3|MADD|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr11</td>\n",
       "      <td>6604360</td>\n",
       "      <td>6604726</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166333.14_1|ILK|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr11</td>\n",
       "      <td>125982109</td>\n",
       "      <td>125983871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000064309.16_2|CDON|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr12</td>\n",
       "      <td>6844792</td>\n",
       "      <td>6846724</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000111665.12_4|CDCA3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000186047.11_2|DLEU7|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr13</td>\n",
       "      <td>87583688</td>\n",
       "      <td>87585080</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228824.8_4|MIR4500HG|bleedthrough|cry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr14</td>\n",
       "      <td>60808428</td>\n",
       "      <td>60809015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000020426.11_2|MNAT1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr14</td>\n",
       "      <td>100845533</td>\n",
       "      <td>100847301</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000214548.18_1|MEG3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr14</td>\n",
       "      <td>23051580</td>\n",
       "      <td>23051969</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000139880.21_1|CDH24|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr14</td>\n",
       "      <td>23051613</td>\n",
       "      <td>23051969</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000139880.21_1|CDH24|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23079945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1|USP31|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr16</td>\n",
       "      <td>15047215</td>\n",
       "      <td>15047441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000157045.9_1|NTAN1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr17</td>\n",
       "      <td>75648864</td>\n",
       "      <td>75650516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000108469.15_2|RECQL5|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr18</td>\n",
       "      <td>79467582</td>\n",
       "      <td>79470014</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131196.18_2|NFATC1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr18</td>\n",
       "      <td>8809561</td>\n",
       "      <td>8810002</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000168502.18_3|MTCL1|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000101624.11_3|CEP76|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr19</td>\n",
       "      <td>21375964</td>\n",
       "      <td>21376755</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000172687.14_3|ZNF738|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr21</td>\n",
       "      <td>42849225</td>\n",
       "      <td>42850040</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000160193.12_1|WDR4|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825221</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 45013791  | 45013889  | .          | +            | ...   |\n",
       "| chr1         | 1616614   | 1619210   | .          | +            | ...   |\n",
       "| chr1         | 112540178 | 112540778 | .          | -            | ...   |\n",
       "| chr1         | 112540872 | 112541950 | .          | -            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chr19        | 21375964  | 21376755  | .          | +            | ...   |\n",
       "| chr21        | 42849225  | 42850040  | .          | -            | ...   |\n",
       "| chrX         | 40653641  | 40654363  | .          | -            | ...   |\n",
       "| chrX         | 131823777 | 131825221 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 51 rows and 6 columns from 19 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bleed_out_bed.subset(lambda df: df.Name.str.contains(\"cryptic$\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"../processed/iclip_regions\"):\n",
    "    os.makedirs(\"../processed/iclip_regions\")\n",
    "\n",
    "#### TEMPORARY - work out why duplicate regions/multiple PAS coords for some last exons. To prevent duplciats slect longest per interval\n",
    "spliced_out_bed = select_rep_prox_site(spliced_out_bed, \"Name\")\n",
    "bleed_out_bed = select_rep_prox_site(bleed_out_bed, \"Name\")\n",
    "\n",
    "spliced_out_bed.to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "spliced_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.pas.bed\")\n",
    "# also first coordinate of exon (i.e. interval ending at this coordinate = splice site end, or substract 1 (strand-aware) to get final nucleotide of splice site)\n",
    "spliced_out_bed.five_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_spliced.le_start.bed\")\n",
    "\n",
    "bleed_out_bed.to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "bleed_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.pas.bed\")\n",
    "# also first coordinate of exon (i.e. 1st nucleotide of intron\n",
    "bleed_out_bed.five_end().to_bed(\"../processed/iclip_regions/2023-06-26_papa_cryptic_bleedthrough.le_start.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
