{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# approach\n",
    "\n",
    "### Input last exons - selecting representative PAS/last exon:\n",
    "\n",
    "le_ids can have multiple predictions, need to select a single one.\n",
    "Strategy:\n",
    "(starting with novel last exons used for quantification) \n",
    "1. Prefer events that match atlas\n",
    "2. if multiple, select min atlas distance\n",
    "3. If matches, s\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Because quant regions \n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
      "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
      "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
      "       'le_number', 'le_id'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074713_S59.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_3.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24475256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24473773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-B_S2.334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24474852</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_3.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36889</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074729_S12.25635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36890</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36891</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.26168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36892</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234583</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_0_2.21889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36893</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234594</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.24449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36894 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24475256  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24473773  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234583  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234594  | 13234826  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 36,894 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(quant_uniq_le.columns)\n",
    "quant_uniq_le.subset(lambda df: ~df.transcript_id.str.startswith(\"ENST\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['experiment_name', 'binID', 'groupID', 'featureID', 'exonBaseMean',\n",
       "       'dispersion', 'stat', 'pvalue', 'padj', 'UsageCoefficient_base',\n",
       "       'UsageCoefficient_contrast', 'log2fold_contrast_base', 'le_id',\n",
       "       'gene.qvalue', 'contrast_name', 'gene_id', 'gene_name', 'event_type',\n",
       "       'annot_status', 'transcript_id', 'chromosome', 'strand', 'start', 'end',\n",
       "       'mean_PPAU_base', 'mean_PPAU_treatment', 'delta_PPAU_treatment_control',\n",
       "       'simple_event_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
       "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
       "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
       "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
       "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
       "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_rep_site(df):\n",
    "\n",
    "\n",
    "    # First select sites of le_id that match atlas site (if applicable)\n",
    "    if df[\"atlas_filter\"].astype(int).sum() > 0:\n",
    "\n",
    "        out_df = df[df[\"atlas_filter\"].eq(\"1\")]\n",
    "\n",
    "        # 2. Within these, see if multiple different 3'ends predicted\n",
    "        ### TODO: use Name (atlas ID) instead of 3' coordinate to match\n",
    "\n",
    "        if out_df[\"Strand\"].unique()[0] == \"+\":\n",
    "            \n",
    "            # plus strand - End coord = 3'end of last exon\n",
    "            n_atlas = df[\"End\"].nunique()\n",
    "\n",
    "            if n_atlas > 1:\n",
    "                # multiple different atlas sites predicted\n",
    "                # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                end3_c = out_df[\"End\"].value_counts()\n",
    "                end3_c = end3_c[end3_c.values == end3_c.max()]\n",
    "\n",
    "                if len(end3_c) > 1:\n",
    "                    # multiple 3'ends/ atlas sites predicted\n",
    "                    raise Exception(f\"{out_df['le_id_quant'].drop_duplicates()[0]} - le_id has multiple predicted atlas sites\")\n",
    "                                    \n",
    "                else:\n",
    "                    # 1 pas remaining\n",
    "                    end3_coord = end3_c.index[0]\n",
    "\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df[out_df[\"End\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "\n",
    "                    decision_str = \"atlas_max_datasets\"\n",
    "\n",
    "\n",
    "            else:\n",
    "                # only single atlas site predicted - just select a single row\n",
    "                end3_coord = out_df[\"End\"].drop_duplicates().index[0]\n",
    "                # Select a single row for le_id corresponding to selected PAS\n",
    "                out_df = out_df[out_df[\"End\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "\n",
    "                decision_str = \"atlas_1_pred\"\n",
    "                            \n",
    "        else:\n",
    "            # minus strand - Start coord = 3'end of last exon\n",
    "            n_atlas = out_df[\"Start\"].nunique()\n",
    "            if n_atlas > 1:\n",
    "                # multiple different atlas sites predicted\n",
    "                # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                end3_c = out_df[\"Start\"].value_counts()\n",
    "                end3_c = end3_c[end3_c.values == end3_c.max()]\n",
    "\n",
    "                if len(end3_c) > 1:\n",
    "                    # multiple 3'ends predicted\n",
    "                    raise Exception(f\"{out_df['le_id_quant'].drop_duplicates().iloc[0]} - le_id has multiple predicted atlas sites in same number of datasets\")\n",
    "                                    \n",
    "                else:\n",
    "                    # 1 pas remaining\n",
    "                    end3_coord = end3_c.index[0]\n",
    "\n",
    "                    # Select a single row for le_id corresponding to selected PAS\n",
    "                    out_df = out_df[out_df[\"Start\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "\n",
    "                    decision_str = \"atlas_max_datasets\"\n",
    "\n",
    "            else:\n",
    "                # only single atlas site predicted - just select a single row\n",
    "                end3_coord = out_df[\"Start\"].drop_duplicates().index[0]\n",
    "                # Select a single row for le_id corresponding to selected PAS\n",
    "                out_df = out_df[out_df[\"Start\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "\n",
    "                decision_str = \"atlas_1_pred\"\n",
    "\n",
    "        return out_df, decision_str    \n",
    "    \n",
    "    else:\n",
    "        # TODO: implement motif filtering\n",
    "        return df, \"motif_only\"\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000000003.15_2\n",
      "ENSG00000000419.14_2\n",
      "ENSG00000001036.14_3\n",
      "ENSG00000001084.13_4\n",
      "ENSG00000001167.15_2\n",
      "ENSG00000001460.18_6\n",
      "ENSG00000001461.17_6\n",
      "ENSG00000001561.7_1\n",
      "ENSG00000001629.10_3\n",
      "ENSG00000001631.17_3\n"
     ]
    },
    {
     "ename": "Exception",
     "evalue": "ENSG00000001631.17_3 - le_id has multiple predicted atlas sites in same number of datasets",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[72], line 20\u001b[0m\n\u001b[1;32m     18\u001b[0m le_id, df \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39m(le_grpd)\n\u001b[1;32m     19\u001b[0m \u001b[39mprint\u001b[39m(le_id)\n\u001b[0;32m---> 20\u001b[0m out_df, decision_str \u001b[39m=\u001b[39m select_rep_site(df)\n\u001b[1;32m     21\u001b[0m sel_dfs\u001b[39m.\u001b[39mappend(out_df)\n\u001b[1;32m     22\u001b[0m sel_dict[decision_str]\u001b[39m.\u001b[39mappend(le_id)\n",
      "Cell \u001b[0;32mIn[71], line 54\u001b[0m, in \u001b[0;36mselect_rep_site\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m     50\u001b[0m end3_c \u001b[39m=\u001b[39m end3_c[end3_c\u001b[39m.\u001b[39mvalues \u001b[39m==\u001b[39m end3_c\u001b[39m.\u001b[39mmax()]\n\u001b[1;32m     52\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(end3_c) \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m     53\u001b[0m     \u001b[39m# multiple 3'ends predicted\u001b[39;00m\n\u001b[0;32m---> 54\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mout_df[\u001b[39m'\u001b[39m\u001b[39mle_id\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mdrop_duplicates()\u001b[39m.\u001b[39miloc[\u001b[39m0\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m - le_id has multiple predicted atlas sites in same number of datasets\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     56\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m     57\u001b[0m     \u001b[39m# 1 pas remaining\u001b[39;00m\n\u001b[1;32m     58\u001b[0m     end3_coord \u001b[39m=\u001b[39m end3_c\u001b[39m.\u001b[39mindex[\u001b[39m0\u001b[39m]\n",
      "\u001b[0;31mException\u001b[0m: ENSG00000001631.17_3 - le_id has multiple predicted atlas sites in same number of datasets"
     ]
    }
   ],
   "source": [
    "#### SELECTING REPRESENTATIVE PAS\n",
    "# LEs have to be quantified\n",
    "novel_le_df = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"])).as_df()\n",
    "\n",
    "\n",
    "i = 0\n",
    "\n",
    "sel_dfs = []\n",
    "sel_dict = {\"atlas_1_pred\": [],\n",
    "            \"atlas_max_datasets\": [],\n",
    "            \"motif_only\": []}\n",
    "\n",
    "le_grpd = iter(novel_le_df.groupby(\"le_id_quant\"))\n",
    "\n",
    "\n",
    "\n",
    "while i < 100:\n",
    "        le_id, df = next(le_grpd)\n",
    "        print(le_id)\n",
    "        out_df, decision_str = select_rep_site(df)\n",
    "        sel_dfs.append(out_df)\n",
    "        sel_dict[decision_str].append(le_id)\n",
    "        i +=1  \n",
    "\n",
    "\n",
    "\n",
    "                \n",
    "    \n",
    "print({dec: len(ids) for dec, ids in sel_dict.items()})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
      "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
      "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
      "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
      "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
      "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU',\n",
      "       'le_id_quant'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>atlas_filter</th>\n",
       "      <th>Name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14815</th>\n",
       "      <td>92198195</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92197498:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14816</th>\n",
       "      <td>92198602</td>\n",
       "      <td>92200804</td>\n",
       "      <td>0</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14817</th>\n",
       "      <td>92199253</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14818</th>\n",
       "      <td>92199257</td>\n",
       "      <td>92200804</td>\n",
       "      <td>1</td>\n",
       "      <td>7:92199243:-</td>\n",
       "      <td>ENSG00000001631.17_3</td>\n",
       "      <td>KRIT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Start       End atlas_filter          Name                 le_id  \\\n",
       "14815  92198195  92200804            0  7:92197498:-  ENSG00000001631.17_3   \n",
       "14816  92198602  92200804            0  7:92199243:-  ENSG00000001631.17_3   \n",
       "14817  92199253  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "14818  92199257  92200804            1  7:92199243:-  ENSG00000001631.17_3   \n",
       "\n",
       "      gene_name_ref  \n",
       "14815         KRIT1  \n",
       "14816         KRIT1  \n",
       "14817         KRIT1  \n",
       "14818         KRIT1  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"].columns)\n",
    "novel_le_df[novel_le_df[\"le_id_quant\"] == \"ENSG00000001631.17_3\"][[\"Start\", \"End\", \"atlas_filter\", \"Name\", \"le_id\", \"gene_name_ref\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for le_id, df in novel_le_df.groupby(\"le_id_quant\"):\n",
    "        while i < 5:\n",
    "                print(le_id)\n",
    "                #1. select PAS passing atlas filter only (if applicable)\n",
    "                print(df)\n",
    "                print(df.Strand.unique()[0])\n",
    "                print(len(df))\n",
    "\n",
    "                if df[\"atlas_filter\"].astype(int).sum() > 0:\n",
    "                        out_df = df[df[\"atlas_filter\"].eq(\"1\")]\n",
    "\n",
    "                        # 2. Within these, see if multiple different 3'ends predicted\n",
    "                        if out_df[\"Strand\"].unique()[0] == \"+\":\n",
    "                                n_atlas = df[\"End\"].nunique()\n",
    "\n",
    "                                if n_atlas > 1:\n",
    "                                        # multiple different atlas sites predicted\n",
    "                                        # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                                        end3_c = out_df[\"End\"].value_counts()\n",
    "                                        print(\"end3_c\")\n",
    "                                        print(end3_c)\n",
    "                                        end3_c = end3_c[end3_c.values == end3_c.max()]\n",
    "                                        print(end3_c)\n",
    "\n",
    "                                else:\n",
    "                                        # only single atlas site predicted - just select a single row\n",
    "                                        pass\n",
    "                            \n",
    "                        else:\n",
    "                                # minus strand - Start coord = 3'end of last exon\n",
    "                                n_atlas = out_df[\"Start\"].nunique()\n",
    "                                if n_atlas > 1:\n",
    "                                        # multiple different atlas sites predicted\n",
    "                                        # see if somes sites were called in > 1 dataset - select those with max number of datasets called\n",
    "                                        end3_c = out_df[\"Start\"].value_counts()\n",
    "                                        print(\"end3_c\")\n",
    "                                        print(end3_c)\n",
    "                                        end3_c = end3_c[end3_c.values == end3_c.max()]\n",
    "                                        print(end3_c)\n",
    "\n",
    "                                        if len(end3_c) > 1:\n",
    "                                            # multiple 3'ends predicted\n",
    "                                            pass\n",
    "                                    \n",
    "                                        else:\n",
    "                                            # 1 pas remaining\n",
    "                                            end3_coord = end3_c.index[0]\n",
    "\n",
    "                                            # Select a single row for le_id corresponding to selected PAS\n",
    "                                            out_df = out_df[out_df[\"Start\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "\n",
    "\n",
    "                                else:\n",
    "                                        # only single atlas site predicted - just select a single row\n",
    "                                        end3_coord = out_df[\"Start\"].drop_duplicates().index[0]\n",
    "                                        # Select a single row for le_id corresponding to selected PAS\n",
    "                                        out_df = out_df[out_df[\"Start\"] == end3_coord].sort_values(by=\"experiment_id\").drop_duplicates()\n",
    "                \n",
    "                print(len(out_df))\n",
    "                \n",
    "                i +=1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
