{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from define_helpers import select_rep_site, _define_cryptic_status, select_rep_prox_site, select_rep_five_end\n",
    "import os\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# approach\n",
    "\n",
    "### Input last exons - selecting representative PAS/last exon:\n",
    "\n",
    "le_ids can have multiple predictions, need to select a single one.\n",
    "Strategy:\n",
    "(starting with novel last exons used for quantification) \n",
    "1. Prefer events that match atlas\n",
    "2. if multiple, select min atlas distance\n",
    "3. If matches, s\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## To-dos:\n",
    "- fix updating to atlas for -ve strand matching pas\n",
    "- run selection for all novel 3'UTR extensions\n",
    "- Implement selection for annotated proximal sites of 3'UTR extensions (i.e. most distal annotated site)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input novel last exons used to generate combined reference of last exons\n",
    "novel_le = pr.read_gtf(\"../data/papa/2023-03-29_papa_i3_cortical_upf1_zanovello_overlap_annotated.gtf\")\n",
    "# dexseq results df (used to extract cryptic events of each class)\n",
    "dexseq_df = pd.read_csv(\"../data/papa/2023-05-24_i3_cortical_zanovello.all_datasets.dexseq_apa.results.processed.cleaned.tsv\", sep=\"\\t\")\n",
    "\n",
    "cryptics_df = pd.read_csv(\"../../preprocessing/processed/cryptics_summary_all_events.tsv\", sep=\"\\t\")\n",
    "\n",
    "# refeence GTF used to identify novel last exons, quantify vs ref\n",
    "ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n",
    "# \n",
    "tx2le = pd.read_csv(\"../data/papa/novel_ref_combined.tx2le.tsv\", sep=\"\\t\")\n",
    "\n",
    "# last exon quantification regions used as input to Salmon\n",
    "quant_uniq_le = pr.read_gtf(\"../data/papa/novel_ref_combined.quant.last_exons.gtf\")\n",
    "\n",
    "\n",
    "info_df = pd.read_csv(\"../data/papa/novel_ref_combined.info.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
      "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
      "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
      "       'le_number', 'le_id'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074713_S59.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_3.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24475256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24473773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-B_S2.334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24474852</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_3.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000001461.17_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36889</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074729_S12.25635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36890</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36891</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.26168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36892</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234583</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_0_2.21889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36893</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13234594</td>\n",
       "      <td>13234826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.24449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000183878.16</td>\n",
       "      <td>UTY,UTY,UTY,UTY,UTY,UTY,UTY,UTY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000183878.16_7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36894 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24476667  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24475256  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24473773  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234576  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234583  | 13234826  | ...   |\n",
       "| chrY         | .          | exon         | 13234594  | 13234826  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 36,894 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(quant_uniq_le.columns)\n",
    "quant_uniq_le.subset(lambda df: ~df.transcript_id.str.startswith(\"ENST\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['experiment_name', 'binID', 'groupID', 'featureID', 'exonBaseMean',\n",
       "       'dispersion', 'stat', 'pvalue', 'padj', 'UsageCoefficient_base',\n",
       "       'UsageCoefficient_contrast', 'log2fold_contrast_base', 'le_id',\n",
       "       'gene.qvalue', 'contrast_name', 'gene_id', 'gene_name', 'event_type',\n",
       "       'annot_status', 'transcript_id', 'chromosome', 'strand', 'start', 'end',\n",
       "       'mean_PPAU_base', 'mean_PPAU_treatment', 'delta_PPAU_treatment_control',\n",
       "       'simple_event_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dexseq_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'transcript_id', 'event_type',\n",
       "       '3p_extension_length', 'atlas_filter', 'nearest_atlas_distance', 'Name',\n",
       "       'motif_filter', 'pas_motifs', 'condition_id', 'le_id', 'gene_name_ref',\n",
       "       'gene_id_ref', 'mean_PPAU_base', 'mean_PPAU_treatment',\n",
       "       'delta_PPAU_treatment_control', 'experiment_id', 'Cluster',\n",
       "       'experiment_count', 'multiple_datasets', 'max_mean_PPAU'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3369"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get le_ids that are novel 3'UTR ectensions\n",
    "le_id_d3utr = set(dexseq_df[dexseq_df.simple_event_type == \"distal_3utr_extension\"].le_id)\n",
    "len(le_id_d3utr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic 3'UTR extensions - 104\n",
      "Number of cryptic 3'UTR extension-containing genes - 104\n"
     ]
    }
   ],
   "source": [
    "cryp_le_ids = set(cryptics_df.loc[cryptics_df[\"simple_event_type\"] == \"distal_3utr_extension\", \"le_id\"])\n",
    "cryp_gene_names = set(cryptics_df.loc[cryptics_df[\"simple_event_type\"] == \"distal_3utr_extension\", \"gene_name\"])\n",
    "\n",
    "print(f\"Number of cryptic 3'UTR extensions - {len(cryp_le_ids)}\")\n",
    "print(f\"Number of cryptic 3'UTR extension-containing genes - {len(cryp_gene_names)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add le_id used in downstream analysis\n",
    "novel_le = novel_le.apply(lambda df: df.merge(tx2le, on=\"transcript_id\", how=\"inner\", suffixes=[None, \"_quant\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of novel 3'UTR extenstion intervals before selecting representative LEs - 15810\n",
      "Number of novel 3'UTR extension intervals after selecting representative LEs - 3398\n",
      "{'atlas_1_pred': 1549, 'atlas_max_datasets': 534, 'atlas_max_datasets_shortest': 344, 'motif_1_min': 889, 'motif_shortest_min': 53}\n"
     ]
    }
   ],
   "source": [
    "# Select representative PAS for distal 3'UTR extension events (novel)\n",
    "novel_le_rep_d3utr, rep_choices_d3utr = select_rep_site(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_d3utr))), id_col=\"le_id_quant\")\n",
    "\n",
    "\n",
    "print(f\"Number of novel 3'UTR extenstion intervals before selecting representative LEs - {len(novel_le.subset(lambda df: df.le_id_quant.isin(set(le_id_d3utr))))}\")\n",
    "print(f\"Number of novel 3'UTR extension intervals after selecting representative LEs - {len(novel_le_rep_d3utr)}\")\n",
    "print({dec: len(ids) for dec, ids in rep_choices_d3utr.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       0  1\n",
      "6     ENSG00000001461.17  6\n",
      "85    ENSG00000023902.14  3\n",
      "98    ENSG00000031698.13  3\n",
      "102   ENSG00000033122.21  3\n",
      "143    ENSG00000053372.5  3\n",
      "...                  ... ..\n",
      "1137  ENSG00000129824.16  2\n",
      "1798   ENSG00000154620.6  2\n",
      "2091  ENSG00000165246.15  3\n",
      "60    ENSG00000012817.16  3\n",
      "3081   ENSG00000230663.1  2\n",
      "\n",
      "[3398 rows x 2 columns]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6       ENSG00000001461.17_5\n",
       "85      ENSG00000023902.14_2\n",
       "98      ENSG00000031698.13_2\n",
       "102     ENSG00000033122.21_2\n",
       "143      ENSG00000053372.5_2\n",
       "                ...         \n",
       "1137    ENSG00000129824.16_1\n",
       "1798     ENSG00000154620.6_1\n",
       "2091    ENSG00000165246.15_2\n",
       "60      ENSG00000012817.16_2\n",
       "3081     ENSG00000230663.1_1\n",
       "Name: 0, Length: 3398, dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to select representative proximal site for each distal site, since quantification approach lumps together all annotated proximal sites (and compares them to extension)\n",
    "# since le_ids are annotated sequentially, know that immediately preceding le_number is the partner le_id\n",
    "d3utr_le_id_split = novel_le_rep_d3utr.le_id.str.split(\"_\", regex=False, expand=True)\n",
    "print(d3utr_le_id_split)\n",
    "d3utr_le_id_split[1] = d3utr_le_id_split[1].astype(int).subtract(1)\n",
    "\n",
    "# reconstruct le_id (corresponding to proximal le_id)\n",
    "d3utr_prox_le_id = d3utr_le_id_split[0].str.cat(d3utr_le_id_split[1].astype(str), sep=\"_\")\n",
    "d3utr_prox_le_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150159404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150159497</td>\n",
       "      <td>150160065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150159404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>133.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705416</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11575</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11576</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19705424</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11577</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>17574730</td>\n",
       "      <td>17575367</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.22119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000230663.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11578 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 150158818 | 150159404 | ...   |\n",
       "| chr1         | .          | exon         | 150159497 | 150160065 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 19705416  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19705419  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19705424  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 17574730  | 17575367  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 11,578 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset for annotated last exons with 3'end proximal to 3'UTR extensions\n",
    "quant_uniq_le_d3utr_prox = quant_uniq_le.subset(lambda df: df.le_id.isin(set(d3utr_prox_le_id.values)))\n",
    "quant_uniq_le_d3utr_prox\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>150159497</td>\n",
       "      <td>150160065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>150159404</td>\n",
       "      <td>150159497</td>\n",
       "      <td>ENST00000502767.1</td>\n",
       "      <td>568.0</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000023902.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>109237730</td>\n",
       "      <td>109238182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>109237373</td>\n",
       "      <td>109237730</td>\n",
       "      <td>ENST00000468588.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000031698.13</td>\n",
       "      <td>SARS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000031698.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>69873466</td>\n",
       "      <td>69875004</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000033122.21</td>\n",
       "      <td>LRRC7</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>69838283</td>\n",
       "      <td>69931506</td>\n",
       "      <td>ENST00000415775.2</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000033122.21</td>\n",
       "      <td>LRRC7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000033122.21_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4289</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>13704435</td>\n",
       "      <td>13705224</td>\n",
       "      <td>ENST00000284856.4</td>\n",
       "      <td>NULL</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000154620.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4290</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845650</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000165246.15</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>14830519,14830519,14830519,14830519</td>\n",
       "      <td>14840412,14840412,14840412,14840412</td>\n",
       "      <td>ENST00000339174.9,ENST00000382868.5,ENST000003...</td>\n",
       "      <td>NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000165246.15</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000165246.15_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4291</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4292</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>19706345,19706345,19706345,19706345,19706345</td>\n",
       "      <td>19706440,19706440,19706440,19706440,19706440</td>\n",
       "      <td>ENST00000317961.9,ENST00000382806.6,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000012817.16_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4293</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>17574730</td>\n",
       "      <td>17575367</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.22119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000230663.1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000230663.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4294 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 24468985  | 24472976  | ...   |\n",
       "| chr1         | .          | exon         | 150159497 | 150160065 | ...   |\n",
       "| chr1         | .          | exon         | 109237730 | 109238182 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 14840412  | 14845650  | ...   |\n",
       "| chrY         | .          | exon         | 19703864  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 19703864  | 19706345  | ...   |\n",
       "| chrY         | .          | exon         | 17574730  | 17575367  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 4,294 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select representative proximal site for partner annotated ALEs (most distal of annotated PAS)\n",
    "sel_le_d3utr_prox = select_rep_prox_site(quant_uniq_le_d3utr_prox, id_col=\"le_id\")\n",
    "sel_le_d3utr_prox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptic 3'UTR extension-containing genes - 95\n",
      "Number of ns 3'UTR extension-containing genes - 1624\n",
      "Number of cryptic 3'UTR extensions - 104\n"
     ]
    }
   ],
   "source": [
    "# now want to combine the two with minimal info required for making maps\n",
    "# 3'end coordinates, gene name, le_id, site type, regulation status\n",
    "\n",
    "# define site_type columns\n",
    "sel_le_d3utr_prox.site_type = \"proximal\"\n",
    "novel_le_rep_d3utr.site_type = \"distal\"\n",
    "\n",
    "\n",
    "# background - ns in any dataset (to do this, get list of sig in any dataset)\n",
    "# get gene IDs with no regulated ALEs in any dataset\n",
    "# returns pd.Series\n",
    "ns_gene_ids = (dexseq_df.assign(reg_status=lambda df: np.where(df[\"padj\"].le(0.05), 1, 0))\n",
    "               .groupby(\"groupID\")\n",
    "               [\"reg_status\"]\n",
    "               .sum()\n",
    "               .loc[lambda x: x == 0]\n",
    "             )\n",
    "\n",
    "# if one isoform of pair is called as cryptic, label both as cryptic\n",
    "cryp_gene_ids_d3utr = (pd.concat([sel_le_d3utr_prox.as_df()[[\"le_id\", \"gene_id\"]].drop_duplicates(),\n",
    "                                  novel_le_rep_d3utr.as_df()[[\"le_id\", \"gene_id_ref\"]].rename(columns={\"gene_id_ref\": \"gene_id\"}).drop_duplicates()],\n",
    "           ignore_index=True)\n",
    "           .assign(cryptic_status=lambda df: np.where(df[\"le_id\"].isin(cryp_le_ids), 1, 0))\n",
    "           # for each gene_name, if any isoforms (prox/distal are called as cryptic, annotated as cryptic)\n",
    "           .groupby(\"gene_id\")\n",
    "           [\"cryptic_status\"].sum()\n",
    "           .loc[lambda x: x > 0]\n",
    ")\n",
    "\n",
    "ns_gene_ids_d3utr = (pd.concat([sel_le_d3utr_prox.as_df()[[\"le_id\", \"gene_id\"]].drop_duplicates(),\n",
    "           novel_le_rep_d3utr.as_df()[[\"le_id\", \"gene_id_ref\"]].rename(columns={\"gene_id_ref\": \"gene_id\"}).drop_duplicates()],\n",
    "           ignore_index=True)\n",
    "           .assign(cryptic_status=lambda df: np.where(df[\"gene_id\"].isin(ns_gene_ids.index), 1, 0))\n",
    "           # for each gene_name, if any isoforms (prox/distal are called as cryptic, annotated as cryptic)\n",
    "           .groupby(\"gene_id\")\n",
    "           [\"cryptic_status\"].sum()\n",
    "           .loc[lambda x: x > 0]\n",
    ")\n",
    "\n",
    "\n",
    "print(f\"Number of cryptic 3'UTR extension-containing genes - {len(cryp_gene_ids_d3utr)}\")\n",
    "print(f\"Number of ns 3'UTR extension-containing genes - {len(ns_gene_ids_d3utr)}\")\n",
    "print(f\"Number of cryptic 3'UTR extensions - {len(cryp_le_ids)}\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# sel_le_d3utr_prox\n",
    "# sel_novel_le_d3utr\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# le_ids of novel extensions with no change in any experiment\n",
    "ns_le_ids_d3utr = set(dexseq_df.loc[(dexseq_df[\"groupID\"].isin(ns_gene_ids.index)) & (dexseq_df[\"simple_event_type\"] == \"distal_3utr_extension\"), \"le_id\"].drop_duplicates())\n",
    "# now get le_ids of partner proximal les\n",
    "ns_le_ids_d3utr_prtnr = set([\"_\".join([le_id.split(\"_\")[0], str(int(le_id.split(\"_\")[1]) - 1)]) for le_id in ns_le_ids_d3utr])\n",
    "\n",
    "ns_le_ids_d3utr_comb = ns_le_ids_d3utr.union(ns_le_ids_d3utr_prtnr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure proximal partner of cryptic les are labelled\n",
    "cryp_le_ids_prtnr = set([\"_\".join([le_id.split(\"_\")[0], str(int(le_id.split(\"_\")[1]) - 1)]) for le_id in list(cryp_le_ids)])\n",
    "cryp_le_ids_comb = cryp_le_ids.union(cryp_le_ids_prtnr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids_comb, ns_le_ids_d3utr_comb, id_col=\"le_id\"))\n",
    "novel_le_rep_d3utr = novel_le_rep_d3utr.assign(\"cryptic_status\", lambda df: _define_cryptic_status(df, cryp_le_ids_comb, ns_le_ids_d3utr_comb, id_col=\"le_id_quant\"))\n",
    "\n",
    "\n",
    "\n",
    "# sel_le_d3utr_prox = sel_le_d3utr_prox.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))\n",
    "# sel_novel_le_d3utr = sel_novel_le_d3utr.assign(\"cryptic_status\", lambda df: pd.Series(np.where(df.gene_id_ref.isin(cryp_gene_ids_d3utr.index), \"cryptic\", \"background\"), index=df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cryptic_status\n",
       "NULL          1657\n",
       "background    1637\n",
       "cryptic        104\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_le_rep_d3utr.cryptic_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct 'name' string with minimal annotation information\n",
    "# le_id|gene_name|site_type|cryptic_status\n",
    "\n",
    "sel_le_d3utr_prox = (sel_le_d3utr_prox\n",
    "                     .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"ref_gene_name\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         # some ref_gene_name entries have multiple gene names\n",
    "         lambda df: df[\"ref_gene_name\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"gene_name_common\",\n",
    "                          lambda df: pd.Series(np.where(df[\"gene_name\"].isna(),\n",
    "                                                        df[\"ref_gene_name\"],\n",
    "                                                        df[\"gene_name\"])))\n",
    " .assign(\"Name\",\n",
    "                          lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_common\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    "                    \n",
    " )\n",
    "\n",
    "\n",
    "novel_le_rep_d3utr = (novel_le_rep_d3utr\n",
    "                      .subset(lambda df: df[\"cryptic_status\"].ne(\"NULL\"))\n",
    " .assign(\"gene_name_ref\",\n",
    "         # list(dict.fromkeys(x.split(\",\"))) - drops duplicates whilst preserving order\n",
    "         lambda df: df[\"gene_name_ref\"].apply(lambda x: \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "         )\n",
    " .assign(\"Name\",\n",
    "         lambda df: df[\"le_id\"].str.cat(df[[\"gene_name_ref\", \"site_type\", \"cryptic_status\"]], sep=\"|\"))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3855\n"
     ]
    }
   ],
   "source": [
    "# generate output bed file of proximal and distal PAS\n",
    "utr3_out_bed = pr.concat([sel_le_d3utr_prox[[\"Score\", \"Name\"]],\n",
    "                          novel_le_rep_d3utr[[\"Score\", \"Name\"]]]\n",
    "                          )\n",
    "\n",
    "print(len(utr3_out_bed))\n",
    "# drop duplicate entries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155738258</td>\n",
       "      <td>155739010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132676.16_8|DAP3|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155738608</td>\n",
       "      <td>155739010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132676.16_8|DAP3|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182828723</td>\n",
       "      <td>182830384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135838.14_3|NPL|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182828723</td>\n",
       "      <td>182830384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135838.14_3|NPL|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182828723</td>\n",
       "      <td>182830384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135838.14_3|NPL|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr22</td>\n",
       "      <td>21469641</td>\n",
       "      <td>21469935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000206140.12_1|TMEM191C|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50622753</td>\n",
       "      <td>50623773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100299.18_1|ARSA|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50622753</td>\n",
       "      <td>50623027</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100299.18_1|ARSA|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127449</td>\n",
       "      <td>130129628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3|AIFM1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127459</td>\n",
       "      <td>130129628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3|AIFM1|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 155738258 | 155739010 | .          | +            | ...   |\n",
       "| chr1         | 155738608 | 155739010 | .          | +            | ...   |\n",
       "| chr1         | 182828723 | 182830384 | .          | +            | ...   |\n",
       "| chr1         | 182828723 | 182830384 | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chr22        | 21469616  | 21469935  | .          | +            | ...   |\n",
       "| chr22        | 21469641  | 21469935  | .          | +            | ...   |\n",
       "| chr22        | 50622753  | 50623773  | .          | -            | ...   |\n",
       "| chr22        | 50622753  | 50623027  | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 89 rows and 6 columns from 16 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr3_out_bed.subset(lambda df: df.Name.isin(utr3_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select representative 5'end coord (the shoretest exon) for a given le_id\n",
    "utr3_out_bed = select_rep_five_end(utr3_out_bed, id_col = \"Name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70439100</td>\n",
       "      <td>70445530</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116761.12_2|CTH|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70439100</td>\n",
       "      <td>70445536</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116761.12_2|CTH|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44779122</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142937.12_2|RPS8|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44779128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142937.12_2|RPS8|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42775208</td>\n",
       "      <td>42781977</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164008.17_2|C1orf50|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42775208</td>\n",
       "      <td>42781983</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164008.17_2|C1orf50|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173606034</td>\n",
       "      <td>173610215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000183831.7_3|ANKRD45|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173606030</td>\n",
       "      <td>173610215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000183831.7_3|ANKRD45|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr5</td>\n",
       "      <td>168262048</td>\n",
       "      <td>168264155</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000145934.17_3|TENM2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr5</td>\n",
       "      <td>168262048</td>\n",
       "      <td>168264157</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000145934.17_3|TENM2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr7</td>\n",
       "      <td>100092240</td>\n",
       "      <td>100093133</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000166508.18_3|MCM7|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>100092232</td>\n",
       "      <td>100093133</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000166508.18_3|MCM7|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr11</td>\n",
       "      <td>36229159</td>\n",
       "      <td>36232257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179241.13_2|LDLRAD3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr11</td>\n",
       "      <td>36229159</td>\n",
       "      <td>36232253</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179241.13_2|LDLRAD3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr13</td>\n",
       "      <td>102678226</td>\n",
       "      <td>102679165</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134900.12_3|TPP2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr13</td>\n",
       "      <td>102678226</td>\n",
       "      <td>102679173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134900.12_3|TPP2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54538062</td>\n",
       "      <td>54545302</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100532.13_3|CGRRF1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54538062</td>\n",
       "      <td>54545300</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100532.13_3|CGRRF1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr15</td>\n",
       "      <td>73769469</td>\n",
       "      <td>73770615</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260469.3_2|INSYN1-AS1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr15</td>\n",
       "      <td>73769469</td>\n",
       "      <td>73770611</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260469.3_2|INSYN1-AS1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr17</td>\n",
       "      <td>34637176</td>\n",
       "      <td>34639316</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181291.8_2|TMEM132E|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr17</td>\n",
       "      <td>34637176</td>\n",
       "      <td>34639318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181291.8_2|TMEM132E|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127449</td>\n",
       "      <td>130129628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3|AIFM1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127459</td>\n",
       "      <td>130129628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3|AIFM1|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 70439100  | 70445530  | .          | +            | ...   |\n",
       "| chr1         | 70439100  | 70445536  | .          | +            | ...   |\n",
       "| chr1         | 44778575  | 44779122  | .          | +            | ...   |\n",
       "| chr1         | 44778575  | 44779128  | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chr17        | 34637176  | 34639316  | .          | +            | ...   |\n",
       "| chr17        | 34637176  | 34639318  | .          | +            | ...   |\n",
       "| chrX         | 130127449 | 130129628 | .          | -            | ...   |\n",
       "| chrX         | 130127459 | 130129628 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 24 rows and 6 columns from 9 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr3_out_bed.subset(lambda df: df.Name.isin(utr3_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({None: 48, 'atlas_1_pred': 7, 'atlas_max_datasets': 5})"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many le_ids have distinct coordinates? which collapsing category do they belong to?\n",
    "from collections import Counter\n",
    "Counter([k if le_id.split(\"|\")[0] in v else None for le_id in utr3_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index for k,v in rep_choices_d3utr.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70439100</td>\n",
       "      <td>70445536</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116761.12_2|CTH|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44779128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142937.12_2|RPS8|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42775208</td>\n",
       "      <td>42781983</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164008.17_2|C1orf50|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173606030</td>\n",
       "      <td>173610215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000183831.7_3|ANKRD45|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>168262048</td>\n",
       "      <td>168264157</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000145934.17_3|TENM2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>100092232</td>\n",
       "      <td>100093133</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000166508.18_3|MCM7|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>36229159</td>\n",
       "      <td>36232257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000179241.13_2|LDLRAD3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr13</td>\n",
       "      <td>102678226</td>\n",
       "      <td>102679173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134900.12_3|TPP2|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54538062</td>\n",
       "      <td>54545302</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100532.13_3|CGRRF1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr15</td>\n",
       "      <td>73769469</td>\n",
       "      <td>73770615</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260469.3_2|INSYN1-AS1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr17</td>\n",
       "      <td>34637176</td>\n",
       "      <td>34639318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000181291.8_2|TMEM132E|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>130127449</td>\n",
       "      <td>130129628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000156709.15_3|AIFM1|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 70439100  | 70445536  | .          | +            | ...   |\n",
       "| chr1         | 44778575  | 44779128  | .          | +            | ...   |\n",
       "| chr1         | 42775208  | 42781983  | .          | +            | ...   |\n",
       "| chr1         | 173606030 | 173610215 | .          | -            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chr14        | 54538062  | 54545302  | .          | +            | ...   |\n",
       "| chr15        | 73769469  | 73770615  | .          | +            | ...   |\n",
       "| chr17        | 34637176  | 34639318  | .          | +            | ...   |\n",
       "| chrX         | 130127449 | 130129628 | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 6 columns from 9 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all these events appear to be equidistant from nearest atlas site. In this case pick the most distal as representative (already have function)\n",
    "select_rep_prox_site(utr3_out_bed, id_col=\"Name\").subset(lambda df: df.Name.isin(utr3_out_bed.drop_duplicate_positions().as_df()[\"Name\"].value_counts().loc[lambda x: x > 1].index))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_5|NIPAL3|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_5|NIPAL3|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24476667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17_6|NIPAL3|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150159497</td>\n",
       "      <td>150160065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000023902.14_2|PLEKHO1|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150160329</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000023902.14_3|PLEKHO1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3801</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_1|TMSB4Y|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3802</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_2|TMSB4Y|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3803</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845650</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246.15_2|NLGN4Y|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3804</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845666</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246.15_3|NLGN4Y|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3805</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_2|FAM224B|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3806 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 24468985  | 24472976  | .          | +            | ...   |\n",
       "| chr1         | 24468985  | 24472976  | .          | +            | ...   |\n",
       "| chr1         | 24468985  | 24476667  | .          | +            | ...   |\n",
       "| chr1         | 150159497 | 150160065 | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrY         | 13705224  | 13706471  | .          | +            | ...   |\n",
       "| chrY         | 14840412  | 14845650  | .          | +            | ...   |\n",
       "| chrY         | 14840412  | 14845666  | .          | +            | ...   |\n",
       "| chrY         | 17551679  | 17553954  | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 3,806 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr3_out_bed = select_rep_prox_site(utr3_out_bed, id_col=\"Name\")\n",
    "utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_1|ATAD3B|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1498495</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2|ATAD3B|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3775794</td>\n",
       "      <td>3775982</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235169.11_2|SMIM1|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3775794</td>\n",
       "      <td>3776724</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235169.11_3|SMIM1|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6220805</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000158286.13_3|RNF207|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3344</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_1|TMSB4Y|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3345</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6_2|TMSB4Y|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3346</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845650</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246.15_2|NLGN4Y|proximal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3347</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14840412</td>\n",
       "      <td>14845666</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246.15_3|NLGN4Y|distal|background</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_2|FAM224B|distal|background</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3349 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Score      | Strand       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr1         | 1495484   | 1497848   | .          | +            | ...   |\n",
       "| chr1         | 1495484   | 1498495   | .          | +            | ...   |\n",
       "| chr1         | 3775794   | 3775982   | .          | +            | ...   |\n",
       "| chr1         | 3775794   | 3776724   | .          | +            | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...          | ...   |\n",
       "| chrY         | 13705224  | 13706471  | .          | +            | ...   |\n",
       "| chrY         | 14840412  | 14845650  | .          | +            | ...   |\n",
       "| chrY         | 14840412  | 14845666  | .          | +            | ...   |\n",
       "| chrY         | 17551679  | 17553954  | .          | -            | ...   |\n",
       "+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 3,349 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicate entries\n",
    "utr3_out_bed = utr3_out_bed.apply(lambda df: df.drop_duplicates()).sort()\n",
    "\n",
    "utr3_out_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"../processed/iclip_regions\"):\n",
    "    os.makedirs(\"../processed/iclip_regions\")\n",
    "    \n",
    "utr3_out_bed.to_bed(\"../processed/iclip_regions/2023-07-04_papa_cryptic_d3utr.last_exons.bed\")\n",
    "# also make one with just polYA sites\n",
    "utr3_out_bed.three_end().to_bed(\"../processed/iclip_regions/2023-07-04_papa_cryptic_d3utr.pas.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
