{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_exon(gr: pr.PyRanges, fasta: str = \"data/GRCh38.primary_assembly.genome.fa\", drop_seqs: bool = True) -> pr.PyRanges:\n",
    "    '''Generate a peptide sequence for a continuous genomic interval (e.g. exon)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    gr : pr.PyRanges\n",
    "        _description_\n",
    "    fasta : str, optional\n",
    "        path to FASTA file of genome sequence, by default \"data/GRCh38.primary_assembly.genome.fa\"\n",
    "    drop_seqs : bool, optional\n",
    "        whether to drop columns containing extracted genomic sequence, by default True\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pr.PyRanges\n",
    "        input PyRanges object with added 'peptide_seq' column containing str of peptide sequences\n",
    "    '''\n",
    "\n",
    "    assert \"Frame\" in gr.columns\n",
    "    assert \"dna_seq\" not in gr.columns\n",
    "\n",
    "    # extract genomic sequences for intervals, add as column\n",
    "    seqs_gr = pr.get_sequence(gr, fasta)\n",
    "    gr.dna_seq = seqs_gr\n",
    "\n",
    "    # convert dna_seq to a Seq object to translate (BioPython)\n",
    "    # remove nucleotides from start if necessary to get complete codon\n",
    "    gr = gr.assign(\"seq_dna_seq\",\n",
    "                    lambda df: df.apply(lambda row: Seq(row[\"dna_seq\"][int(row[\"Frame\"]):]),\n",
    "                                        axis=1)\n",
    "                                        )\n",
    "    \n",
    "    # generate peptide sequence\n",
    "    gr = gr.assign(\"peptide_seq\",\n",
    "                   lambda df: df[\"seq_dna_seq\"].apply(lambda x: str(x.translate(to_stop=True)))\n",
    "                   )\n",
    "\n",
    "    if drop_seqs:\n",
    "        return gr.drop([\"dna_seq\", \"seq_dna_seq\"])\n",
    "    \n",
    "    else:\n",
    "        return gr\n",
    "    \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exper_cryp</th>\n",
       "      <th>event_manual_validation</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>brown_i3_cortical,seddighi_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>coverage quite low, but reads consistently up ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>seddighi - event looks real but v lowly expres...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>brown_i3_cortical,humphrey_i3_cortical,seddigh...</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>looks like real 3'UTR, but unclear whether cry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>humphrey_i3_cortical,seddighi_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>reads in cryptic also downstream of stops in a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>genuine bleedthrough, not particularly cryptic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>very lowly expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FIRRE</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>lncRNA. Looks real but lowly expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HECW1</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>definitely upregulated but also basal IR. Even...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>genuine event but not cryptic (overlapping txs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>expressed in ctls but likely cryptic by ratio ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>expression in ctls, likely cryptic by ratio (e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>lowly included. Whole exon is cryptic (novel SJ)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>extension of ALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>IR from upstream exon through to bleedthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>very clean, obvious downreg of RNA downstream....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>brown_i3_cortical,humphrey_i3_cortical,seddigh...</td>\n",
       "      <td>yes</td>\n",
       "      <td>cryptic CDS terminates 1 codon downstream of SJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>annotated, also low level basal IR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name                 le_id  \\\n",
       "0       ACER3  ENSG00000078124.13_1   \n",
       "1       ADCY8  ENSG00000155897.10_1   \n",
       "2     ANKRD27  ENSG00000105186.16_3   \n",
       "9       CEP76  ENSG00000101624.11_3   \n",
       "10      CNPY3  ENSG00000137161.18_1   \n",
       "11       CTBS  ENSG00000117151.13_1   \n",
       "12      DLEU7  ENSG00000186047.11_2   \n",
       "14  FBXO38-DT   ENSG00000247199.6_1   \n",
       "15      FIRRE   ENSG00000213468.7_1   \n",
       "17      HECW1  ENSG00000002746.15_3   \n",
       "18       LIPA  ENSG00000107798.18_2   \n",
       "20      NBPF3  ENSG00000142794.19_3   \n",
       "23      NYAP2  ENSG00000144460.13_2   \n",
       "24    PCDH11X  ENSG00000102290.23_3   \n",
       "26      PLPBP  ENSG00000147471.12_1   \n",
       "27      PLPBP  ENSG00000147471.12_1   \n",
       "28      PLPBP  ENSG00000147471.12_1   \n",
       "30      RFTN2  ENSG00000162944.11_1   \n",
       "31      SIN3B  ENSG00000127511.10_2   \n",
       "34      TMED4  ENSG00000158604.15_1   \n",
       "35      USP31  ENSG00000103404.15_1   \n",
       "\n",
       "                                           exper_cryp event_manual_validation  \\\n",
       "0              brown_i3_cortical,seddighi_i3_cortical                     yes   \n",
       "1                                humphrey_i3_cortical                     yes   \n",
       "2   brown_i3_cortical,humphrey_i3_cortical,seddigh...                     yes   \n",
       "9                                   brown_i3_cortical                     yes   \n",
       "10          humphrey_i3_cortical,seddighi_i3_cortical                     yes   \n",
       "11                               humphrey_i3_cortical                     yes   \n",
       "12                               humphrey_i3_cortical                     yes   \n",
       "14                               humphrey_i3_cortical                     yes   \n",
       "15                                  brown_i3_cortical                     yes   \n",
       "17                                  brown_i3_cortical                     yes   \n",
       "18                               humphrey_i3_cortical                     yes   \n",
       "20                               humphrey_i3_cortical                     yes   \n",
       "23                               humphrey_i3_cortical                     yes   \n",
       "24                               humphrey_i3_cortical                     yes   \n",
       "26                               humphrey_i3_cortical                     yes   \n",
       "27                               humphrey_i3_cortical                     yes   \n",
       "28                               humphrey_i3_cortical                     yes   \n",
       "30                               humphrey_i3_cortical                     yes   \n",
       "31  brown_i3_cortical,humphrey_i3_cortical,seddigh...                     yes   \n",
       "34                               humphrey_i3_cortical                     yes   \n",
       "35                               humphrey_i3_cortical                     yes   \n",
       "\n",
       "                                                notes  \n",
       "0   coverage quite low, but reads consistently up ...  \n",
       "1   seddighi - event looks real but v lowly expres...  \n",
       "2                                                 NaN  \n",
       "9   looks like real 3'UTR, but unclear whether cry...  \n",
       "10  reads in cryptic also downstream of stops in a...  \n",
       "11  genuine bleedthrough, not particularly cryptic...  \n",
       "12                                                NaN  \n",
       "14                               very lowly expressed  \n",
       "15             lncRNA. Looks real but lowly expressed  \n",
       "17  definitely upregulated but also basal IR. Even...  \n",
       "18  genuine event but not cryptic (overlapping txs...  \n",
       "20  expressed in ctls but likely cryptic by ratio ...  \n",
       "23  expression in ctls, likely cryptic by ratio (e...  \n",
       "24   lowly included. Whole exon is cryptic (novel SJ)  \n",
       "26                                   extension of ALE  \n",
       "27                                                NaN  \n",
       "28      IR from upstream exon through to bleedthrough  \n",
       "30  very clean, obvious downreg of RNA downstream....  \n",
       "31    cryptic CDS terminates 1 codon downstream of SJ  \n",
       "34                 annotated, also low level basal IR  \n",
       "35                                                NaN  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First define set of IDs & last exons for which to extract peptide seqs\n",
    "bleeds = pd.read_csv(\"data/riboseq_manual_verification_of_i3_cortical_cryptic_bleedthroughs.tsv\", sep=\"\\t\", usecols=[\"gene_name\", \"le_id\", \"exper_cryp\", \"event_manual_validation\", \"notes\"])\n",
    "\n",
    "bleeds_y = bleeds[bleeds[\"event_manual_validation\"] == \"yes\"]\n",
    "bleeds_y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 22 s, sys: 5.14 s, total: 27.1 s\n",
      "Wall time: 1min 5s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21483467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21440348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001546346.2</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21444947</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131755597</td>\n",
       "      <td>131755746</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSE00001752569.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131749305</td>\n",
       "      <td>131749458</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00001674554.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1538</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131711650</td>\n",
       "      <td>131711720</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00001594498.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131709497</td>\n",
       "      <td>131709556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSE00001734033.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131702651</td>\n",
       "      <td>131703518</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSE00001658652.2</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1541 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +19   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | transcript | 21440127  | 21483467  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21440127  | 21440348  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21444947  | 21445219  | ...   |\n",
       "| chr1         | HAVANA     | CDS        | 21445086  | 21445219  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131749305 | 131749458 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131711650 | 131711720 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131709497 | 131709556 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131702651 | 131703518 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,541 rows and 24 columns from 13 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# read in reference GTF (used to define last exons), subsetting only to cryptic-containing genes to save memory\n",
    "ref_gtf = pr.read_gtf(\"data/reference_filtered.gtf\").subset(lambda df: df.gene_name.isin(set(bleeds_y.gene_name)))\n",
    "ref_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24419640</td>\n",
       "      <td>24420128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419290</td>\n",
       "      <td>24419640</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>488.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000001461.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24421813</td>\n",
       "      <td>24422110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419640,24419640,24419640</td>\n",
       "      <td>24440171,24440171,24433103</td>\n",
       "      <td>ENST00000358028.8,ENST00000374399.9,ENST000000...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000001461.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24454053</td>\n",
       "      <td>24454824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24453504,24453504</td>\n",
       "      <td>24456137,24456137</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000001461.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24464120</td>\n",
       "      <td>24466378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464025</td>\n",
       "      <td>24464120</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>2258.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000001461.17_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880307</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21039044</td>\n",
       "      <td>21042268</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>23936831</td>\n",
       "      <td>23939185</td>\n",
       "      <td>ENST00000631331.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129930 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24419640  | 24420128  | ...   |\n",
       "| chr1         | .          | exon         | 24421813  | 24422110  | ...   |\n",
       "| chr1         | .          | exon         | 24454053  | 24454824  | ...   |\n",
       "| chr1         | .          | exon         | 24464120  | 24466378  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 21880075  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21880307  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21038288  | 21039044  | ...   |\n",
       "| chrY         | .          | exon         | 23936726  | 23936831  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 129,930 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in last exon references\n",
    "# 'quant last exons', which is unique regions of last exons only, will first be used to define 'annotated bleedthroughs'\n",
    "quant_le = pr.read_gtf(\"data/novel_ref_combined.quant.last_exons.gtf\")\n",
    "full_le = pr.read_gtf(\"data/novel_ref_combined.last_exons.gtf\")\n",
    "\n",
    "# subset to bleedthroughs only\n",
    "quant_le_bld = quant_le.subset(lambda df: df[\"le_id\"].isin(set(bleeds_y[\"le_id\"])))\n",
    "full_le_bld =  full_le.subset(lambda df: df[\"le_id\"].isin(set(bleeds_y[\"le_id\"])))\n",
    "quant_le"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1 - identify bleedthroughs that have an overlapping/annotated CDS\n",
    "\n",
    "In some cases, bleedthrough events are defined by the pipeline as novel extensions of annotated last exons (e.g. SIN3B). These events will already have a CDS annotated for the coding region of the last exon, so can use that to define the cryptic peptide\n",
    "\n",
    "Strategy:\n",
    "1. Overlap 'quantification' last exons with CDSs, checking that the parent transcripts match\n",
    "2. Keep these CDSs for these bleedthroughs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDS and stop codons are annotated separately in GTFs - still want to report full coding region for bleedthrough including stop codon\n",
    "# merge CDS & stop codons into single interval by transcript ID\n",
    "ref_gtf_cds = ref_gtf.subset(lambda df: df.Feature == \"CDS\")\n",
    "ref_gtf_sc = ref_gtf.subset(lambda df: df.Feature == \"stop_codon\")\n",
    "# ref_gtf_cds_m = pr.concat([ref_gtf_cds, ref_gtf_sc]).merge(strand=True, by=\"transcript_id\")\n",
    "# ref_gtf_cds_m\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>Start_cds</th>\n",
       "      <th>End_cds</th>\n",
       "      <th>Strand_cds</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44581068</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44581068</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50823316</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50823316</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673367</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673367</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673367</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673367</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854262</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32639304</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +11   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr7         | 44579896  | 44581092  | -            | ...   |\n",
       "| chr7         | 44579903  | 44581092  | -            | ...   |\n",
       "| chr13        | 50814965  | 50823520  | -            | ...   |\n",
       "| chr13        | 50815080  | 50823520  | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr18        | 12672624  | 12673387  | -            | ...   |\n",
       "| chr18        | 12672624  | 12673387  | -            | ...   |\n",
       "| chr18        | 12673096  | 12673387  | -            | ...   |\n",
       "| chr18        | 12673096  | 12673387  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 10 rows and 15 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: ref_gene_name, le_id, transcript_id, transcript_id_ref, ... (+ 7 more.)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bld_cds_anno = quant_le_bld[[\"ref_gene_name\", \"le_id\", \"transcript_id\", \"transcript_id_ref\"]].join(ref_gtf_cds[[\"gene_name\", \"gene_id\", \"transcript_id\", \"Frame\"]], strandedness=\"same\", suffix=\"_cds\")\n",
    "bld_cds_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>Start_le</th>\n",
       "      <th>End_le</th>\n",
       "      <th>Strand_le</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854265</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32639301</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +11   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 10 rows and 15 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: ref_gene_name, le_id, transcript_id, transcript_id_ref, ... (+ 7 more.)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make the CDS coordinates the main coordinate range\n",
    "cds_bld_anno = bld_cds_anno.new_position(\"swap\").apply(lambda df: df.rename(columns={\"Start_cds\": \"Start_le\", \"End_cds\": \"End_le\", \"Strand_cds\": \"Strand_le\"}))\n",
    "# as CDS entries lack the stop codon, extend by 3nt in each case to add in stop codon\n",
    "cds_bld_anno = cds_bld_anno.extend({\"3\": 3})\n",
    "cds_bld_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>Start_le</th>\n",
       "      <th>End_le</th>\n",
       "      <th>Strand_le</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "      <th>dna_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854265</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>0</td>\n",
       "      <td>GTCCGCCGGGTGCTGAAGAGCCAGGAGGTGTATGAAAACTTCCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32639301</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>1</td>\n",
       "      <td>GATGGCAAATTTGAGTTACATCAAAAACTTCAGGTTTAGCAGCTTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +12   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 10 rows and 16 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "12 hidden columns: ref_gene_name, le_id, transcript_id, transcript_id_ref, ... (+ 8 more.)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract sequence for CDS sequences\n",
    "seqs_cds_bld_anno = pr.get_sequence(cds_bld_anno, \"data/GRCh38.primary_assembly.genome.fa\")\n",
    "cds_bld_anno.dna_seq = seqs_cds_bld_anno\n",
    "\n",
    "cds_bld_anno"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Translating CDSs\n",
    "\n",
    "Minimal information needed is:\n",
    "- DNA sequence\n",
    "- Frame - how many bases to remove from start of sequence to get the first base of a codon\n",
    "\n",
    "To translate all (or all but the first incomplete codon) codons, will need to ensure DNA sequence starts at a complete codon. From there on it should be a simple translation.\n",
    "- Convert to biopython Seq object (stripping frame positions from start of dna_sequence)\n",
    "- seq.translate()\n",
    "- convert back to string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>Start_le</th>\n",
       "      <th>End_le</th>\n",
       "      <th>Strand_le</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "      <th>dna_seq</th>\n",
       "      <th>seq_dna_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "      <td>(C, G, G, G, T, G, C, A, T, C, T, C, G, A, C, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "      <td>(C, G, G, G, T, G, C, A, T, C, T, C, G, A, C, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "      <td>(G, A, T, A, G, T, G, T, T, G, A, G, T, T, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "      <td>(G, A, T, A, G, T, G, T, T, G, A, G, T, T, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854265</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>0</td>\n",
       "      <td>GTCCGCCGGGTGCTGAAGAGCCAGGAGGTGTATGAAAACTTCCTCC...</td>\n",
       "      <td>(G, T, C, C, G, C, C, G, G, G, T, G, C, T, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32639301</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>1</td>\n",
       "      <td>GATGGCAAATTTGAGTTACATCAAAAACTTCAGGTTTAGCAGCTTG...</td>\n",
       "      <td>(A, T, G, G, C, A, A, A, T, T, T, G, A, G, T, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +13   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 10 rows and 17 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: ref_gene_name, le_id, transcript_id, transcript_id_ref, ... (+ 9 more.)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_bld_anno = cds_bld_anno.assign(\"seq_dna_seq\",\n",
    "                    lambda df: df.apply(lambda row: Seq(row[\"dna_seq\"][int(row[\"Frame\"]):]),\n",
    "                                        axis=1)\n",
    "                                        )\n",
    "\n",
    "cds_bld_anno\n",
    "                                                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>Start_le</th>\n",
       "      <th>End_le</th>\n",
       "      <th>Strand_le</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "      <th>dna_seq</th>\n",
       "      <th>seq_dna_seq</th>\n",
       "      <th>peptide_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "      <td>(C, G, G, G, T, G, C, A, T, C, T, C, G, A, C, ...</td>\n",
       "      <td>RVHLDIQVGEHANNYPEIAAKDKLTELQLRARQLLDQVEQIQKEQD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44581065</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581092</td>\n",
       "      <td>-</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>0</td>\n",
       "      <td>CGGGTGCATCTCGACATCCAGGTTGGGGAGCATGCCAACAACTACC...</td>\n",
       "      <td>(C, G, G, G, T, G, C, A, T, C, T, C, G, A, C, ...</td>\n",
       "      <td>RVHLDIQVGEHANNYPEIAAKDKLTELQLRARQLLDQVEQIQKEQD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "      <td>(G, A, T, A, G, T, G, T, T, G, A, G, T, T, T, ...</td>\n",
       "      <td>DSVEFRNICSHLALQIEGQQFDRDLNAAHQCLKTIVKKLIQSLANF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50823313</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>0</td>\n",
       "      <td>GATAGTGTTGAGTTTAGAAACATCTGCAGTCATTTGGCTCTACAGA...</td>\n",
       "      <td>(G, A, T, A, G, T, G, T, T, G, A, G, T, T, T, ...</td>\n",
       "      <td>DSVEFRNICSHLALQIEGQQFDRDLNAAHQCLKTIVKKLIQSLANF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "      <td>SPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "      <td>SPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "      <td>SPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673364</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673387</td>\n",
       "      <td>-</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCTCCTTTCTGTGAAGAAATAATCTGTTGCCGTGGAGACCAAGTG...</td>\n",
       "      <td>(T, C, T, C, C, T, T, T, C, T, G, T, G, A, A, ...</td>\n",
       "      <td>SPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854265</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>0</td>\n",
       "      <td>GTCCGCCGGGTGCTGAAGAGCCAGGAGGTGTATGAAAACTTCCTCC...</td>\n",
       "      <td>(G, T, C, C, G, C, C, G, G, G, T, G, C, T, G, ...</td>\n",
       "      <td>VRRVLKSQEVYENFLRCIALFNQELVSGSELLQLVSPFLG*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32639301</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>-</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>1</td>\n",
       "      <td>GATGGCAAATTTGAGTTACATCAAAAACTTCAGGTTTAGCAGCTTG...</td>\n",
       "      <td>(A, T, G, G, C, A, A, A, T, T, T, G, A, G, T, ...</td>\n",
       "      <td>MANLSYIKNFRFSSLAKDELGYCLTSFEAAIEYIRQGSLSAKPPVR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +14   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr7         | 44581065  | 44581239  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| chr13        | 50823313  | 50823520  | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "| chr18        | 12673364  | 12673503  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 10 rows and 18 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "14 hidden columns: ref_gene_name, le_id, transcript_id, transcript_id_ref, ... (+ 10 more.)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_bld_anno.peptide_seq = cds_bld_anno.seq_dna_seq.apply(lambda x: str(x.translate()))\n",
    "cds_bld_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TMED4\tRVHLDIQVGEHANNYPEIAAKDKLTELQLRARQLLDQVEQIQKEQDYQRASAYLLVI*\n",
      "TMED4\tRVHLDIQVGEHANNYPEIAAKDKLTELQLRARQLLDQVEQIQKEQDYQRASAYLLVI*\n",
      "DLEU7\tDSVEFRNICSHLALQIEGQQFDRDLNAAHQCLKTIVKKLIQSLANFPSDAHMVACASLRQILQNLPDI*\n",
      "DLEU7\tDSVEFRNICSHLALQIEGQQFDRDLNAAHQCLKTIVKKLIQSLANFPSDAHMVACASLRQILQNLPDI*\n",
      "CEP76\tSPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*\n",
      "CEP76\tSPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*\n",
      "CEP76\tSPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*\n",
      "CEP76\tSPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL*\n",
      "SIN3B\tVRRVLKSQEVYENFLRCIALFNQELVSGSELLQLVSPFLG*\n",
      "ANKRD27\tMANLSYIKNFRFSSLAKDELGYCLTSFEAAIEYIRQGSLSAKPPVRSHPCPGLPLWASWFP*\n"
     ]
    }
   ],
   "source": [
    "for _, df in cds_bld_anno.as_df().iterrows():\n",
    "    print(\"\\t\".join(df[[\"gene_name\", \"peptide_seq\"]]))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "I manually checked the above and they are correct - now to define full cryptic peptides for those with no annotated CDS overlapping the unique region. To do this:\n",
    "1. set of le_ids with overlapping/already defined peptides\n",
    "2. For remaining les, take the full last exon sequences and find overlapping annotated CDSs. Make sure they exactly match at the 5'ends of the exon\n",
    "    - This gives us the frame/phase for the exon \n",
    "3. Extract sequence and translate as before\n",
    "\n",
    "## Translating last exons without an annotated CDS extending into unique region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.4038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.7930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377695</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-D_S4.6266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148378366</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.6788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.8982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr7</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>43444214</td>\n",
       "      <td>43449982</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.7990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>HECW1,HECW1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr7</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>43444214</td>\n",
       "      <td>43446033</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-C_S3.7757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>HECW1,HECW1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43448981</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766741</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766355</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>386.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>37766481</td>\n",
       "      <td>37767739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.10983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced,last_exon_spliced</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>PLPBP,PLPBP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>37766481</td>\n",
       "      <td>37766716</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>37766355</td>\n",
       "      <td>37772754</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.11594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.11524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.14569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89331008</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.13899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89332056</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.13417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89334602</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.10969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-D_S4.11799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23075386</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.19771</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23076572</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.20251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079037</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_2.14903</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079058</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.16462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079174</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.15356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079300</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_1.14929</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.14367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | .          | exon         | 84561927  | 84563418  | ...   |\n",
       "| chr2         | .          | exon         | 225651431 | 225658565 | ...   |\n",
       "| chr2         | .          | exon         | 197630416 | 197631220 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr16        | .          | exon         | 23079058  | 23080171  | ...   |\n",
       "| chr16        | .          | exon         | 23079174  | 23080171  | ...   |\n",
       "| chr16        | .          | exon         | 23079300  | 23080171  | ...   |\n",
       "| chr16        | .          | exon         | 23079754  | 23080171  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 37 rows and 22 columns from 10 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get full coordinates of bleedthough last exons without an annotated peptide \n",
    "full_le_bld_nol = full_le_bld.subset(lambda df: ~df[\"le_id\"].isin(set(cds_bld_anno.le_id))).drop_duplicate_positions(strand=True)\n",
    "full_le_bld_nol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>Start_cds</th>\n",
       "      <th>End_cds</th>\n",
       "      <th>Strand_cds</th>\n",
       "      <th>gene_name_cds</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>84563256</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225651535</td>\n",
       "      <td>+</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP-4.4038.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP-4.8982.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42935670</td>\n",
       "      <td>+</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43448981</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10456.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43445570</td>\n",
       "      <td>+</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>ENST00000453890.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43448981</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10456.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43445570</td>\n",
       "      <td>+</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>ENST00000395891.7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766741</td>\n",
       "      <td>+</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "      <td>ENST00000520073.5</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766355</td>\n",
       "      <td>+</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-1.11594.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-1.11594.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-6.11524.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-6.11524.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.Cont-D_S4.11799.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.Cont-D_S4.11799.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23075386</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP-4.19771.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23076572</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.20251.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079037</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_2.14903.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079058</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.16462.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079174</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-F_S6.15356.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079300</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_1.14929.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.14367.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | gene_name   | +11   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------------+-------|\n",
       "| chr1         | 84561927  | 84563418  | -            | nan         | ...   |\n",
       "| chr2         | 225651431 | 225658565 | +            | nan         | ...   |\n",
       "| chr2         | 197630416 | 197631220 | -            | nan         | ...   |\n",
       "| chr2         | 197630635 | 197631220 | -            | nan         | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...         | ...   |\n",
       "| chr16        | 23079058  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079174  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079300  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079754  | 23080171  | -            | nan         | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 31 rows and 16 columns from 7 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: le_id, transcript_id, transcript_id_ref, ref_gene_name, Start_cds, ... (+ 6 more.)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overlap join last exons with CDSs, filtering for exact 5'matches\n",
    "cds_bld_anno_nol = (full_le_bld_nol[[\"gene_name\", \"le_id\", \"transcript_id\", \"transcript_id_ref\", \"ref_gene_name\"]].join(ref_gtf_cds[[\"gene_name\", \"gene_id\", \"transcript_id\", \"Frame\"]],\n",
    "                                                                                                     strandedness=\"same\",\n",
    "                                                                                                     suffix=\"_cds\")\n",
    ".subset(lambda df: ((df.Strand == \"+\") & (df.Start == df.Start_cds)) | ((df.Strand == \"-\") & (df.End == df.End_cds))\n",
    "        )\n",
    ")\n",
    "\n",
    "cds_bld_anno_nol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>Start_cds</th>\n",
       "      <th>End_cds</th>\n",
       "      <th>Strand_cds</th>\n",
       "      <th>gene_name_cds</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_cds</th>\n",
       "      <th>Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>84563256</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225651535</td>\n",
       "      <td>+</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP-4.4038.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP-4.8982.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42935670</td>\n",
       "      <td>+</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43448981</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10456.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43445570</td>\n",
       "      <td>+</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>ENST00000453890.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43448981</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000002746.15_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.10456.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>43444217</td>\n",
       "      <td>43445570</td>\n",
       "      <td>+</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>ENSG00000002746.15</td>\n",
       "      <td>ENST00000395891.7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766741</td>\n",
       "      <td>+</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12_1</td>\n",
       "      <td>ENST00000520073.5</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>37766279</td>\n",
       "      <td>37766355</td>\n",
       "      <td>+</td>\n",
       "      <td>PLPBP</td>\n",
       "      <td>ENSG00000147471.12</td>\n",
       "      <td>ENST00000328195.8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-1.11594.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-1.11594.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-6.11524.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-6.11524.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>130903771</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.Cont-D_S4.11799.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.Cont-D_S4.11799.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76868156</td>\n",
       "      <td>+</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23075386</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP-4.19771.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23076572</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.20251.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079037</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_2.14903.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079058</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.16462.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079174</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-F_S6.15356.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079300</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_1.14929.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.14367.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USP31</td>\n",
       "      <td>23079945</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | gene_name   | +11   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------------+-------|\n",
       "| chr1         | 84561927  | 84563418  | -            | nan         | ...   |\n",
       "| chr2         | 225651431 | 225658565 | +            | nan         | ...   |\n",
       "| chr2         | 197630416 | 197631220 | -            | nan         | ...   |\n",
       "| chr2         | 197630635 | 197631220 | -            | nan         | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...         | ...   |\n",
       "| chr16        | 23079058  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079174  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079300  | 23080171  | -            | nan         | ...   |\n",
       "| chr16        | 23079754  | 23080171  | -            | nan         | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 31 rows and 16 columns from 7 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: le_id, transcript_id, transcript_id_ref, ref_gene_name, Start_cds, ... (+ 6 more.)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overlap join last exons with CDSs, filtering for exact 5'matches\n",
    "cds_bld_anno_nol = (full_le_bld_nol[[\"gene_name\", \"le_id\", \"transcript_id\", \"transcript_id_ref\", \"ref_gene_name\"]].join(ref_gtf_cds[[\"gene_name\", \"gene_id\", \"transcript_id\", \"Frame\"]],\n",
    "                                                                                                     strandedness=\"same\",\n",
    "                                                                                                     suffix=\"_cds\")\n",
    ".subset(lambda df: ((df.Strand == \"+\") & (df.Start == df.Start_cds)) | ((df.Strand == \"-\") & (df.End == df.End_cds))\n",
    "        )\n",
    ")\n",
    "\n",
    "cds_bld_anno_nol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTBS\tDHVCTIAKVPFRGAPCSDAAGRQVPYKTIMKQINSSISGNLWDKDQRAPYYNYKVRLFVSYEHLFY\n",
      "NYAP2\tPKVSCKLGRSASTSGVPPPSVTPLRQSSDLQQSQVPSSLANRD\n",
      "RFTN2\tSDNKLYTVFNAFDDDSTSWAYQEGILSMKVTRKGSVISTLDADWLELTTFYYKQGLSLIDSFVFWETSKGKFYVNDILYLRNLNLYQNQ\n",
      "CNPY3\tDLRLIEVTETICKRLLDYSLHKERTGSNRFAKVGFGIVLHPLWGQACMYLSVSAGVSVI\n",
      "HECW1\tDEEISLSTEPESAQIQDSPMNNLMESGSGEPRSEAPESSESWKPEQLGEGSVPDGPGNQSIELSRPAEEAAVITEAGDQGMVSVGPEGAGELLAQVQKDIQPAPSAEELAEQLDLGEEASALLLEDGEAPASTKEEPLEEEATTQSRAGREEEEKEQEEEGDVSTLEQGEGRLQLRASVKRKSRPCSLPVSELETVIASACGDPETPRTHYIRIHTLLHSMPSAQGGSAAEEEDGAEEESTLKDSSEKDGLSEVDTVAADPSALEEDREEPEGATPGTAHPGHSGGHFPSLANGAAQDGDTHPSTGSESDSSPRQGGDHSCEGCDASCCSPSCYSSSCYSTSCYSSSCYSASCYSPSCYNGNRFASHTRFSSVDSAKISESTVFSSQDDEEEENSAFESVPDSMQSPELDPESTNGAGPWQDELAAPSGHVERSPEGLESPVAGPSNRREG\n",
      "PLPBP\tILSLCPEIKWHFIGHLQKQNVNKLMGKIKLNMKTKLFCHCIASTTLCGRESNF\n",
      "ADCY8\tRIHISKATLDCLNGDYNVEEGHGKERNEFLRKHNIETYLIKQPEDSLLSLPEDIVKESVSSSDRRNSGATFTEGSWSPELPFDNIVGKQNVSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTISLSPQDRINLVCNFRKCQKCQRVPNTLSLSLFFFL\n",
      "ACER3\tSFLPSSLKSNLKLVSTHQASFLVLGLWAGRGDIPVI\n",
      "USP31\tEGDRRMKLQNMVKFPLTGLDMTPHVVKRSQSSWSLPSHWSPWRRPYGLGRDPEDYIYDLYAVCNHHGTMQGGHYTGLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHPLDLCPRANKRTFFLLFLSWVVQ\n",
      "USP31\tEGDRRMKLQNMVKFPLTGLDMTPHVVKRSQSSWSLPSHWSPWRRPYGLGRDPEDYIYDLYAVCNHHGTMQGGHYTGLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHPLDLCPRANKRTFFLLFLS\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Now have full last exons with exactly matched annotated CDS (and frame to get complete CDS)\n",
    "cds_bld_anno_nol = translate_exon(cds_bld_anno_nol)\n",
    "\n",
    "for _, df in cds_bld_anno_nol.as_df().drop_duplicates(subset=[\"ref_gene_name\", \"peptide_seq\"]).iterrows():\n",
    "    print(\"\\t\".join(df[[\"ref_gene_name\", \"peptide_seq\"]]))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually checking a few:\n",
    "- USP31 - There is a TSL5 transcript so can cross-check, is spot on! There is possibly a shortened one without a stop codon :O\n",
    "- CNPY3 - also correct (refseq has extension partially annotated)\n",
    "- ACER3 - also correct\n",
    "- ADCY8 - also correct\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.7930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377695</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-D_S4.6266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148378366</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.6788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.14569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89331008</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.13899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89332056</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.13417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89334602</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.10969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 21453372  | 21457150  | ...   |\n",
       "| chr5         | .          | exon         | 148377588 | 148378845 | ...   |\n",
       "| chr5         | .          | exon         | 148377695 | 148378845 | ...   |\n",
       "| chr5         | .          | exon         | 148378366 | 148378845 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr10        | .          | exon         | 89330996  | 89335534  | ...   |\n",
       "| chr10        | .          | exon         | 89331008  | 89335534  | ...   |\n",
       "| chr10        | .          | exon         | 89332056  | 89335534  | ...   |\n",
       "| chr10        | .          | exon         | 89334602  | 89335534  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 11 rows and 22 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which ones are missing?\n",
    "full_le_bld_nol.subset(lambda df: ~df.le_id.isin(cds_bld_anno_nol.le_id))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FIRRE is definitely a lncRNA, the others not so sure (TO CHECK ref GTF) - maybe they don't have any annotated CDS if protein-coding?\n",
    "\n",
    "### Extracting unique regions of peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing with strings: abcdefg abcxyz\n",
      "Original function result: 2\n",
      "Original function result extracting matched string1: abc\n",
      "Original function result extracting unique part of string1: defg\n"
     ]
    }
   ],
   "source": [
    "# function to match two peptide sequences & find position where they stop matching (assuming they start at the same position)\n",
    "def longest_matching_substring(str1: str, str2: str) -> int:\n",
    "    '''Return the index of the final position in longest exactly matched substring from the beginning of two strings\n",
    "\n",
    "    this is intended for strings that begin with identical values. it only checks the longest matching substring from the first position in each string. if there is one mismatch, the function terminates\n",
    "    \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    str1 : str\n",
    "        first string of pair wish to find the longest matching substring from the beginning of the string\n",
    "    str2 : str\n",
    "        second string of pair wish to find the longest matching substring from the beginning of the string\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    int\n",
    "        index of the final position (in either string) of the longest matched substring between the two. If you wish to slice a string to retain/exclude the longest match, remember to add one to returned value\n",
    "    '''\n",
    "    longest_substring = 0 \n",
    "    for i, (char1, char2) in enumerate(zip(str1, str2)):\n",
    "        if char1 == char2:\n",
    "            longest_substring = i\n",
    "        else:\n",
    "            break\n",
    "\n",
    "    return longest_substring \n",
    "\n",
    "string1 = \"abcdefg\"\n",
    "string2 = \"abcxyz\"\n",
    "\n",
    "print(\"Testing with strings:\", string1, string2)\n",
    "print(\"Original function result:\", longest_matching_substring(string1, string2))\n",
    "print(\"Original function result extracting matched string1:\", string1[:longest_matching_substring(string1, string2) + 1])\n",
    "print(\"Original function result extracting unique part of string1:\", string1[longest_matching_substring(string1, string2) + 1:])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
