{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from peptide_helpers import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cryptics - 18\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>simple_event_type</th>\n",
       "      <th>n_exper_cryptic</th>\n",
       "      <th>experiment_name</th>\n",
       "      <th>annot_status</th>\n",
       "      <th>dataset</th>\n",
       "      <th>event_manual_validation</th>\n",
       "      <th>cryptic_riboseq_reads</th>\n",
       "      <th>riboseq_downstream_change</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>13</td>\n",
       "      <td>humphrey_i3_cortical,seddighi_i3_cortical,zano...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>unclear</td>\n",
       "      <td>reads in cryptic also downstream of stops in a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>USP31</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>8</td>\n",
       "      <td>humphrey_i3_cortical,zanovello_i3_cortical_upf...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>unclear</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>5</td>\n",
       "      <td>brown_i3_cortical,humphrey_i3_cortical,seddigh...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>4</td>\n",
       "      <td>brown_skndz,humphrey_i3_cortical,zanovello_shs...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>unclear</td>\n",
       "      <td>unclear</td>\n",
       "      <td>expressed in ctls but likely cryptic by ratio ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>3</td>\n",
       "      <td>brown_i3_cortical,seddighi_i3_cortical,zanovel...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>few reads</td>\n",
       "      <td>unclear</td>\n",
       "      <td>coverage quite low, but reads consistently up ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>3</td>\n",
       "      <td>zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...</td>\n",
       "      <td>novel</td>\n",
       "      <td>zanovello_shsy5y_curve_0075</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3'end of bleedthroughs are truncated. lowly ex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>3</td>\n",
       "      <td>brown_i3_cortical,humphrey_i3_cortical,seddigh...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>cryptic CDS terminates 1 codon downstream of SJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>2</td>\n",
       "      <td>humphrey_i3_cortical,zanovello_i3_cortical_upf...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>seddighi - event looks real but v lowly expres...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>2</td>\n",
       "      <td>humphrey_i3_cortical,zanovello_i3_cortical_upf...</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>very clean, obvious downreg of RNA downstream....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>annotated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>looks like real 3'UTR, but unclear whether cry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>lowly included. Whole exon is cryptic (novel SJ)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genuine event but not cryptic (overlapping txs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>genuine bleedthrough, not particularly cryptic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>expression in ctls, likely cryptic by ratio (e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>annotated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>annotated, also low level basal IR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>brown_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>lncRNA. Looks real but lowly expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>bleedthrough</td>\n",
       "      <td>1</td>\n",
       "      <td>humphrey_i3_cortical</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>very lowly expressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   le_id  gene_name simple_event_type  n_exper_cryptic  \\\n",
       "0   ENSG00000137161.18_1      CNPY3      bleedthrough               13   \n",
       "2   ENSG00000103404.15_1      USP31      bleedthrough                8   \n",
       "6   ENSG00000105186.16_3    ANKRD27      bleedthrough                5   \n",
       "9   ENSG00000142794.19_3      NBPF3      bleedthrough                4   \n",
       "10  ENSG00000078124.13_1      ACER3      bleedthrough                3   \n",
       "11  ENSG00000119392.16_1       GLE1      bleedthrough                3   \n",
       "12  ENSG00000127511.10_2      SIN3B      bleedthrough                3   \n",
       "17  ENSG00000155897.10_1      ADCY8      bleedthrough                2   \n",
       "18  ENSG00000162944.11_1      RFTN2      bleedthrough                2   \n",
       "24  ENSG00000101624.11_3      CEP76      bleedthrough                1   \n",
       "25  ENSG00000102290.23_3    PCDH11X      bleedthrough                1   \n",
       "28  ENSG00000107798.18_2       LIPA      bleedthrough                1   \n",
       "30  ENSG00000117151.13_1       CTBS      bleedthrough                1   \n",
       "36  ENSG00000144460.13_2      NYAP2      bleedthrough                1   \n",
       "39  ENSG00000158604.15_1      TMED4      bleedthrough                1   \n",
       "49  ENSG00000186047.11_2      DLEU7      bleedthrough                1   \n",
       "51   ENSG00000213468.7_1      FIRRE      bleedthrough                1   \n",
       "53   ENSG00000247199.6_1  FBXO38-DT      bleedthrough                1   \n",
       "\n",
       "                                      experiment_name annot_status  \\\n",
       "0   humphrey_i3_cortical,seddighi_i3_cortical,zano...        novel   \n",
       "2   humphrey_i3_cortical,zanovello_i3_cortical_upf...        novel   \n",
       "6   brown_i3_cortical,humphrey_i3_cortical,seddigh...        novel   \n",
       "9   brown_skndz,humphrey_i3_cortical,zanovello_shs...        novel   \n",
       "10  brown_i3_cortical,seddighi_i3_cortical,zanovel...        novel   \n",
       "11  zanovello_i3_cortical_upf1_tdp_tdpkd_upf1ctl_v...        novel   \n",
       "12  brown_i3_cortical,humphrey_i3_cortical,seddigh...        novel   \n",
       "17  humphrey_i3_cortical,zanovello_i3_cortical_upf...        novel   \n",
       "18  humphrey_i3_cortical,zanovello_i3_cortical_upf...        novel   \n",
       "24                                  brown_i3_cortical    annotated   \n",
       "25                               humphrey_i3_cortical        novel   \n",
       "28                               humphrey_i3_cortical        novel   \n",
       "30                               humphrey_i3_cortical        novel   \n",
       "36                               humphrey_i3_cortical        novel   \n",
       "39                               humphrey_i3_cortical    annotated   \n",
       "49                               humphrey_i3_cortical        novel   \n",
       "51                                  brown_i3_cortical        novel   \n",
       "53                               humphrey_i3_cortical        novel   \n",
       "\n",
       "                        dataset event_manual_validation cryptic_riboseq_reads  \\\n",
       "0                           NaN                     yes                   yes   \n",
       "2                           NaN                     yes                    no   \n",
       "6                           NaN                     yes                   yes   \n",
       "9                           NaN                     yes               unclear   \n",
       "10                          NaN                     yes             few reads   \n",
       "11  zanovello_shsy5y_curve_0075                     yes                   NaN   \n",
       "12                          NaN                     yes                    no   \n",
       "17                          NaN                     yes                    no   \n",
       "18                          NaN                     yes                    no   \n",
       "24                          NaN                     yes                    no   \n",
       "25                          NaN                     yes                    no   \n",
       "28                          NaN                     yes                   NaN   \n",
       "30                          NaN                     yes                    no   \n",
       "36                          NaN                     yes                    no   \n",
       "39                          NaN                     yes                   yes   \n",
       "49                          NaN                     yes                    no   \n",
       "51                          NaN                     yes                    no   \n",
       "53                          NaN                     yes                    no   \n",
       "\n",
       "   riboseq_downstream_change  \\\n",
       "0                    unclear   \n",
       "2                    unclear   \n",
       "6                        yes   \n",
       "9                    unclear   \n",
       "10                   unclear   \n",
       "11                       NaN   \n",
       "12                       yes   \n",
       "17                        no   \n",
       "18                        no   \n",
       "24                        no   \n",
       "25                        no   \n",
       "28                       NaN   \n",
       "30                        no   \n",
       "36                        no   \n",
       "39                       yes   \n",
       "49                        no   \n",
       "51                        no   \n",
       "53                        no   \n",
       "\n",
       "                                                notes  \n",
       "0   reads in cryptic also downstream of stops in a...  \n",
       "2                                                 NaN  \n",
       "6                                                 NaN  \n",
       "9   expressed in ctls but likely cryptic by ratio ...  \n",
       "10  coverage quite low, but reads consistently up ...  \n",
       "11  3'end of bleedthroughs are truncated. lowly ex...  \n",
       "12    cryptic CDS terminates 1 codon downstream of SJ  \n",
       "17  seddighi - event looks real but v lowly expres...  \n",
       "18  very clean, obvious downreg of RNA downstream....  \n",
       "24  looks like real 3'UTR, but unclear whether cry...  \n",
       "25   lowly included. Whole exon is cryptic (novel SJ)  \n",
       "28  genuine event but not cryptic (overlapping txs...  \n",
       "30  genuine bleedthrough, not particularly cryptic...  \n",
       "36  expression in ctls, likely cryptic by ratio (e...  \n",
       "39                 annotated, also low level basal IR  \n",
       "49                                                NaN  \n",
       "51             lncRNA. Looks real but lowly expressed  \n",
       "53                               very lowly expressed  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First define set of IDs & last exons for which to extract peptide seqs\n",
    "bleeds = pd.read_csv(\"data/bleedthrough_manual_validation.tsv\", sep=\"\\t\")\n",
    "\n",
    "bleeds_y = bleeds[bleeds[\"event_manual_validation\"] == \"yes\"]\n",
    "# filter for at least 1 i3 cortical dataset\n",
    "bleeds_y = bleeds_y[bleeds_y.experiment_name.str.contains(\"i3_cortical\", regex=False)]\n",
    "bleeds_gn = bleeds_y.gene_name.unique().tolist()\n",
    "bleeds_le = bleeds_y.le_id.unique().tolist()\n",
    "\n",
    "print(f\"Number of cryptics - {bleeds_y.le_id.nunique()}\")\n",
    "bleeds_y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['brown_skndz,humphrey_i3_cortical,zanovello_shsy5y_curve_0021,zanovello_shsy5y_curve_0075']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bleeds_y.loc[bleeds_y.gene_name == \"NBPF3\", \"experiment_name\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147622531</td>\n",
       "      <td>147623356</td>\n",
       "      <td>ENSG00000116128.12_2|BCL9|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "      <td>BCL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45013791</td>\n",
       "      <td>45013889</td>\n",
       "      <td>ENSG00000126088.14_1|UROD|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000126088.14_1</td>\n",
       "      <td>UROD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>NBPF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>156139102</td>\n",
       "      <td>156140091</td>\n",
       "      <td>ENSG00000160789.24_11|LMNA|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160789.24_11</td>\n",
       "      <td>LMNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616614</td>\n",
       "      <td>1619210</td>\n",
       "      <td>ENSG00000197530.13_1|MIB2|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000197530.13_1</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr21</td>\n",
       "      <td>42849225</td>\n",
       "      <td>42850040</td>\n",
       "      <td>ENSG00000160193.12_1|WDR4|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000160193.12_1</td>\n",
       "      <td>WDR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chrX</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>ENSG00000102290.23_3|PCDH11X|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "      <td>PCDH11X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chrX</td>\n",
       "      <td>107087378</td>\n",
       "      <td>107088435</td>\n",
       "      <td>ENSG00000089682.17_2|RBM41|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000089682.17_2</td>\n",
       "      <td>RBM41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "      <td>MED14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>FIRRE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +5    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 147622531 | 147623356 | ...   |\n",
       "| chr1         | 45013791  | 45013889  | ...   |\n",
       "| chr1         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | 156139102 | 156140091 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 91882906  | 91891321  | ...   |\n",
       "| chrX         | 107087378 | 107088435 | ...   |\n",
       "| chrX         | 40653641  | 40654363  | ...   |\n",
       "| chrX         | 131823777 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 74 rows and 8 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Name, Score, Strand, le_id, gene_name"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in consolidated bleedthrough coordinates \n",
    "# NB: THESE ARE THE UNIQUE REGION COORDINATES, so not used to extract peptides (but instead used to provide coordinates)\n",
    "bleeds_le = pr.read_bed(\"data/2023-07-04_papa_cryptic_bleedthrough.last_exons.bed\")\n",
    "bleeds_le_cryp = bleeds_le[bleeds_le.Name.str.contains(\"cryptic\",regex=False)]\n",
    "# pull out gene name and le_id from the Name field\n",
    "bleeds_le_cryp = bleeds_le_cryp.apply(lambda df: pd.concat([df, df.Name.str.split(\"\\\\|\", expand=True)[[0,1]].rename(columns={0: \"le_id\", 1: \"gene_name\"})], axis=1))\n",
    "bleeds_le_cryp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# bleeds_le_cryp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.4038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.7930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.6392</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-C_S3.6203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148377695</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-D_S4.6266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr5</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>148378366</td>\n",
       "      <td>148378845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.6788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.8982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.8671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.DZ_curves_1_1.7410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_1.9014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.7018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.7006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr6</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.6691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr7</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581239</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>44581092</td>\n",
       "      <td>44581239</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>1196.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr7</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581239</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>44581092</td>\n",
       "      <td>44581239</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>1189.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.11594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.11524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr8</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.12288</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr9</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128516666</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.9453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr9</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128515777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065405_S46.13067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr9</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128516262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.12557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.14569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.14306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-2.13552</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89331008</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.13899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89332056</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-6.13417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>89334602</td>\n",
       "      <td>89335534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.10969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.Cont-D_S4.11799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-G_S7.12058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.14768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr11</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr13</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr13</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23075386</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.19771</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23076572</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.20251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079037</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_2.14903</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079058</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-E_S5.16462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079174</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.15356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079300</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_1.14929</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.14367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr16</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_4.15637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr18</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>12673387</td>\n",
       "      <td>12673503</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>763.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr18</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>12673387</td>\n",
       "      <td>12673503</td>\n",
       "      <td>ENST00000590143.5</td>\n",
       "      <td>291.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr19</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr19</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | .          | exon         | 84561927  | 84563418  | ...   |\n",
       "| chr2         | .          | exon         | 225651431 | 225658565 | ...   |\n",
       "| chr2         | .          | exon         | 197630416 | 197631220 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr19        | .          | exon         | 16854142  | 16857420  | ...   |\n",
       "| chr19        | .          | exon         | 32635458  | 32639488  | ...   |\n",
       "| chrX         | .          | exon         | 131823775 | 131825365 | ...   |\n",
       "| chrX         | .          | exon         | 131823777 | 131825365 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 56 rows and 22 columns from 14 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alt approach so have full last exon sequence\n",
    "full_bleeds_le = pr.read_gtf(\"data/novel_ref_combined.last_exons.gtf\")\n",
    "full_bleeds_le_mv =  full_bleeds_le.subset(lambda df: df[\"le_id\"].isin(set(bleeds_y[\"le_id\"])))\n",
    "\n",
    "full_bleeds_le_mv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>ref_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>PAPA.TDP-1.345.1</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091.9</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451.2</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119.3</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP-4.4038.5</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr5</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.7930.4</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr5</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>PAPA.TDP43-G_S7.6392.4</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr5</td>\n",
       "      <td>148377588</td>\n",
       "      <td>148378845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>PAPA.Cont-C_S3.6203.3</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr5</td>\n",
       "      <td>148377695</td>\n",
       "      <td>148378845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>PAPA.Cont-D_S4.6266.4</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr5</td>\n",
       "      <td>148378366</td>\n",
       "      <td>148378845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000247199.6_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.6788.4</td>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>ENSG00000247199.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP-4.8982.1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.8671.1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.DZ_curves_1_1.7410.3</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_1.9014.3</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP43-F_S6.7018.3</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.7006.2</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.6691.2</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44579896</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44579903</td>\n",
       "      <td>44581239</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>ENST00000477639.5</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130899857</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.12351.12</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130900799</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-1.11594.5</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP-6.11524.6</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr8</td>\n",
       "      <td>130901663</td>\n",
       "      <td>130904042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.12288.5</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr9</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128516666</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.9453.3</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr9</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128515777</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>PAPA.TDP43_19065405_S46.13067.1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr9</td>\n",
       "      <td>128515528</td>\n",
       "      <td>128516262</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>PAPA.TDP-6.12557.1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.14569.8</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.14306.12</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP-2.13552.8</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89331008</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.13899.16</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89332056</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP-6.13417.11</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89334602</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43-G_S7.10969.1</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.Cont-D_S4.11799.1</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869050</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_1.11919.4</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-G_S7.12058.3</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869405</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.12818.2</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76871079</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.15683.2</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869066</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.14768.6</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr11</td>\n",
       "      <td>76868119</td>\n",
       "      <td>76869450</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>PAPA.TDP43_19065413_S19.15054.2</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50814965</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.18282.3</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr13</td>\n",
       "      <td>50815080</td>\n",
       "      <td>50823520</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.17822.2</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23075386</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP-4.19771.4</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23076572</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.20251.8</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079037</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_2.14903.3</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079058</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-E_S5.16462.3</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079174</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.TDP43-F_S6.15356.5</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079300</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_CTRL_ctrl_1.14929.4</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_3.14367.6</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23079754</td>\n",
       "      <td>23080171</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_4.15637.5</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12672624</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr18</td>\n",
       "      <td>12673096</td>\n",
       "      <td>12673503</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chrX</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26532.4</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505.8</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 21453372  | 21457150  | +            | ...   |\n",
       "| chr1         | 84561927  | 84563418  | -            | ...   |\n",
       "| chr2         | 225651431 | 225658565 | +            | ...   |\n",
       "| chr2         | 197630416 | 197631220 | -            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr19        | 16854142  | 16857420  | +            | ...   |\n",
       "| chr19        | 32635458  | 32639488  | -            | ...   |\n",
       "| chrX         | 131823775 | 131825365 | -            | ...   |\n",
       "| chrX         | 131823777 | 131825365 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 56 rows and 8 columns from 14 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: le_id, transcript_id, ref_gene_name, ref_gene_id"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_bleeds_le_mv = full_bleeds_le_mv[[\"le_id\",\"transcript_id\", \"ref_gene_name\", \"ref_gene_id\"]]\n",
    "full_bleeds_le_mv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 29.4 s, sys: 4.94 s, total: 34.3 s\n",
      "Wall time: 1min 6s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21483467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21440348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001546346.2</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21444947</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131755597</td>\n",
       "      <td>131755746</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSE00001752569.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131749305</td>\n",
       "      <td>131749458</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00001674554.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1417</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131711650</td>\n",
       "      <td>131711720</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00001594498.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1418</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131709497</td>\n",
       "      <td>131709556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSE00001734033.1</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1419</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131702651</td>\n",
       "      <td>131703518</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-201</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000493690.2</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSE00001658652.2</td>\n",
       "      <td>HGNC:49627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1420 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +19   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | transcript | 21440127  | 21483467  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21440127  | 21440348  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21444947  | 21445219  | ...   |\n",
       "| chr1         | HAVANA     | CDS        | 21445086  | 21445219  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131749305 | 131749458 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131711650 | 131711720 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131709497 | 131709556 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131702651 | 131703518 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,420 rows and 24 columns from 14 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# read in reference GTF (used to define last exons), subsetting only to cryptic-containing genes to save memory\n",
    "ref_gtf = pr.read_gtf(\"data/reference_filtered.gtf\").subset(lambda df: df.gene_name.isin(set(bleeds_y.gene_name)))\n",
    "ref_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21470631</td>\n",
       "      <td>21470734</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003734729.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21471568</td>\n",
       "      <td>21471783</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003723776.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21472842</td>\n",
       "      <td>21472915</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003753961.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21473379</td>\n",
       "      <td>21473585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003747501.1</td>\n",
       "      <td>HGNC:25076</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>91876780</td>\n",
       "      <td>91879273</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCDH11X-207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000359297.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003757792.1</td>\n",
       "      <td>HGNC:8656</td>\n",
       "      <td>ENSP00000384758.1</td>\n",
       "      <td>CCDS55459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>92201374</td>\n",
       "      <td>92201455</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCDH11X-207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000359297.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001300322.1</td>\n",
       "      <td>HGNC:8656</td>\n",
       "      <td>ENSP00000384758.1</td>\n",
       "      <td>CCDS55459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>92263113</td>\n",
       "      <td>92263143</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCDH11X-207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000359297.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001327755.1</td>\n",
       "      <td>HGNC:8656</td>\n",
       "      <td>ENSP00000384758.1</td>\n",
       "      <td>CCDS55459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>92387734</td>\n",
       "      <td>92387933</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCDH11X-207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000359297.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001718872.1</td>\n",
       "      <td>HGNC:8656</td>\n",
       "      <td>ENSP00000384758.1</td>\n",
       "      <td>CCDS55459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>92618263</td>\n",
       "      <td>92618937</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCDH11X-207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic,appris_alternative_2,CCDS</td>\n",
       "      <td>OTTHUMT00000359297.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001459488.2</td>\n",
       "      <td>HGNC:8656</td>\n",
       "      <td>ENSP00000384758.1</td>\n",
       "      <td>CCDS55459.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>426 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +19   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | CDS        | 21445086  | 21445219  | ...   |\n",
       "| chr1         | HAVANA     | CDS        | 21470631  | 21470734  | ...   |\n",
       "| chr1         | HAVANA     | CDS        | 21471568  | 21471783  | ...   |\n",
       "| chr1         | HAVANA     | CDS        | 21472842  | 21472915  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | CDS        | 92201374  | 92201455  | ...   |\n",
       "| chrX         | HAVANA     | CDS        | 92263113  | 92263143  | ...   |\n",
       "| chrX         | HAVANA     | CDS        | 92387734  | 92387933  | ...   |\n",
       "| chrX         | HAVANA     | CDS        | 92618263  | 92618937  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 426 rows and 24 columns from 13 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract CDS annotations\n",
    "ref_cds = ref_gtf.subset(lambda df: df.Feature == \"CDS\")\n",
    "ref_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'havana_gene',\n",
       "       'transcript_id', 'transcript_type', 'transcript_name',\n",
       "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
       "       'exon_id', 'hgnc_id', 'protein_id', 'ccdsid'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check ref_cds has a strand aware exon_number/region_number\n",
    "ref_cds.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445219</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000454000.6</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2</td>\n",
       "      <td>21470631</td>\n",
       "      <td>21470734</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000454000.6</td>\n",
       "      <td>3</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>21471568</td>\n",
       "      <td>21471783</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000454000.6</td>\n",
       "      <td>4</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2</td>\n",
       "      <td>21472842</td>\n",
       "      <td>21472915</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000454000.6</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>21473379</td>\n",
       "      <td>21473585</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000454000.6</td>\n",
       "      <td>6</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>91876780</td>\n",
       "      <td>91879273</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000406881.3</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>92201374</td>\n",
       "      <td>92201455</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000406881.3</td>\n",
       "      <td>3</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>92263113</td>\n",
       "      <td>92263143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000406881.3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>92387734</td>\n",
       "      <td>92387933</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000406881.3</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>chrX</td>\n",
       "      <td>2</td>\n",
       "      <td>92618263</td>\n",
       "      <td>92618937</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000406881.3</td>\n",
       "      <td>6</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>426 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Frame      | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 0          | 21445086  | 21445219  | +            | ...   |\n",
       "| chr1         | 2          | 21470631  | 21470734  | +            | ...   |\n",
       "| chr1         | 1          | 21471568  | 21471783  | +            | ...   |\n",
       "| chr1         | 2          | 21472842  | 21472915  | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 0          | 92201374  | 92201455  | +            | ...   |\n",
       "| chrX         | 0          | 92263113  | 92263143  | +            | ...   |\n",
       "| chrX         | 0          | 92387734  | 92387933  | +            | ...   |\n",
       "| chrX         | 2          | 92618263  | 92618937  | +            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 426 rows and 10 columns from 13 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_id, gene_name, transcript_id, exon_number, Feature"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_cds = ref_cds[[\"Frame\",\"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"Feature\"]]\n",
    "ref_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_bl</th>\n",
       "      <th>End_bl</th>\n",
       "      <th>Strand_bl</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id_bl</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>ref_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>84563256</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>PAPA.ctrl_ctrl_1.1091.9</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225651535</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>7</td>\n",
       "      <td>CDS</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.4451.2</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.4119.3</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>197631010</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>PAPA.TDP-4.4038.5</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42935670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>PAPA.TDP-4.8982.1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854261</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854262</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16854261</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.24194.4</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr19</td>\n",
       "      <td>1</td>\n",
       "      <td>32639355</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>12</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr19</td>\n",
       "      <td>1</td>\n",
       "      <td>32639304</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>12</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>PAPA.TDP43_19065411_S54.22523.3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Frame      | Start     | End       | Strand       | +12   |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 0          | 84563256  | 84563418  | -            | ...   |\n",
       "| chr2         | 2          | 225651431 | 225651535 | +            | ...   |\n",
       "| chr2         | 2          | 197631010 | 197631220 | -            | ...   |\n",
       "| chr2         | 2          | 197631010 | 197631220 | -            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chr19        | 0          | 16854142  | 16854262  | +            | ...   |\n",
       "| chr19        | 0          | 16854142  | 16854261  | +            | ...   |\n",
       "| chr19        | 1          | 32639355  | 32639488  | -            | ...   |\n",
       "| chr19        | 1          | 32639304  | 32639488  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 62 rows and 17 columns from 11 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "12 hidden columns: gene_id, gene_name, transcript_id, exon_number, Feature, Start_bl, ... (+ 6 more.)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# step 1 - find the annotated CDS regions that correspond to the bleedthrough alst exon\n",
    "# Definition requires that 5'ends exactly match\n",
    "# filter for exact 3'ends, then update the 3'end position of the annotated CDS object to the bleedthrough 3'end\n",
    "\n",
    "ref_cds_bld = (ref_cds.join(full_bleeds_le_mv, strandedness=\"same\",suffix=\"_bl\")\n",
    " .subset(lambda df: ((df.Strand == \"+\") & (df.Start == df[\"Start\" + \"_bl\"])) | ((df.Strand == \"-\") & (df.End == df[\"End\" + \"_bl\"]))))\n",
    "\n",
    "ref_cds_bld\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Feature</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>225651431</td>\n",
       "      <td>225658565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>7</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>197630416</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>197630635</td>\n",
       "      <td>197631220</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>42935573</td>\n",
       "      <td>42936378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16857420</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr19</td>\n",
       "      <td>1</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>12</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr19</td>\n",
       "      <td>1</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>12</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Frame      | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 0          | 84561927  | 84563418  | -            | ...   |\n",
       "| chr2         | 2          | 225651431 | 225658565 | +            | ...   |\n",
       "| chr2         | 2          | 197630416 | 197631220 | -            | ...   |\n",
       "| chr2         | 2          | 197630635 | 197631220 | -            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chr19        | 0          | 16854142  | 16857420  | +            | ...   |\n",
       "| chr19        | 0          | 16854142  | 16857420  | +            | ...   |\n",
       "| chr19        | 1          | 32635458  | 32639488  | -            | ...   |\n",
       "| chr19        | 1          | 32635458  | 32639488  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 62 rows and 11 columns from 11 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: gene_id, gene_name, transcript_id, exon_number, Feature, le_id"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Swap Start and End coordinates to the bleedthrough coordinates, then drop remaining joined columns\n",
    "# i.e. the 5'end coordinate remains the same, but update the 3'end to the end of the bleedthrough event\n",
    "ref_cds_bld_swap = ref_cds_bld.new_position(\"swap\").drop(like=\"^ref|bl$\")\n",
    "ref_cds_bld_swap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exon_number_le</th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000272907.8;ENSG00000144460.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000295049.9;ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000372836.5;ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000457408.7;ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000481238.1;ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000286355.10;ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000377928.7;ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENST00000309971.9</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000309971.9;ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENST00000372770.4</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000372770.4;ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000525194.5;ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000531352.5;ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000504404.2;ENSG00000186047.11_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>13</td>\n",
       "      <td>ENST00000219689.12;ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000262127.7;ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000423709.6;ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name       transcript_id                 le_id  \\\n",
       "0   ENSG00000117151.13      CTBS   ENST00000370630.6  ENSG00000117151.13_1   \n",
       "1   ENSG00000144460.13     NYAP2   ENST00000272907.8  ENSG00000144460.13_2   \n",
       "2   ENSG00000162944.11     RFTN2   ENST00000295049.9  ENSG00000162944.11_1   \n",
       "4   ENSG00000137161.18     CNPY3   ENST00000372836.5  ENSG00000137161.18_1   \n",
       "11  ENSG00000158604.15     TMED4   ENST00000457408.7  ENSG00000158604.15_1   \n",
       "13  ENSG00000158604.15     TMED4   ENST00000481238.1  ENSG00000158604.15_1   \n",
       "15  ENSG00000155897.10     ADCY8  ENST00000286355.10  ENSG00000155897.10_1   \n",
       "19  ENSG00000155897.10     ADCY8   ENST00000377928.7  ENSG00000155897.10_1   \n",
       "23  ENSG00000119392.16      GLE1   ENST00000309971.9  ENSG00000119392.16_1   \n",
       "26  ENSG00000119392.16      GLE1   ENST00000372770.4  ENSG00000119392.16_1   \n",
       "29  ENSG00000078124.13     ACER3   ENST00000525194.5  ENSG00000078124.13_1   \n",
       "36  ENSG00000078124.13     ACER3   ENST00000531352.5  ENSG00000078124.13_1   \n",
       "43  ENSG00000186047.11     DLEU7   ENST00000504404.2  ENSG00000186047.11_2   \n",
       "45  ENSG00000103404.15     USP31  ENST00000219689.12  ENSG00000103404.15_1   \n",
       "53  ENSG00000101624.11     CEP76   ENST00000262127.7  ENSG00000101624.11_3   \n",
       "55  ENSG00000101624.11     CEP76   ENST00000423709.6  ENSG00000101624.11_3   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "60  ENSG00000105186.16   ANKRD27   ENST00000306065.9  ENSG00000105186.16_3   \n",
       "61  ENSG00000105186.16   ANKRD27   ENST00000587352.5  ENSG00000105186.16_3   \n",
       "\n",
       "   exon_number_le                    cryptic_transcript_id  \n",
       "0               6   ENST00000370630.6;ENSG00000117151.13_1  \n",
       "1               7   ENST00000272907.8;ENSG00000144460.13_2  \n",
       "2               5   ENST00000295049.9;ENSG00000162944.11_1  \n",
       "4               3   ENST00000372836.5;ENSG00000137161.18_1  \n",
       "11              4   ENST00000457408.7;ENSG00000158604.15_1  \n",
       "13              4   ENST00000481238.1;ENSG00000158604.15_1  \n",
       "15              7  ENST00000286355.10;ENSG00000155897.10_1  \n",
       "19              7   ENST00000377928.7;ENSG00000155897.10_1  \n",
       "23              3   ENST00000309971.9;ENSG00000119392.16_1  \n",
       "26              3   ENST00000372770.4;ENSG00000119392.16_1  \n",
       "29              2   ENST00000525194.5;ENSG00000078124.13_1  \n",
       "36              2   ENST00000531352.5;ENSG00000078124.13_1  \n",
       "43              2   ENST00000504404.2;ENSG00000186047.11_2  \n",
       "45             13  ENST00000219689.12;ENSG00000103404.15_1  \n",
       "53             12   ENST00000262127.7;ENSG00000101624.11_3  \n",
       "55             11   ENST00000423709.6;ENSG00000101624.11_3  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2  \n",
       "60             12   ENST00000306065.9;ENSG00000105186.16_3  \n",
       "61             12   ENST00000587352.5;ENSG00000105186.16_3  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# construct a 'cryptic transcript ID' - combine annotated transcript with the le_id\n",
    "cryp_txid = ref_cds_bld_swap.as_df()[[\"gene_id\", \"gene_name\", \"transcript_id\", \"le_id\", \"exon_number\"]].rename(columns={\"exon_number\": \"exon_number_le\"})\n",
    "cryp_txid[\"cryptic_transcript_id\"] = cryp_txid[\"transcript_id\"].str.cat(cryp_txid[\"le_id\"], sep=\";\")\n",
    "cryp_txid.drop_duplicates(inplace=True)\n",
    "cryp_txid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exon_number_le</th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name       transcript_id                 le_id  \\\n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "57  ENSG00000127511.10     SIN3B  ENST00000248054.10  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "58  ENSG00000127511.10     SIN3B   ENST00000596802.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "59  ENSG00000127511.10     SIN3B   ENST00000379803.5  ENSG00000127511.10_2   \n",
       "\n",
       "   exon_number_le                    cryptic_transcript_id  exon_number  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            1  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            2  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            3  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            4  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            5  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            6  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            7  \n",
       "57              8  ENST00000248054.10;ENSG00000127511.10_2            8  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            1  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            2  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            3  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            4  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            5  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            6  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            7  \n",
       "58              8   ENST00000596802.5;ENSG00000127511.10_2            8  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            1  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            2  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            3  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            4  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            5  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            6  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            7  \n",
       "59              8   ENST00000379803.5;ENSG00000127511.10_2            8  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand cryp_txid such that n*exon_number_le rows repeated for each cryptic_tx_id\n",
    "# and label these with exon number 1..n\n",
    "# Essentially, this gives you a 'truncated' transcript's exon structure, so can pull out the CDS annotations up to and including the bleedthrough\n",
    "# --> peptide sequence for cryptic containing transcript\n",
    "\n",
    "# generate n copies of rows for each cryptic-matching transcript\n",
    "trnc_cryp_tx = cryp_txid.drop_duplicates().loc[lambda df: df.index.repeat(df[\"exon_number_le\"])]\n",
    "# add exon number from 1..n (n = exon_number_le)\n",
    "trnc_cryp_tx[\"exon_number\"] = trnc_cryp_tx.groupby('cryptic_transcript_id').cumcount() + 1\n",
    "trnc_cryp_tx[trnc_cryp_tx.gene_name == \"SIN3B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct a gr of CDS objects\n",
    "# should be all annotated cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exon_number_le</th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>84561927</td>\n",
       "      <td>84563418</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>84563734</td>\n",
       "      <td>84563832</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>84565840</td>\n",
       "      <td>84566012</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>84569930</td>\n",
       "      <td>84570139</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>84570581</td>\n",
       "      <td>84570720</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>0</td>\n",
       "      <td>32646458</td>\n",
       "      <td>32646615</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "      <td>4</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>0</td>\n",
       "      <td>32649681</td>\n",
       "      <td>32649792</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>0</td>\n",
       "      <td>32649681</td>\n",
       "      <td>32649792</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>0</td>\n",
       "      <td>32658913</td>\n",
       "      <td>32659015</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>0</td>\n",
       "      <td>32658913</td>\n",
       "      <td>32659015</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+------------+-----------+-----------+--------------+------------+-------+\n",
       "| Frame      | Start     | End       | Strand       | Feature    | +8    |\n",
       "| (object)   | (int64)   | (int64)   | (category)   | (object)   | ...   |\n",
       "|------------+-----------+-----------+--------------+------------+-------|\n",
       "| 0          | 84561927  | 84563418  | -            | CDS        | ...   |\n",
       "| 2          | 84563734  | 84563832  | -            | CDS        | ...   |\n",
       "| 0          | 84565840  | 84566012  | -            | CDS        | ...   |\n",
       "| 2          | 84569930  | 84570139  | -            | CDS        | ...   |\n",
       "| ...        | ...       | ...       | ...          | ...        | ...   |\n",
       "| 0          | 32649681  | 32649792  | -            | CDS        | ...   |\n",
       "| 0          | 32649681  | 32649792  | -            | CDS        | ...   |\n",
       "| 0          | 32658913  | 32659015  | -            | CDS        | ...   |\n",
       "| 0          | 32658913  | 32659015  | -            | CDS        | ...   |\n",
       "+------------+-----------+-----------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 176 rows and 13 columns from 11 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: gene_id, gene_name, transcript_id, le_id, exon_number_le, cryptic_transcript_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# construct a gr of CDSs up until bleedthrough last exon\n",
    "trnc_cryp_tx = trnc_cryp_tx.astype({\"exon_number\": str, \"exon_number_le\": str})\n",
    "\n",
    "# first, extract CDS coordinates for the bleedthrough last exon\n",
    "trnc_cryp_tx_bl = trnc_cryp_tx.loc[lambda df: df[\"exon_number\"] == df[\"exon_number_le\"],:]\n",
    "\n",
    "# now join in CDS coordinates fro bleedthrough for each truncated transcript\n",
    "trnc_cds_bl = trnc_cryp_tx_bl.merge(ref_cds_bld_swap.as_df().astype({\"exon_number\": str}).drop(columns=[\"gene_id\", \"gene_name\"]),\n",
    "                      on=[\"transcript_id\", \"le_id\", \"exon_number\"],\n",
    "                      how=\"left\")\n",
    "\n",
    "# make sure all had a CDS/coordinates joined\n",
    "assert len(trnc_cds_bl.loc[lambda df: df[\"Chromosome\"].isna(), :]) == 0\n",
    "\n",
    "# repeat for all CDS exons before the bleedthrough\n",
    "trnc_cryp_tx_bf = trnc_cryp_tx.loc[lambda df: df[\"exon_number\"] != df[\"exon_number_le\"],:]\n",
    "# join with annotated CDS to get coordinates\n",
    "trnc_cds_bf = trnc_cryp_tx_bf.merge(ref_cds.as_df().astype({\"exon_number\": str}).drop(columns=[\"gene_id\", \"gene_name\"]),\n",
    "                      on=[\"exon_number\", \"transcript_id\"],\n",
    "                      ) # since some exons may not contain CDS, only join where matches (at true start of CDS)\n",
    "\n",
    "# make sure all had a CDS/coordinates joined\n",
    "try:\n",
    "    assert len(trnc_cds_bf.loc[lambda df: df[\"Chromosome\"].isna(), :]) == 0\n",
    "except AssertionError:\n",
    "    print(trnc_cds_bf.loc[lambda df: df[\"Chromosome\"].isna(), :])\n",
    "\n",
    "# # # combine the two dfs and convert to pyranges object\n",
    "trnc_cds_comb = pd.concat([trnc_cds_bl, trnc_cds_bf])\n",
    "new_order = trnc_cds_comb.columns.tolist()[-5:] + trnc_cds_comb.columns.tolist()[:-5]\n",
    "trnc_cds_comb = trnc_cds_comb[new_order]\n",
    "trnc_cds_comb_gr = pr.PyRanges(trnc_cds_comb).sort()\n",
    "\n",
    "trnc_cds_comb_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Frame</th>\n",
       "      <th>peptide_seq</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exon_number_le</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000219689.12;ENSG00000103404.15_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000262127.7;ENSG00000101624.11_3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000272907.8;ENSG00000144460.13_2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000286355.10;ENSG00000155897.10_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000295049.9;ENSG00000162944.11_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000309971.9;ENSG00000119392.16_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENST00000309971.9</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000372770.4;ENSG00000119392.16_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENST00000372770.4</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000372836.5;ENSG00000137161.18_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000377928.7;ENSG00000155897.10_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000423709.6;ENSG00000101624.11_3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000457408.7;ENSG00000158604.15_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000481238.1;ENSG00000158604.15_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000504404.2;ENSG00000186047.11_2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000525194.5;ENSG00000078124.13_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000531352.5;ENSG00000078124.13_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      cryptic_transcript_id  exon_number  Frame  \\\n",
       "0   ENST00000219689.12;ENSG00000103404.15_1            1      0   \n",
       "1   ENST00000248054.10;ENSG00000127511.10_2            1      0   \n",
       "2    ENST00000262127.7;ENSG00000101624.11_3            1      0   \n",
       "3    ENST00000272907.8;ENSG00000144460.13_2            3      0   \n",
       "4   ENST00000286355.10;ENSG00000155897.10_1            1      0   \n",
       "5    ENST00000295049.9;ENSG00000162944.11_1            1      0   \n",
       "6    ENST00000306065.9;ENSG00000105186.16_3            2      0   \n",
       "7    ENST00000309971.9;ENSG00000119392.16_1            1      0   \n",
       "8    ENST00000370630.6;ENSG00000117151.13_1            1      0   \n",
       "9    ENST00000372770.4;ENSG00000119392.16_1            1      0   \n",
       "10   ENST00000372836.5;ENSG00000137161.18_1            1      0   \n",
       "11   ENST00000377928.7;ENSG00000155897.10_1            1      0   \n",
       "12   ENST00000379803.5;ENSG00000127511.10_2            1      0   \n",
       "13   ENST00000423709.6;ENSG00000101624.11_3            1      0   \n",
       "14   ENST00000457408.7;ENSG00000158604.15_1            1      0   \n",
       "15   ENST00000481238.1;ENSG00000158604.15_1            1      0   \n",
       "16   ENST00000504404.2;ENSG00000186047.11_2            1      0   \n",
       "17   ENST00000525194.5;ENSG00000078124.13_1            1      0   \n",
       "18   ENST00000531352.5;ENSG00000078124.13_1            1      0   \n",
       "19   ENST00000587352.5;ENSG00000105186.16_3            2      0   \n",
       "20   ENST00000596802.5;ENSG00000127511.10_2            1      0   \n",
       "\n",
       "                                          peptide_seq             gene_id  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...  ENSG00000103404.15   \n",
       "1   MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "2   MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...  ENSG00000101624.11   \n",
       "3   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...  ENSG00000144460.13   \n",
       "4   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...  ENSG00000155897.10   \n",
       "5   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...  ENSG00000162944.11   \n",
       "6   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...  ENSG00000105186.16   \n",
       "7   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...  ENSG00000119392.16   \n",
       "8   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...  ENSG00000117151.13   \n",
       "9   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...  ENSG00000119392.16   \n",
       "10  MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...  ENSG00000137161.18   \n",
       "11  MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...  ENSG00000155897.10   \n",
       "12  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "13  MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...  ENSG00000101624.11   \n",
       "14  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...  ENSG00000158604.15   \n",
       "15  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...  ENSG00000158604.15   \n",
       "16  MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...  ENSG00000186047.11   \n",
       "17  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...  ENSG00000078124.13   \n",
       "18  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...  ENSG00000078124.13   \n",
       "19  MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...  ENSG00000105186.16   \n",
       "20  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "\n",
       "   gene_name       transcript_id                 le_id exon_number_le  \n",
       "0      USP31  ENST00000219689.12  ENSG00000103404.15_1             13  \n",
       "1      SIN3B  ENST00000248054.10  ENSG00000127511.10_2              8  \n",
       "2      CEP76   ENST00000262127.7  ENSG00000101624.11_3             12  \n",
       "3      NYAP2   ENST00000272907.8  ENSG00000144460.13_2              7  \n",
       "4      ADCY8  ENST00000286355.10  ENSG00000155897.10_1              7  \n",
       "5      RFTN2   ENST00000295049.9  ENSG00000162944.11_1              5  \n",
       "6    ANKRD27   ENST00000306065.9  ENSG00000105186.16_3             12  \n",
       "7       GLE1   ENST00000309971.9  ENSG00000119392.16_1              3  \n",
       "8       CTBS   ENST00000370630.6  ENSG00000117151.13_1              6  \n",
       "9       GLE1   ENST00000372770.4  ENSG00000119392.16_1              3  \n",
       "10     CNPY3   ENST00000372836.5  ENSG00000137161.18_1              3  \n",
       "11     ADCY8   ENST00000377928.7  ENSG00000155897.10_1              7  \n",
       "12     SIN3B   ENST00000379803.5  ENSG00000127511.10_2              8  \n",
       "13     CEP76   ENST00000423709.6  ENSG00000101624.11_3             11  \n",
       "14     TMED4   ENST00000457408.7  ENSG00000158604.15_1              4  \n",
       "15     TMED4   ENST00000481238.1  ENSG00000158604.15_1              4  \n",
       "16     DLEU7   ENST00000504404.2  ENSG00000186047.11_2              2  \n",
       "17     ACER3   ENST00000525194.5  ENSG00000078124.13_1              2  \n",
       "18     ACER3   ENST00000531352.5  ENSG00000078124.13_1              2  \n",
       "19   ANKRD27   ENST00000587352.5  ENSG00000105186.16_3             12  \n",
       "20     SIN3B   ENST00000596802.5  ENSG00000127511.10_2              8  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# translate truncated transcripts\n",
    "# Note: exon number col must be strand aware and int dtype to allow proper sorting\n",
    "trnc_cds_comb_gr = trnc_cds_comb_gr.apply(lambda df: df.astype({\"exon_number\": int}))\n",
    "\n",
    "\n",
    "trnc_cds_peptides = translate_tx(trnc_cds_comb_gr, \"cryptic_transcript_id\", add_stop=True)\n",
    "# return additional annotation information\n",
    "trnc_cds_peptides = trnc_cds_peptides.merge(cryp_txid, on=\"cryptic_transcript_id\")\n",
    "trnc_cds_peptides\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Frame</th>\n",
       "      <th>peptide_seq</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>exon_number_le</th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>ENSG00000103404.15</td>\n",
       "      <td>USP31</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>13</td>\n",
       "      <td>ENST00000219689.12;ENSG00000103404.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000262127.7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000262127.7;ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>ENSG00000144460.13</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000272907.8;ENSG00000144460.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000286355.10;ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>ENSG00000162944.11</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000295049.9;ENSG00000162944.11_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000309971.9</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000309971.9;ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>ENSG00000117151.13</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000372770.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>ENSG00000119392.16</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000372770.4;ENSG00000119392.16_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>ENSG00000137161.18</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000372836.5;ENSG00000137161.18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000377928.7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>ENSG00000155897.10</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000377928.7;ENSG00000155897.10_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000423709.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>ENSG00000101624.11</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>ENSG00000101624.11_3</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000423709.6;ENSG00000101624.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000457408.7;ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000481238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>ENSG00000158604.15</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000481238.1;ENSG00000158604.15_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000504404.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...</td>\n",
       "      <td>ENSG00000186047.11</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>ENSG00000186047.11_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000504404.2;ENSG00000186047.11_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000525194.5;ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000531352.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>ENSG00000078124.13</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000531352.5;ENSG00000078124.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000587352.5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>ENSG00000105186.16</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>ENSG00000127511.10</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENSG00000127511.10_2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id  exon_number  Frame  \\\n",
       "0   ENST00000219689.12            1      0   \n",
       "1   ENST00000248054.10            1      0   \n",
       "2    ENST00000262127.7            1      0   \n",
       "3    ENST00000272907.8            3      0   \n",
       "4   ENST00000286355.10            1      0   \n",
       "5    ENST00000295049.9            1      0   \n",
       "6    ENST00000306065.9            2      0   \n",
       "7    ENST00000309971.9            1      0   \n",
       "8    ENST00000370630.6            1      0   \n",
       "9    ENST00000372770.4            1      0   \n",
       "10   ENST00000372836.5            1      0   \n",
       "11   ENST00000377928.7            1      0   \n",
       "12   ENST00000379803.5            1      0   \n",
       "13   ENST00000423709.6            1      0   \n",
       "14   ENST00000457408.7            1      0   \n",
       "15   ENST00000481238.1            1      0   \n",
       "16   ENST00000504404.2            1      0   \n",
       "17   ENST00000525194.5            1      0   \n",
       "18   ENST00000531352.5            1      0   \n",
       "19   ENST00000587352.5            2      0   \n",
       "20   ENST00000596802.5            1      0   \n",
       "\n",
       "                                          peptide_seq             gene_id  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...  ENSG00000103404.15   \n",
       "1   MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "2   MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...  ENSG00000101624.11   \n",
       "3   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...  ENSG00000144460.13   \n",
       "4   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...  ENSG00000155897.10   \n",
       "5   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...  ENSG00000162944.11   \n",
       "6   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...  ENSG00000105186.16   \n",
       "7   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...  ENSG00000119392.16   \n",
       "8   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...  ENSG00000117151.13   \n",
       "9   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...  ENSG00000119392.16   \n",
       "10  MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...  ENSG00000137161.18   \n",
       "11  MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...  ENSG00000155897.10   \n",
       "12  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "13  MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...  ENSG00000101624.11   \n",
       "14  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...  ENSG00000158604.15   \n",
       "15  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...  ENSG00000158604.15   \n",
       "16  MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...  ENSG00000186047.11   \n",
       "17  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...  ENSG00000078124.13   \n",
       "18  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...  ENSG00000078124.13   \n",
       "19  MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...  ENSG00000105186.16   \n",
       "20  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...  ENSG00000127511.10   \n",
       "\n",
       "   gene_name                 le_id exon_number_le  \\\n",
       "0      USP31  ENSG00000103404.15_1             13   \n",
       "1      SIN3B  ENSG00000127511.10_2              8   \n",
       "2      CEP76  ENSG00000101624.11_3             12   \n",
       "3      NYAP2  ENSG00000144460.13_2              7   \n",
       "4      ADCY8  ENSG00000155897.10_1              7   \n",
       "5      RFTN2  ENSG00000162944.11_1              5   \n",
       "6    ANKRD27  ENSG00000105186.16_3             12   \n",
       "7       GLE1  ENSG00000119392.16_1              3   \n",
       "8       CTBS  ENSG00000117151.13_1              6   \n",
       "9       GLE1  ENSG00000119392.16_1              3   \n",
       "10     CNPY3  ENSG00000137161.18_1              3   \n",
       "11     ADCY8  ENSG00000155897.10_1              7   \n",
       "12     SIN3B  ENSG00000127511.10_2              8   \n",
       "13     CEP76  ENSG00000101624.11_3             11   \n",
       "14     TMED4  ENSG00000158604.15_1              4   \n",
       "15     TMED4  ENSG00000158604.15_1              4   \n",
       "16     DLEU7  ENSG00000186047.11_2              2   \n",
       "17     ACER3  ENSG00000078124.13_1              2   \n",
       "18     ACER3  ENSG00000078124.13_1              2   \n",
       "19   ANKRD27  ENSG00000105186.16_3             12   \n",
       "20     SIN3B  ENSG00000127511.10_2              8   \n",
       "\n",
       "                      cryptic_transcript_id  \n",
       "0   ENST00000219689.12;ENSG00000103404.15_1  \n",
       "1   ENST00000248054.10;ENSG00000127511.10_2  \n",
       "2    ENST00000262127.7;ENSG00000101624.11_3  \n",
       "3    ENST00000272907.8;ENSG00000144460.13_2  \n",
       "4   ENST00000286355.10;ENSG00000155897.10_1  \n",
       "5    ENST00000295049.9;ENSG00000162944.11_1  \n",
       "6    ENST00000306065.9;ENSG00000105186.16_3  \n",
       "7    ENST00000309971.9;ENSG00000119392.16_1  \n",
       "8    ENST00000370630.6;ENSG00000117151.13_1  \n",
       "9    ENST00000372770.4;ENSG00000119392.16_1  \n",
       "10   ENST00000372836.5;ENSG00000137161.18_1  \n",
       "11   ENST00000377928.7;ENSG00000155897.10_1  \n",
       "12   ENST00000379803.5;ENSG00000127511.10_2  \n",
       "13   ENST00000423709.6;ENSG00000101624.11_3  \n",
       "14   ENST00000457408.7;ENSG00000158604.15_1  \n",
       "15   ENST00000481238.1;ENSG00000158604.15_1  \n",
       "16   ENST00000504404.2;ENSG00000186047.11_2  \n",
       "17   ENST00000525194.5;ENSG00000078124.13_1  \n",
       "18   ENST00000531352.5;ENSG00000078124.13_1  \n",
       "19   ENST00000587352.5;ENSG00000105186.16_3  \n",
       "20   ENST00000596802.5;ENSG00000127511.10_2  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# translate the annotated transcript\n",
    "ref_cds = ref_cds.apply(lambda df: df.astype({\"exon_number\": int}))\n",
    "ref_cds_peptides = translate_tx(ref_cds.subset(lambda df: df.transcript_id.isin(cryp_txid[\"transcript_id\"])), \"transcript_id\", add_stop=True)\n",
    "# return additional annotation information\n",
    "ref_cds_peptides = ref_cds_peptides.merge(cryp_txid, on=\"transcript_id\")\n",
    "ref_cds_peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>peptide_seq_cryptic</th>\n",
       "      <th>peptide_seq_full</th>\n",
       "      <th>longest_match_slice</th>\n",
       "      <th>peptide_seq_cryptic_uniq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000219689.12;ENSG00000103404.15_1</td>\n",
       "      <td>USP31</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>725</td>\n",
       "      <td>GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000248054.10;ENSG00000127511.10_2</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>353</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000262127.7;ENSG00000101624.11_3</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>660</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000272907.8;ENSG00000144460.13_2</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>645</td>\n",
       "      <td>PSSLANRD*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000286355.10;ENSG00000155897.10_1</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>637</td>\n",
       "      <td>VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000295049.9;ENSG00000162944.11_1</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>310</td>\n",
       "      <td>KFYVNDILYLRNLNLYQNQ*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>372</td>\n",
       "      <td>VRSHPCPGLPLWASWFP*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000309971.9;ENSG00000119392.16_1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>144</td>\n",
       "      <td>VSEPMKEGSLDPASHI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>319</td>\n",
       "      <td>VRLFVSYEHLFY*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000372770.4;ENSG00000119392.16_1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>144</td>\n",
       "      <td>VSEPMKEGSLDPASHI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000372836.5;ENSG00000137161.18_1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>124</td>\n",
       "      <td>VGFGIVLHPLWGQACMYLSVSAGVSVI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000377928.7;ENSG00000155897.10_1</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>637</td>\n",
       "      <td>VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379803.5;ENSG00000127511.10_2</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>353</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000423709.6;ENSG00000101624.11_3</td>\n",
       "      <td>CEP76</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...</td>\n",
       "      <td>585</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000457408.7;ENSG00000158604.15_1</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>178</td>\n",
       "      <td>ASAYLLVI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000481238.1;ENSG00000158604.15_1</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>187</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000504404.2;ENSG00000186047.11_2</td>\n",
       "      <td>DLEU7</td>\n",
       "      <td>MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...</td>\n",
       "      <td>MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...</td>\n",
       "      <td>222</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000525194.5;ENSG00000078124.13_1</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>47</td>\n",
       "      <td>LVSTHQASFLVLGLWAGRGDIPVI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000531352.5;ENSG00000078124.13_1</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>47</td>\n",
       "      <td>LVSTHQASFLVLGLWAGRGDIPVI*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000587352.5;ENSG00000105186.16_3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>390</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000596802.5;ENSG00000127511.10_2</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...</td>\n",
       "      <td>354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      cryptic_transcript_id gene_name  \\\n",
       "0   ENST00000219689.12;ENSG00000103404.15_1     USP31   \n",
       "1   ENST00000248054.10;ENSG00000127511.10_2     SIN3B   \n",
       "2    ENST00000262127.7;ENSG00000101624.11_3     CEP76   \n",
       "3    ENST00000272907.8;ENSG00000144460.13_2     NYAP2   \n",
       "4   ENST00000286355.10;ENSG00000155897.10_1     ADCY8   \n",
       "5    ENST00000295049.9;ENSG00000162944.11_1     RFTN2   \n",
       "6    ENST00000306065.9;ENSG00000105186.16_3   ANKRD27   \n",
       "7    ENST00000309971.9;ENSG00000119392.16_1      GLE1   \n",
       "8    ENST00000370630.6;ENSG00000117151.13_1      CTBS   \n",
       "9    ENST00000372770.4;ENSG00000119392.16_1      GLE1   \n",
       "10   ENST00000372836.5;ENSG00000137161.18_1     CNPY3   \n",
       "11   ENST00000377928.7;ENSG00000155897.10_1     ADCY8   \n",
       "12   ENST00000379803.5;ENSG00000127511.10_2     SIN3B   \n",
       "13   ENST00000423709.6;ENSG00000101624.11_3     CEP76   \n",
       "14   ENST00000457408.7;ENSG00000158604.15_1     TMED4   \n",
       "15   ENST00000481238.1;ENSG00000158604.15_1     TMED4   \n",
       "16   ENST00000504404.2;ENSG00000186047.11_2     DLEU7   \n",
       "17   ENST00000525194.5;ENSG00000078124.13_1     ACER3   \n",
       "18   ENST00000531352.5;ENSG00000078124.13_1     ACER3   \n",
       "19   ENST00000587352.5;ENSG00000105186.16_3   ANKRD27   \n",
       "20   ENST00000596802.5;ENSG00000127511.10_2     SIN3B   \n",
       "\n",
       "                                  peptide_seq_cryptic  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...   \n",
       "1   MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...   \n",
       "2   MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...   \n",
       "3   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...   \n",
       "4   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...   \n",
       "5   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...   \n",
       "6   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...   \n",
       "7   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...   \n",
       "8   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...   \n",
       "9   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...   \n",
       "10  MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...   \n",
       "11  MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...   \n",
       "12  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...   \n",
       "13  MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...   \n",
       "14  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...   \n",
       "15  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...   \n",
       "16  MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...   \n",
       "17  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...   \n",
       "18  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...   \n",
       "19  MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...   \n",
       "20  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...   \n",
       "\n",
       "                                     peptide_seq_full  longest_match_slice  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...                  725   \n",
       "1   MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...                  353   \n",
       "2   MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...                  660   \n",
       "3   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...                  645   \n",
       "4   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...                  637   \n",
       "5   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...                  310   \n",
       "6   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...                  372   \n",
       "7   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...                  144   \n",
       "8   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...                  319   \n",
       "9   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...                  144   \n",
       "10  MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...                  124   \n",
       "11  MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...                  637   \n",
       "12  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...                  353   \n",
       "13  MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQH...                  585   \n",
       "14  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...                  178   \n",
       "15  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...                  187   \n",
       "16  MASPAPLVASISHQMVALQTLQLLQQEWGWGDGPVAPGNPRDPDHV...                  222   \n",
       "17  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...                   47   \n",
       "18  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...                   47   \n",
       "19  MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...                  390   \n",
       "20  MAHAGGGSGGSGAGGPAGRGLSGARWGRSGSAGHEKLPVHVEDALT...                  354   \n",
       "\n",
       "                             peptide_seq_cryptic_uniq  \n",
       "0   GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...  \n",
       "1                                                   *  \n",
       "2                                                      \n",
       "3                                           PSSLANRD*  \n",
       "4   VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...  \n",
       "5                                KFYVNDILYLRNLNLYQNQ*  \n",
       "6                                  VRSHPCPGLPLWASWFP*  \n",
       "7                                   VSEPMKEGSLDPASHI*  \n",
       "8                                       VRLFVSYEHLFY*  \n",
       "9                                   VSEPMKEGSLDPASHI*  \n",
       "10                       VGFGIVLHPLWGQACMYLSVSAGVSVI*  \n",
       "11  VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...  \n",
       "12                                                  *  \n",
       "13                                                     \n",
       "14                                          ASAYLLVI*  \n",
       "15                                                     \n",
       "16                                                     \n",
       "17                          LVSTHQASFLVLGLWAGRGDIPVI*  \n",
       "18                          LVSTHQASFLVLGLWAGRGDIPVI*  \n",
       "19                                                     \n",
       "20                                                     "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get a slice of the point where cryptic peptide and overlapping normal peptide match/align\n",
    "pep_comb = trnc_cds_peptides[[\"cryptic_transcript_id\",\"gene_name\", \"peptide_seq\"]].merge(ref_cds_peptides[[\"cryptic_transcript_id\", \"peptide_seq\"]], on=\"cryptic_transcript_id\", suffixes=[\"_cryptic\", \"_full\"])\n",
    "pep_comb[\"longest_match_slice\"] = pep_comb.apply(lambda row: longest_matching_substring(row[\"peptide_seq_cryptic\"], row[\"peptide_seq_full\"]) + 1, axis=1)\n",
    "\n",
    "\n",
    "# slice the cruptic peptide to get unique region\n",
    "pep_comb[\"peptide_seq_cryptic_uniq\"] = pep_comb.apply(lambda row: row[\"peptide_seq_cryptic\"][row[\"longest_match_slice\"]:], axis=1)\n",
    "out_order = [\"cryptic_transcript_id\", \"gene_name\", \"peptide_seq_cryptic\", \"peptide_seq_full\", \"peptide_seq_cryptic_uniq\", \"longest_match_slice\"]\n",
    "\n",
    "\n",
    "pep_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output to TSV\n",
    "pep_comb[out_order].to_csv(\"2023-11-08_i3cortical_cryptic_bleedthrough_fullpeptides.tsv\", sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cryptic_transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>peptide_seq_cryptic</th>\n",
       "      <th>peptide_seq_full</th>\n",
       "      <th>longest_match_slice</th>\n",
       "      <th>peptide_seq_cryptic_uniq</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>position</th>\n",
       "      <th>fa_name</th>\n",
       "      <th>peptide_seq_cryptic_uniq_nostop</th>\n",
       "      <th>peptide_seq_cryptic_nostop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000219689.12;ENSG00000103404.15_1</td>\n",
       "      <td>USP31</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "      <td>725</td>\n",
       "      <td>GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...</td>\n",
       "      <td>ENST00000219689.12</td>\n",
       "      <td>ENSG00000103404.15_1</td>\n",
       "      <td>chr16:23079754:23079945:-</td>\n",
       "      <td>ENST00000219689.12|USP31|ENSG00000103404.15_1|...</td>\n",
       "      <td>GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...</td>\n",
       "      <td>MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000272907.8;ENSG00000144460.13_2</td>\n",
       "      <td>NYAP2</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "      <td>645</td>\n",
       "      <td>PSSLANRD*</td>\n",
       "      <td>ENST00000272907.8</td>\n",
       "      <td>ENSG00000144460.13_2</td>\n",
       "      <td>chr2:225651535:225658565:+</td>\n",
       "      <td>ENST00000272907.8|NYAP2|ENSG00000144460.13_2|c...</td>\n",
       "      <td>PSSLANRD</td>\n",
       "      <td>MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000286355.10;ENSG00000155897.10_1</td>\n",
       "      <td>ADCY8</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "      <td>637</td>\n",
       "      <td>VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...</td>\n",
       "      <td>ENST00000286355.10</td>\n",
       "      <td>ENSG00000155897.10_1</td>\n",
       "      <td>chr8:130901663:130903771:-</td>\n",
       "      <td>ENST00000286355.10|ADCY8|ENSG00000155897.10_1|...</td>\n",
       "      <td>VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...</td>\n",
       "      <td>MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000295049.9;ENSG00000162944.11_1</td>\n",
       "      <td>RFTN2</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "      <td>310</td>\n",
       "      <td>KFYVNDILYLRNLNLYQNQ*</td>\n",
       "      <td>ENST00000295049.9</td>\n",
       "      <td>ENSG00000162944.11_1</td>\n",
       "      <td>chr2:197630635:197631010:-</td>\n",
       "      <td>ENST00000295049.9|RFTN2|ENSG00000162944.11_1|c...</td>\n",
       "      <td>KFYVNDILYLRNLNLYQNQ</td>\n",
       "      <td>MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000306065.9;ENSG00000105186.16_3</td>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "      <td>372</td>\n",
       "      <td>VRSHPCPGLPLWASWFP*</td>\n",
       "      <td>ENST00000306065.9</td>\n",
       "      <td>ENSG00000105186.16_3</td>\n",
       "      <td>chr19:32635458:32639355:-</td>\n",
       "      <td>ENST00000306065.9|ANKRD27|ENSG00000105186.16_3...</td>\n",
       "      <td>VRSHPCPGLPLWASWFP</td>\n",
       "      <td>MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000309971.9;ENSG00000119392.16_1</td>\n",
       "      <td>GLE1</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "      <td>144</td>\n",
       "      <td>VSEPMKEGSLDPASHI*</td>\n",
       "      <td>ENST00000309971.9</td>\n",
       "      <td>ENSG00000119392.16_1</td>\n",
       "      <td>chr9:128515639:128516262:+</td>\n",
       "      <td>ENST00000309971.9|GLE1|ENSG00000119392.16_1|ch...</td>\n",
       "      <td>VSEPMKEGSLDPASHI</td>\n",
       "      <td>MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000370630.6;ENSG00000117151.13_1</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "      <td>319</td>\n",
       "      <td>VRLFVSYEHLFY*</td>\n",
       "      <td>ENST00000370630.6</td>\n",
       "      <td>ENSG00000117151.13_1</td>\n",
       "      <td>chr1:84561927:84563256:-</td>\n",
       "      <td>ENST00000370630.6|CTBS|ENSG00000117151.13_1|ch...</td>\n",
       "      <td>VRLFVSYEHLFY</td>\n",
       "      <td>MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000372836.5;ENSG00000137161.18_1</td>\n",
       "      <td>CNPY3</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "      <td>124</td>\n",
       "      <td>VGFGIVLHPLWGQACMYLSVSAGVSVI*</td>\n",
       "      <td>ENST00000372836.5</td>\n",
       "      <td>ENSG00000137161.18_1</td>\n",
       "      <td>chr6:42935670:42936378:+</td>\n",
       "      <td>ENST00000372836.5|CNPY3|ENSG00000137161.18_1|c...</td>\n",
       "      <td>VGFGIVLHPLWGQACMYLSVSAGVSVI</td>\n",
       "      <td>MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000457408.7;ENSG00000158604.15_1</td>\n",
       "      <td>TMED4</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "      <td>178</td>\n",
       "      <td>ASAYLLVI*</td>\n",
       "      <td>ENST00000457408.7</td>\n",
       "      <td>ENSG00000158604.15_1</td>\n",
       "      <td>chr7:44579896:44581092:-</td>\n",
       "      <td>ENST00000457408.7|TMED4|ENSG00000158604.15_1|c...</td>\n",
       "      <td>ASAYLLVI</td>\n",
       "      <td>MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000525194.5;ENSG00000078124.13_1</td>\n",
       "      <td>ACER3</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "      <td>47</td>\n",
       "      <td>LVSTHQASFLVLGLWAGRGDIPVI*</td>\n",
       "      <td>ENST00000525194.5</td>\n",
       "      <td>ENSG00000078124.13_1</td>\n",
       "      <td>chr11:76868156:76869066:+</td>\n",
       "      <td>ENST00000525194.5|ACER3|ENSG00000078124.13_1|c...</td>\n",
       "      <td>LVSTHQASFLVLGLWAGRGDIPVI</td>\n",
       "      <td>MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      cryptic_transcript_id gene_name  \\\n",
       "0   ENST00000219689.12;ENSG00000103404.15_1     USP31   \n",
       "1    ENST00000272907.8;ENSG00000144460.13_2     NYAP2   \n",
       "2   ENST00000286355.10;ENSG00000155897.10_1     ADCY8   \n",
       "3    ENST00000295049.9;ENSG00000162944.11_1     RFTN2   \n",
       "4    ENST00000306065.9;ENSG00000105186.16_3   ANKRD27   \n",
       "5    ENST00000309971.9;ENSG00000119392.16_1      GLE1   \n",
       "6    ENST00000370630.6;ENSG00000117151.13_1      CTBS   \n",
       "8    ENST00000372836.5;ENSG00000137161.18_1     CNPY3   \n",
       "10   ENST00000457408.7;ENSG00000158604.15_1     TMED4   \n",
       "11   ENST00000525194.5;ENSG00000078124.13_1     ACER3   \n",
       "\n",
       "                                  peptide_seq_cryptic  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...   \n",
       "1   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...   \n",
       "2   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...   \n",
       "3   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...   \n",
       "4   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...   \n",
       "5   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...   \n",
       "6   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...   \n",
       "8   MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...   \n",
       "10  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...   \n",
       "11  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...   \n",
       "\n",
       "                                     peptide_seq_full  longest_match_slice  \\\n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...                  725   \n",
       "1   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...                  645   \n",
       "2   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...                  637   \n",
       "3   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...                  310   \n",
       "4   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...                  372   \n",
       "5   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...                  144   \n",
       "6   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...                  319   \n",
       "8   MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...                  124   \n",
       "10  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...                  178   \n",
       "11  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...                   47   \n",
       "\n",
       "                             peptide_seq_cryptic_uniq   ref_transcript_id  \\\n",
       "0   GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...  ENST00000219689.12   \n",
       "1                                           PSSLANRD*   ENST00000272907.8   \n",
       "2   VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...  ENST00000286355.10   \n",
       "3                                KFYVNDILYLRNLNLYQNQ*   ENST00000295049.9   \n",
       "4                                  VRSHPCPGLPLWASWFP*   ENST00000306065.9   \n",
       "5                                   VSEPMKEGSLDPASHI*   ENST00000309971.9   \n",
       "6                                       VRLFVSYEHLFY*   ENST00000370630.6   \n",
       "8                        VGFGIVLHPLWGQACMYLSVSAGVSVI*   ENST00000372836.5   \n",
       "10                                          ASAYLLVI*   ENST00000457408.7   \n",
       "11                          LVSTHQASFLVLGLWAGRGDIPVI*   ENST00000525194.5   \n",
       "\n",
       "                   le_id                    position  \\\n",
       "0   ENSG00000103404.15_1   chr16:23079754:23079945:-   \n",
       "1   ENSG00000144460.13_2  chr2:225651535:225658565:+   \n",
       "2   ENSG00000155897.10_1  chr8:130901663:130903771:-   \n",
       "3   ENSG00000162944.11_1  chr2:197630635:197631010:-   \n",
       "4   ENSG00000105186.16_3   chr19:32635458:32639355:-   \n",
       "5   ENSG00000119392.16_1  chr9:128515639:128516262:+   \n",
       "6   ENSG00000117151.13_1    chr1:84561927:84563256:-   \n",
       "8   ENSG00000137161.18_1    chr6:42935670:42936378:+   \n",
       "10  ENSG00000158604.15_1    chr7:44579896:44581092:-   \n",
       "11  ENSG00000078124.13_1   chr11:76868156:76869066:+   \n",
       "\n",
       "                                              fa_name  \\\n",
       "0   ENST00000219689.12|USP31|ENSG00000103404.15_1|...   \n",
       "1   ENST00000272907.8|NYAP2|ENSG00000144460.13_2|c...   \n",
       "2   ENST00000286355.10|ADCY8|ENSG00000155897.10_1|...   \n",
       "3   ENST00000295049.9|RFTN2|ENSG00000162944.11_1|c...   \n",
       "4   ENST00000306065.9|ANKRD27|ENSG00000105186.16_3...   \n",
       "5   ENST00000309971.9|GLE1|ENSG00000119392.16_1|ch...   \n",
       "6   ENST00000370630.6|CTBS|ENSG00000117151.13_1|ch...   \n",
       "8   ENST00000372836.5|CNPY3|ENSG00000137161.18_1|c...   \n",
       "10  ENST00000457408.7|TMED4|ENSG00000158604.15_1|c...   \n",
       "11  ENST00000525194.5|ACER3|ENSG00000078124.13_1|c...   \n",
       "\n",
       "                      peptide_seq_cryptic_uniq_nostop  \\\n",
       "0   GLQCERLRVCAGESFRRVCLWQGDLTLEPHFPSGRQCANCEGSFHP...   \n",
       "1                                            PSSLANRD   \n",
       "2   VSPLFLLLLGHECMLVCISNPDTSEKTDSSLTFSPMNTIIANDQTI...   \n",
       "3                                 KFYVNDILYLRNLNLYQNQ   \n",
       "4                                   VRSHPCPGLPLWASWFP   \n",
       "5                                    VSEPMKEGSLDPASHI   \n",
       "6                                        VRLFVSYEHLFY   \n",
       "8                         VGFGIVLHPLWGQACMYLSVSAGVSVI   \n",
       "10                                           ASAYLLVI   \n",
       "11                           LVSTHQASFLVLGLWAGRGDIPVI   \n",
       "\n",
       "                           peptide_seq_cryptic_nostop  \n",
       "0   MSKVTAPGSGPPAAASGKEKRSFSKRLFRSGRAGGGGAGGPGASGP...  \n",
       "1   MISSKMMSSNPEEDPLDTFLQYIEDMGMKAYDGLVIQNASDIAREN...  \n",
       "2   MELSDVRCLTGSEELYTIHPTPPAGDGRSASRPQRLLWQTAVRHIT...  \n",
       "3   MGCGLRKLEDPDDSSPGKIFSTLKRPQVETKTEFAYEYVLLDFTLQ...  \n",
       "4   MALYDEDLLKNPFYLALQKCRPDLCSKVAQIHGIVLVPCKGSLSSS...  \n",
       "5   MPSEGRCWETLKALRSSDKGRLCYYRDWLLRREDVLEECMSLPKLS...  \n",
       "6   MSRPQLRRWRLVSSPPSGVPGLALLALLALLALRLAAGTDCPCPEP...  \n",
       "8   MDSMPEPASRCLLLLPLLLLLLLLLPAPELGPSQAGAEENDWVRLP...  \n",
       "10  MAGVGAGPLRAMGRQALLLLALCATGAQGLYFHIGETEKRCFIEEI...  \n",
       "11  MAPAADREGYWGPTTSTLDWCEENYSVTWYIAEFCSFLPSSLKSNL...  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# outptu as FASTA\n",
    "# Note that need a minimum sequence for this to be viable\n",
    "pep_comb_fa = pep_comb.loc[pep_comb.peptide_seq_cryptic_uniq.str.len() > 1,:]\n",
    "\n",
    "# construct header - transcript_id|gene_name|le_id|coords\n",
    "# first add le_id\n",
    "pep_comb_fa = pd.concat([pep_comb_fa,\n",
    "                         pep_comb_fa.cryptic_transcript_id.str.split(\";\", expand=True).rename(columns={0: \"ref_transcript_id\", 1: \"le_id\"})],\n",
    "                         axis=\"columns\")\n",
    "\n",
    "# create the representative bleedthrough coordinate string\n",
    "bleeds_le_cryp_df = bleeds_le_cryp.as_df()\n",
    "bleeds_le_cryp_df.loc[:, \"position\"] = bleeds_le_cryp_df[\"Chromosome\"].str.cat(bleeds_le_cryp_df[[\"Start\", \"End\", \"Strand\"]].astype(str), sep=\":\")\n",
    "# bleeds_le_cryp_df\n",
    "\n",
    "# now merge with rep bleedthrough coordinates\n",
    "pep_comb_fa = pep_comb_fa.merge(bleeds_le_cryp_df[[\"le_id\", \"position\"]], on=\"le_id\", how=\"left\")\n",
    "pep_comb_fa.drop_duplicates(subset=[\"le_id\", \"peptide_seq_cryptic_uniq\"], inplace=True)\n",
    "\n",
    "# now to make final header - transcript_id|gene_name|le_id|coords\n",
    "pep_comb_fa.loc[:, \"fa_name\"] = pep_comb_fa[\"ref_transcript_id\"].str.cat(pep_comb_fa[[\"gene_name\", \"le_id\", \"position\"]], sep=\"|\")\n",
    "\n",
    "# remove stop codon from seq string\n",
    "pep_comb_fa.loc[:, \"peptide_seq_cryptic_uniq_nostop\"] = pep_comb_fa[\"peptide_seq_cryptic_uniq\"].str.rstrip(\"*\")\n",
    "pep_comb_fa.loc[:, \"peptide_seq_cryptic_nostop\"] = pep_comb_fa[\"peptide_seq_cryptic\"].str.rstrip(\"*\")\n",
    "\n",
    "pep_comb_fa\n",
    "\n",
    "\n",
    "# pep_comb_fa.cryptic_transcript_id.str.split(\";\", expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"2023-11-10_i3cortical_cryptic_bleedthrough.cryptic_uniq_peptide.fa\", \"w\") as out:\n",
    "    for _, df in pep_comb_fa.iterrows():\n",
    "        out.write(f\">{df['fa_name']}\\n{df['peptide_seq_cryptic_uniq_nostop']}\\n\")\n",
    "\n",
    "\n",
    "with open(\"2023-11-10_i3cortical_cryptic_bleedthrough.cryptic_full_peptide.fa\", \"w\") as out:\n",
    "    for _, df in pep_comb_fa.iterrows():\n",
    "        out.write(f\">{df['fa_name']}\\n{df['peptide_seq_cryptic_nostop']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FBXO38-DT', 'FIRRE', 'LIPA', 'NBPF3', 'PCDH11X'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which bleedthroughs are missing?\n",
    "bleeds_missing_gn = set(bleeds_gn).difference(set(pep_comb.gene_name))\n",
    "bleeds_missing_gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>FBXO38-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>LIPA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>LIPA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name       gene_type          transcript_type\n",
       "0        NBPF3  protein_coding           protein_coding\n",
       "68       NBPF3  protein_coding  nonsense_mediated_decay\n",
       "109      NBPF3  protein_coding     processed_transcript\n",
       "202  FBXO38-DT          lncRNA                   lncRNA\n",
       "208       LIPA  protein_coding           protein_coding\n",
       "277       LIPA  protein_coding     processed_transcript\n",
       "290    PCDH11X  protein_coding     processed_transcript\n",
       "297    PCDH11X  protein_coding          retained_intron\n",
       "300    PCDH11X  protein_coding           protein_coding\n",
       "398      FIRRE          lncRNA                   lncRNA"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# are they annotated as protein coding at gene/transcript level?\n",
    "ref_gtf.subset(lambda df: df.gene_name.isin(bleeds_missing_gn)).as_df()[[\"gene_name\", \"gene_type\", \"transcript_type\"]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Feature</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>Start_bl</th>\n",
       "      <th>End_bl</th>\n",
       "      <th>Strand_bl</th>\n",
       "      <th>le_id</th>\n",
       "      <th>transcript_id_bl</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>ref_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21454306</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000478653.6</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>PAPA.TDP-1.345.1</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.ctrl_ctrl_4.14569.8</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.14306.12</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89330996</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP-2.13552.8</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89331008</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43_19065403_S23.13899.16</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89332056</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP-6.13417.11</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>89335335</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENST00000489359.1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>89334602</td>\n",
       "      <td>89335534</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107798.18_2</td>\n",
       "      <td>PAPA.TDP43-G_S7.10969.1</td>\n",
       "      <td>LIPA</td>\n",
       "      <td>ENSG00000107798.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91884007</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENST00000298274.12</td>\n",
       "      <td>6</td>\n",
       "      <td>exon</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>91882906</td>\n",
       "      <td>91891321</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102290.23_3</td>\n",
       "      <td>PAPA.TDP43_ctrl_4.26532.4</td>\n",
       "      <td>PCDH11X</td>\n",
       "      <td>ENSG00000102290.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Frame      |     Start |       End | Strand       | +13   |\n",
       "| (category)   | (object)   |   (int64) |   (int64) | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | .          |  21453372 |  21454306 | +            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chr10        | .          |  89335335 |  89335534 | -            | ...   |\n",
       "| chrX         | .          |  91882906 |  91884007 | +            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 18 columns from 3 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: gene_id, gene_name, transcript_id, exon_number, Feature, transcript_type, ... (+ 7 more.)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# can exclude FIRRE & FBCO38-DT as they are lncRNAs. But NBPF3, LIPA & PCDH11X could all potentially be coding.\n",
    "# for NBPF3, LIPA and PCDH11X, check if the exon that they extend is protein-coding/has an annotated CDS\n",
    "bleeds_missing_gn_pc = [\"NBPF3\", \"LIPA\", \"PCDH11X\"]\n",
    "ref_gtf_bleeds_missing_pc_ex = ref_gtf.subset(lambda df: (df.gene_name.isin(bleeds_missing_gn_pc)) & (df.Feature.isin([\"CDS\", \"exon\"])))[[\"Frame\",\"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"Feature\", \"transcript_type\"]]\n",
    "\n",
    "\n",
    "(ref_gtf_bleeds_missing_pc_ex.join(full_bleeds_le_mv, strandedness=\"same\",suffix=\"_bl\")\n",
    " .subset(lambda df: ((df.Strand == \"+\") & (df.Start == df[\"Start\" + \"_bl\"])) | \n",
    " ((df.Strand == \"-\") & (df.End == df[\"End\" + \"_bl\"])))\n",
    " )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
