{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1 µs, sys: 1e+03 ns, total: 2 µs\n",
      "Wall time: 5.01 µs\n"
     ]
    }
   ],
   "source": [
    "%time\n",
    "gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "le_gtf = pr.read_gtf(\"../processed/2023-06-20_cryptic_last_exons.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.224.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40421975</td>\n",
       "      <td>40423342</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000084070.12</td>\n",
       "      <td>SMAP2</td>\n",
       "      <td>PAPA.TDP-6.631.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147623356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.1123.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147622693</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.1186.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147623701</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.1139.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>RAI2,RAI2</td>\n",
       "      <td>PAPA.TDP43-G_S7.20262.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40653676</td>\n",
       "      <td>40654556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.26410.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>870 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+--------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Feature      | Source     | Start     | End       | +5    |\n",
       "| (category)   | (category)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+--------------+------------+-----------+-----------+-------|\n",
       "| chr1         | exon         | .          | 15939295  | 15940477  | ...   |\n",
       "| chr1         | exon         | .          | 40421975  | 40423342  | ...   |\n",
       "| chr1         | exon         | .          | 147622270 | 147623356 | ...   |\n",
       "| chr1         | exon         | .          | 147622270 | 147622693 | ...   |\n",
       "| ...          | ...          | ...        | ...       | ...       | ...   |\n",
       "| chrX         | exon         | .          | 40653641  | 40654556  | ...   |\n",
       "| chrX         | exon         | .          | 40653676  | 40654556  | ...   |\n",
       "| chrX         | exon         | .          | 131823775 | 131825365 | ...   |\n",
       "| chrX         | exon         | .          | 131823777 | 131825365 | ...   |\n",
       "+--------------+--------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 870 rows and 10 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Score, Strand, gene_id, gene_name, transcript_id"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# minimal info required from GTF - gene_id (representing REF gene_id) & transcript id\n",
    "# only need to add novel last exons\n",
    "le_gtf_min = (le_gtf.subset(lambda df: df.annot_status == \"novel\")\n",
    " [[\"Feature\", \"Source\", \"Score\", \"ref_gene_id\", \"ref_gene_name\", \"transcript_id\", \"annot_status\"]]\n",
    " .apply(lambda df: df.rename(columns={\"ref_gene_id\": \"gene_id\", \"ref_gene_name\": \"gene_name\"}))\n",
    " .drop(\"annot_status\"))\n",
    "\n",
    "le_gtf_min\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.224.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40421975</td>\n",
       "      <td>40423342</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000084070.12</td>\n",
       "      <td>SMAP2</td>\n",
       "      <td>PAPA.TDP-6.631.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147623356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.1123.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147622693</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.1186.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>147622270</td>\n",
       "      <td>147623701</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.1139.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>PAPA.TDP43-G_S7.20262.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>PAPA.TDP43-G_S7.20350.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>40653676</td>\n",
       "      <td>40654556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.26410.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>870 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+--------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Feature      | Source     | Start     | End       | +5    |\n",
       "| (category)   | (category)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+--------------+------------+-----------+-----------+-------|\n",
       "| chr1         | exon         | .          | 15939295  | 15940477  | ...   |\n",
       "| chr1         | exon         | .          | 40421975  | 40423342  | ...   |\n",
       "| chr1         | exon         | .          | 147622270 | 147623356 | ...   |\n",
       "| chr1         | exon         | .          | 147622270 | 147622693 | ...   |\n",
       "| ...          | ...          | ...        | ...       | ...       | ...   |\n",
       "| chrX         | exon         | .          | 40653641  | 40654556  | ...   |\n",
       "| chrX         | exon         | .          | 40653676  | 40654556  | ...   |\n",
       "| chrX         | exon         | .          | 131823775 | 131825365 | ...   |\n",
       "| chrX         | exon         | .          | 131823777 | 131825365 | ...   |\n",
       "+--------------+--------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 870 rows and 10 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Score, Strand, gene_id, gene_name, transcript_id"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sometimes I have gene_names duplicated - remove duplicates, returning to comma-separatd string in same insertion order as before in case still unqiue values\n",
    "le_gtf_min.gene_name = le_gtf_min.gene_name.apply(lambda x:  \",\".join(list(dict.fromkeys(x.split(\",\")))))\n",
    "le_gtf_min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>358871</td>\n",
       "      <td>365510</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000412666</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence</td>\n",
       "      <td>OTTHUMT00000007987.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>358871</td>\n",
       "      <td>358957</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000412666</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence</td>\n",
       "      <td>OTTHUMT00000007987.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001700697.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>365170</td>\n",
       "      <td>365510</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000412666</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence</td>\n",
       "      <td>OTTHUMT00000007987.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001745103.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>360056</td>\n",
       "      <td>366052</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000635159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>non_submitted_evidence,basic</td>\n",
       "      <td>OTTHUMT00000488661.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>360056</td>\n",
       "      <td>360168</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000635159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>non_submitted_evidence,basic</td>\n",
       "      <td>OTTHUMT00000488661.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003784495.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556647</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>25386916</td>\n",
       "      <td>25387085</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000228786.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LINC00266-4P-201</td>\n",
       "      <td>3.0</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000104946.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001653980.1</td>\n",
       "      <td>HGNC:38800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556648</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>25378299</td>\n",
       "      <td>25378878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000228786.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LINC00266-4P-201</td>\n",
       "      <td>3.0</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000104946.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001596444.1</td>\n",
       "      <td>HGNC:38800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556649</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence,basic,Ensembl_cano...</td>\n",
       "      <td>OTTHUMT00000058824.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:38513</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556650</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203181</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence,basic,Ensembl_cano...</td>\n",
       "      <td>OTTHUMT00000058824.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>HGNC:38513</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556651</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>not_best_in_genome_evidence,basic,Ensembl_cano...</td>\n",
       "      <td>OTTHUMT00000058824.3</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002145073.1</td>\n",
       "      <td>HGNC:38513</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1556652 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +19   |\n",
       "| (object)     | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | transcript | 358871    | 365510    | ...   |\n",
       "| chr1         | HAVANA     | exon       | 358871    | 358957    | ...   |\n",
       "| chr1         | HAVANA     | exon       | 365170    | 365510    | ...   |\n",
       "| chr1         | HAVANA     | transcript | 360056    | 366052    | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrY         | HAVANA     | exon       | 25378299  | 25378878  | ...   |\n",
       "| chrY         | HAVANA     | transcript | 57201142  | 57203357  | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57203181  | 57203357  | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57201142  | 57202020  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,556,652 rows and 24 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine the two gtfs\n",
    "comb_gtf = pr.concat([gtf, le_gtf_min])\n",
    "comb_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000412666.1</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000635159.1</td>\n",
       "      <td>ENSG00000236601.2</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000450696.1</td>\n",
       "      <td>ENSG00000235146.2</td>\n",
       "      <td>ENSG00000235146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000412115.2</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>LINC01409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000429505.6</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>LINC01409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556622</th>\n",
       "      <td>ENST00000618574.1</td>\n",
       "      <td>ENSG00000185894.8</td>\n",
       "      <td>BPY2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556634</th>\n",
       "      <td>ENST00000456123.1</td>\n",
       "      <td>ENSG00000228296.1</td>\n",
       "      <td>TTTY4C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556639</th>\n",
       "      <td>ENST00000421387.2</td>\n",
       "      <td>ENSG00000223641.2</td>\n",
       "      <td>TTTY17C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556645</th>\n",
       "      <td>ENST00000427373.5</td>\n",
       "      <td>ENSG00000228786.5</td>\n",
       "      <td>LINC00266-4P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556649</th>\n",
       "      <td>ENST00000399966.9_PAR_Y</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103601 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   transcript_id                   gene_id        gene_name\n",
       "0              ENST00000412666.1         ENSG00000236601.2  ENSG00000236601\n",
       "3              ENST00000635159.1         ENSG00000236601.2  ENSG00000236601\n",
       "6              ENST00000450696.1         ENSG00000235146.2  ENSG00000235146\n",
       "9              ENST00000412115.2        ENSG00000237491.10        LINC01409\n",
       "15             ENST00000429505.6        ENSG00000237491.10        LINC01409\n",
       "...                          ...                       ...              ...\n",
       "1556622        ENST00000618574.1         ENSG00000185894.8            BPY2C\n",
       "1556634        ENST00000456123.1         ENSG00000228296.1           TTTY4C\n",
       "1556639        ENST00000421387.2         ENSG00000223641.2          TTTY17C\n",
       "1556645        ENST00000427373.5         ENSG00000228786.5     LINC00266-4P\n",
       "1556649  ENST00000399966.9_PAR_Y  ENSG00000185203.12_PAR_Y           WASIR1\n",
       "\n",
       "[103601 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate tx2gene for subsequent merging\n",
    "tx2gene2name = comb_gtf.as_df()[[\"transcript_id\", \"gene_id\", \"gene_name\"]].drop_duplicates()\n",
    "tx2gene2name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_gtf.to_gtf(\"../processed/2023-08-23_reference_filtered_cryptic_le_combined.gtf\")\n",
    "tx2gene2name.to_csv(\"../processed/2023-08-23_reference_filtered_cryptic_le_combined.tx2gene2name.tsv\", sep=\"\\t\", header=True, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
