{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Extract splice junctions for AS-ALE (/ last exon spliced) events.\n","\n","Due to a pretty silly move on my part, my pipeline does not output splice junctions for novel spliced last exons (nor last exons)\n","Given want to use SJs to quantify these events in NYGC data, need to extract using only last exon coordinates.\n","\n","In pipeline, approach to define last exons:\n","- 3'end doesn't overlap any annotated exon\n","- terminal splice junction must match an annotated splice junction at the 5'end. \n","\n","Can exploit these criteria to grab SJs 'in reverse'\n","\n","Strategy for novel last exons:\n","- Extract introns (SJs) from annotated transcripts\n","- find introns in which last exons are completely contained\n","- (w/ some form of pr.join), construct a SJ with start coordinate of annotated intron and end coordinate = start of exon coordinate\n","- As a sanity check, compare with AL's provided list of cryptic SJs inferred with MAJIQ - for common genes, do we get the same SJ coordinates\n","\n","Strategy for annotated last exons:\n","- Any spliced last exon that is not captured with novel LE strategy\n","- pr.join with slack=1 (allow bookended intervals to overlap) annotated SJs with last exon coordinates\n","- Subset for exact matches at SJ 3'end/last exon 5'end\n","- Retain the SJs for downstream analysis"]},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":["import pyranges as pr\n","import pandas as pd\n","import numpy as np"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["CPU times: user 1e+03 ns, sys: 1 µs, total: 2 µs\n","Wall time: 3.58 µs\n"]}],"source":["majiq_sj = pr.read_bed(\"../data/Everything_minus_ferguson_liu_recent_new_cryptics.junctions.bed\")\n","ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n","le = pr.read_bed(\"../data/2023-07-04_papa_cryptic_spliced.last_exons.bed\") # generated in motifs/ subdir, all cryptic last exon coords used for iCLIP analysis"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":["# ref gtf has lots of attributes don't need, subset to minimal cols for analysis\n","ref_gtf = ref_gtf[[\"Feature\", \"gene_id\", \"transcript_id\", \"gene_name\"]]"]},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>76871267</td>\n","      <td>76871821</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>54634687</td>\n","      <td>54639192</td>\n","      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>245464258</td>\n","      <td>245471621</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>102</th>\n","      <td>chr22</td>\n","      <td>37464524</td>\n","      <td>37465718</td>\n","      <td>ENSG00000100060.18_2|MFNG|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>103</th>\n","      <td>chrX</td>\n","      <td>102721363</td>\n","      <td>102724864</td>\n","      <td>ENSG00000198908.12_1|BHLHB9|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>104</th>\n","      <td>chrX</td>\n","      <td>98679426</td>\n","      <td>98679978</td>\n","      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>105</th>\n","      <td>chrX</td>\n","      <td>17835910</td>\n","      <td>17837395</td>\n","      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>106</th>\n","      <td>chrX</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>107 rows × 6 columns</p>\n","</div>"],"text/plain":["+--------------+-----------+-----------+-------+\n","| Chromosome   | Start     | End       | +3    |\n","| (category)   | (int64)   | (int64)   | ...   |\n","|--------------+-----------+-----------+-------|\n","| chr1         | 76871267  | 76871821  | ...   |\n","| chr1         | 61824444  | 61825501  | ...   |\n","| chr1         | 54634687  | 54639192  | ...   |\n","| chr1         | 245464258 | 245471621 | ...   |\n","| ...          | ...       | ...       | ...   |\n","| chrX         | 102721363 | 102724864 | ...   |\n","| chrX         | 98679426  | 98679978  | ...   |\n","| chrX         | 17835910  | 17837395  | ...   |\n","| chrX         | 108267641 | 108269152 | ...   |\n","+--------------+-----------+-----------+-------+\n","Stranded PyRanges object has 107 rows and 6 columns from 23 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","3 hidden columns: Name, Score, Strand"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["# subset le for cryptic LEs only\n","le_cryp = le.subset(lambda df: df.Name.str.contains(\"cryptic\", regex=False))\n","ids_le_cryp = set(le_cryp.Name)\n","le_cryp"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":["# extract introns (SJs) for each transcript\n","ref_introns = ref_gtf.features.introns(by=\"transcript\")\n","# store a df noting which transcript IDs share the same intron coordinates (useful if collapse later and want to track/retain)\n","intron2tx = ref_introns.as_df()[[\"Chromosome\", \"Start\", \"End\", \"Strand\", \"transcript_id\"]]\n","\n"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1492</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1493</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1494</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1495</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1496</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1497 rows × 13 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +8    |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,497 rows and 13 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","8 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 1 more.)"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["# first extract LEs completely contained within annotated introns\n","le_cryp_cont = le_cryp.overlap(ref_introns, strandedness=\"same\", how=\"containment\")\n","\n","# now overlap-join introns with LEs\n","ref_introns_le = ref_introns.join(le_cryp_cont, strandedness=\"same\")\n","ref_introns_le"]},{"cell_type":"code","execution_count":23,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1492</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1493</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1494</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1495</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1496</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1497 rows × 15 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +10   |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,497 rows and 15 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":23,"metadata":{},"output_type":"execute_result"}],"source":["# If + strand\n","# Start coord = intron start (i.e. 'Start' column), End coord = exon start (i.e. 'Start_b' column)\n","# If - strand\n","# Start coord = Exon start (i.e. 'End_b' column), End coord = intron start (i.e. 'End' column)\n","\n","def _df_intron_exon_to_sj(df: pd.DataFrame) -> pd.DataFrame:\n","    '''Convert joined intron + overlapping exon df of PyRanges object to splice junction coordinates (the first bin between the two intervals)\n","\n","    If + strand\n","    Start coord = intron start (i.e. 'Start' column), End coord = exon start (i.e. 'Start_b' column)\n","    If - strand\n","    Start coord = Exon start (i.e. 'End_b' column), End coord = intron start (i.e. 'End' column)\n","    \n","    Parameters\n","    ----------\n","    df : pd.DataFrame\n","        _description_\n","\n","    Returns\n","    -------\n","    pd.DataFrame\n","        _description_\n","\n","    Raises\n","    ------\n","    Exception\n","        Strand column of df doesn't contain all '+' or '-'\n","    '''\n","\n","    df[[\"Start_intron\", \"End_intron\"]] = df[[\"Start\", \"End\"]]\n","\n","    if (df.Strand == \"+\").all():\n","        df[\"End\"] = df[\"Start_b\"]\n","\n","    elif (df.Strand == \"-\").all():\n","        df[\"Start\"] = df[\"End_b\"]\n","\n","    else:\n","        raise Exception(\"Strand column must only contain all '+' or '-'\")\n","    \n","    return df\n","\n","\n","le_sj = ref_introns_le.apply(_df_intron_exon_to_sj)\n","le_sj"]},{"cell_type":"code","execution_count":25,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>20776596</td>\n","      <td>20776596</td>\n","      <td>-</td>\n","      <td>ENSG00000127483.19</td>\n","      <td>ENST00000312239.10</td>\n","      <td>HP1BP3</td>\n","      <td>20775449</td>\n","      <td>20776596</td>\n","      <td>ENSG00000127483.19_1|HP1BP3|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>20776596</td>\n","      <td>20776596</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    |     Start |       End | Strand       | +10   |\n","| (object)     | (object)   |   (int64) |   (int64) | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     |   1616614 |   1616614 | +            | ...   |\n","| chr1         | intron     |  20776596 |  20776596 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 2 rows and 15 columns from 1 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":25,"metadata":{},"output_type":"execute_result"}],"source":["le_sj.subset(lambda df: df.End - df.Start == 0).drop_duplicate_positions() # perhaps these two are annotated in some way..."]},{"cell_type":"code","execution_count":26,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","      <td>+</td>\n","      <td>ENSG00000162849.16</td>\n","      <td>ENST00000407071.7</td>\n","      <td>KIF26B</td>\n","      <td>245464258</td>\n","      <td>245471621</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","      <td>+</td>\n","      <td>ENSG00000117069.15</td>\n","      <td>ENST00000477717.6</td>\n","      <td>ST6GALNAC5</td>\n","      <td>76871267</td>\n","      <td>76871821</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1358</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1359</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1360</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1361</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1362</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1363 rows × 15 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +10   |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 245419745 | 245464258 | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,363 rows and 15 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":26,"metadata":{},"output_type":"execute_result"}],"source":["# remove 0 length SJs for now as likely some annotation issue - check back at end after mined annotated SJs\n","le_sj = le_sj.subset(lambda df: df.Start != df.End)\n","le_sj"]},{"cell_type":"code","execution_count":36,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","      <td>+</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","      <td>+</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1358</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1359</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1360</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1361</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1362</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1363 rows × 6 columns</p>\n","</div>"],"text/plain":["+--------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Start     | End       | Strand       | +2    |\n","| (object)     | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+-----------+-----------+--------------+-------|\n","| chr1         | 245419745 | 245464258 | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| ...          | ...       | ...       | ...          | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","+--------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,363 rows and 6 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","2 hidden columns: Name, Score"]},"execution_count":36,"metadata":{},"output_type":"execute_result"}],"source":["# finally generate a cleaned BED-ready file of SJs\n","le_sj_clean_nov = le_sj[[\"Name\"]]\n","le_sj_clean_nov.Score = \".\"\n","le_sj_clean_nov"]},{"cell_type":"code","execution_count":33,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>54630886</td>\n","      <td>54634687</td>\n","      <td>+</td>\n","      <td>ENSG00000162390.18</td>\n","      <td>ENST00000371316.3</td>\n","      <td>ACOT11</td>\n","      <td>54634687</td>\n","      <td>54639192</td>\n","      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>112542090</td>\n","      <td>112550600</td>\n","      <td>-</td>\n","      <td>ENSG00000007341.19</td>\n","      <td>ENST00000343210.11</td>\n","      <td>ST7L</td>\n","      <td>112540178</td>\n","      <td>112542090</td>\n","      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>112542090</td>\n","      <td>112555867</td>\n","      <td>-</td>\n","      <td>ENSG00000007341.19</td>\n","      <td>ENST00000360743.8</td>\n","      <td>ST7L</td>\n","      <td>112540178</td>\n","      <td>112542090</td>\n","      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>33592489</td>\n","      <td>33600864</td>\n","      <td>-</td>\n","      <td>ENSG00000121904.19</td>\n","      <td>ENST00000373380.5</td>\n","      <td>CSMD2</td>\n","      <td>33591806</td>\n","      <td>33592489</td>\n","      <td>ENSG00000121904.19_3|CSMD2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>163305313</td>\n","      <td>163306232</td>\n","      <td>-</td>\n","      <td>ENSG00000232995.7</td>\n","      <td>ENST00000439699.1</td>\n","      <td>ENSG00000232995</td>\n","      <td>163304210</td>\n","      <td>163305313</td>\n","      <td>ENSG00000232995.7_2|ENSG00000232995|spliced|cr...</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>chr2</td>\n","      <td>intron</td>\n","      <td>205766818</td>\n","      <td>205776245</td>\n","      <td>+</td>\n","      <td>ENSG00000118257.17</td>\n","      <td>ENST00000272849.7</td>\n","      <td>NRP2</td>\n","      <td>205776245</td>\n","      <td>205779714</td>\n","      <td>ENSG00000118257.17_5|NRP2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>chr2</td>\n","      <td>intron</td>\n","      <td>205766803</td>\n","      <td>205776245</td>\n","      <td>+</td>\n","      <td>ENSG00000118257.17</td>\n","      <td>ENST00000357118.8</td>\n","      <td>NRP2</td>\n","      <td>205776245</td>\n","      <td>205779714</td>\n","      <td>ENSG00000118257.17_5|NRP2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>chr2</td>\n","      <td>intron</td>\n","      <td>3457693</td>\n","      <td>3457794</td>\n","      <td>+</td>\n","      <td>ENSG00000171853.16</td>\n","      <td>ENST00000441099.5</td>\n","      <td>TRAPPC12</td>\n","      <td>3457794</td>\n","      <td>3458181</td>\n","      <td>ENSG00000171853.16_4|TRAPPC12|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>chr3</td>\n","      <td>intron</td>\n","      <td>3129188</td>\n","      <td>3129918</td>\n","      <td>+</td>\n","      <td>ENSG00000072756.17</td>\n","      <td>ENST00000339437.11</td>\n","      <td>TRNT1</td>\n","      <td>3129918</td>\n","      <td>3130679</td>\n","      <td>ENSG00000072756.17_1|TRNT1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>chr3</td>\n","      <td>intron</td>\n","      <td>114339426</td>\n","      <td>114350273</td>\n","      <td>-</td>\n","      <td>ENSG00000181722.17</td>\n","      <td>ENST00000357258.8</td>\n","      <td>ZBTB20</td>\n","      <td>114314500</td>\n","      <td>114339426</td>\n","      <td>ENSG00000181722.17_5|ZBTB20|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>chr3</td>\n","      <td>intron</td>\n","      <td>8195073</td>\n","      <td>8494680</td>\n","      <td>-</td>\n","      <td>ENSG00000227110.7</td>\n","      <td>ENST00000452802.6</td>\n","      <td>LMCD1-AS1</td>\n","      <td>8193136</td>\n","      <td>8195073</td>\n","      <td>ENSG00000227110.7_3|LMCD1-AS1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>chr4</td>\n","      <td>intron</td>\n","      <td>4544907</td>\n","      <td>4560393</td>\n","      <td>+</td>\n","      <td>ENSG00000247708.8</td>\n","      <td>ENST00000507244.6</td>\n","      <td>STX18-AS1</td>\n","      <td>4560393</td>\n","      <td>4562914</td>\n","      <td>ENSG00000247708.8_2|STX18-AS1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>chr5</td>\n","      <td>intron</td>\n","      <td>171265847</td>\n","      <td>171271150</td>\n","      <td>+</td>\n","      <td>ENSG00000204764.14</td>\n","      <td>ENST00000521834.1</td>\n","      <td>RANBP17</td>\n","      <td>171271150</td>\n","      <td>171271619</td>\n","      <td>ENSG00000204764.14_9|RANBP17|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>chr6</td>\n","      <td>intron</td>\n","      <td>24128200</td>\n","      <td>24128716</td>\n","      <td>+</td>\n","      <td>ENSG00000152954.12</td>\n","      <td>ENST00000475830.5</td>\n","      <td>NRSN1</td>\n","      <td>24128716</td>\n","      <td>24130138</td>\n","      <td>ENSG00000152954.12_1|NRSN1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>chr6</td>\n","      <td>intron</td>\n","      <td>99412318</td>\n","      <td>99412550</td>\n","      <td>-</td>\n","      <td>ENSG00000132424.17</td>\n","      <td>ENST00000498075.1</td>\n","      <td>PNISR</td>\n","      <td>99411735</td>\n","      <td>99412318</td>\n","      <td>ENSG00000132424.17_1|PNISR|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>chr7</td>\n","      <td>intron</td>\n","      <td>117222062</td>\n","      <td>117222860</td>\n","      <td>+</td>\n","      <td>ENSG00000004866.22</td>\n","      <td>ENST00000265437.9</td>\n","      <td>ST7</td>\n","      <td>117222860</td>\n","      <td>117223907</td>\n","      <td>ENSG00000004866.22_2|ST7|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>chr7</td>\n","      <td>intron</td>\n","      <td>154566916</td>\n","      <td>154587649</td>\n","      <td>+</td>\n","      <td>ENSG00000130226.17</td>\n","      <td>ENST00000406326.5</td>\n","      <td>DPP6</td>\n","      <td>154587649</td>\n","      <td>154588631</td>\n","      <td>ENSG00000130226.17_3|DPP6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>chr7</td>\n","      <td>intron</td>\n","      <td>74798265</td>\n","      <td>74801984</td>\n","      <td>-</td>\n","      <td>ENSG00000196275.16</td>\n","      <td>ENST00000451013.7</td>\n","      <td>GTF2IRD2</td>\n","      <td>74796150</td>\n","      <td>74798265</td>\n","      <td>ENSG00000196275.16_6|GTF2IRD2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>chr7</td>\n","      <td>intron</td>\n","      <td>151409053</td>\n","      <td>151409426</td>\n","      <td>-</td>\n","      <td>ENSG00000187260.16</td>\n","      <td>ENST00000484630.1</td>\n","      <td>WDR86</td>\n","      <td>151408559</td>\n","      <td>151409053</td>\n","      <td>ENSG00000187260.16_1|WDR86|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>19</th>\n","      <td>chr8</td>\n","      <td>intron</td>\n","      <td>37766355</td>\n","      <td>37766481</td>\n","      <td>+</td>\n","      <td>ENSG00000147471.12</td>\n","      <td>ENST00000523994.1</td>\n","      <td>PLPBP</td>\n","      <td>37766481</td>\n","      <td>37767739</td>\n","      <td>ENSG00000147471.12_1|PLPBP|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>20</th>\n","      <td>chr9</td>\n","      <td>intron</td>\n","      <td>95111345</td>\n","      <td>95111462</td>\n","      <td>-</td>\n","      <td>ENSG00000158169.14</td>\n","      <td>ENST00000490972.7</td>\n","      <td>FANCC</td>\n","      <td>95110511</td>\n","      <td>95111345</td>\n","      <td>ENSG00000158169.14_3|FANCC|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>21</th>\n","      <td>chr10</td>\n","      <td>intron</td>\n","      <td>132351753</td>\n","      <td>132352881</td>\n","      <td>+</td>\n","      <td>ENSG00000148814.18</td>\n","      <td>ENST00000625755.2</td>\n","      <td>LRRC27</td>\n","      <td>132352881</td>\n","      <td>132353396</td>\n","      <td>ENSG00000148814.18_3|LRRC27|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>22</th>\n","      <td>chr10</td>\n","      <td>intron</td>\n","      <td>49103860</td>\n","      <td>49107593</td>\n","      <td>-</td>\n","      <td>ENSG00000165633.13</td>\n","      <td>ENST00000298454.3</td>\n","      <td>VSTM4</td>\n","      <td>49102171</td>\n","      <td>49103860</td>\n","      <td>ENSG00000165633.13_1|VSTM4|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>23</th>\n","      <td>chr10</td>\n","      <td>intron</td>\n","      <td>48575658</td>\n","      <td>48582952</td>\n","      <td>-</td>\n","      <td>ENSG00000128805.15</td>\n","      <td>ENST00000511570.1</td>\n","      <td>ARHGAP22</td>\n","      <td>48573060</td>\n","      <td>48575658</td>\n","      <td>ENSG00000128805.15_2|ARHGAP22|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>24</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66644140</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21</td>\n","      <td>ENST00000310092.12</td>\n","      <td>RBM4</td>\n","      <td>66646026</td>\n","      <td>66646468</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>25</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>71455170</td>\n","      <td>71458427</td>\n","      <td>+</td>\n","      <td>ENSG00000172890.13</td>\n","      <td>ENST00000319023.7</td>\n","      <td>NADSYN1</td>\n","      <td>71458427</td>\n","      <td>71460949</td>\n","      <td>ENSG00000172890.13_2|NADSYN1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>26</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66640123</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21</td>\n","      <td>ENST00000398692.8</td>\n","      <td>RBM4</td>\n","      <td>66646026</td>\n","      <td>66646468</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>27</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66617057</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000248643.5</td>\n","      <td>ENST00000500635.2</td>\n","      <td>RBM14-RBM4</td>\n","      <td>66646026</td>\n","      <td>66646468</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>28</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66644743</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21</td>\n","      <td>ENST00000515838.2</td>\n","      <td>RBM4</td>\n","      <td>66646026</td>\n","      <td>66646468</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>29</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>134482730</td>\n","      <td>134502943</td>\n","      <td>+</td>\n","      <td>ENSG00000255545.8</td>\n","      <td>ENST00000532886.2</td>\n","      <td>B3GAT1-DT</td>\n","      <td>134502943</td>\n","      <td>134505661</td>\n","      <td>ENSG00000255545.8_2|B3GAT1-DT|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>30</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66521466</td>\n","      <td>66526935</td>\n","      <td>-</td>\n","      <td>ENSG00000174165.8</td>\n","      <td>ENST00000526986.5</td>\n","      <td>ZDHHC24</td>\n","      <td>66520636</td>\n","      <td>66521466</td>\n","      <td>ENSG00000174165.8_3|ZDHHC24|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>31</th>\n","      <td>chr11</td>\n","      <td>intron</td>\n","      <td>66521466</td>\n","      <td>66524154</td>\n","      <td>-</td>\n","      <td>ENSG00000174165.8</td>\n","      <td>ENST00000534073.5</td>\n","      <td>ZDHHC24</td>\n","      <td>66520636</td>\n","      <td>66521466</td>\n","      <td>ENSG00000174165.8_3|ZDHHC24|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>32</th>\n","      <td>chr12</td>\n","      <td>intron</td>\n","      <td>123156908</td>\n","      <td>123160780</td>\n","      <td>-</td>\n","      <td>ENSG00000051825.15</td>\n","      <td>ENST00000302373.8</td>\n","      <td>MPHOSPH9</td>\n","      <td>123156635</td>\n","      <td>123156908</td>\n","      <td>ENSG00000051825.15_5|MPHOSPH9|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>33</th>\n","      <td>chr15</td>\n","      <td>intron</td>\n","      <td>74333967</td>\n","      <td>74334974</td>\n","      <td>+</td>\n","      <td>ENSG00000140481.15</td>\n","      <td>ENST00000268082.4</td>\n","      <td>CCDC33</td>\n","      <td>74334974</td>\n","      <td>74341262</td>\n","      <td>ENSG00000140481.15_6|CCDC33|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>34</th>\n","      <td>chr15</td>\n","      <td>intron</td>\n","      <td>75732152</td>\n","      <td>75738386</td>\n","      <td>-</td>\n","      <td>ENSG00000246877.2</td>\n","      <td>ENST00000501931.1</td>\n","      <td>DNM1P35</td>\n","      <td>75727496</td>\n","      <td>75732152</td>\n","      <td>ENSG00000246877.2_2|DNM1P35|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>35</th>\n","      <td>chr16</td>\n","      <td>intron</td>\n","      <td>29904965</td>\n","      <td>29905787</td>\n","      <td>+</td>\n","      <td>ENSG00000174939.11</td>\n","      <td>ENST00000308748.10</td>\n","      <td>ASPHD1</td>\n","      <td>29905787</td>\n","      <td>29906416</td>\n","      <td>ENSG00000174939.11_1|ASPHD1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>36</th>\n","      <td>chr16</td>\n","      <td>intron</td>\n","      <td>86541198</td>\n","      <td>86541696</td>\n","      <td>-</td>\n","      <td>ENSG00000103248.19</td>\n","      <td>ENST00000562940.6</td>\n","      <td>MTHFSD</td>\n","      <td>86540719</td>\n","      <td>86541198</td>\n","      <td>ENSG00000103248.19_2|MTHFSD|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>37</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>20232414</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19</td>\n","      <td>ENST00000395522.6</td>\n","      <td>SPECC1</td>\n","      <td>20236859</td>\n","      <td>20239427</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>38</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>20232405</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19</td>\n","      <td>ENST00000395525.7</td>\n","      <td>SPECC1</td>\n","      <td>20236859</td>\n","      <td>20239427</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>39</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>80429831</td>\n","      <td>80436132</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17</td>\n","      <td>ENST00000517795.5</td>\n","      <td>ENDOV</td>\n","      <td>80436132</td>\n","      <td>80438086</td>\n","      <td>ENSG00000173818.17_5|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>40</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>80425620</td>\n","      <td>80436132</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17</td>\n","      <td>ENST00000518907.5</td>\n","      <td>ENDOV</td>\n","      <td>80436132</td>\n","      <td>80438086</td>\n","      <td>ENSG00000173818.17_5|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>41</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>80419720</td>\n","      <td>80420049</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17</td>\n","      <td>ENST00000523999.5</td>\n","      <td>ENDOV</td>\n","      <td>80420049</td>\n","      <td>80420533</td>\n","      <td>ENSG00000173818.17_2|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>42</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>20232451</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19</td>\n","      <td>ENST00000584527.1</td>\n","      <td>SPECC1</td>\n","      <td>20236859</td>\n","      <td>20239427</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>43</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>38545459</td>\n","      <td>38547758</td>\n","      <td>-</td>\n","      <td>ENSG00000277363.5</td>\n","      <td>ENST00000613927.1</td>\n","      <td>SRCIN1</td>\n","      <td>38544577</td>\n","      <td>38545459</td>\n","      <td>ENSG00000277363.5_2|SRCIN1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>44</th>\n","      <td>chr17</td>\n","      <td>intron</td>\n","      <td>38545459</td>\n","      <td>38546460</td>\n","      <td>-</td>\n","      <td>ENSG00000277363.5</td>\n","      <td>ENST00000621275.1</td>\n","      <td>SRCIN1</td>\n","      <td>38544577</td>\n","      <td>38545459</td>\n","      <td>ENSG00000277363.5_2|SRCIN1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>45</th>\n","      <td>chr19</td>\n","      <td>intron</td>\n","      <td>53166347</td>\n","      <td>53175444</td>\n","      <td>-</td>\n","      <td>ENSG00000197497.11</td>\n","      <td>ENST00000396424.5</td>\n","      <td>ZNF665</td>\n","      <td>53162423</td>\n","      <td>53166347</td>\n","      <td>ENSG00000197497.11_3|ZNF665|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>46</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000461897.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +8    |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 54630886  | 54634687  | +            | ...   |\n","| chr1         | intron     | 112542090 | 112550600 | -            | ...   |\n","| chr1         | intron     | 112542090 | 112555867 | -            | ...   |\n","| chr1         | intron     | 33592489  | 33600864  | -            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chr17        | intron     | 80419720  | 80420049  | +            | ...   |\n","| chr17        | intron     | 20232451  | 20236859  | +            | ...   |\n","| chr17        | intron     | 38545459  | 38547758  | -            | ...   |\n","| chr17        | intron     | 38545459  | 38546460  | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 47 rows and 13 columns from 17 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","8 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 1 more.)"]},"execution_count":33,"metadata":{},"output_type":"execute_result"}],"source":["# now to get SJs for annotated ALEs\n","# overlap-join SJs and LEs, allowing bookended/directly adjacent to overlap\n","# subset for direct match at intron 3'end, last exon 5'EncodingWarning\n","le_sj_ref = (ref_introns.join(le_cryp, strandedness=\"same\", slack=1)\n"," .subset(lambda df: ((df.Strand == \"+\") & (df[\"End\"] == df[\"Start_b\"])) |\n","         ((df.Strand == \"-\") & (df[\"Start\"] == df[\"End_b\"]))\n","         )\n","         .drop_duplicate_positions()\n"," )\n","\n","le_sj_ref"]},{"cell_type":"code","execution_count":34,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>54630886</td>\n","      <td>54634687</td>\n","      <td>+</td>\n","      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>112542090</td>\n","      <td>112550600</td>\n","      <td>-</td>\n","      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>112542090</td>\n","      <td>112555867</td>\n","      <td>-</td>\n","      <td>ENSG00000007341.19_4|ST7L|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>33592489</td>\n","      <td>33600864</td>\n","      <td>-</td>\n","      <td>ENSG00000121904.19_3|CSMD2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>163305313</td>\n","      <td>163306232</td>\n","      <td>-</td>\n","      <td>ENSG00000232995.7_2|ENSG00000232995|spliced|cr...</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>chr2</td>\n","      <td>205766818</td>\n","      <td>205776245</td>\n","      <td>+</td>\n","      <td>ENSG00000118257.17_5|NRP2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>chr2</td>\n","      <td>205766803</td>\n","      <td>205776245</td>\n","      <td>+</td>\n","      <td>ENSG00000118257.17_5|NRP2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>chr2</td>\n","      <td>3457693</td>\n","      <td>3457794</td>\n","      <td>+</td>\n","      <td>ENSG00000171853.16_4|TRAPPC12|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>chr3</td>\n","      <td>3129188</td>\n","      <td>3129918</td>\n","      <td>+</td>\n","      <td>ENSG00000072756.17_1|TRNT1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>chr3</td>\n","      <td>114339426</td>\n","      <td>114350273</td>\n","      <td>-</td>\n","      <td>ENSG00000181722.17_5|ZBTB20|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>chr3</td>\n","      <td>8195073</td>\n","      <td>8494680</td>\n","      <td>-</td>\n","      <td>ENSG00000227110.7_3|LMCD1-AS1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>chr4</td>\n","      <td>4544907</td>\n","      <td>4560393</td>\n","      <td>+</td>\n","      <td>ENSG00000247708.8_2|STX18-AS1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>chr5</td>\n","      <td>171265847</td>\n","      <td>171271150</td>\n","      <td>+</td>\n","      <td>ENSG00000204764.14_9|RANBP17|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>chr6</td>\n","      <td>24128200</td>\n","      <td>24128716</td>\n","      <td>+</td>\n","      <td>ENSG00000152954.12_1|NRSN1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>chr6</td>\n","      <td>99412318</td>\n","      <td>99412550</td>\n","      <td>-</td>\n","      <td>ENSG00000132424.17_1|PNISR|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>chr7</td>\n","      <td>117222062</td>\n","      <td>117222860</td>\n","      <td>+</td>\n","      <td>ENSG00000004866.22_2|ST7|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>chr7</td>\n","      <td>154566916</td>\n","      <td>154587649</td>\n","      <td>+</td>\n","      <td>ENSG00000130226.17_3|DPP6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>chr7</td>\n","      <td>74798265</td>\n","      <td>74801984</td>\n","      <td>-</td>\n","      <td>ENSG00000196275.16_6|GTF2IRD2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>chr7</td>\n","      <td>151409053</td>\n","      <td>151409426</td>\n","      <td>-</td>\n","      <td>ENSG00000187260.16_1|WDR86|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>19</th>\n","      <td>chr8</td>\n","      <td>37766355</td>\n","      <td>37766481</td>\n","      <td>+</td>\n","      <td>ENSG00000147471.12_1|PLPBP|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>20</th>\n","      <td>chr9</td>\n","      <td>95111345</td>\n","      <td>95111462</td>\n","      <td>-</td>\n","      <td>ENSG00000158169.14_3|FANCC|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>21</th>\n","      <td>chr10</td>\n","      <td>132351753</td>\n","      <td>132352881</td>\n","      <td>+</td>\n","      <td>ENSG00000148814.18_3|LRRC27|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>22</th>\n","      <td>chr10</td>\n","      <td>49103860</td>\n","      <td>49107593</td>\n","      <td>-</td>\n","      <td>ENSG00000165633.13_1|VSTM4|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>23</th>\n","      <td>chr10</td>\n","      <td>48575658</td>\n","      <td>48582952</td>\n","      <td>-</td>\n","      <td>ENSG00000128805.15_2|ARHGAP22|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>24</th>\n","      <td>chr11</td>\n","      <td>66644140</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>25</th>\n","      <td>chr11</td>\n","      <td>71455170</td>\n","      <td>71458427</td>\n","      <td>+</td>\n","      <td>ENSG00000172890.13_2|NADSYN1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>26</th>\n","      <td>chr11</td>\n","      <td>66640123</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>27</th>\n","      <td>chr11</td>\n","      <td>66617057</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>28</th>\n","      <td>chr11</td>\n","      <td>66644743</td>\n","      <td>66646026</td>\n","      <td>+</td>\n","      <td>ENSG00000173933.21_2|RBM4|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>29</th>\n","      <td>chr11</td>\n","      <td>134482730</td>\n","      <td>134502943</td>\n","      <td>+</td>\n","      <td>ENSG00000255545.8_2|B3GAT1-DT|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>30</th>\n","      <td>chr11</td>\n","      <td>66521466</td>\n","      <td>66526935</td>\n","      <td>-</td>\n","      <td>ENSG00000174165.8_3|ZDHHC24|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>31</th>\n","      <td>chr11</td>\n","      <td>66521466</td>\n","      <td>66524154</td>\n","      <td>-</td>\n","      <td>ENSG00000174165.8_3|ZDHHC24|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>32</th>\n","      <td>chr12</td>\n","      <td>123156908</td>\n","      <td>123160780</td>\n","      <td>-</td>\n","      <td>ENSG00000051825.15_5|MPHOSPH9|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>33</th>\n","      <td>chr15</td>\n","      <td>74333967</td>\n","      <td>74334974</td>\n","      <td>+</td>\n","      <td>ENSG00000140481.15_6|CCDC33|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>34</th>\n","      <td>chr15</td>\n","      <td>75732152</td>\n","      <td>75738386</td>\n","      <td>-</td>\n","      <td>ENSG00000246877.2_2|DNM1P35|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>35</th>\n","      <td>chr16</td>\n","      <td>29904965</td>\n","      <td>29905787</td>\n","      <td>+</td>\n","      <td>ENSG00000174939.11_1|ASPHD1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>36</th>\n","      <td>chr16</td>\n","      <td>86541198</td>\n","      <td>86541696</td>\n","      <td>-</td>\n","      <td>ENSG00000103248.19_2|MTHFSD|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>37</th>\n","      <td>chr17</td>\n","      <td>20232414</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>38</th>\n","      <td>chr17</td>\n","      <td>20232405</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>39</th>\n","      <td>chr17</td>\n","      <td>80429831</td>\n","      <td>80436132</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17_5|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>40</th>\n","      <td>chr17</td>\n","      <td>80425620</td>\n","      <td>80436132</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17_5|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>41</th>\n","      <td>chr17</td>\n","      <td>80419720</td>\n","      <td>80420049</td>\n","      <td>+</td>\n","      <td>ENSG00000173818.17_2|ENDOV|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>42</th>\n","      <td>chr17</td>\n","      <td>20232451</td>\n","      <td>20236859</td>\n","      <td>+</td>\n","      <td>ENSG00000128487.19_1|SPECC1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>43</th>\n","      <td>chr17</td>\n","      <td>38545459</td>\n","      <td>38547758</td>\n","      <td>-</td>\n","      <td>ENSG00000277363.5_2|SRCIN1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>44</th>\n","      <td>chr17</td>\n","      <td>38545459</td>\n","      <td>38546460</td>\n","      <td>-</td>\n","      <td>ENSG00000277363.5_2|SRCIN1|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>45</th>\n","      <td>chr19</td>\n","      <td>53166347</td>\n","      <td>53175444</td>\n","      <td>-</td>\n","      <td>ENSG00000197497.11_3|ZNF665|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>46</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Start     | End       | Strand       | +2    |\n","| (object)     | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+-----------+-----------+--------------+-------|\n","| chr1         | 54630886  | 54634687  | +            | ...   |\n","| chr1         | 112542090 | 112550600 | -            | ...   |\n","| chr1         | 112542090 | 112555867 | -            | ...   |\n","| chr1         | 33592489  | 33600864  | -            | ...   |\n","| ...          | ...       | ...       | ...          | ...   |\n","| chr17        | 80419720  | 80420049  | +            | ...   |\n","| chr17        | 20232451  | 20236859  | +            | ...   |\n","| chr17        | 38545459  | 38547758  | -            | ...   |\n","| chr17        | 38545459  | 38546460  | -            | ...   |\n","+--------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 47 rows and 6 columns from 17 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","2 hidden columns: Name, Score"]},"execution_count":34,"metadata":{},"output_type":"execute_result"}],"source":["# finally generate a cleaned BED-ready file of ref SJs\n","le_sj_clean_ref = le_sj_ref[[\"Name\"]]\n","le_sj_clean_ref.Score = \".\"\n","le_sj_clean_ref"]},{"cell_type":"code","execution_count":38,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","      <td>+</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","      <td>+</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>54630886</td>\n","      <td>54634687</td>\n","      <td>+</td>\n","      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>243613034</td>\n","      <td>243613670</td>\n","      <td>-</td>\n","      <td>ENSG00000117020.19_2|AKT3|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>153</th>\n","      <td>chrX</td>\n","      <td>98644574</td>\n","      <td>98679426</td>\n","      <td>+</td>\n","      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>154</th>\n","      <td>chrX</td>\n","      <td>17837395</td>\n","      <td>17837536</td>\n","      <td>-</td>\n","      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>155</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>156</th>\n","      <td>chrX</td>\n","      <td>17837395</td>\n","      <td>17861097</td>\n","      <td>-</td>\n","      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>157</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>158 rows × 6 columns</p>\n","</div>"],"text/plain":["+--------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Start     | End       | Strand       | +2    |\n","| (object)     | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+-----------+-----------+--------------+-------|\n","| chr1         | 245419745 | 245464258 | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| chr1         | 76868742  | 76871267  | +            | ...   |\n","| chr1         | 54630886  | 54634687  | +            | ...   |\n","| ...          | ...       | ...       | ...          | ...   |\n","| chrX         | 17837395  | 17837536  | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 17837395  | 17861097  | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","+--------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 158 rows and 6 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","2 hidden columns: Name, Score"]},"execution_count":38,"metadata":{},"output_type":"execute_result"}],"source":["# now just need to check that all les are represented\n","le_sj_clean = pr.concat([le_sj_clean_nov.drop_duplicate_positions(), le_sj_clean_ref])\n","le_sj_clean"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"code","execution_count":55,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Fraction of cryptics where gene has SJ found by MAJIQ - 0.7526881720430108\n"]}],"source":["# check overlap with AL's cryptics BED\n","# get gene names from le_sj_id_clean\n","# Name field = ENSG00000162849.16_2|KIF26B|spliced|cryptic\t\n","le_cryp_gn = set(le_sj_clean.Name.str.split(\"|\", expand=True)[1])\n","\n","# Name field = PTPRU|novel_donor|2\t\n","majiq_sj_gn = set(majiq_sj.Name.str.split(\"|\", expand=True)[0])\n","\n","# get matching genes\n","le_majiq_m_gn = le_cryp_gn.intersection(majiq_sj_gn)\n","\n","print(f\"Fraction of cryptics where gene has SJ found by MAJIQ - {len(le_majiq_m_gn) / len(le_cryp_gn)}\")\n","#\n"]},{"cell_type":"code","execution_count":66,"metadata":{},"outputs":[{"data":{"text/plain":["count    7.400000e+01\n","mean     1.789569e+07\n","std      4.010586e+07\n","min      0.000000e+00\n","0%       0.000000e+00\n","10%      0.000000e+00\n","20%      0.000000e+00\n","30%      1.000000e+00\n","40%      1.000000e+00\n","50%      1.000000e+00\n","60%      1.689400e+03\n","70%      1.410880e+04\n","80%      2.557911e+07\n","90%      7.430823e+07\n","100%     2.057762e+08\n","max      2.057762e+08\n","Name: end3_diff, dtype: float64"]},"execution_count":66,"metadata":{},"output_type":"execute_result"}],"source":["# extract gene_name as column\n","le_sj_clean = le_sj_clean.assign(\"gene_name\", lambda df: df.Name.str.split(\"|\", expand=True)[1])\n","majiq_sj = majiq_sj.assign(\"gene_name\", lambda df: df.Name.str.split(\"|\", expand=True)[0])\n","\n","# overlap-join le inferred SJs with MAJIQ SJs \n","sj_le_majiq = (le_sj_clean.subset(lambda df: df.gene_name.isin(le_majiq_m_gn))\n"," .join(majiq_sj.subset(lambda df: df.gene_name.isin(le_majiq_m_gn)),\n","       how=\"left\",\n","       strandedness=\"same\"\n","        )\n"," )\n","\n","# calculate difference between 3' coordinates\n","(sj_le_majiq.assign(\"end3_diff\",\n","                    lambda df: abs(df.End_b - df.End) if (df.Strand == \"+\").all() else abs(df.Start_b - df.Start))\n","                    .apply(lambda df: df.loc[df.groupby(\"Name\")[\"end3_diff\"].idxmin(), :])\n","                    .end3_diff.describe(percentiles=[0.1 * i for i in range(0,11,1)])\n",")\n"]},{"cell_type":"code","execution_count":68,"metadata":{},"outputs":[{"data":{"text/plain":["count    7.400000e+01\n","mean     1.789371e+07\n","std      4.010648e+07\n","min      0.000000e+00\n","0%       0.000000e+00\n","10%      0.000000e+00\n","20%      0.000000e+00\n","30%      0.000000e+00\n","40%      1.000000e+00\n","50%      1.000000e+00\n","60%      1.000000e+00\n","70%      2.867100e+03\n","80%      2.557979e+07\n","90%      7.430702e+07\n","100%     2.057668e+08\n","max      2.057668e+08\n","Name: end5_diff, dtype: float64"]},"execution_count":68,"metadata":{},"output_type":"execute_result"}],"source":["(sj_le_majiq.assign(\"end5_diff\",\n","                    lambda df: abs(df.Start_b - df.Start) if (df.Strand == \"+\").all() else abs(df.End_b - df.End))\n","                    .apply(lambda df: df.loc[df.groupby(\"Name\")[\"end5_diff\"].idxmin(), :]) # select smallest per gene (so prioritises match)\n","                    .end5_diff.describe(percentiles=[0.1 * i for i in range(0,11,1)])\n",")"]},{"cell_type":"markdown","metadata":{},"source":["20-30 % are exactly matching at 3'end (start of last exon), but further 20-30 % differ by 1 nucleotide... One off error?\n","\n","5'end matching is slightly better...\n","\n","Let's IGV browse a few "]},{"cell_type":"code","execution_count":70,"metadata":{},"outputs":[],"source":["# Add 5' and 3' difference to df\n","sj_le_majiq = (sj_le_majiq.assign(\"end3_diff\",\n","                                                      lambda df: abs(df.End_b - df.End) if (df.Strand == \"+\").all() else abs(df.Start_b - df.Start))\n","                    .assign(\"end5_diff\",\n","                                                lambda df: abs(df.Start_b - df.Start) if (df.Strand == \"+\").all() else abs(df.End_b - df.End))\n","                    )"]},{"cell_type":"code","execution_count":87,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name_b</th>\n","      <th>Score_b</th>\n","      <th>Strand_b</th>\n","      <th>gene_name_b</th>\n","      <th>end3_diff</th>\n","      <th>end5_diff</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr6</td>\n","      <td>158017290</td>\n","      <td>158019983</td>\n","      <td>+</td>\n","      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>SYNJ2</td>\n","      <td>158017290</td>\n","      <td>158019984</td>\n","      <td>SYNJ2|novel_acceptor|5</td>\n","      <td>0</td>\n","      <td>+</td>\n","      <td>SYNJ2</td>\n","      <td>1</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr8</td>\n","      <td>79611214</td>\n","      <td>79616821</td>\n","      <td>+</td>\n","      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>STMN2</td>\n","      <td>79611214</td>\n","      <td>79616822</td>\n","      <td>STMN2|novel_acceptor|23</td>\n","      <td>0</td>\n","      <td>+</td>\n","      <td>STMN2</td>\n","      <td>1</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr8</td>\n","      <td>79611791</td>\n","      <td>79616821</td>\n","      <td>+</td>\n","      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>STMN2</td>\n","      <td>79611791</td>\n","      <td>79616822</td>\n","      <td>STMN2|novel_acceptor|1</td>\n","      <td>0</td>\n","      <td>+</td>\n","      <td>STMN2</td>\n","      <td>1</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr11</td>\n","      <td>128992046</td>\n","      <td>128998318</td>\n","      <td>-</td>\n","      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>ARHGAP32</td>\n","      <td>128992046</td>\n","      <td>128998319</td>\n","      <td>ARHGAP32|novel_acceptor|21</td>\n","      <td>0</td>\n","      <td>-</td>\n","      <td>ARHGAP32</td>\n","      <td>0</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr15</td>\n","      <td>52777690</td>\n","      <td>52788779</td>\n","      <td>-</td>\n","      <td>ENSG00000169856.9_1|ONECUT1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>ONECUT1</td>\n","      <td>52777690</td>\n","      <td>52788780</td>\n","      <td>ONECUT1|novel_acceptor|5</td>\n","      <td>0</td>\n","      <td>-</td>\n","      <td>ONECUT1</td>\n","      <td>0</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+-----------+-----------+--------------+-------+\n","| Chromosome   |     Start |       End | Strand       | +11   |\n","| (object)     |   (int64) |   (int64) | (category)   | ...   |\n","|--------------+-----------+-----------+--------------+-------|\n","| chr6         | 158017290 | 158019983 | +            | ...   |\n","| chr8         |  79611214 |  79616821 | +            | ...   |\n","| chr8         |  79611791 |  79616821 | +            | ...   |\n","| chr11        | 128992046 | 128998318 | -            | ...   |\n","| chr15        |  52777690 |  52788779 | -            | ...   |\n","+--------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 5 rows and 15 columns from 4 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","11 hidden columns: Name, Score, gene_name, Start_b, End_b, Name_b, ... (+ 5 more.)"]},"execution_count":87,"metadata":{},"output_type":"execute_result"}],"source":["sj_le_majiq.subset(lambda df: df.end3_diff + df.end5_diff <= 1).subset(lambda df: df.gene_name.isin([\"STMN2\", \"ARHGAP32\", \"SYNJ2\", \"GLE1\", \"ONECUT1\"])).drop_duplicate_positions()"]},{"cell_type":"code","execution_count":84,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>count</th>\n","      <th>mean</th>\n","      <th>std</th>\n","      <th>min</th>\n","      <th>0%</th>\n","      <th>10%</th>\n","      <th>20%</th>\n","      <th>30%</th>\n","      <th>40%</th>\n","      <th>50%</th>\n","      <th>60%</th>\n","      <th>70%</th>\n","      <th>80%</th>\n","      <th>90%</th>\n","      <th>100%</th>\n","      <th>max</th>\n","    </tr>\n","    <tr>\n","      <th>Strand</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>+</th>\n","      <td>21.0</td>\n","      <td>1.0</td>\n","      <td>0.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","    </tr>\n","    <tr>\n","      <th>-</th>\n","      <td>20.0</td>\n","      <td>1.0</td>\n","      <td>0.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["        count  mean  std  min   0%  10%  20%  30%  40%  50%  60%  70%  80%  \\\n","Strand                                                                       \n","+        21.0   1.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   \n","-        20.0   1.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   \n","\n","        90%  100%  max  \n","Strand                  \n","+       1.0   1.0  1.0  \n","-       1.0   1.0  1.0  "]},"execution_count":84,"metadata":{},"output_type":"execute_result"}],"source":["# matches seem to have a difference of 1 at either 5'/3' end\n","# I think this is due to End coordinates being systematically shifted\n","# If so, then when subset for matches the one-off difference should be perfectly split by strand\n","# (+ strand - 3'end = End coordinate, - strand - 5'end = End coordinate)\n","(sj_le_majiq.subset(lambda df: df.end3_diff + df.end5_diff <= 1)\n"," .drop_duplicate_positions()\n"," .assign(\"end_diff\", lambda df: abs(df.End_b - df.End))\n"," .as_df()[[\"Strand\", \"Name\", \"end_diff\"]]\n"," .groupby(\"Strand\")\n"," .end_diff.describe(percentiles=[0.1 * i for i in range(0,11,1)])\n"," )"]},{"cell_type":"code","execution_count":86,"metadata":{},"outputs":[],"source":["# output current SJ BED just so I can vis in IGV\n","le_sj_clean.drop(\"gene_name\").to_bed(\"../processed/2023-09-01_putative_as_cle_sjs.bed\")"]},{"cell_type":"code","execution_count":81,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>count</th>\n","      <th>mean</th>\n","      <th>std</th>\n","      <th>min</th>\n","      <th>0%</th>\n","      <th>10%</th>\n","      <th>20%</th>\n","      <th>30%</th>\n","      <th>40%</th>\n","      <th>50%</th>\n","      <th>60%</th>\n","      <th>70%</th>\n","      <th>80%</th>\n","      <th>90%</th>\n","      <th>100%</th>\n","      <th>max</th>\n","    </tr>\n","    <tr>\n","      <th>Strand</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>+</th>\n","      <td>35.0</td>\n","      <td>1.555124e+07</td>\n","      <td>4.182312e+07</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1478.0</td>\n","      <td>8860.4</td>\n","      <td>647852.6</td>\n","      <td>73130860.8</td>\n","      <td>205776246.0</td>\n","      <td>205776246.0</td>\n","    </tr>\n","    <tr>\n","      <th>-</th>\n","      <td>39.0</td>\n","      <td>1.999886e+07</td>\n","      <td>3.892956e+07</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>15867.0</td>\n","      <td>45014407.6</td>\n","      <td>77149927.4</td>\n","      <td>151409427.0</td>\n","      <td>151409427.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["        count          mean           std  min   0%  10%  20%  30%  40%  50%  \\\n","Strand                                                                         \n","+        35.0  1.555124e+07  4.182312e+07  1.0  1.0  1.0  1.0  1.0  1.0  1.0   \n","-        39.0  1.999886e+07  3.892956e+07  0.0  0.0  1.0  1.0  1.0  1.0  1.0   \n","\n","           60%      70%         80%         90%         100%          max  \n","Strand                                                                     \n","+       1478.0   8860.4    647852.6  73130860.8  205776246.0  205776246.0  \n","-          1.0  15867.0  45014407.6  77149927.4  151409427.0  151409427.0  "]},"execution_count":81,"metadata":{},"output_type":"execute_result"}],"source":["# # potentially a strand & End coordinate problem? AL's BED is always +1?\n","# (sj_le_majiq.assign(\"end_diff\", lambda df: abs(df.End_b - df.End))\n","#  .apply(lambda df: df.loc[df.groupby(\"Name\")[\"end_diff\"].idxmin(), :]) # select smallest per gene (so prioritises match)\n","#  .as_df()\n","#  [[\"Name\", \"end_diff\",\"Strand\"]]\n","#  .groupby(\"Strand\")\n","#  .end_diff.describe(percentiles=[0.1 * i for i in range(0,11,1)])\n","# )"]},{"cell_type":"code","execution_count":42,"metadata":{},"outputs":[{"data":{"text/plain":["{'ENSG00000002746.15_3|HECW1|spliced|cryptic',\n"," 'ENSG00000100060.18_2|MFNG|spliced|cryptic',\n"," 'ENSG00000111665.12_4|CDCA3|spliced|cryptic',\n"," 'ENSG00000127483.19_1|HP1BP3|spliced|cryptic',\n"," 'ENSG00000151692.15_6|RNF144A|spliced|cryptic',\n"," 'ENSG00000151692.15_7|RNF144A|spliced|cryptic',\n"," 'ENSG00000183570.17_3|PCBP3|spliced|cryptic',\n"," 'ENSG00000184347.15_7|SLIT3|spliced|cryptic',\n"," 'ENSG00000197530.13_1|MIB2|spliced|cryptic',\n"," 'ENSG00000216895.9_4|ENSG00000216895|spliced|cryptic'}"]},"execution_count":42,"metadata":{},"output_type":"execute_result"}],"source":["# which le_ids do not have a SJ?\n","ids_le_cryp.difference(set(le_sj_clean.Name))"]},{"cell_type":"code","execution_count":44,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand</th>\n","      <th>Feature</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Strand_b</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000378712.5</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000479659.5</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000514363.1</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000489635.5</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000504599.6</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000505820.7</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000507229.5</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000511502.5</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000518681.6</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623388</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000520777.6</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1617077</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000463492.1</td>\n","      <td>MIB2</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>chr1</td>\n","      <td>20775449</td>\n","      <td>20776596</td>\n","      <td>ENSG00000127483.19_1|HP1BP3|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>intron</td>\n","      <td>20773610</td>\n","      <td>20776596</td>\n","      <td>-</td>\n","      <td>ENSG00000127483.19</td>\n","      <td>ENST00000312239.10</td>\n","      <td>HP1BP3</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>chr1</td>\n","      <td>20775449</td>\n","      <td>20776596</td>\n","      <td>ENSG00000127483.19_1|HP1BP3|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>intron</td>\n","      <td>20776005</td>\n","      <td>20776596</td>\n","      <td>-</td>\n","      <td>ENSG00000127483.19</td>\n","      <td>ENST00000375000.5</td>\n","      <td>HP1BP3</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>chr7</td>\n","      <td>43444214</td>\n","      <td>43446033</td>\n","      <td>ENSG00000002746.15_3|HECW1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>43442629</td>\n","      <td>43444217</td>\n","      <td>+</td>\n","      <td>ENSG00000002746.15</td>\n","      <td>ENST00000395891.7</td>\n","      <td>HECW1</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>chr7</td>\n","      <td>43444214</td>\n","      <td>43446033</td>\n","      <td>ENSG00000002746.15_3|HECW1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>43442629</td>\n","      <td>43444217</td>\n","      <td>+</td>\n","      <td>ENSG00000002746.15</td>\n","      <td>ENST00000453890.5</td>\n","      <td>HECW1</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>chr7</td>\n","      <td>43444214</td>\n","      <td>43446033</td>\n","      <td>ENSG00000002746.15_3|HECW1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>43445570</td>\n","      <td>43456296</td>\n","      <td>+</td>\n","      <td>ENSG00000002746.15</td>\n","      <td>ENST00000453890.5</td>\n","      <td>HECW1</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>chr7</td>\n","      <td>43444214</td>\n","      <td>43446033</td>\n","      <td>ENSG00000002746.15_3|HECW1|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>intron</td>\n","      <td>43445570</td>\n","      <td>43450827</td>\n","      <td>+</td>\n","      <td>ENSG00000002746.15</td>\n","      <td>ENST00000395891.7</td>\n","      <td>HECW1</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>chr12</td>\n","      <td>6844792</td>\n","      <td>6846724</td>\n","      <td>ENSG00000111665.12_4|CDCA3|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>intron</td>\n","      <td>6846367</td>\n","      <td>6846724</td>\n","      <td>-</td>\n","      <td>ENSG00000111665.12</td>\n","      <td>ENST00000603043.1</td>\n","      <td>CDCA3</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+-----------+-----------+-------+\n","| Chromosome   | Start     | End       | +10   |\n","| (category)   | (int64)   | (int64)   | ...   |\n","|--------------+-----------+-----------+-------|\n","| chr1         | 1616614   | 1619210   | ...   |\n","| chr1         | 1616614   | 1619210   | ...   |\n","| chr1         | 1616614   | 1619210   | ...   |\n","| chr1         | 1616614   | 1619210   | ...   |\n","| ...          | ...       | ...       | ...   |\n","| chr7         | 43444214  | 43446033  | ...   |\n","| chr7         | 43444214  | 43446033  | ...   |\n","| chr7         | 43444214  | 43446033  | ...   |\n","| chr7         | 43444214  | 43446033  | ...   |\n","+--------------+-----------+-----------+-------+\n","Stranded PyRanges object has 19 rows and 13 columns from 3 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: Name, Score, Strand, Feature, Start_b, ... (+ 5 more.)"]},"execution_count":44,"metadata":{},"output_type":"execute_result"}],"source":["le_cryp.subset(lambda df: df.Name.isin(ids_le_cryp.difference(set(le_sj_clean.Name)))).join(ref_introns, strandedness=\"same\", slack=1)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# check overlap"]}],"metadata":{"kernelspec":{"display_name":"pybioinfo","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.11"},"orig_nbformat":4},"nbformat":4,"nbformat_minor":2}
