{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Extract splice junctions for AS-ALE (/ last exon spliced) events.\n","\n","Due to a pretty silly move on my part, my pipeline does not output splice junctions for novel spliced last exons (nor last exons)\n","Given want to use SJs to quantify these events in NYGC data, need to extract using only last exon coordinates.\n","\n","In pipeline, approach to define last exons:\n","- 3'end doesn't overlap any annotated exon\n","- terminal splice junction must match an annotated splice junction at the 5'end. \n","\n","Can exploit these criteria to grab SJs 'in reverse'\n","\n","Strategy for novel last exons:\n","- Extract introns (SJs) from annotated transcripts\n","- find introns in which last exons are completely contained\n","- (w/ some form of pr.join), construct a SJ with start coordinate of annotated intron and end coordinate = start of exon coordinate\n","- As a sanity check, compare with AL's provided list of cryptic SJs inferred with MAJIQ - for common genes, do we get the same SJ coordinates\n","\n","Strategy for annotated last exons:\n","- Any spliced last exon that is not captured with novel LE strategy\n","- pr.join with slack=1 (allow bookended intervals to overlap) annotated SJs with last exon coordinates\n","- Subset for exact matches at SJ 3'end/last exon 5'end\n","- Retain the SJs for downstream analysis"]},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":["import pyranges as pr\n","import pandas as pd\n","import numpy as np"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["CPU times: user 1e+03 ns, sys: 1 µs, total: 2 µs\n","Wall time: 3.58 µs\n"]}],"source":["majiq_sj = pr.read_bed(\"../data/Everything_minus_ferguson_liu_recent_new_cryptics.junctions.bed\")\n","ref_gtf = pr.read_gtf(\"../data/reference_filtered.gtf\")\n","le = pr.read_bed(\"../data/2023-07-04_papa_cryptic_spliced.last_exons.bed\") # generated in motifs/ subdir, all cryptic last exon coords used for iCLIP analysis"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":["# ref gtf has lots of attributes don't need, subset to minimal cols for analysis\n","ref_gtf = ref_gtf[[\"Feature\", \"gene_id\", \"transcript_id\", \"gene_name\"]]"]},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>76871267</td>\n","      <td>76871821</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>54634687</td>\n","      <td>54639192</td>\n","      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>245464258</td>\n","      <td>245471621</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>102</th>\n","      <td>chr22</td>\n","      <td>37464524</td>\n","      <td>37465718</td>\n","      <td>ENSG00000100060.18_2|MFNG|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>103</th>\n","      <td>chrX</td>\n","      <td>102721363</td>\n","      <td>102724864</td>\n","      <td>ENSG00000198908.12_1|BHLHB9|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>104</th>\n","      <td>chrX</td>\n","      <td>98679426</td>\n","      <td>98679978</td>\n","      <td>ENSG00000281566.3_1|ENSG00000281566|spliced|cr...</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>105</th>\n","      <td>chrX</td>\n","      <td>17835910</td>\n","      <td>17837395</td>\n","      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>106</th>\n","      <td>chrX</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>107 rows × 6 columns</p>\n","</div>"],"text/plain":["+--------------+-----------+-----------+-------+\n","| Chromosome   | Start     | End       | +3    |\n","| (category)   | (int64)   | (int64)   | ...   |\n","|--------------+-----------+-----------+-------|\n","| chr1         | 76871267  | 76871821  | ...   |\n","| chr1         | 61824444  | 61825501  | ...   |\n","| chr1         | 54634687  | 54639192  | ...   |\n","| chr1         | 245464258 | 245471621 | ...   |\n","| ...          | ...       | ...       | ...   |\n","| chrX         | 102721363 | 102724864 | ...   |\n","| chrX         | 98679426  | 98679978  | ...   |\n","| chrX         | 17835910  | 17837395  | ...   |\n","| chrX         | 108267641 | 108269152 | ...   |\n","+--------------+-----------+-----------+-------+\n","Stranded PyRanges object has 107 rows and 6 columns from 23 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","3 hidden columns: Name, Score, Strand"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["# subset le for cryptic LEs only\n","le_cryp = le.subset(lambda df: df.Name.str.contains(\"cryptic\", regex=False))\n","ids_le_cryp = set(le_cryp.Name)\n","le_cryp"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":["# extract introns (SJs) for each transcript\n","ref_introns = ref_gtf.features.introns(by=\"transcript\")\n","# store a df noting which transcript IDs share the same intron coordinates (useful if collapse later and want to track/retain)\n","intron2tx = ref_introns.as_df()[[\"Chromosome\", \"Start\", \"End\", \"Strand\", \"transcript_id\"]]\n","\n"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1623430</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1492</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1493</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1494</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1495</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","    <tr>\n","      <th>1496</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108221374</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1497 rows × 13 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +8    |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1623430   | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108221374 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,497 rows and 13 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","8 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 1 more.)"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["# first extract LEs completely contained within annotated introns\n","le_cryp_cont = le_cryp.overlap(ref_introns, strandedness=\"same\", how=\"containment\")\n","\n","# now overlap-join introns with LEs\n","ref_introns_le = ref_introns.join(le_cryp_cont, strandedness=\"same\")\n","ref_introns_le"]},{"cell_type":"code","execution_count":23,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1492</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1493</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1494</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1495</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1496</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1497 rows × 15 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +10   |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| chr1         | intron     | 1616614   | 1616614   | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,497 rows and 15 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":23,"metadata":{},"output_type":"execute_result"}],"source":["# If + strand\n","# Start coord = intron start (i.e. 'Start' column), End coord = exon start (i.e. 'Start_b' column)\n","# If - strand\n","# Start coord = Exon start (i.e. 'End_b' column), End coord = intron start (i.e. 'End' column)\n","\n","def _df_intron_exon_to_sj(df: pd.DataFrame) -> pd.DataFrame:\n","    '''Convert joined intron + overlapping exon df of PyRanges object to splice junction coordinates (the first bin between the two intervals)\n","\n","    If + strand\n","    Start coord = intron start (i.e. 'Start' column), End coord = exon start (i.e. 'Start_b' column)\n","    If - strand\n","    Start coord = Exon start (i.e. 'End_b' column), End coord = intron start (i.e. 'End' column)\n","    \n","    Parameters\n","    ----------\n","    df : pd.DataFrame\n","        _description_\n","\n","    Returns\n","    -------\n","    pd.DataFrame\n","        _description_\n","\n","    Raises\n","    ------\n","    Exception\n","        Strand column of df doesn't contain all '+' or '-'\n","    '''\n","\n","    df[[\"Start_intron\", \"End_intron\"]] = df[[\"Start\", \"End\"]]\n","\n","    if (df.Strand == \"+\").all():\n","        df[\"End\"] = df[\"Start_b\"]\n","\n","    elif (df.Strand == \"-\").all():\n","        df[\"Start\"] = df[\"End_b\"]\n","\n","    else:\n","        raise Exception(\"Strand column must only contain all '+' or '-'\")\n","    \n","    return df\n","\n","\n","le_sj = ref_introns_le.apply(_df_intron_exon_to_sj)\n","le_sj"]},{"cell_type":"code","execution_count":25,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","      <td>+</td>\n","      <td>ENSG00000197530.13</td>\n","      <td>ENST00000355826.10</td>\n","      <td>MIB2</td>\n","      <td>1616614</td>\n","      <td>1619210</td>\n","      <td>ENSG00000197530.13_1|MIB2|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>1616614</td>\n","      <td>1616614</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>20776596</td>\n","      <td>20776596</td>\n","      <td>-</td>\n","      <td>ENSG00000127483.19</td>\n","      <td>ENST00000312239.10</td>\n","      <td>HP1BP3</td>\n","      <td>20775449</td>\n","      <td>20776596</td>\n","      <td>ENSG00000127483.19_1|HP1BP3|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>20776596</td>\n","      <td>20776596</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    |     Start |       End | Strand       | +10   |\n","| (object)     | (object)   |   (int64) |   (int64) | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     |   1616614 |   1616614 | +            | ...   |\n","| chr1         | intron     |  20776596 |  20776596 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 2 rows and 15 columns from 1 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":25,"metadata":{},"output_type":"execute_result"}],"source":["le_sj.subset(lambda df: df.End - df.Start == 0).drop_duplicate_positions() # perhaps these two are annotated in some way..."]},{"cell_type":"code","execution_count":26,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Feature</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>gene_name</th>\n","      <th>Start_b</th>\n","      <th>End_b</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","      <th>Strand_b</th>\n","      <th>Start_intron</th>\n","      <th>End_intron</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","      <td>+</td>\n","      <td>ENSG00000162849.16</td>\n","      <td>ENST00000407071.7</td>\n","      <td>KIF26B</td>\n","      <td>245464258</td>\n","      <td>245471621</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22</td>\n","      <td>ENST00000459752.5</td>\n","      <td>PATJ</td>\n","      <td>61824444</td>\n","      <td>61825501</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>intron</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","      <td>+</td>\n","      <td>ENSG00000117069.15</td>\n","      <td>ENST00000477717.6</td>\n","      <td>ST6GALNAC5</td>\n","      <td>76871267</td>\n","      <td>76871821</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>+</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1358</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1359</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000538570.5</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1360</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1361</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","    <tr>\n","      <th>1362</th>\n","      <td>chrX</td>\n","      <td>intron</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17</td>\n","      <td>ENST00000621266.4</td>\n","      <td>COL4A6</td>\n","      <td>108267641</td>\n","      <td>108269152</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","      <td>-</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1363 rows × 15 columns</p>\n","</div>"],"text/plain":["+--------------+------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Feature    | Start     | End       | Strand       | +10   |\n","| (object)     | (object)   | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+------------+-----------+-----------+--------------+-------|\n","| chr1         | intron     | 245419745 | 245464258 | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| chr1         | intron     | 61823079  | 61824444  | +            | ...   |\n","| ...          | ...        | ...       | ...       | ...          | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","| chrX         | intron     | 108269152 | 108310747 | -            | ...   |\n","+--------------+------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,363 rows and 15 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","10 hidden columns: gene_id, transcript_id, gene_name, Start_b, End_b, Name, Score, ... (+ 3 more.)"]},"execution_count":26,"metadata":{},"output_type":"execute_result"}],"source":["# remove 0 length SJs for now as likely some annotation issue - check back at end after mined annotated SJs\n","le_sj = le_sj.subset(lambda df: df.Start != df.End)\n","le_sj"]},{"cell_type":"code","execution_count":28,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chromosome</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Name</th>\n","      <th>Score</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>245419745</td>\n","      <td>245464258</td>\n","      <td>+</td>\n","      <td>ENSG00000162849.16_2|KIF26B|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>61823079</td>\n","      <td>61824444</td>\n","      <td>+</td>\n","      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>76868742</td>\n","      <td>76871267</td>\n","      <td>+</td>\n","      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1358</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1359</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1360</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1361</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1362</th>\n","      <td>chrX</td>\n","      <td>108269152</td>\n","      <td>108310747</td>\n","      <td>-</td>\n","      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n","      <td>.</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1363 rows × 6 columns</p>\n","</div>"],"text/plain":["+--------------+-----------+-----------+--------------+-------+\n","| Chromosome   | Start     | End       | Strand       | +2    |\n","| (object)     | (int64)   | (int64)   | (category)   | ...   |\n","|--------------+-----------+-----------+--------------+-------|\n","| chr1         | 245419745 | 245464258 | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| chr1         | 61823079  | 61824444  | +            | ...   |\n","| ...          | ...       | ...       | ...          | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","| chrX         | 108269152 | 108310747 | -            | ...   |\n","+--------------+-----------+-----------+--------------+-------+\n","Stranded PyRanges object has 1,363 rows and 6 columns from 21 chromosomes.\n","For printing, the PyRanges was sorted on Chromosome and Strand.\n","2 hidden columns: Name, Score"]},"execution_count":28,"metadata":{},"output_type":"execute_result"}],"source":["# finally generate a cleaned BED-ready file of SJs\n","le_sj_clean = le_sj[[\"Name\"]]\n","le_sj_clean.Score = \".\"\n","le_sj_clean"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"pybioinfo","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.11"},"orig_nbformat":4},"nbformat":4,"nbformat_minor":2}
