{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from helpers import get_terminal_regions\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>cryptic_status</th>\n",
       "      <th>bed_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000138083.5_1|SIX3|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000138083.5_2|SIX3|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>USP31</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>USP31|annotated|ENST00000219689.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>USP31</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000103404.15_1|USP31|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ACER3|annotated|ENST00000532485.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000078124.13_1|ACER3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>NBPF3|annotated|ENST00000318249.10;ENST0000034...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DLGAP1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>DLGAP1|annotated|ENST00000315677.8;ENST0000040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DLGAP1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000170579.17_1|DLGAP1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000169856.9_3|ONECUT1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000169856.9_1|ONECUT1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ETV5</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ETV5|annotated|ENST00000306376.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ETV5</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000244405.8_3|ETV5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>KIF26B</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>KIF26B|annotated|ENST00000407071.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>KIF26B</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000162849.16_2|KIF26B|spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>STMN2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>STMN2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PHF2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PHF2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name event_type cryptic_status  \\\n",
       "0       ELK1      3'Ext      annotated   \n",
       "1       ELK1      3'Ext        cryptic   \n",
       "2       SIX3      3'Ext      annotated   \n",
       "3       SIX3      3'Ext        cryptic   \n",
       "4       TLX1      3'Ext      annotated   \n",
       "5       TLX1      3'Ext        cryptic   \n",
       "6    ANKRD27        IPA      annotated   \n",
       "7    ANKRD27        IPA        cryptic   \n",
       "8      SIN3B        IPA      annotated   \n",
       "9      SIN3B        IPA        cryptic   \n",
       "10     CNPY3        IPA      annotated   \n",
       "11     CNPY3        IPA        cryptic   \n",
       "12     USP31        IPA      annotated   \n",
       "13     USP31        IPA        cryptic   \n",
       "14     ACER3        IPA      annotated   \n",
       "15     ACER3        IPA        cryptic   \n",
       "16     NBPF3        IPA      annotated   \n",
       "17     NBPF3        IPA        cryptic   \n",
       "18    DLGAP1        ALE      annotated   \n",
       "19    DLGAP1        ALE        cryptic   \n",
       "20   ONECUT1        ALE      annotated   \n",
       "21   ONECUT1        ALE        cryptic   \n",
       "22      ETV5        ALE      annotated   \n",
       "23      ETV5        ALE        cryptic   \n",
       "24    KIF26B        ALE      annotated   \n",
       "25    KIF26B        ALE        cryptic   \n",
       "26     SYNJ2        ALE      annotated   \n",
       "27     SYNJ2        ALE        cryptic   \n",
       "28     STMN2        ALE      annotated   \n",
       "29     STMN2        ALE        cryptic   \n",
       "30  ARHGAP32        ALE      annotated   \n",
       "31  ARHGAP32        ALE        cryptic   \n",
       "32      PHF2        ALE      annotated   \n",
       "33      PHF2        ALE        cryptic   \n",
       "\n",
       "                                             bed_name  \n",
       "0          ENSG00000126767.18_1|ELK1|proximal|cryptic  \n",
       "1            ENSG00000126767.18_2|ELK1|distal|cryptic  \n",
       "2           ENSG00000138083.5_1|SIX3|proximal|cryptic  \n",
       "3             ENSG00000138083.5_2|SIX3|distal|cryptic  \n",
       "4                                  ENST00000370196.11  \n",
       "5            ENSG00000107807.13_2|TLX1|distal|cryptic  \n",
       "6                 ANKRD27|annotated|ENST00000306065.9  \n",
       "7   ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic  \n",
       "8                   SIN3B|annotated|ENST00000379803.5  \n",
       "9     ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic  \n",
       "10                  CNPY3|annotated|ENST00000372836.5  \n",
       "11    ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic  \n",
       "12                 USP31|annotated|ENST00000219689.12  \n",
       "13    ENSG00000103404.15_1|USP31|bleedthrough|cryptic  \n",
       "14                  ACER3|annotated|ENST00000532485.6  \n",
       "15    ENSG00000078124.13_1|ACER3|bleedthrough|cryptic  \n",
       "16  NBPF3|annotated|ENST00000318249.10;ENST0000034...  \n",
       "17    ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic  \n",
       "18  DLGAP1|annotated|ENST00000315677.8;ENST0000040...  \n",
       "19        ENSG00000170579.17_1|DLGAP1|spliced|cryptic  \n",
       "20         ENSG00000169856.9_3|ONECUT1|distal|cryptic  \n",
       "21        ENSG00000169856.9_1|ONECUT1|spliced|cryptic  \n",
       "22                  ETV5|annotated|ENST00000306376.10  \n",
       "23           ENSG00000244405.8_3|ETV5|spliced|cryptic  \n",
       "24                 KIF26B|annotated|ENST00000407071.7  \n",
       "25                ENSG00000162849.16_2|KIF26B|spliced  \n",
       "26  SYNJ2|annotated|ENST00000355585.9;ENST00000638...  \n",
       "27         ENSG00000078269.15_1|SYNJ2|spliced|cryptic  \n",
       "28                 STMN2|annotated|ENST00000220876.12  \n",
       "29         ENSG00000104435.14_1|STMN2|spliced|cryptic  \n",
       "30  ARHGAP32|annotated|ENST00000310343.13;ENST0000...  \n",
       "31      ENSG00000134909.19_1|ARHGAP32|spliced|cryptic  \n",
       "32  PHF2|annotated|ENST00000359246.9;ENST000003753...  \n",
       "33          ENSG00000197724.11_1|PHF2|spliced|cryptic  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df = pd.read_csv(\"data/regions/seddighi_i3_cryptic_annot_pairs.csv\", index_col=False)\n",
    "pairs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6220805</td>\n",
       "      <td>ENSG00000158286.13_3|RNF207|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6221220</td>\n",
       "      <td>ENSG00000158286.13_4|RNF207|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11296049</td>\n",
       "      <td>ENSG00000120942.14_2|UBIAD1|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11299544</td>\n",
       "      <td>ENSG00000120942.14_3|UBIAD1|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149923199</td>\n",
       "      <td>149930081</td>\n",
       "      <td>ENSG00000197021.9_4|EOLA2|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1764</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154505228</td>\n",
       "      <td>154506906</td>\n",
       "      <td>ENSG00000071889.17_3|FAM3A|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154506158</td>\n",
       "      <td>154506906</td>\n",
       "      <td>ENSG00000071889.17_2|FAM3A|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1766</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155491666</td>\n",
       "      <td>ENSG00000185973.12_5|TMLHE|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1767</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489010</td>\n",
       "      <td>155491666</td>\n",
       "      <td>ENSG00000185973.12_4|TMLHE|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1768 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 6219235   | 6220805   | ...   |\n",
       "| chr1         | 6219235   | 6221220   | ...   |\n",
       "| chr1         | 11294872  | 11296049  | ...   |\n",
       "| chr1         | 11294872  | 11299544  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 154505228 | 154506906 | ...   |\n",
       "| chrX         | 154506158 | 154506906 | ...   |\n",
       "| chrX         | 155478419 | 155491666 | ...   |\n",
       "| chrX         | 155489010 | 155491666 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,768 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d3utr = pr.read_bed(\"data/regions/2023-12-15_papa_cryptic_d3utr.background_shsy5y.last_exons.bed\")\n",
    "d3utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"data/regions/gencode.v40.annotation.TLX1.gtf\"):\n",
    "    os.system(\"grep 'TLX1' data/regions/gencode.v40.annotation.gtf > data/regions/gencode.v40.annotation.TLX1.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131299</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001452047.7</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000721324.1</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001452043.6</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131303</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001868083.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000721324.1</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136667</td>\n",
       "      <td>101137497</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001899820.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131727</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002478430.2</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134138</td>\n",
       "      <td>101134225</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001907822.1</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101132781</td>\n",
       "      <td>101133074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002180022.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134215</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002154572.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134313</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002145719.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101135353</td>\n",
       "      <td>101135516</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170599.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101130785</td>\n",
       "      <td>101131126</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001779824.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101126959</td>\n",
       "      <td>101127098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001733617.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101089320</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001660194.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101130756</td>\n",
       "      <td>101131126</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001778812.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101126959</td>\n",
       "      <td>101127098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001733617.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101090307</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001690224.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +19   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | HAVANA     | exon         | 101131299 | 101132109 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134174 | 101134376 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101131303 | 101132109 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101089320 | 101090702 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101130756 | 101131126 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101126959 | 101127098 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101090307 | 101090702 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 18 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract exons of TLX1 transcripts from GTF\n",
    "tlx1_gtf = pr.read_gtf(\"data/regions/gencode.v40.annotation.TLX1.gtf\")\n",
    "tlx1_gtf =tlx1_gtf.subset(lambda df: df.Feature == \"exon\")\n",
    "tlx1_gtf.exon_number = tlx1_gtf.exon_number.astype(int)\n",
    "tlx1_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001452043.6</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134138</td>\n",
       "      <td>101134225</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001907822.1</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136667</td>\n",
       "      <td>101137497</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001899820.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101135353</td>\n",
       "      <td>101135516</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170599.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134215</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002154572.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101090307</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001690224.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101089320</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001660194.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +19   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | HAVANA     | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134138 | 101134225 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101136667 | 101137497 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101135353 | 101135516 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134174 | 101134215 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101090307 | 101090702 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101089320 | 101090702 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 7 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get last exons for TLX1 transripts\n",
    "tlx1_les = get_terminal_regions(tlx1_gtf)\n",
    "tlx1_les"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>cryptic_status</th>\n",
       "      <th>bed_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000138083.5_1|SIX3|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000138083.5_2|SIX3|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>USP31</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>USP31|annotated|ENST00000219689.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>USP31</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000103404.15_1|USP31|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ACER3|annotated|ENST00000532485.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000078124.13_1|ACER3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>NBPF3|annotated|ENST00000318249.10;ENST0000034...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NBPF3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DLGAP1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>DLGAP1|annotated|ENST00000315677.8;ENST0000040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DLGAP1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000170579.17_1|DLGAP1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000169856.9_3|ONECUT1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000169856.9_1|ONECUT1|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ETV5</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ETV5|annotated|ENST00000306376.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ETV5</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000244405.8_3|ETV5|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>KIF26B</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>KIF26B|annotated|ENST00000407071.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>KIF26B</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000162849.16_2|KIF26B|spliced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>STMN2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>STMN2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PHF2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>annotated</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PHF2</td>\n",
       "      <td>ALE</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name event_type cryptic_status  \\\n",
       "0       ELK1      3'Ext      annotated   \n",
       "1       ELK1      3'Ext        cryptic   \n",
       "2       SIX3      3'Ext      annotated   \n",
       "3       SIX3      3'Ext        cryptic   \n",
       "4       TLX1      3'Ext      annotated   \n",
       "5       TLX1      3'Ext        cryptic   \n",
       "6    ANKRD27        IPA      annotated   \n",
       "7    ANKRD27        IPA        cryptic   \n",
       "8      SIN3B        IPA      annotated   \n",
       "9      SIN3B        IPA        cryptic   \n",
       "10     CNPY3        IPA      annotated   \n",
       "11     CNPY3        IPA        cryptic   \n",
       "12     USP31        IPA      annotated   \n",
       "13     USP31        IPA        cryptic   \n",
       "14     ACER3        IPA      annotated   \n",
       "15     ACER3        IPA        cryptic   \n",
       "16     NBPF3        IPA      annotated   \n",
       "17     NBPF3        IPA        cryptic   \n",
       "18    DLGAP1        ALE      annotated   \n",
       "19    DLGAP1        ALE        cryptic   \n",
       "20   ONECUT1        ALE      annotated   \n",
       "21   ONECUT1        ALE        cryptic   \n",
       "22      ETV5        ALE      annotated   \n",
       "23      ETV5        ALE        cryptic   \n",
       "24    KIF26B        ALE      annotated   \n",
       "25    KIF26B        ALE        cryptic   \n",
       "26     SYNJ2        ALE      annotated   \n",
       "27     SYNJ2        ALE        cryptic   \n",
       "28     STMN2        ALE      annotated   \n",
       "29     STMN2        ALE        cryptic   \n",
       "30  ARHGAP32        ALE      annotated   \n",
       "31  ARHGAP32        ALE        cryptic   \n",
       "32      PHF2        ALE      annotated   \n",
       "33      PHF2        ALE        cryptic   \n",
       "\n",
       "                                             bed_name  \n",
       "0          ENSG00000126767.18_1|ELK1|proximal|cryptic  \n",
       "1            ENSG00000126767.18_2|ELK1|distal|cryptic  \n",
       "2           ENSG00000138083.5_1|SIX3|proximal|cryptic  \n",
       "3             ENSG00000138083.5_2|SIX3|distal|cryptic  \n",
       "4                                  ENST00000370196.11  \n",
       "5            ENSG00000107807.13_2|TLX1|distal|cryptic  \n",
       "6                 ANKRD27|annotated|ENST00000306065.9  \n",
       "7   ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic  \n",
       "8                   SIN3B|annotated|ENST00000379803.5  \n",
       "9     ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic  \n",
       "10                  CNPY3|annotated|ENST00000372836.5  \n",
       "11    ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic  \n",
       "12                 USP31|annotated|ENST00000219689.12  \n",
       "13    ENSG00000103404.15_1|USP31|bleedthrough|cryptic  \n",
       "14                  ACER3|annotated|ENST00000532485.6  \n",
       "15    ENSG00000078124.13_1|ACER3|bleedthrough|cryptic  \n",
       "16  NBPF3|annotated|ENST00000318249.10;ENST0000034...  \n",
       "17    ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic  \n",
       "18  DLGAP1|annotated|ENST00000315677.8;ENST0000040...  \n",
       "19        ENSG00000170579.17_1|DLGAP1|spliced|cryptic  \n",
       "20         ENSG00000169856.9_3|ONECUT1|distal|cryptic  \n",
       "21        ENSG00000169856.9_1|ONECUT1|spliced|cryptic  \n",
       "22                  ETV5|annotated|ENST00000306376.10  \n",
       "23           ENSG00000244405.8_3|ETV5|spliced|cryptic  \n",
       "24                 KIF26B|annotated|ENST00000407071.7  \n",
       "25                ENSG00000162849.16_2|KIF26B|spliced  \n",
       "26  SYNJ2|annotated|ENST00000355585.9;ENST00000638...  \n",
       "27         ENSG00000078269.15_1|SYNJ2|spliced|cryptic  \n",
       "28                 STMN2|annotated|ENST00000220876.12  \n",
       "29         ENSG00000104435.14_1|STMN2|spliced|cryptic  \n",
       "30  ARHGAP32|annotated|ENST00000310343.13;ENST0000...  \n",
       "31      ENSG00000134909.19_1|ARHGAP32|spliced|cryptic  \n",
       "32  PHF2|annotated|ENST00000359246.9;ENST000003753...  \n",
       "33          ENSG00000197724.11_1|PHF2|spliced|cryptic  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>TLX1|annotated|ENST00000370196.11</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr10        | 101136690 | 101137789 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Strand, Score"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create 3'UTR extension BED\n",
    "# first - make sure select TLX1 event from annotation, as proximal PAS was mis-annotated\n",
    "\n",
    "# for TLX1, extract 'ENST00000370196.11' from refernce GTF (proximal is currently mis-annotated)\n",
    "# get tx id\n",
    "tlx1_prox_tx = pairs_df.loc[(pairs_df.gene_name == \"TLX1\") & ~(pairs_df.bed_name.str.contains(\"distal\", regex=False)), \"bed_name\"].iloc[0]\n",
    "\n",
    "# extract last exon for selected transcript, \n",
    "tlx1_annot_le = (tlx1_les.subset(lambda df: df.transcript_id == tlx1_prox_tx)                            \n",
    " .assign(\"annot_status\", lambda df: pd.Series([\"annotated\"]*len(df), index=df.index))\n",
    " .assign(\"Score\", lambda df: pd.Series([\".\"]*len(df), index=df.index))\n",
    " .assign(\"Name\", lambda df: df.gene_name.str.cat(df[[\"annot_status\", \"transcript_id\"]].astype(str), sep=\"|\"))\n",
    " [[\"Name\", \"Score\"]]\n",
    ".sort()\n",
    ")\n",
    "\n",
    "tlx1_annot_le\n",
    "#tlx1_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>44944567</td>\n",
       "      <td>44946071</td>\n",
       "      <td>ENSG00000138083.5_1|SIX3|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>44944567</td>\n",
       "      <td>44948085</td>\n",
       "      <td>ENSG00000138083.5_2|SIX3|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>TLX1|annotated|ENST00000370196.11</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47635520</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr2         |  44944567 |  44946071 | ...   |\n",
       "| chr2         |  44944567 |  44948085 | ...   |\n",
       "| chr10        | 101136690 | 101137789 | ...   |\n",
       "| chr10        | 101136690 | 101138531 | ...   |\n",
       "| chrX         |  47631618 |  47636927 | ...   |\n",
       "| chrX         |  47635520 |  47636927 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 6 columns from 3 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#1. Select proximal and distal pairs for SIX3 and ELK1\n",
    "d3utr_region_nms = pairs_df.loc[pairs_df.gene_name.isin([\"ELK1\", \"SIX3\", \"TLX1\"]), \"bed_name\"].to_list()\n",
    "d3utr_sel = d3utr.subset(lambda df: df.Name.isin(d3utr_region_nms))\n",
    "d3utr_sel = pr.concat([d3utr_sel, tlx1_annot_le]).sort()\n",
    "d3utr_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "d3utr_sel.to_bed(\"processed/validation_region_beds/2024-04-08_cryptic_and_annotated.selected.3ext.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935670</td>\n",
       "      <td>42936378</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42938567</td>\n",
       "      <td>42939294</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880353</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32597005</td>\n",
       "      <td>32598378</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         |  42935670 |  42936378 | ...   |\n",
       "| chr6         |  42938567 |  42939294 | ...   |\n",
       "| chr19        |  16854261 |  16857420 | ...   |\n",
       "| chr19        |  16878496 |  16880353 | ...   |\n",
       "| chr19        |  32597005 |  32598378 | ...   |\n",
       "| chr19        |  32635458 |  32639355 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 6 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## IPA events\n",
    "ipa = pr.read_bed(\"processed/validation_region_beds/2024-04-05_cryptic_and_annotated.ipa.bed\")\n",
    "ipa_region_gns = pairs_df.loc[pairs_df.gene_name.isin([\"SIN3B\", \"ANKRD27\", \"CNPY3\"]), \"bed_name\"].to_list()\n",
    "ipa_sel = ipa.subset(lambda df: df.Name.isin(ipa_region_gns))\n",
    "ipa_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935670</td>\n",
       "      <td>42936378</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42938567</td>\n",
       "      <td>42939294</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857610</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880353</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32597005</td>\n",
       "      <td>32598378</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         |  42935670 |  42936378 | ...   |\n",
       "| chr6         |  42938567 |  42939294 | ...   |\n",
       "| chr19        |  16854261 |  16857610 | ...   |\n",
       "| chr19        |  16878496 |  16880353 | ...   |\n",
       "| chr19        |  32597005 |  32598378 | ...   |\n",
       "| chr19        |  32635458 |  32639355 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 6 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "# IPA - Update SIN3B's PAS - slight extension of ~ 200 bp (by coverage and jnc read support)\n",
    "# name: chr19:+:16857609:16857610 location: chr19:16857607-16857612 (+) score: 18.0\n",
    "# name is in BED format, so just update end of SIN3B cryptic to 16857610\n",
    "ipa_sel = ipa_sel.assign(\"End\",\n",
    "               lambda df: pd.Series(np.where(df[\"Name\"] == \"ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic\", 16857610, df[\"End\"])))\n",
    "ipa_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ipa_sel.sort().to_bed(\"processed/validation_region_beds/2024-05-13_cryptic_and_annotated.selected.ipa.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158019983</td>\n",
       "      <td>158024104</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158095617</td>\n",
       "      <td>158099176</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93662001</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93677587</td>\n",
       "      <td>93679587</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128965059</td>\n",
       "      <td>128971159</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991449</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         | 158019983 | 158024104 | ...   |\n",
       "| chr6         | 158095617 | 158099176 | ...   |\n",
       "| chr8         |  79616821 |  79617071 | ...   |\n",
       "| chr8         |  79664814 |  79666158 | ...   |\n",
       "| chr9         |  93661347 |  93662001 | ...   |\n",
       "| chr9         |  93677587 |  93679587 | ...   |\n",
       "| chr11        | 128965059 | 128971159 | ...   |\n",
       "| chr11        | 128991449 | 128992046 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 6 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ALEs\n",
    "ale = pr.read_bed(\"processed/validation_region_beds/2024-04-05_cryptic_and_annotated.ales.bed\")\n",
    "ale_region_gns = pairs_df.loc[pairs_df.gene_name.isin([\"STMN2\", \"ARHGAP32\", \"PHF2\", \"SYNJ2\"]), \"bed_name\"].to_list()\n",
    "ale_sel = ale.subset(lambda df: df.Name.isin(ale_region_gns))\n",
    "ale_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158019983</td>\n",
       "      <td>158024104</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158095617</td>\n",
       "      <td>158099176</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93662001</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93677587</td>\n",
       "      <td>93679587</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128965059</td>\n",
       "      <td>128971159</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991449</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128988274</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1;11:128988275:-|ARHGAP32|s...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         | 158019983 | 158024104 | ...   |\n",
       "| chr6         | 158095617 | 158099176 | ...   |\n",
       "| chr8         | 79616821  | 79617071  | ...   |\n",
       "| chr8         | 79664814  | 79666158  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr9         | 93661347  | 93662001  | ...   |\n",
       "| chr9         | 93677587  | 93679587  | ...   |\n",
       "| chr11        | 128991449 | 128992046 | ...   |\n",
       "| chr11        | 128988274 | 128992046 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 9 rows and 6 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ARHGAP32\n",
    "# ARHGAP32 has two PAS, one major (in current BED + supported by 99 junction reads (name: chr11:-:128991449:128991450,  location: chr11:128991448-128991461 (-), # score: 99.0)\n",
    "# and 1 more distal, minor PAS found in i3Ns which closely aligns with PolyASite site\n",
    "# name: 11:128988275:- , location: chr11:128988275-128988275 (-) score: 0.0011\n",
    "# \n",
    "# Need to add minor PAS\n",
    "arhgap32_extra = ale_sel[ale_sel.Name == \"ENSG00000134909.19_1|ARHGAP32|spliced|cryptic\"]\n",
    "arhgap32_extra.Start = 128988275 -1 # PolyASite uses UCSC position format, so subtract 1 to get BED-format coord\n",
    "arhgap32_extra.Name = \"ENSG00000134909.19_1;11:128988275:-|ARHGAP32|spliced|cryptic\"\n",
    "arhgap32_extra\n",
    "\n",
    "ale_sel = pr.concat([ale_sel, arhgap32_extra])\n",
    "ale_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158019983</td>\n",
       "      <td>158024104</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158095617</td>\n",
       "      <td>158099176</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617207</td>\n",
       "      <td>ENSG00000104435.14_1;chr8:+:79617206:79617207|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93662001</td>\n",
       "      <td>ENSG00000197724.11_1|PHF2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93677587</td>\n",
       "      <td>93679587</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128965059</td>\n",
       "      <td>128971159</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991449</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128988274</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1;11:128988275:-|ARHGAP32|s...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         | 158019983 | 158024104 | ...   |\n",
       "| chr6         | 158095617 | 158099176 | ...   |\n",
       "| chr8         | 79616821  | 79617071  | ...   |\n",
       "| chr8         | 79664814  | 79666158  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr9         | 93661347  | 93662001  | ...   |\n",
       "| chr9         | 93677587  | 93679587  | ...   |\n",
       "| chr11        | 128991449 | 128992046 | ...   |\n",
       "| chr11        | 128988274 | 128992046 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 10 rows and 6 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STMN2\n",
    "# 1. Current representative event in BED (+ assembled) - name: chr8:+:79617076:79617077 location: chr8:79617065-79617080 (+) score: 224.0\n",
    "# 2. Additional slightly lower supported PAS approx 130bp downstream - name: chr8:+:79617206:79617207 location: chr8:79617199-79617213 (+) score: 143.0\n",
    "stmn2_extra = ale_sel[ale_sel.Name == \"ENSG00000104435.14_1|STMN2|spliced|cryptic\"]\n",
    "stmn2_extra.End = 79617207\n",
    "stmn2_extra.Name = \"ENSG00000104435.14_1;chr8:+:79617206:79617207|STMN2|spliced|cryptic\"\n",
    "ale_sel = pr.concat([ale_sel, stmn2_extra])\n",
    "ale_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93661959</td>\n",
       "      <td>ENSG00000197724.11_1;chr9:+:93661958:93661959|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93661876</td>\n",
       "      <td>ENSG00000197724.11_1;chr9:+:93661875:93661876|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr9         |  93661347 |  93661959 | ...   |\n",
       "| chr9         |  93661347 |  93661876 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PHF2\n",
    "# 1. 1 site supported by junction reads ~ 33bp upstream - name: chr9:+:93661958:93661959 location: chr9:93661957-93661967 (+) score: 5.0\n",
    "# 2. 1 junction read site ~ 150 bp upstream - name: chr9:+:93661875:93661876 location: chr9:93661865-93661876 (+) score: 9.0\n",
    "phf2_extra1 = ale_sel[ale_sel.Name == \"ENSG00000197724.11_1|PHF2|spliced|cryptic\"]\n",
    "phf2_extra2 = phf2_extra1.copy()\n",
    "phf2_extra1.End = 93661959\n",
    "phf2_extra1.Name = \"ENSG00000197724.11_1;chr9:+:93661958:93661959|PHF2|spliced|cryptic\"\n",
    "phf2_extra2.End = 93661876\n",
    "phf2_extra2.Name = \"ENSG00000197724.11_1;chr9:+:93661875:93661876|PHF2|spliced|cryptic\"\n",
    "phf2_extras = pr.concat([phf2_extra1, phf2_extra2])\n",
    "phf2_extras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158019983</td>\n",
       "      <td>158024104</td>\n",
       "      <td>ENSG00000078269.15_1|SYNJ2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158095617</td>\n",
       "      <td>158099176</td>\n",
       "      <td>SYNJ2|annotated|ENST00000355585.9;ENST00000638...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617071</td>\n",
       "      <td>ENSG00000104435.14_1|STMN2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617207</td>\n",
       "      <td>ENSG00000104435.14_1;chr8:+:79617206:79617207|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>STMN2|annotated|ENST00000220876.12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93661876</td>\n",
       "      <td>ENSG00000197724.11_1;chr9:+:93661875:93661876|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93661347</td>\n",
       "      <td>93661959</td>\n",
       "      <td>ENSG00000197724.11_1;chr9:+:93661958:93661959|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr9</td>\n",
       "      <td>93677587</td>\n",
       "      <td>93679587</td>\n",
       "      <td>PHF2|annotated|ENST00000359246.9;ENST000003753...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128965059</td>\n",
       "      <td>128971159</td>\n",
       "      <td>ARHGAP32|annotated|ENST00000310343.13;ENST0000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128988274</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1;11:128988275:-|ARHGAP32|s...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr11</td>\n",
       "      <td>128991449</td>\n",
       "      <td>128992046</td>\n",
       "      <td>ENSG00000134909.19_1|ARHGAP32|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         | 158019983 | 158024104 | ...   |\n",
       "| chr6         | 158095617 | 158099176 | ...   |\n",
       "| chr8         | 79616821  | 79617071  | ...   |\n",
       "| chr8         | 79616821  | 79617207  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr9         | 93661347  | 93661876  | ...   |\n",
       "| chr9         | 93661347  | 93661959  | ...   |\n",
       "| chr9         | 93677587  | 93679587  | ...   |\n",
       "| chr11        | 128991449 | 128992046 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 11 rows and 6 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ale_sel = pr.concat([ale_sel, phf2_extras]).subset(lambda df: ~(df.Name == \"ENSG00000197724.11_1|PHF2|spliced|cryptic\")).sort()\n",
    "ale_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "ale_sel.to_bed(\"processed/validation_region_beds/2024-05-13_cryptic_and_annotated.selected.ale.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
