{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from helpers import get_terminal_regions\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>cryptic_status</th>\n",
       "      <th>bed_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENSG00000138083.5_1|SIX3|proximal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIX3</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000138083.5_2|SIX3|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TLX1</td>\n",
       "      <td>3'Ext</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ANKRD27</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SIN3B</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>annotated</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>IPA</td>\n",
       "      <td>cryptic</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name event_type cryptic_status  \\\n",
       "0       ELK1      3'Ext      annotated   \n",
       "1       ELK1      3'Ext        cryptic   \n",
       "2       SIX3      3'Ext      annotated   \n",
       "3       SIX3      3'Ext        cryptic   \n",
       "4       TLX1      3'Ext      annotated   \n",
       "5       TLX1      3'Ext        cryptic   \n",
       "6    ANKRD27        IPA      annotated   \n",
       "7    ANKRD27        IPA        cryptic   \n",
       "8      SIN3B        IPA      annotated   \n",
       "9      SIN3B        IPA        cryptic   \n",
       "10     CNPY3        IPA      annotated   \n",
       "11     CNPY3        IPA        cryptic   \n",
       "\n",
       "                                             bed_name  \n",
       "0          ENSG00000126767.18_1|ELK1|proximal|cryptic  \n",
       "1            ENSG00000126767.18_2|ELK1|distal|cryptic  \n",
       "2           ENSG00000138083.5_1|SIX3|proximal|cryptic  \n",
       "3             ENSG00000138083.5_2|SIX3|distal|cryptic  \n",
       "4                                  ENST00000370196.11  \n",
       "5            ENSG00000107807.13_2|TLX1|distal|cryptic  \n",
       "6                 ANKRD27|annotated|ENST00000306065.9  \n",
       "7   ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic  \n",
       "8                   SIN3B|annotated|ENST00000379803.5  \n",
       "9     ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic  \n",
       "10                  CNPY3|annotated|ENST00000372836.5  \n",
       "11    ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df = pd.read_csv(\"data/regions/seddighi_i3_cryptic_annot_pairs.csv\", index_col=False)\n",
    "pairs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6220805</td>\n",
       "      <td>ENSG00000158286.13_3|RNF207|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6219235</td>\n",
       "      <td>6221220</td>\n",
       "      <td>ENSG00000158286.13_4|RNF207|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11296049</td>\n",
       "      <td>ENSG00000120942.14_2|UBIAD1|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11294872</td>\n",
       "      <td>11299544</td>\n",
       "      <td>ENSG00000120942.14_3|UBIAD1|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149923199</td>\n",
       "      <td>149930081</td>\n",
       "      <td>ENSG00000197021.9_4|EOLA2|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1764</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154505228</td>\n",
       "      <td>154506906</td>\n",
       "      <td>ENSG00000071889.17_3|FAM3A|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154506158</td>\n",
       "      <td>154506906</td>\n",
       "      <td>ENSG00000071889.17_2|FAM3A|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1766</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155491666</td>\n",
       "      <td>ENSG00000185973.12_5|TMLHE|distal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1767</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489010</td>\n",
       "      <td>155491666</td>\n",
       "      <td>ENSG00000185973.12_4|TMLHE|proximal|background</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1768 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 6219235   | 6220805   | ...   |\n",
       "| chr1         | 6219235   | 6221220   | ...   |\n",
       "| chr1         | 11294872  | 11296049  | ...   |\n",
       "| chr1         | 11294872  | 11299544  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 154505228 | 154506906 | ...   |\n",
       "| chrX         | 154506158 | 154506906 | ...   |\n",
       "| chrX         | 155478419 | 155491666 | ...   |\n",
       "| chrX         | 155489010 | 155491666 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,768 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d3utr = pr.read_bed(\"data/regions/2023-12-15_papa_cryptic_d3utr.background_shsy5y.last_exons.bed\")\n",
    "d3utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not os.path.exists(\"data/regions/gencode.v40.annotation.TLX1.gtf\"):\n",
    "    os.system(\"grep 'TLX1' data/regions/gencode.v40.annotation.gtf > data/regions/gencode.v40.annotation.TLX1.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/IPython/core/completer.py\", line 3168, in _complete\n",
      "    matcher(text), type=_UNKNOWN_TYPE\n",
      "  File \"/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/IPython/core/completer.py\", line 2339, in python_matches\n",
      "    matches = self.global_matches(text)\n",
      "  File \"/home/sam/mambaforge-pypy3/envs/pybioinfo/lib/python3.10/site-packages/IPython/core/completer.py\", line 1113, in global_matches\n",
      "    for word in lst:\n",
      "RuntimeError: dictionary changed size during iteration\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131299</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001452047.7</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000721324.1</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001452043.6</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131303</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001868083.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000721324.1</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136667</td>\n",
       "      <td>101137497</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001899820.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101131727</td>\n",
       "      <td>101132109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002478430.2</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134138</td>\n",
       "      <td>101134225</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001907822.1</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101132781</td>\n",
       "      <td>101133074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002180022.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134215</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002154572.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134313</td>\n",
       "      <td>101134376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002145719.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101135353</td>\n",
       "      <td>101135516</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170599.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101130785</td>\n",
       "      <td>101131126</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001779824.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101126959</td>\n",
       "      <td>101127098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001733617.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101089320</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001660194.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101130756</td>\n",
       "      <td>101131126</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001778812.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101126959</td>\n",
       "      <td>101127098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001733617.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101090307</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001690224.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +19   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | HAVANA     | exon         | 101131299 | 101132109 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134174 | 101134376 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101131303 | 101132109 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101089320 | 101090702 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101130756 | 101131126 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101126959 | 101127098 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101090307 | 101090702 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 18 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract exons of TLX1 transcripts from GTF\n",
    "tlx1_gtf = pr.read_gtf(\"data/regions/gencode.v40.annotation.TLX1.gtf\")\n",
    "tlx1_gtf =tlx1_gtf.subset(lambda df: df.Feature == \"exon\")\n",
    "tlx1_gtf.exon_number = tlx1_gtf.exon_number.astype(int)\n",
    "tlx1_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000370196.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000051193.6</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001452043.6</td>\n",
       "      <td>ENSP00000359215.6</td>\n",
       "      <td>CCDS7510.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134138</td>\n",
       "      <td>101134225</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000463716.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000051194.5</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001907822.1</td>\n",
       "      <td>ENSP00000434358.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136667</td>\n",
       "      <td>101137497</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000467928.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TLX1-203</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000409763.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001899820.2</td>\n",
       "      <td>ENSP00000434914.2</td>\n",
       "      <td>CCDS55725.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101135353</td>\n",
       "      <td>101135516</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000525019.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-204</td>\n",
       "      <td>5</td>\n",
       "      <td>low_sequence_quality</td>\n",
       "      <td>OTTHUMT00000394897.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170599.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101134215</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000533319.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>TLX1-205</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000394896.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002154572.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101090307</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000425505.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000049926.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001690224.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>101089320</td>\n",
       "      <td>101090702</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000236311.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000445873.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TLX1NB-202</td>\n",
       "      <td>1</td>\n",
       "      <td>exp_conf</td>\n",
       "      <td>OTTHUMT00000049925.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001660194.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +19   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | HAVANA     | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134138 | 101134225 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101136667 | 101137497 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101135353 | 101135516 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101134174 | 101134215 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101090307 | 101090702 | ...   |\n",
       "| chr10        | HAVANA     | exon         | 101089320 | 101090702 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 7 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 12 more.)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get last exons for TLX1 transripts\n",
    "tlx1_les = get_terminal_regions(tlx1_gtf)\n",
    "tlx1_les"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>TLX1|annotated|ENST00000370196.11</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr10        | 101136690 | 101137789 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Strand, Score"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create 3'UTR extension BED\n",
    "# first - make sure select TLX1 event from annotation, as proximal PAS was mis-annotated\n",
    "\n",
    "# for TLX1, extract 'ENST00000370196.11' from refernce GTF (proximal is currently mis-annotated)\n",
    "# get tx id\n",
    "tlx1_prox_tx = pairs_df.loc[(pairs_df.gene_name == \"TLX1\") & ~(pairs_df.bed_name.str.contains(\"distal\")), \"bed_name\"].iloc[0]\n",
    "\n",
    "# extract last exon for selected transcript, \n",
    "tlx1_annot_le = (tlx1_les.subset(lambda df: df.transcript_id == tlx1_prox_tx)                            \n",
    " .assign(\"annot_status\", lambda df: pd.Series([\"annotated\"]*len(df), index=df.index))\n",
    " .assign(\"Score\", lambda df: pd.Series([\".\"]*len(df), index=df.index))\n",
    " .assign(\"Name\", lambda df: df.gene_name.str.cat(df[[\"annot_status\", \"transcript_id\"]].astype(str), sep=\"|\"))\n",
    " [[\"Name\", \"Score\"]]\n",
    ".sort()\n",
    ")\n",
    "\n",
    "tlx1_annot_le\n",
    "#tlx1_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>44944567</td>\n",
       "      <td>44946071</td>\n",
       "      <td>ENSG00000138083.5_1|SIX3|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>44944567</td>\n",
       "      <td>44948085</td>\n",
       "      <td>ENSG00000138083.5_2|SIX3|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>TLX1|annotated|ENST00000370196.11</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47635520</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr2         |  44944567 |  44946071 | ...   |\n",
       "| chr2         |  44944567 |  44948085 | ...   |\n",
       "| chr10        | 101136690 | 101137789 | ...   |\n",
       "| chr10        | 101136690 | 101138531 | ...   |\n",
       "| chrX         |  47631618 |  47636927 | ...   |\n",
       "| chrX         |  47635520 |  47636927 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 6 columns from 3 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#1. Select proximal and distal pairs for SIX3 and ELK1\n",
    "d3utr_region_nms = pairs_df.loc[pairs_df.gene_name.isin([\"ELK1\", \"SIX3\", \"TLX1\"]), \"bed_name\"].to_list()\n",
    "d3utr_sel = d3utr.subset(lambda df: df.Name.isin(d3utr_region_nms))\n",
    "d3utr_sel = pr.concat([d3utr_sel, tlx1_annot_le]).sort()\n",
    "d3utr_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "d3utr_sel.to_bed(\"processed/validation_region_beds/2024-04-08_cryptic_and_annotated.selected.3ext.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42935670</td>\n",
       "      <td>42936378</td>\n",
       "      <td>ENSG00000137161.18_1|CNPY3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>42938567</td>\n",
       "      <td>42939294</td>\n",
       "      <td>CNPY3|annotated|ENST00000372836.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854261</td>\n",
       "      <td>16857420</td>\n",
       "      <td>ENSG00000127511.10_2|SIN3B|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880353</td>\n",
       "      <td>SIN3B|annotated|ENST00000379803.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32597005</td>\n",
       "      <td>32598378</td>\n",
       "      <td>ANKRD27|annotated|ENST00000306065.9</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19</td>\n",
       "      <td>32635458</td>\n",
       "      <td>32639355</td>\n",
       "      <td>ENSG00000105186.16_3|ANKRD27|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr6         |  42935670 |  42936378 | ...   |\n",
       "| chr6         |  42938567 |  42939294 | ...   |\n",
       "| chr19        |  16854261 |  16857420 | ...   |\n",
       "| chr19        |  16878496 |  16880353 | ...   |\n",
       "| chr19        |  32597005 |  32598378 | ...   |\n",
       "| chr19        |  32635458 |  32639355 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 6 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## IPA events\n",
    "ipa = pr.read_bed(\"processed/validation_region_beds/2024-04-05_cryptic_and_annotated.ipa.bed\")\n",
    "ipa_region_gns = pairs_df.loc[pairs_df.event_type == \"IPA\", \"bed_name\"].to_list()\n",
    "ipa_sel = ipa.subset(lambda df: df.Name.isin(ipa_region_gns))\n",
    "ipa_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ipa_sel.to_bed(\"processed/validation_region_beds/2024-04-08_cryptic_and_annotated.selected.ipa.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
