{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get last exon coordinates for experimental validation with 3'RACE\n",
    "\n",
    "Need:\n",
    "- Cryptic last exon coordinates (with 3'end precision/resolution)\n",
    "- Annotated alternative last exon (primary/dominant isoform, again with 3'end precision)\n",
    "\n",
    "Since my last exon BEDs used for iCLIP/motif analyses are unlikely to contain these for the same gene (for ALE & IPA events), need to go back and grab them from reference annotation.\n",
    "- Extract all annotated alternative last exons from reference\n",
    "- Output to BED file, check in i3Neurons for major expressed isoform\n",
    "\n",
    "3'Ext events should be straightforward - just use selected proximal and distal isoforms, with a check that proximal/distal coords are valid\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from typing import Literal\n",
    "import os\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pr.read_gtf(\"data/regions/gencode.v40.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in last exon beds\n",
    "ipa = pr.read_bed(\"data/regions/2023-12-14_papa_cryptic_bleedthrough_uniq.background_shsy5y.last_exons.bed\")\n",
    "ale = pr.read_bed(\"data/regions/2023-12-14_papa_cryptic_spliced.background_shsy5y.last_exons.bed\")\n",
    "three_ext = pr.read_bed(\"data/regions/2023-12-15_papa_cryptic_d3utr.background_shsy5y.last_exons.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract gene names/ids for cryptic IPA and ALE events\n",
    "ipa_cryp = ipa.subset(lambda df: df.Name.str.contains(\"cryptic\", regex=False))\n",
    "ale_cryp = ale.subset(lambda df: df.Name.str.contains(\"cryptic\", regex=False))\n",
    "three_ext_cryp = three_ext.subset(lambda df: df.Name.str.contains(\"cryptic\", regex=False))\n",
    "\n",
    "# e.g. ENSG00000119392.16_1|GLE1|bleedthrough|cryptic\n",
    "ipa_cryp_gn = ipa_cryp.Name.str.split(\"|\", expand=True)[1].tolist()\n",
    "ale_cryp_gn = ale_cryp.Name.str.split(\"|\", expand=True)[1].tolist()\n",
    "cryp_gn = ipa_cryp_gn + ale_cryp_gn\n",
    "\n",
    "len(cryp_gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21485005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21483467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21440348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001546346.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21444947</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21445086</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25631</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>131823763</td>\n",
       "      <td>131830862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000513644.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25632</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131830291</td>\n",
       "      <td>131830862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000513644.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003877682.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25633</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823763</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000513644.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003853674.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25634</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>131829923</td>\n",
       "      <td>131830622</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000505674.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25635</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131829923</td>\n",
       "      <td>131830622</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000505674.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003869744.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25636 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +20   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | gene       | 21440127  | 21485005  | ...   |\n",
       "| chr1         | HAVANA     | transcript | 21440127  | 21483467  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21440127  | 21440348  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21444947  | 21445219  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131830291 | 131830862 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131823763 | 131825365 | ...   |\n",
       "| chrX         | HAVANA     | transcript | 131829923 | 131830622 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131829923 | 131830622 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 25,636 rows and 25 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cryp = gtf.subset(lambda df: df.gene_name.isin(cryp_gn))\n",
    "gtf_cryp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21440127</td>\n",
       "      <td>21440348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001546346.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21444947</td>\n",
       "      <td>21445219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003642335.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21470631</td>\n",
       "      <td>21470734</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003734729.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21471568</td>\n",
       "      <td>21471783</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003723776.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21472842</td>\n",
       "      <td>21472915</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-205</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476522.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003753961.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000415711.2</td>\n",
       "      <td>CCDS57977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12258</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131825221</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000507677.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001659046.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12259</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131804594</td>\n",
       "      <td>131805037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000507677.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003882991.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12260</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131830291</td>\n",
       "      <td>131830862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000513644.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003877682.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12261</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823763</td>\n",
       "      <td>131825365</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000513644.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003853674.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12262</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131829923</td>\n",
       "      <td>131830622</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000505674.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003869744.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12263 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +20   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | exon       | 21440127  | 21440348  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21444947  | 21445219  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21470631  | 21470734  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21471568  | 21471783  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131804594 | 131805037 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131830291 | 131830862 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131823763 | 131825365 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131829923 | 131830622 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 12,263 rows and 25 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cryp_ex = gtf_cryp.subset(lambda df: df.Feature == \"exon\")\n",
    "gtf_cryp_ex.exon_number = gtf_cryp_ex.exon_number.astype(int)\n",
    "gtf_cryp_ex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: make sure type hinting for get_terminal_regions is correct (by default stranded, but can be one of stranded/unstranded)\n",
    "def get_terminal_regions(gr: pr.PyRanges,\n",
    "                         feature_col = \"Feature\",\n",
    "                         feature_key = \"exon\",\n",
    "                         id_col = \"transcript_id\",\n",
    "                         region_number_col = \"exon_number\",\n",
    "                         number_type: Literal[\"stranded\", \"unstranded\"] = \"stranded\",\n",
    "                         which_region: str = \"last\",\n",
    "                         filter_single = False,\n",
    "                         ):\n",
    "    '''Return the first/last interval in group of intervals\n",
    "\n",
    "    Requires a column that provides a 1..n numbering of intervals within each group (can be generated by add_region_number). Extraction will always be with respect to strand (can handle strand-aware/non-strand aware ranking in region_number_col) \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    gr : pr.PyRanges\n",
    "        _description_\n",
    "    feature_col : str, optional\n",
    "        _description_, by default \"Feature\"\n",
    "    feature_key : str, optional\n",
    "        _description_, by default \"exon\"\n",
    "    id_col : str, optional\n",
    "        _description_, by default \"transcript_id\"\n",
    "    region_number_col : str, optional\n",
    "        _description_, by default \"exon_number\"\n",
    "    number_type : Literal[&quot;stranded&quot;, &quot;unstranded&quot;], optional\n",
    "        _description_, by default \"stranded\"\n",
    "    which_region : str, optional\n",
    "        _description_, by default \"last\"\n",
    "    filter_single : bool, optional\n",
    "        _description_, by default False\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    _type_\n",
    "        _description_\n",
    "    '''\n",
    "\n",
    "    assert number_type in [\"stranded\", \"unstranded\"]\n",
    "    assert which_region in [\"first\", \"last\"]\n",
    "    assert region_number_col in gr.columns.tolist()\n",
    "    assert feature_col in gr.columns.tolist()\n",
    "    assert id_col in gr.columns.tolist()\n",
    "\n",
    "    # Make sure only 'exon' features are in the gr\n",
    "    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], \"only {} entries should be present in gr\".format(feature_key)\n",
    "    # Make sure region_number_col is int\n",
    "    # assert gr.as_df()[region_number_col].dtype\n",
    "\n",
    "    # TODO: Make this into a separate function\n",
    "    # Make sure region_number_col is int\n",
    "    # try:\n",
    "    #     mod_gr = (gr.assign(region_number_col,\n",
    "    #                         lambda df: df[region_number_col].astype(float).astype(int),\n",
    "    #                         nb_cpu=1)\n",
    "    #               )\n",
    "    # except KeyError:\n",
    "    #     # Currently getting weird KeyError with assign for certain chromosome\n",
    "    #     # Mostly non-std chrom names\n",
    "    #     # No error if do '.<exon_number>' to assign, but this makes inflexible to colname\n",
    "    #     # Also no error if gr -> df assign -> gr\n",
    "    #     print(\"pr.assign returned KeyError. Converting {} to int via pandas df conversion\".format(region_number_col))\n",
    "\n",
    "    #     mod_gr = gr.as_df()\n",
    "    #     mod_gr[region_number_col] = mod_gr[region_number_col].astype(float).astype(int)\n",
    "    #     mod_gr = pr.PyRanges(mod_gr)\n",
    "\n",
    "\n",
    "    # Make sure gr is sorted by transcript_id & 'region number' (ascending order so 1..n)\n",
    "    mod_gr = gr.apply(lambda df: df.sort_values(by=[id_col, region_number_col], ascending=True),\n",
    "                          nb_cpu=1)\n",
    "\n",
    "\n",
    "    # Filter out single-exon transcripts\n",
    "    if filter_single:\n",
    "        print(\"Filtering for multi-exon transcripts...\")\n",
    "        print(\"Before: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "\n",
    "        # Setting to 'False' marks all duplicates as True (so keep these)\n",
    "        mod_gr = mod_gr.subset(lambda df: df.duplicated(subset=[id_col], keep=False), nb_cpu=1)\n",
    "\n",
    "        print(\"After: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "\n",
    "\n",
    "    if number_type == \"stranded\":\n",
    "        # source = None means that 1 = first region of group regardless of strand\n",
    "        # Pick last region entry by max region number for each transcript (id_col)\n",
    "        # Pick first region entry by min region number for each transcript (id_col)\n",
    "\n",
    "        # keep=\"last\" sets last in ID to 'False' and all others true (negate to keep last only)\n",
    "        # keep=\"first\" sets first in ID to 'False'\n",
    "\n",
    "        out_gr = mod_gr.subset(lambda df: ~(df.duplicated(subset=[id_col], keep=which_region)),\n",
    "                               nb_cpu=1\n",
    "                              )\n",
    "\n",
    "\n",
    "    else:\n",
    "        # Numbering doesn't respect strand\n",
    "        # Need to flip selecting first/last in group depending on strand\n",
    "        # minus strand - pick min if Minus strand, max if plus strand\n",
    "\n",
    "        if which_region == \"first\":\n",
    "            # + strand - pick first in group, - strand - pick last in group\n",
    "\n",
    "            out_gr = (mod_gr.subset(lambda df:\n",
    "                                    #1. plus strand & first in group/ID\n",
    "                                    (df[\"Strand\"] == \"+\") & ~(df.duplicated(subset=[id_col],\n",
    "                                                                            keep=\"first\")) |\n",
    "                                    #2. minus strand & last in group/ID\n",
    "                                    (df[\"Strand\"] == \"-\") & ~(df.duplicated(subset=[id_col],\n",
    "                                                                            keep=\"last\")),\n",
    "                                    nb_cpu=1)\n",
    "                     )\n",
    "\n",
    "        elif which_region == \"last\":\n",
    "            # + strand - pick last in group/ID\n",
    "            # - strand - pick first in group/ID\n",
    "            out_gr = (mod_gr.subset(lambda df:\n",
    "                                    #1. plus strand & last in group/ID\n",
    "                                    (df[\"Strand\"] == \"+\") & ~(df.duplicated(subset=[id_col],\n",
    "                                                                            keep=\"last\")) |\n",
    "                                    #2. minus strand & first in group/ID\n",
    "                                    (df[\"Strand\"] == \"-\") & ~(df.duplicated(subset=[id_col],\n",
    "                                                                            keep=\"first\")),\n",
    "                                    nb_cpu=1)\n",
    "                     )\n",
    "\n",
    "\n",
    "    return out_gr\n",
    "\n",
    "\n",
    "def _df_collapse_metadata(df, id_col, standard_cols, collapse_cols, collapse_uniq_cols, collapse_sep):\n",
    "    '''\n",
    "    Intended to be applied to internal dfs of PyRanges objects\n",
    "    '''\n",
    "\n",
    "    found_collapsed = [col for col in collapse_cols if col in df.columns]\n",
    "\n",
    "    not_found_collapsed = set(collapse_cols) - set(found_collapsed)\n",
    "\n",
    "    if len(not_found_collapsed) > 0:\n",
    "        chr_strand = f\"{df.Chromosome.drop_duplicates()[0]},{df.Strand.drop_duplicates()[0]}\"\n",
    "        print(f\"following 'collapse_cols' columns not found in df (chr/strand) - {chr_strand} - {', '.join(not_found_collapsed)}\")\n",
    "\n",
    "    grouped = df.groupby(id_col)\n",
    "\n",
    "    # Pick first entry for all standard_cols, these should be same for all rows of id_col\n",
    "    # Leaves a df with id_col values as index\n",
    "    std_collapsed = grouped[standard_cols].first()\n",
    "\n",
    "    # For collapse cols, collapse to single row of delimited strings for each column\n",
    "    # Again leave a df with id_col values as index labels\n",
    "    clp_collapsed = grouped[found_collapsed].agg(lambda col: collapse_sep.join(col.astype(str)))\n",
    "\n",
    "    if collapse_uniq_cols is not None:\n",
    "        # Collapse these cols to single row of delimited strings whilst dropping duplicates\n",
    "        # Again leave a df with id_col values as index labels\n",
    "        clp_uniq_collapsed = grouped[collapse_uniq_cols].agg(lambda col: collapse_sep.join(list(dict.fromkeys(col.astype(str)))))\n",
    "\n",
    "        int_collapsed = clp_collapsed.merge(clp_uniq_collapsed, left_index=True, right_index=True)\n",
    "\n",
    "        collapsed = std_collapsed.merge(int_collapsed, left_index=True, right_index=True)\n",
    "\n",
    "    else:\n",
    "        # combine by id_col\n",
    "        collapsed = std_collapsed.merge(clp_collapsed, left_index=True, right_index=True)\n",
    "\n",
    "    return collapsed\n",
    "\n",
    "\n",
    "def collapse_metadata(gr,\n",
    "                      id_col=\"transcript_id\",\n",
    "                      standard_cols=[\"Chromosome\", \"Start\", \"End\", \"Strand\"],\n",
    "                      collapse_cols=None,\n",
    "                      collapse_uniq_cols=None,\n",
    "                      collapse_sep=\",\"):\n",
    "    '''\n",
    "    Collapse to a single entry/row per ID entry whilst retaining/collapsing metadata on duplicate rows\n",
    "    standard_cols: list of column labels that have the same value for all entries of id_col and do not need to be collapsed.\n",
    "    This is essential for PyRanges standard columns, as you do not want to be changing their dtypes to string. All columns labels in this list retain their dtype, and the first value is retained\n",
    "\n",
    "    collapse_cols: list of column labels containing metadata you'd like to collapse to a single row (separated by collapse_sep)\n",
    "        If None, then all columns in gr except for standard_cols, id_col & collapse_uniq_cols will be collapsed\n",
    "    collapse_uniq_cols: list of column labels containing metadata you'd like to collapse to a single row whilst dropping duplicate values. Values will maintain order of appearance in df\n",
    "    '''\n",
    "\n",
    "    assert all([True if col in gr.columns else False for col in standard_cols])\n",
    "\n",
    "    if collapse_uniq_cols is not None:\n",
    "        # Here just checking the columns are found in df\n",
    "        assert all([True if col in gr.columns else False for col in collapse_uniq_cols])\n",
    "\n",
    "    if collapse_cols is None:\n",
    "        if collapse_uniq_cols is not None:\n",
    "            def_cols = standard_cols + [id_col] + collapse_uniq_cols\n",
    "\n",
    "        else:\n",
    "            def_cols = standard_cols + [id_col]\n",
    "\n",
    "        collapse_cols = [col for col in gr.columns if col not in def_cols]\n",
    "\n",
    "    else:\n",
    "        assert all([True if col in gr.columns else False for col in collapse_cols])\n",
    "\n",
    "\n",
    "    return gr.apply(lambda df: _df_collapse_metadata(df,\n",
    "                                                     id_col,\n",
    "                                                     standard_cols,\n",
    "                                                     collapse_cols,\n",
    "                                                     collapse_uniq_cols,\n",
    "                                                     collapse_sep\n",
    "                                                     ))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>62117131</td>\n",
       "      <td>62117393</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000132849.22</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PATJ-201</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000488269.2</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSE00003788988.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000307496.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>NBPF3-201</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000008190.3</td>\n",
       "      <td>18</td>\n",
       "      <td>ENSE00001464927.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000316739.7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000476523.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSE00001814163.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000316782.5</td>\n",
       "      <td>CCDS216.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>77062974</td>\n",
       "      <td>77064058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000117069.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>ST6GALNAC5-201</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000026693.4</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001674568.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436263.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NBPF3-203</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000008193.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSE00001814163.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000340336.5</td>\n",
       "      <td>CCDS57976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131794113</td>\n",
       "      <td>131794466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000514677.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003865783.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689737</td>\n",
       "      <td>131689840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000512454.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00003872410.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689754</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000521770.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00003880552.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131691522</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000508518.1</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSE00003853517.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1417</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>131793752</td>\n",
       "      <td>131794466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FIRRE-217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>OTTHUMT00000510743.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003881923.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1418 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +20   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | exon       | 62117131  | 62117393  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21483142  | 21485005  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 21483142  | 21484900  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 77062974  | 77064058  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131689737 | 131689840 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131689754 | 131692636 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131691522 | 131692636 | ...   |\n",
       "| chrX         | HAVANA     | exon       | 131793752 | 131794466 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,418 rows and 25 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cryp_ex_le = get_terminal_regions(gtf_cryp_ex)\n",
    "gtf_cryp_ex_le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62117131</td>\n",
       "      <td>62117393</td>\n",
       "      <td>+</td>\n",
       "      <td>PATJ</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>77062974</td>\n",
       "      <td>77064058</td>\n",
       "      <td>+</td>\n",
       "      <td>ST6GALNAC5</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131794113</td>\n",
       "      <td>131794466</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131689737</td>\n",
       "      <td>131689840</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131689754</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131691522</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1417</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131793752</td>\n",
       "      <td>131794466</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1418 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | gene_name   | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------------+-------|\n",
       "| chr1         | 62117131  | 62117393  | +            | PATJ        | ...   |\n",
       "| chr1         | 21483142  | 21485005  | +            | NBPF3       | ...   |\n",
       "| chr1         | 21483142  | 21484900  | +            | NBPF3       | ...   |\n",
       "| chr1         | 77062974  | 77064058  | +            | ST6GALNAC5  | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...         | ...   |\n",
       "| chrX         | 131689737 | 131689840 | -            | FIRRE       | ...   |\n",
       "| chrX         | 131689754 | 131692636 | -            | FIRRE       | ...   |\n",
       "| chrX         | 131691522 | 131692636 | -            | FIRRE       | ...   |\n",
       "| chrX         | 131793752 | 131794466 | -            | FIRRE       | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 1,418 rows and 9 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: transcript_type, transcript_support_level, tag, exon_number"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cryp_ex_le[[\"gene_name\", \"transcript_type\", \"transcript_support_level\", \"tag\", \"exon_number\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16855745</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000596802.5;ENST00000596638.1;ENST000005...</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | +2    |\n",
       "| (category)   |   (int64) |   (int64) | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr19        |  16854142 |  16855745 | +            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: transcript_id, gene_name"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_gr = gtf_cryp_ex_le.subset(lambda df: df.gene_name == \"SIN3B\")\n",
    "tmp_gr = tmp_gr.cluster(by=[\"gene_name\", \"Start\", \"End\"], strand=True)\n",
    "tmp_gr_orig = collapse_metadata(tmp_gr, id_col=\"gene_name\", collapse_cols=[\"transcript_id\"], collapse_uniq_cols=[\"gene_name\"], collapse_sep=\";\")\n",
    "tmp_gr_cluster = collapse_metadata(tmp_gr, id_col=\"Cluster\", collapse_cols=[\"transcript_id\"], collapse_uniq_cols=[\"gene_name\"], collapse_sep=\";\")\n",
    "tmp_gr_orig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16855745</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16862883</td>\n",
       "      <td>16863286</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000596638.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16865409</td>\n",
       "      <td>16865590</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000599880.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16869459</td>\n",
       "      <td>16869623</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000594372.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16871228</td>\n",
       "      <td>16871638</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000602204.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16872467</td>\n",
       "      <td>16872666</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000595900.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878706</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000595049.5</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878942</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000594235.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878961</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000595541.1</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16879060</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000601141.5</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880349</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880353</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>SIN3B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr19        | 16854142  | 16855745  | +            | ...   |\n",
       "| chr19        | 16862883  | 16863286  | +            | ...   |\n",
       "| chr19        | 16865409  | 16865590  | +            | ...   |\n",
       "| chr19        | 16869459  | 16869623  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr19        | 16878496  | 16878961  | +            | ...   |\n",
       "| chr19        | 16878496  | 16879060  | +            | ...   |\n",
       "| chr19        | 16878496  | 16880349  | +            | ...   |\n",
       "| chr19        | 16878496  | 16880353  | +            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: transcript_id, gene_name"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_gr_cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16854142</td>\n",
       "      <td>16855745</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596802.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16862883</td>\n",
       "      <td>16863286</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000596638.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16865409</td>\n",
       "      <td>16865590</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000599880.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16869459</td>\n",
       "      <td>16869623</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000594372.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16871228</td>\n",
       "      <td>16871638</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000602204.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16872467</td>\n",
       "      <td>16872666</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000595900.1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878706</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000595049.5</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878942</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000594235.1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16878961</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000595541.1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16879060</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000601141.5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880349</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000248054.10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr19</td>\n",
       "      <td>16878496</td>\n",
       "      <td>16880353</td>\n",
       "      <td>+</td>\n",
       "      <td>SIN3B</td>\n",
       "      <td>ENST00000379803.5</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | gene_name   | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------------+-------|\n",
       "| chr19        | 16854142  | 16855745  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16862883  | 16863286  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16865409  | 16865590  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16869459  | 16869623  | +            | SIN3B       | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...         | ...   |\n",
       "| chr19        | 16878496  | 16878961  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16878496  | 16879060  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16878496  | 16880349  | +            | SIN3B       | ...   |\n",
       "| chr19        | 16878496  | 16880353  | +            | SIN3B       | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: transcript_id, Cluster"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_gr[[\"gene_name\", \"transcript_id\", \"Cluster\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>54597258</td>\n",
       "      <td>54598858</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ACOT11-205</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000027353.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00001846834.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>54599295</td>\n",
       "      <td>54599658</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ACOT11-203</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000027354.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001886392.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610124</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ACOT11-204</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000027352.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSE00001886798.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ACOT11-201</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000027351.1</td>\n",
       "      <td>16</td>\n",
       "      <td>ENSE00001385652.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000340260.2</td>\n",
       "      <td>CCDS593.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ACOT11-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000027356.1</td>\n",
       "      <td>17</td>\n",
       "      <td>ENSE00001454930.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000360366.3</td>\n",
       "      <td>CCDS592.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17800048</td>\n",
       "      <td>17802034</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>RAI2-204</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000504349.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003849071.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000401323.1</td>\n",
       "      <td>CCDS14183.1</td>\n",
       "      <td>1183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17800050</td>\n",
       "      <td>17802034</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>RAI2-201</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000055937.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001300453.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000333456.1</td>\n",
       "      <td>CCDS14183.1</td>\n",
       "      <td>1184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17800050</td>\n",
       "      <td>17802034</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>RAI2-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000055938.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001300453.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000353106.1</td>\n",
       "      <td>CCDS14183.1</td>\n",
       "      <td>1184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>17800050</td>\n",
       "      <td>17802034</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>RAI2-206</td>\n",
       "      <td>4</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001300453.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000444210.1</td>\n",
       "      <td>CCDS14183.1</td>\n",
       "      <td>1184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1417</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17801663</td>\n",
       "      <td>17801773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>RAI2-205</td>\n",
       "      <td>3</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMT00000471628.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002037351.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000473562.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1185</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1418 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +21   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | exon       | 54597258  | 54598858  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 54599295  | 54599658  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 54608956  | 54610124  | ...   |\n",
       "| chr1         | HAVANA     | exon       | 54608956  | 54610329  | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrX         | HAVANA     | exon       | 17800050  | 17802034  | ...   |\n",
       "| chrX         | HAVANA     | exon       | 17800050  | 17802034  | ...   |\n",
       "| chrX         | ENSEMBL    | exon       | 17800050  | 17802034  | ...   |\n",
       "| chrX         | HAVANA     | exon       | 17801663  | 17801773  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,418 rows and 26 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "21 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 14 more.)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To create BED file, easier if have unique sequences\n",
    "# Simplicity - cluster if identical last exon coordinates\n",
    "gtf_cryp_ex_le = gtf_cryp_ex_le.cluster(by=[\"gene_name\", \"Start\", \"End\"], strand=True)\n",
    "gtf_cryp_ex_le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54597258</td>\n",
       "      <td>54598858</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000498228.1</td>\n",
       "      <td>ACOT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54599295</td>\n",
       "      <td>54599658</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000479837.1</td>\n",
       "      <td>ACOT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610124</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000481208.5</td>\n",
       "      <td>ACOT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000343744.7</td>\n",
       "      <td>ACOT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000371316.3</td>\n",
       "      <td>ACOT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40726527</td>\n",
       "      <td>40726851</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000463072.1</td>\n",
       "      <td>MED14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800048</td>\n",
       "      <td>17801725</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000415486.7</td>\n",
       "      <td>RAI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800048</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000451717.6</td>\n",
       "      <td>RAI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800050</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000331511.5;ENST00000360011.5;ENST000005...</td>\n",
       "      <td>RAI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17801663</td>\n",
       "      <td>17801773</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000509491.2</td>\n",
       "      <td>RAI2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1185 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 54597258  | 54598858  | +            | ...   |\n",
       "| chr1         | 54599295  | 54599658  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610124  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610329  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 17800048  | 17801725  | -            | ...   |\n",
       "| chrX         | 17800048  | 17802034  | -            | ...   |\n",
       "| chrX         | 17800050  | 17802034  | -            | ...   |\n",
       "| chrX         | 17801663  | 17801773  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,185 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: transcript_id, gene_name"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collapse redundant last exon coordinates, collapsing corresponding transcript_ids and the gene name\n",
    "gtf_cryp_ex_le_clpsd = collapse_metadata(gtf_cryp_ex_le, id_col=\"Cluster\", collapse_cols=[\"transcript_id\"], collapse_uniq_cols=[\"gene_name\"], collapse_sep=\";\")\n",
    "gtf_cryp_ex_le_clpsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>annot_status</th>\n",
       "      <th>Score</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54597258</td>\n",
       "      <td>54598858</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000498228.1</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>ACOT11|annotated|ENST00000498228.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54599295</td>\n",
       "      <td>54599658</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000479837.1</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>ACOT11|annotated|ENST00000479837.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610124</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000481208.5</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>ACOT11|annotated|ENST00000481208.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000343744.7</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>ACOT11|annotated|ENST00000343744.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000371316.3</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>ACOT11|annotated|ENST00000371316.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40726527</td>\n",
       "      <td>40726851</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000463072.1</td>\n",
       "      <td>MED14</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>MED14|annotated|ENST00000463072.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800048</td>\n",
       "      <td>17801725</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000415486.7</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>RAI2|annotated|ENST00000415486.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800048</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000451717.6</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>RAI2|annotated|ENST00000451717.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17800050</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000331511.5;ENST00000360011.5;ENST000005...</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>RAI2|annotated|ENST00000331511.5;ENST000003600...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17801663</td>\n",
       "      <td>17801773</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000509491.2</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>annotated</td>\n",
       "      <td>.</td>\n",
       "      <td>RAI2|annotated|ENST00000509491.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1185 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 54597258  | 54598858  | +            | ...   |\n",
       "| chr1         | 54599295  | 54599658  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610124  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610329  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 17800048  | 17801725  | -            | ...   |\n",
       "| chrX         | 17800048  | 17802034  | -            | ...   |\n",
       "| chrX         | 17800050  | 17802034  | -            | ...   |\n",
       "| chrX         | 17801663  | 17801773  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,185 rows and 9 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: transcript_id, gene_name, annot_status, Score, Name"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate a name column, minimal info required for BEd\n",
    "gtf_cryp_ex_le_clpsd = (gtf_cryp_ex_le_clpsd\n",
    "                            .assign(\"annot_status\", lambda df: pd.Series([\"annotated\"]*len(df), index=df.index))\n",
    "                            .assign(\"Score\", lambda df: pd.Series([\".\"]*len(df), index=df.index))\n",
    "                            .assign(\"Name\", lambda df: df.gene_name.str.cat(df[[\"annot_status\", \"transcript_id\"]].astype(str), sep=\"|\")))\n",
    "\n",
    "gtf_cryp_ex_le_clpsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54597258</td>\n",
       "      <td>54598858</td>\n",
       "      <td>ACOT11|annotated|ENST00000498228.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54599295</td>\n",
       "      <td>54599658</td>\n",
       "      <td>ACOT11|annotated|ENST00000479837.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610124</td>\n",
       "      <td>ACOT11|annotated|ENST00000481208.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>ACOT11|annotated|ENST00000343744.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1098</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108155613</td>\n",
       "      <td>108157260</td>\n",
       "      <td>COL4A6|annotated|ENST00000334504.12</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108161630</td>\n",
       "      <td>108161735</td>\n",
       "      <td>COL4A6|annotated|ENST00000487645.1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108213791</td>\n",
       "      <td>108214228</td>\n",
       "      <td>COL4A6|annotated|ENST00000468338.5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>COL4A6|annotated|ENST00000461897.5;ENST0000047...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1103 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (object)     | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 54597258  | 54598858  | ...   |\n",
       "| chr1         | 54599295  | 54599658  | ...   |\n",
       "| chr1         | 54608956  | 54610124  | ...   |\n",
       "| chr1         | 54608956  | 54610329  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 108161630 | 108161735 | ...   |\n",
       "| chrX         | 108213791 | 108214228 | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,103 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create separate BED file for cryptic ALEs and IPAs\n",
    "ale_cryp_annot_bed = pr.concat([ale_cryp, gtf_cryp_ex_le_clpsd.subset(lambda df: df.gene_name.isin(ale_cryp_gn))[[\"Name\", \"Score\"]]]).sort()\n",
    "ipa_cryp_annot_bed = pr.concat([ipa_cryp, gtf_cryp_ex_le_clpsd.subset(lambda df: df.gene_name.isin(ipa_cryp_gn))[[\"Name\", \"Score\"]]]).sort()\n",
    "\n",
    "ale_cryp_annot_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21442239</td>\n",
       "      <td>21442358</td>\n",
       "      <td>NBPF3|annotated|ENST00000475869.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21446036</td>\n",
       "      <td>21446482</td>\n",
       "      <td>NBPF3|annotated|ENST00000485941.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21451766</td>\n",
       "      <td>21451858</td>\n",
       "      <td>NBPF3|annotated|ENST00000467103.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21454306</td>\n",
       "      <td>NBPF3|annotated|ENST00000478653.6</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131795049</td>\n",
       "      <td>131796331</td>\n",
       "      <td>FIRRE|annotated|ENST00000650184.1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131804594</td>\n",
       "      <td>131805037</td>\n",
       "      <td>FIRRE|annotated|ENST00000653136.1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823763</td>\n",
       "      <td>131825365</td>\n",
       "      <td>FIRRE|annotated|ENST00000657242.1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131829923</td>\n",
       "      <td>131830622</td>\n",
       "      <td>FIRRE|annotated|ENST00000658060.1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (object)     | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 21442239  | 21442358  | ...   |\n",
       "| chr1         | 21446036  | 21446482  | ...   |\n",
       "| chr1         | 21451766  | 21451858  | ...   |\n",
       "| chr1         | 21453372  | 21454306  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 131804594 | 131805037 | ...   |\n",
       "| chrX         | 131823763 | 131825365 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "| chrX         | 131829923 | 131830622 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 194 rows and 6 columns from 15 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ipa_cryp_annot_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = \"processed/validation_region_beds\"\n",
    "if not os.path.exists(outdir):\n",
    "        os.makedirs(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ale_cryp_annot_bed.to_bed(os.path.join(outdir, \"2024-04-05_cryptic_and_annotated.ales.bed\"))\n",
    "ipa_cryp_annot_bed.to_bed(os.path.join(outdir, \"2024-04-05_cryptic_and_annotated.ipa.bed\"))\n",
    "three_ext_cryp.to_bed(os.path.join(outdir, \"2024-04-05_cryptic_and_annotated.3ext.bed\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
