import pandas as pd
import os

# Extract values from sample table based on the wildcard
def get_param(wildcards, param_name):
    sample = wildcards.sample
    if sample in sample_table['comparison_name'].values:
        index = sample_table[sample_table['comparison_name'] == sample].index[0]
        return sample_table.at[index, param_name]
    else:
        raise ValueError(f"Sample {sample} not found in sample_table")


# Read the sample table using pandas
sample_table = pd.read_csv(config["sample_table"])

# Define the list of sample names based on the 'comparison_name' column
sample_names = sample_table['comparison_name'].tolist()

# base output directory
base_outdir = config["base_output_dir"]


# Define a rule to generate the target files using the 'sample' wildcard
rule all:
    input:
        expand(os.path.join(base_outdir, "{sample}", "cv_coverage_{group}", "cv_coverage.done"), sample=sample_names, group=["foreground", "background"]) if config["run_cv_coverage"] else expand(os.path.join(base_outdir, "{sample}", "peka.done"), sample=sample_names),
        expand(os.path.join(base_outdir, "{sample}", "peka.done"), sample=sample_names)


rule pas_to_peka_beds:
    input:
        pas=lambda wildcards: get_param(wildcards, 'regions_bed')
    output:
        comb_bed=os.path.join(base_outdir, "peka_beds", "{sample}.all.bed"),
        reg_bed=os.path.join(base_outdir, "peka_beds", "{sample}.foreground.bed"),
        bg_bed=os.path.join(base_outdir, "peka_beds", "{sample}.background.bed")
    params:
        script="scripts/pas_to_peka_beds.py",
        foreground_key=lambda wildcards: get_param(wildcards, 'foreground_key'),
        output_prefix = os.path.join(base_outdir, "peka_beds", "{sample}")

    log:
        stdout=os.path.join(base_outdir, "logs", "pas_to_peka_beds.{sample}.stdout.txt"),
        stderr=os.path.join(base_outdir, "logs", "pas_to_peka_beds.{sample}.stderr.txt")

    conda:
        "envs/peka_fork.yaml"

    shell:
       """python {params.script} \
       -f {params.foreground_key} \
       {input.pas} \
       {params.output_prefix} \
       1> {log.stdout} \
       2> {log.stderr}
       """


#Note: PEKA's final output file is quite complicated to track with Snakemake
# f"{sample_name}_{kmer_length}mer_cluster_distribution_{region}.tsv"
# where sample_name is the input comb_bed file with .bed suffix removed
# I'm too lazy to code it out, so instead use an && so that empty file is created only when peka command finishes successfully
rule peka:
    input:
        comb_bed=rules.pas_to_peka_beds.output.comb_bed,
        reg_bed=rules.pas_to_peka_beds.output.reg_bed,
        genome=config["genome"],
        genome_idx=config["genome_idx"]
            
    output:
        os.path.join(base_outdir, "{sample}", "peka.done") 

    params:
        kmer=config["kmer"],
        window=config["window"],
        distal_window=config["distal_window"],
        specific_region=config["specific_region"],
        alloutputs=config["alloutputs"],
        percentile=config["percentile"],
        relpos="-pos " + str(config["relpos"]) if config["relpos"] != "auto" else "",
        regions=config["regions_gtf"],
        outdir = os.path.join(base_outdir, "{sample}", ""),
        tmpdir = os.path.join(base_outdir, "{sample}", "tmp")

    log:
        stdout=os.path.join(base_outdir, "logs", "peka.{sample}.stdout.txt"),
        stderr=os.path.join(base_outdir, "logs", "peka.{sample}.stderr.txt")

    conda:
        "envs/peka_fork.yaml"

    shell:
        """
        mkdir -p {params.tmpdir} && export TMPDIR={params.tmpdir}
        peka -x {input.comb_bed} \
        -i {input.reg_bed} \
        -g {input.genome} \
        -gi {input.genome_idx} \
        -r {params.regions} \
        -k {params.kmer} \
        -p {params.percentile} \
        {params.relpos} \
        -w {params.window} \
        -dw {params.distal_window} \
        --specificregion {params.specific_region} \
        --alloutputs {params.alloutputs} \
        -o {params.outdir} \
        1> {log.stdout} \
        2> {log.stderr} && \
        touch {output}
        """

# plot & get % coverage for specific motifs in given window around foreground regions
# final TSV output to ./results/no_of_txn_{sample_name}' 
# where sample_name is input BED file with .bed suffix removed and any text with '.xl.' removed 
# again too lazy to code out
# outputs in current directory, so have to make and switch inside script
rule cv_coverage_foreground:
    input:
        bed=rules.pas_to_peka_beds.output.reg_bed,
        genome=os.path.abspath(config["genome"]),
        genome_idx=os.path.abspath(config["genome_idx"])

    output:
         os.path.join(base_outdir, "{sample}", "cv_coverage_foreground", "cv_coverage.done")
    
    params:
        script=os.path.abspath("scripts/cv_coverage.py"),
        bed=os.path.abspath(rules.pas_to_peka_beds.output.reg_bed),
        motifs=",".join(config["cv_coverage_motifs"]),
        kmer=config["kmer"],
        percentile=config["percentile"],
        regions=os.path.abspath(config["regions_gtf"]),
        specific_region=config["specific_region"],
        window=config["distal_window"],
        smoothing=config["cv_coverage_smoothing"],
        use_scores=config["cv_coverage_usescores"],
        cap=config["cv_coverage_cap"],
        chunk_size=10000,
        outdir=os.path.join(base_outdir, "{sample}", "cv_coverage_foreground"),
        tmpdir=os.path.abspath(os.path.join(base_outdir, "{sample}", "cv_coverage_foreground", "tmp")),
        out="cv_coverage.done"

    threads:
        config["cv_coverage_n_cores"]


    log:
        stdout=os.path.abspath(os.path.join(base_outdir, "logs", "cv_coverage.{sample}.stdout.txt")),
        stderr=os.path.abspath(os.path.join(base_outdir, "logs", "cv_coverage.{sample}.stderr.txt"))

    conda:
        "envs/peka_fork.yaml"


    shell:
        """
        mkdir -p {params.tmpdir} && export TMPDIR={params.tmpdir} && cd {params.outdir}
        python {params.script} \
        {params.bed} \
        {params.motifs} \
        {params.specific_region} \
        {params.kmer} \
        {input.genome} \
        {input.genome_idx} \
        {params.regions} \
        {params.smoothing} \
        {params.percentile} \
        {params.window} \
        {params.use_scores} \
        {threads} \
        {params.chunk_size} \
        {params.cap} \
        1> {log.stdout} \
        2> {log.stderr} && \
        touch {params.out}
        """


rule cv_coverage_background:
    input:
        bed=rules.pas_to_peka_beds.output.bg_bed,
        genome=os.path.abspath(config["genome"]),
        genome_idx=os.path.abspath(config["genome_idx"])
    
    output:
         os.path.join(base_outdir, "{sample}", "cv_coverage_background", "cv_coverage.done")
    
    params:
        script=os.path.abspath("scripts/cv_coverage.py"),
        bed=os.path.abspath(rules.pas_to_peka_beds.output.bg_bed),
        motifs=",".join(config["cv_coverage_motifs"]),
        kmer=config["kmer"],
        percentile=config["percentile"],
        regions=os.path.abspath(config["regions_gtf"]),
        specific_region=config["specific_region"],
        window=config["distal_window"],
        smoothing=config["cv_coverage_smoothing"],
        use_scores=config["cv_coverage_usescores"],
        cap=config["cv_coverage_cap"],
        chunk_size=10000,
        outdir=os.path.join(base_outdir, "{sample}", "cv_coverage_background"),
        tmpdir=os.path.abspath(os.path.join(base_outdir, "{sample}", "cv_coverage_background", "tmp")),
        out="cv_coverage.done"

    threads:
        config["cv_coverage_n_cores"]


    log:
        stdout=os.path.abspath(os.path.join(base_outdir, "logs", "cv_coverage.{sample}.stdout.txt")),
        stderr=os.path.abspath(os.path.join(base_outdir, "logs", "cv_coverage.{sample}.stderr.txt"))

    conda:
        "envs/peka_fork.yaml"


    shell:
        """
        mkdir -p {params.tmpdir} && export TMPDIR={params.tmpdir} && cd {params.outdir}
        python {params.script} \
        {params.bed} \
        {params.motifs} \
        {params.specific_region} \
        {params.kmer} \
        {input.genome} \
        {input.genome_idx} \
        {params.regions} \
        {params.smoothing} \
        {params.percentile} \
        {params.window} \
        {params.use_scores} \
        {threads} \
        {params.chunk_size} \
        {params.cap} \
        1> {log.stdout} \
        2> {log.stderr} && \
        touch {params.out}
        """