{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from helpers import get_terminal_regions, _df_update_3p, _df_update_5p, add_region_number\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16097</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16098</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16099</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16100</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16101 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 16,101 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Zeng gene names\n",
    "\n",
    "outdir = \"processed/isoform_specific/regions/\"\n",
    "path_zeng_bed = os.path.join(outdir, \"zeng_pas_fig2c.updated.bed.tsv\")\n",
    "zeng_bed = pr.PyRanges(pd.read_csv(path_zeng_bed, sep=\"\\t\"))\n",
    "zeng_gn = set(zeng_bed.gene_name_common)\n",
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of PAPA cryptics - 284\n",
      "{'ENSG00000164609.10_5', 'ENSG00000153094.24_3', 'ENSG00000134909.19_1', 'ENSG00000261221.4_1', 'ENSG00000146733.14_1', 'ENSG00000118922.18_1', 'ENSG00000119509.13_2', 'ENSG00000244405.8_3', 'ENSG00000185352.9_1', 'ENSG00000247708.8_2', 'ENSG00000136144.12_3', 'ENSG00000214562.15_2', 'ENSG00000247199.6_1', 'ENSG00000149577.16_3', 'ENSG00000126767.18_1', 'ENSG00000135090.14_5', 'ENSG00000184347.15_7', 'ENSG00000111450.14_1', 'ENSG00000151692.15_6', 'ENSG00000181090.21_5', 'ENSG00000155897.10_1', 'ENSG00000155093.19_2', 'ENSG00000132485.14_3', 'ENSG00000107960.11_1', 'ENSG00000128274.17_2', 'ENSG00000105926.16_1', 'ENSG00000114904.13_3', 'ENSG00000004866.22_2', 'ENSG00000154319.16_2', 'ENSG00000153094.24_2', 'ENSG00000281566.3_1', 'ENSG00000198797.7_1', 'ENSG00000254602.2_1', 'ENSG00000140750.17_4', 'ENSG00000148814.18_3', 'ENSG00000105948.13_3', 'ENSG00000065526.12_4', 'ENSG00000242686.4_2', 'ENSG00000125676.20_2', 'ENSG00000032219.19_2', 'ENSG00000132394.11_3', 'ENSG00000108587.16_4', 'ENSG00000127995.17_2', 'ENSG00000138802.11_1', 'ENSG00000198797.7_2', 'ENSG00000153558.16_4', 'ENSG00000138083.5_1', 'ENSG00000105808.19_5', 'ENSG00000068784.13_2', 'ENSG00000131375.10_2', 'ENSG00000101624.11_3', 'ENSG00000157500.12_4', 'ENSG00000102678.7_2', 'ENSG00000156171.15_4', 'ENSG00000179796.12_1', 'ENSG00000254602.2_2', 'ENSG00000135953.11_3', 'ENSG00000253741.3_2', 'ENSG00000186908.15_2', 'ENSG00000156171.15_3', 'ENSG00000091009.8_1', 'ENSG00000095203.15_2', 'ENSG00000150760.13_2', 'ENSG00000108587.16_3', 'ENSG00000203485.14_3', 'ENSG00000244115.1_2', 'ENSG00000263069.6_2', 'ENSG00000165633.13_1', 'ENSG00000196689.13_1', 'ENSG00000183570.17_3', 'ENSG00000109472.14_2', 'ENSG00000153558.16_3', 'ENSG00000127511.10_2', 'ENSG00000204314.12_2', 'ENSG00000121578.13_3', 'ENSG00000213020.10_1', 'ENSG00000265808.4_2', 'ENSG00000128487.19_1', 'ENSG00000173818.17_2', 'ENSG00000131375.10_3', 'ENSG00000028310.18_3', 'ENSG00000107960.11_2', 'ENSG00000126767.18_2', 'ENSG00000118257.17_5', 'ENSG00000196275.16_6', 'ENSG00000068383.19_2', 'ENSG00000188191.15_3', 'ENSG00000136144.12_2', 'ENSG00000261879.8_2', 'ENSG00000100299.18_1', 'ENSG00000162849.16_2', 'ENSG00000115839.19_2', 'ENSG00000171853.16_4', 'ENSG00000124370.11_2', 'ENSG00000105808.19_4', 'ENSG00000186615.12_2', 'ENSG00000007047.16_3', 'ENSG00000103404.15_1', 'ENSG00000180182.11_3', 'ENSG00000105186.16_3', 'ENSG00000121578.13_2', 'ENSG00000120594.17_1', 'ENSG00000120868.14_2', 'ENSG00000145349.18_3', 'ENSG00000108599.15_3', 'ENSG00000114541.15_3', 'ENSG00000158169.14_3', 'ENSG00000166135.14_3', 'ENSG00000109118.14_1', 'ENSG00000186908.15_3', 'ENSG00000164849.10_1', 'ENSG00000128708.13_2', 'ENSG00000213468.7_1', 'ENSG00000007047.16_2', 'ENSG00000149050.10_1', 'ENSG00000115665.9_2', 'ENSG00000181722.17_5', 'ENSG00000107798.18_2', 'ENSG00000179965.12_1', 'ENSG00000135953.11_2', 'ENSG00000092140.16_1', 'ENSG00000100557.10_4', 'ENSG00000231312.8_4', 'ENSG00000152954.12_1', 'ENSG00000130226.17_3', 'ENSG00000139767.10_2', 'ENSG00000173208.4_1', 'ENSG00000131831.18_1', 'ENSG00000235173.7_2', 'ENSG00000028310.18_2', 'ENSG00000078269.15_1', 'ENSG00000186615.12_3', 'ENSG00000105948.13_4', 'ENSG00000127995.17_1', 'ENSG00000171984.15_2', 'ENSG00000091009.8_2', 'ENSG00000188026.13_1', 'ENSG00000066032.19_7', 'ENSG00000154319.16_1', 'ENSG00000242686.4_1', 'ENSG00000148120.18_5', 'ENSG00000078124.13_1', 'ENSG00000026652.15_3', 'ENSG00000203546.7,ENSG00000129480.13_2', 'ENSG00000099381.19_2', 'ENSG00000196689.13_2', 'ENSG00000138802.11_2', 'ENSG00000132849.22_1', 'ENSG00000109472.14_3', 'ENSG00000183715.14_3', 'ENSG00000277363.5_2', 'ENSG00000162813.18_3', 'ENSG00000128513.16_4', 'ENSG00000244115.1_1', 'ENSG00000101412.13_1', 'ENSG00000138083.5_2', 'ENSG00000232995.7_2', 'ENSG00000137161.18_1', 'ENSG00000107807.13_2', 'ENSG00000102290.23_3', 'ENSG00000255545.8_2', 'ENSG00000117020.19_2', 'ENSG00000099381.19_3', 'ENSG00000197724.11_1', 'ENSG00000142794.19_3', 'ENSG00000169856.9_2', 'ENSG00000171094.18_1', 'ENSG00000144460.13_2', 'ENSG00000138138.14_2', 'ENSG00000261221.4_2', 'ENSG00000246877.2_2', 'ENSG00000185736.16_1', 'ENSG00000128513.16_3', 'ENSG00000173818.17_5', 'ENSG00000115665.9_1', 'ENSG00000075035.10_3', 'ENSG00000072756.17_1', 'ENSG00000184178.17_1', 'ENSG00000108599.15_4', 'ENSG00000132394.11_2', 'ENSG00000109534.17_4', 'ENSG00000244558.6_1', 'ENSG00000164741.15_3', 'ENSG00000173085.15_2', 'ENSG00000169856.9_1', 'ENSG00000170579.17_1', 'ENSG00000173085.15_1', 'ENSG00000066032.19_6', 'ENSG00000197497.11_3', 'ENSG00000113448.20_2', 'ENSG00000187951.11_3', 'ENSG00000128805.15_4', 'ENSG00000263069.6_3', 'ENSG00000163964.18_2', 'ENSG00000263956.8_3', 'ENSG00000137936.18_1', 'ENSG00000104435.14_1', 'ENSG00000100299.18_2', 'ENSG00000157500.12_3', 'ENSG00000138138.14_1', 'ENSG00000197563.11_3', 'ENSG00000214562.15_1', 'ENSG00000117069.15_2', 'ENSG00000164038.16_1', 'ENSG00000066379.15_2', 'ENSG00000151692.15_7', 'ENSG00000109534.17_3', 'ENSG00000129158.11_4', 'ENSG00000095203.15_3', 'ENSG00000066379.15_1', 'ENSG00000081760.17_1', 'ENSG00000145819.18_3', 'ENSG00000263956.8_2', 'ENSG00000139793.20_1', 'ENSG00000065457.11_1', 'ENSG00000101412.13_2', 'ENSG00000128274.17_1', 'ENSG00000035141.8_3', 'ENSG00000135090.14_4', 'ENSG00000228824.8_4', 'ENSG00000228824.8_3', 'ENSG00000158604.15_1', 'ENSG00000163964.18_3', 'ENSG00000068784.13_1', 'ENSG00000122783.17_4', 'ENSG00000115392.13_1', 'ENSG00000216895.9_4', 'ENSG00000100557.10_3', 'ENSG00000122783.17_3', 'ENSG00000235173.7_1', 'ENSG00000119912.18_3', 'ENSG00000048649.14_1', 'ENSG00000122254.7_1', 'ENSG00000065457.11_2', 'ENSG00000166135.14_2', 'ENSG00000105926.16_2', 'ENSG00000135315.12_2', 'ENSG00000204764.14_9', 'ENSG00000032219.19_1', 'ENSG00000121904.19_3', 'ENSG00000162813.18_2', 'ENSG00000139767.10_3', 'ENSG00000140750.17_5', 'ENSG00000162944.11_1', 'ENSG00000179796.12_2', 'ENSG00000167264.18_1', 'ENSG00000122254.7_2', 'ENSG00000204314.12_1', 'ENSG00000174165.8_3', 'ENSG00000149577.16_4', 'ENSG00000065526.12_3', 'ENSG00000107807.13_1', 'ENSG00000117151.13_1', 'ENSG00000075035.10_4', 'ENSG00000128805.15_2', 'ENSG00000124370.11_1', 'ENSG00000073605.19_4', 'ENSG00000119912.18_4', 'ENSG00000125676.20_3', 'ENSG00000128708.13_3', 'ENSG00000164609.10_4', 'ENSG00000102678.7_1', 'ENSG00000187260.16_1', 'ENSG00000162390.18_4', 'ENSG00000119392.16_1', 'ENSG00000198908.12_1', 'ENSG00000135315.12_1', 'ENSG00000181090.21_6', 'ENSG00000203485.14_4', 'ENSG00000253741.3_1', 'ENSG00000026652.15_2', 'ENSG00000169856.9_3', 'ENSG00000118922.18_2', 'ENSG00000100060.18_2', 'ENSG00000132485.14_2', 'ENSG00000186047.11_2', 'ENSG00000265808.4_3', 'ENSG00000073605.19_5', 'ENSG00000227110.7_3', 'ENSG00000183715.14_2', 'ENSG00000187951.11_2', 'ENSG00000197565.17_1', 'ENSG00000035141.8_2', 'ENSG00000155093.19_3'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177282422</td>\n",
       "      <td>ENSG00000198797.7_1|BRINP2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600573</td>\n",
       "      <td>123601338</td>\n",
       "      <td>ENSG00000125676.20_3|THOC2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940456  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 177280411 | 177282422 | ...   |\n",
       "| chr1         | 177280411 | 177284690 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 17835910  | 17837395  | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 40653641  | 40654363  | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 284 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load in BED file of PAPA cryptics (to get gene names and ids)\n",
    "papa_cryp_bed = pr.read_bed(\"data/2023-12-15_all.last_exons.cryptic.bed\")\n",
    "\n",
    "# get le_ids for cryptics (+ their annotated partners)\n",
    "# e.g. ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic\t\n",
    "papa_name_spl = papa_cryp_bed.Name.str.split(\"\\\\|\", expand=True)\n",
    "cryp_le_ids = set(papa_name_spl[0])\n",
    "cryp_gn = set(papa_name_spl[1])\n",
    "print(f\"Number of PAPA cryptics - {len(cryp_le_ids)}\")\n",
    "print(cryp_le_ids)\n",
    "papa_cryp_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24419640</td>\n",
       "      <td>24420128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419290</td>\n",
       "      <td>24419640</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>488.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000001461.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24421813</td>\n",
       "      <td>24422110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419640,24419640,24419640</td>\n",
       "      <td>24440171,24440171,24433103</td>\n",
       "      <td>ENST00000358028.8,ENST00000374399.9,ENST000000...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000001461.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24454053</td>\n",
       "      <td>24454824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24453504,24453504</td>\n",
       "      <td>24456137,24456137</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000001461.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24464120</td>\n",
       "      <td>24466378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464025</td>\n",
       "      <td>24464120</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>2258.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000001461.17_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880307</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21039044</td>\n",
       "      <td>21042268</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>23936831</td>\n",
       "      <td>23939185</td>\n",
       "      <td>ENST00000631331.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129930 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24419640  | 24420128  | ...   |\n",
       "| chr1         | .          | exon         | 24421813  | 24422110  | ...   |\n",
       "| chr1         | .          | exon         | 24454053  | 24454824  | ...   |\n",
       "| chr1         | .          | exon         | 24464120  | 24466378  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 21880075  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21880307  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21038288  | 21039044  | ...   |\n",
       "| chrY         | .          | exon         | 23936726  | 23936831  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 129,930 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combined reference of novel + annotated last exons used for quantification\n",
    "papa_gtf = pr.read_gtf(\"data/novel_ref_combined.quant.last_exons.gtf\")\n",
    "papa_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24419640</td>\n",
       "      <td>24420128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419290</td>\n",
       "      <td>24419640</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>488.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000001461.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24421813</td>\n",
       "      <td>24422110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419640,24419640,24419640</td>\n",
       "      <td>24440171,24440171,24433103</td>\n",
       "      <td>ENST00000358028.8,ENST00000374399.9,ENST000000...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000001461.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24454053</td>\n",
       "      <td>24454824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24453504,24453504</td>\n",
       "      <td>24456137,24456137</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000001461.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24464120</td>\n",
       "      <td>24466378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464025</td>\n",
       "      <td>24464120</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>2258.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000001461.17_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880307</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21039044</td>\n",
       "      <td>21042268</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>23936831</td>\n",
       "      <td>23939185</td>\n",
       "      <td>ENST00000631331.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129930 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24419640  | 24420128  | ...   |\n",
       "| chr1         | .          | exon         | 24421813  | 24422110  | ...   |\n",
       "| chr1         | .          | exon         | 24454053  | 24454824  | ...   |\n",
       "| chr1         | .          | exon         | 24464120  | 24466378  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 21880075  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21880307  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21038288  | 21039044  | ...   |\n",
       "| chrY         | .          | exon         | 23936726  | 23936831  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 129,930 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some gene names are duplicated in the ref_gene_name column\n",
    "\n",
    "def remove_duplicates_preserve_order(s):\n",
    "    seen = set()\n",
    "    result = []\n",
    "    for item in s.split(','):\n",
    "        if item not in seen:\n",
    "            seen.add(item)\n",
    "            result.append(item)\n",
    "    return ','.join(result)\n",
    "\n",
    "# Apply the function to the series\n",
    "papa_gtf.ref_gene_name = papa_gtf.ref_gene_name.apply(remove_duplicates_preserve_order)\n",
    "papa_gtf\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_id</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11820361</td>\n",
       "      <td>11820779</td>\n",
       "      <td>+</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>ENSG00000011021.23_1</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENST00000376497.7</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENST00000312413.10</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11841207</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENST00000312413.10</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENST00000346436.11</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11841207</td>\n",
       "      <td>11843130</td>\n",
       "      <td>+</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>ENST00000346436.11</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65420</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENST00000382653.6</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65421</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENST00000382680.5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65422</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880307</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>ENST00000418956.2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65423</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65424</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>ENST00000631331.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65425 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 11820361  | 11820779  | +            | ...   |\n",
       "| chr1         | 11840371  | 11841016  | +            | ...   |\n",
       "| chr1         | 11841207  | 11843143  | +            | ...   |\n",
       "| chr1         | 11840371  | 11841016  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 21880075  | 21880652  | -            | ...   |\n",
       "| chrY         | 21880307  | 21880652  | -            | ...   |\n",
       "| chrY         | 21038288  | 21039044  | -            | ...   |\n",
       "| chrY         | 23936726  | 23936831  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 65,425 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: ref_gene_name, le_id, event_type, ref_gene_id, gene_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# exclude cryptic genes or those input from quant-seq\n",
    "papa_gtf_filt = papa_gtf.subset(lambda df: ~df.ref_gene_name.isin(cryp_gn.union(zeng_gn)))[[\"ref_gene_name\", \"le_id\", \"event_type\", \"ref_gene_id\", \"gene_id\", \"transcript_id\", \"le_number\"]]\n",
    "papa_gtf_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673163</td>\n",
       "      <td>196673856</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673962</td>\n",
       "      <td>196675913</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196701317</td>\n",
       "      <td>196701566</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196747110</td>\n",
       "      <td>196747504</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171099804</td>\n",
       "      <td>171100735</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000007933.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32367</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21904194</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000242389.9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32368</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24183433</td>\n",
       "      <td>24183505</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244231.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32369</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32370</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32371</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32372 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+----------------------+\n",
       "| Chromosome   | Start     | End       | Strand       | le_id                |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)             |\n",
       "|--------------+-----------+-----------+--------------+----------------------|\n",
       "| chr1         | 196673163 | 196673856 | +            | ENSG00000000971.17_1 |\n",
       "| chr1         | 196673962 | 196675913 | +            | ENSG00000000971.17_1 |\n",
       "| chr1         | 196701317 | 196701566 | +            | ENSG00000000971.17_2 |\n",
       "| chr1         | 196747110 | 196747504 | +            | ENSG00000000971.17_3 |\n",
       "| ...          | ...       | ...       | ...          | ...                  |\n",
       "| chrY         | 24183433  | 24183505  | -            | ENSG00000244231.1_1  |\n",
       "| chrY         | 21880075  | 21880652  | -            | ENSG00000244395.6_1  |\n",
       "| chrY         | 21038288  | 21039044  | -            | ENSG00000254488.1_1  |\n",
       "| chrY         | 23936726  | 23936831  | -            | ENSG00000280961.1_1  |\n",
       "+--------------+-----------+-----------+--------------+----------------------+\n",
       "Stranded PyRanges object has 32,372 rows and 5 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create merged intervals per last exon ID\n",
    "merged_le_id = papa_gtf_filt.merge(strand=True, by=\"le_id\")\n",
    "merged_le_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>le_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673163</td>\n",
       "      <td>196673856</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "      <td>CFH</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673163</td>\n",
       "      <td>196673856</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "      <td>CFH</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673962</td>\n",
       "      <td>196675913</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "      <td>CFH</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196673962</td>\n",
       "      <td>196675913</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_1</td>\n",
       "      <td>CFH</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196701317</td>\n",
       "      <td>196701566</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.17_2</td>\n",
       "      <td>CFH</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74369</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74370</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74371</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74372</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74373</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74374 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 196673163 | 196673856 | +            | ...   |\n",
       "| chr1         | 196673163 | 196673856 | +            | ...   |\n",
       "| chr1         | 196673962 | 196675913 | +            | ...   |\n",
       "| chr1         | 196673962 | 196675913 | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 21880075  | 21880652  | -            | ...   |\n",
       "| chrY         | 21880075  | 21880652  | -            | ...   |\n",
       "| chrY         | 21038288  | 21039044  | -            | ...   |\n",
       "| chrY         | 23936726  | 23936831  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 74,374 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: le_id, ref_gene_name, event_type, le_number"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# return minimal metadata to each i\n",
    "metadata_le_id = papa_gtf_filt.as_df()[[\"le_id\", \"ref_gene_name\", \"event_type\", \"le_number\"]].astype({\"le_number\": float}).astype({\"le_number\": 'int32'})\n",
    "merged_le_id = merged_le_id.apply(lambda df: df.merge(metadata_le_id, on=\"le_id\", how=\"left\"))\n",
    "merged_le_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3'Exts - generate distal bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1563"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get genes with distal 3'utr extensions\n",
    "# extract last exons, then check event types\n",
    "merged_le_id_le = get_terminal_regions(merged_le_id.assign(\"Feature\", lambda df: pd.Series([\"exon\"]*len(df), index=df.index)),\n",
    "                                       id_col=\"ref_gene_name\", region_number_col=\"le_number\")\n",
    "\n",
    "novel_3ext_gn = set(merged_le_id_le.subset(lambda df: df.event_type == \"last_exon_extension\").ref_gene_name)\n",
    "len(novel_3ext_gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>le_number</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63436822</td>\n",
       "      <td>63440437</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000088035.18_3</td>\n",
       "      <td>ALG6</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150995259</td>\n",
       "      <td>150995885</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000143412.10_2</td>\n",
       "      <td>ANXA9</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190480424</td>\n",
       "      <td>190481783</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237457.6_2</td>\n",
       "      <td>BRINP3-DT</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>178547163</td>\n",
       "      <td>178551764</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000213057.5_2</td>\n",
       "      <td>C1orf220</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1559</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155491666</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.12_5</td>\n",
       "      <td>TMLHE</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>5</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1560</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101006828</td>\n",
       "      <td>101010707</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188917.15_2</td>\n",
       "      <td>TRMT2B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1561</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47444691</td>\n",
       "      <td>47449474</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000147124.13_2</td>\n",
       "      <td>ZNF41</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>305197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1564 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 63436822  | 63440437  | +            | ...   |\n",
       "| chr1         | 150995259 | 150995885 | +            | ...   |\n",
       "| chr1         | 1495484   | 1498495   | +            | ...   |\n",
       "| chr1         | 190480424 | 190481783 | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 101341046 | 101346660 | -            | ...   |\n",
       "| chrX         | 155478419 | 155491666 | -            | ...   |\n",
       "| chrX         | 101006828 | 101010707 | -            | ...   |\n",
       "| chrX         | 47444691  | 47449474  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,564 rows and 9 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: le_id, ref_gene_name, event_type, le_number, Feature"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to genes with 3'exts\n",
    "merged_le_id_3ext = merged_le_id.subset(lambda df: df.ref_gene_name.isin(novel_3ext_gn))\n",
    "\n",
    "# extract last_exons\n",
    "merged_le_id_3ext_le = get_terminal_regions(merged_le_id_3ext.assign(\"Feature\", lambda df: pd.Series([\"exon\"]*len(df), index=df.index)), id_col=\"ref_gene_name\", region_number_col=\"le_number\")\n",
    "\n",
    "# extract non-last exons\n",
    "merged_le_id_3ext_nles = merged_le_id_3ext.subset(lambda df: ~df.le_id.isin(set(merged_le_id_3ext_les.le_id)))\n",
    "\n",
    "merged_le_id_3ext_le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>le_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11820361</td>\n",
       "      <td>11820779</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23_1</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840371</td>\n",
       "      <td>11841016</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23_2</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>first_exon_extension</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9095</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3817527</td>\n",
       "      <td>3817778</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000285756.2_6</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>304528</td>\n",
       "      <td>305197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_1</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9097</th>\n",
       "      <td>chrY</td>\n",
       "      <td>304528</td>\n",
       "      <td>305197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_1</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9098</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17574730</td>\n",
       "      <td>17575367</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_1</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9099</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17552799</td>\n",
       "      <td>17553954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_2</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9100 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 11820361  | 11820779  | +            | ...   |\n",
       "| chr1         | 11840371  | 11841016  | +            | ...   |\n",
       "| chr1         | 11840371  | 11841016  | +            | ...   |\n",
       "| chr1         | 11840371  | 11841016  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 304528    | 305197    | -            | ...   |\n",
       "| chrY         | 304528    | 305197    | -            | ...   |\n",
       "| chrY         | 17574730  | 17575367  | -            | ...   |\n",
       "| chrY         | 17552799  | 17553954  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 9,100 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: le_id, ref_gene_name, event_type, le_number"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_le_id_3ext_nles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>event_type</th>\n",
       "      <th>le_number</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63437133</td>\n",
       "      <td>63440437</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000088035.18_3</td>\n",
       "      <td>ALG6</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150995634</td>\n",
       "      <td>150995885</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000143412.10_2</td>\n",
       "      <td>ANXA9</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1497848</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190480735</td>\n",
       "      <td>190481783</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237457.6_2</td>\n",
       "      <td>BRINP3-DT</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>178548889</td>\n",
       "      <td>178551764</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000213057.5_2</td>\n",
       "      <td>C1orf220</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1485</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155478419</td>\n",
       "      <td>155489010</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.12_5</td>\n",
       "      <td>TMLHE</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>5</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1486</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101006828</td>\n",
       "      <td>101009345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188917.15_2</td>\n",
       "      <td>TRMT2B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1487</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47444691</td>\n",
       "      <td>47445878</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000147124.13_2</td>\n",
       "      <td>ZNF41</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1488</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17552799</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1489</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>304528</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1490 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 63437133  | 63440437  | +            | ...   |\n",
       "| chr1         | 150995634 | 150995885 | +            | ...   |\n",
       "| chr1         | 1497848   | 1498495   | +            | ...   |\n",
       "| chr1         | 190480735 | 190481783 | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 101341046 | 101345042 | -            | ...   |\n",
       "| chrX         | 155478419 | 155489010 | -            | ...   |\n",
       "| chrX         | 101006828 | 101009345 | -            | ...   |\n",
       "| chrX         | 47444691  | 47445878  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,490 rows and 9 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: le_id, ref_gene_name, event_type, le_number, Feature"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract distal bins\n",
    "# define distal specific region (ONLY for UTRs)\n",
    "merged_le_id_3ext_les_uniq = merged_le_id_3ext_le.subtract(merged_le_id_3ext_nles, strandedness=\"same\")\n",
    "merged_le_id_3ext_les_uniq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse regions into featureCounts compliant GTF\n",
    "\n",
    "- each interval reported with 'exon', 'gene' and 'transcript' assigned to Feature column (i.e. reported 3*)\n",
    "- gene_id, transcript_id and le_id minimal attributes ()\n",
    "  - assign le_id to all, except for 3'exts where add a suffix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3'exts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1497848</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1497848</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1497848</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>ENSG00000160072.20_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2505532</td>\n",
       "      <td>2505811</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>PLCH2</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000149527.18_2__distal</td>\n",
       "      <td>ENSG00000149527.18_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2505532</td>\n",
       "      <td>2505811</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>PLCH2</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000149527.18_2__distal</td>\n",
       "      <td>ENSG00000149527.18_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4465</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>304528</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2__distal</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4466</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>304528</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2__distal</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4467</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17552799</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4468</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17552799</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4469</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17552799</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>ENSG00000230663.1_3__distal</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4470 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1497848   | 1498495   | +            | ...   |\n",
       "| chr1         | 1497848   | 1498495   | +            | ...   |\n",
       "| chr1         | 1497848   | 1498495   | +            | ...   |\n",
       "| chr1         | 2505532   | 2505811   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 302836    | 304528    | -            | ...   |\n",
       "| chrY         | 17551679  | 17552799  | -            | ...   |\n",
       "| chrY         | 17551679  | 17552799  | -            | ...   |\n",
       "| chrY         | 17551679  | 17552799  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 4,470 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: le_id, gene_name, le_number, Feature, gene_id, transcript_id, ... (+ 1 more.)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# distal specific bins = le_id + suffix\n",
    "distal_suffix = \"__distal\"\n",
    "merged_le_id_3ext_les_uniq.gene_id = merged_le_id_3ext_les_uniq.le_id + distal_suffix\n",
    "merged_le_id_3ext_les_uniq.transcript_id = merged_le_id_3ext_les_uniq.gene_id\n",
    "merged_le_id_3ext_les_uniq.type = merged_le_id_3ext_les_uniq.Feature\n",
    "merged_le_id_3ext_les_uniq = merged_le_id_3ext_les_uniq.drop(\"event_type\").apply(lambda df: df.rename(columns={\"ref_gene_name\": \"gene_name\"}))\n",
    "\n",
    "# add entries with transcript and gene in the feature column\n",
    "grs = [merged_le_id_3ext_les_uniq.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "merged_le_id_3ext_les_uniqs = pr.concat([merged_le_id_3ext_les_uniq, *grs]).sort()\n",
    "merged_le_id_3ext_les_uniqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1495484</td>\n",
       "      <td>1498495</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>ENSG00000160072.20_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2503921</td>\n",
       "      <td>2505811</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>PLCH2</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2503921</td>\n",
       "      <td>2505811</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>PLCH2</td>\n",
       "      <td>2</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>ENSG00000149527.18_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4687</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>305197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4688</th>\n",
       "      <td>chrY</td>\n",
       "      <td>302836</td>\n",
       "      <td>305197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>ENSG00000178605.13_PAR_Y_2</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4689</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4690</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4691</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17551679</td>\n",
       "      <td>17553954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>FAM224B</td>\n",
       "      <td>3</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>ENSG00000230663.1_3</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4692 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1495484   | 1498495   | +            | ...   |\n",
       "| chr1         | 1495484   | 1498495   | +            | ...   |\n",
       "| chr1         | 1495484   | 1498495   | +            | ...   |\n",
       "| chr1         | 2503921   | 2505811   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 302836    | 305197    | -            | ...   |\n",
       "| chrY         | 17551679  | 17553954  | -            | ...   |\n",
       "| chrY         | 17551679  | 17553954  | -            | ...   |\n",
       "| chrY         | 17551679  | 17553954  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 4,692 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: le_id, gene_name, le_number, Feature, gene_id, transcript_id, ... (+ 1 more.)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# full 3'UTRs\n",
    "merged_le_id_3ext_le.gene_id = merged_le_id_3ext_le.le_id\n",
    "merged_le_id_3ext_le.transcript_id = merged_le_id_3ext_le.le_id\n",
    "merged_le_id_3ext_le = merged_le_id_3ext_le.drop(\"event_type\").apply(lambda df: df.rename(columns={\"ref_gene_name\": \"gene_name\"}))\n",
    "merged_le_id_3ext_le.type = merged_le_id_3ext_le.Feature\n",
    "merged_le_id_3ext_le\n",
    "\n",
    "# add entries with transcript and gene in the feature column\n",
    "grs = [merged_le_id_3ext_le.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "merged_le_id_3ext_les = pr.concat([merged_le_id_3ext_le, *grs]).sort()\n",
    "merged_le_id_3ext_les"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All other events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177712</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25378299</td>\n",
       "      <td>25378878</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>LINC00266-4P</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177713</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25378299</td>\n",
       "      <td>25378878</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>LINC00266-4P</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177714</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177715</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177716</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>177717 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 365170    | 366052    | +            | ...   |\n",
       "| chr1         | 365170    | 366052    | +            | ...   |\n",
       "| chr1         | 365170    | 366052    | +            | ...   |\n",
       "| chr1         | 365170    | 366052    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 25378299  | 25378878  | -            | ...   |\n",
       "| chrY         | 57201142  | 57202020  | -            | ...   |\n",
       "| chrY         | 57201142  | 57202020  | -            | ...   |\n",
       "| chrY         | 57201142  | 57202020  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 177,717 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: le_id, gene_name, le_number, gene_id, transcript_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_le_id_other = merged_le_id.subset(lambda df: ~df.ref_gene_name.isin(novel_3ext_gn))\n",
    "merged_le_id_other\n",
    "\n",
    "# full 3'UTRs\n",
    "merged_le_id_other.gene_id = merged_le_id_other.le_id\n",
    "merged_le_id_other.transcript_id = merged_le_id_other.le_id\n",
    "merged_le_id_other = merged_le_id_other.drop(\"event_type\").apply(lambda df: df.rename(columns={\"ref_gene_name\": \"gene_name\"}))\n",
    "merged_le_id_other.Feature = \"exon\"\n",
    "merged_le_id_other.type = merged_le_id_other.Feature\n",
    "\n",
    "# add entries with transcript and gene in the feature column\n",
    "grs = [merged_le_id_other.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "merged_le_id_others = pr.concat([merged_le_id_other, *grs]).sort()\n",
    "merged_le_id_others"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine all and output to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>365170</td>\n",
       "      <td>366052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>ENSG00000236601.2_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>594234</td>\n",
       "      <td>594574</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235146.2_1</td>\n",
       "      <td>ENSG00000235146</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000235146.2_1</td>\n",
       "      <td>ENSG00000235146.2_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>804966</td>\n",
       "      <td>805270</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237491.10_1</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237491.10_1</td>\n",
       "      <td>ENSG00000237491.10_1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>807323</td>\n",
       "      <td>808378</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237491.10_2</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000237491.10_2</td>\n",
       "      <td>ENSG00000237491.10_2</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>809621</td>\n",
       "      <td>810063</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237491.10_3</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000237491.10_3</td>\n",
       "      <td>ENSG00000237491.10_3</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30638</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25030900</td>\n",
       "      <td>25031222</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185894.8_1</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000185894.8_1</td>\n",
       "      <td>ENSG00000185894.8_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30639</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25063082</td>\n",
       "      <td>25064083</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228296.1_1</td>\n",
       "      <td>TTTY4C</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228296.1_1</td>\n",
       "      <td>ENSG00000228296.1_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30640</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25182276</td>\n",
       "      <td>25182587</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223641.2_1</td>\n",
       "      <td>TTTY17C</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000223641.2_1</td>\n",
       "      <td>ENSG00000223641.2_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30641</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25378299</td>\n",
       "      <td>25378878</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>LINC00266-4P</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>ENSG00000228786.5_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30642</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y_1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30643 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 365170    | 366052    | +            | ...   |\n",
       "| chr1         | 594234    | 594574    | +            | ...   |\n",
       "| chr1         | 804966    | 805270    | +            | ...   |\n",
       "| chr1         | 807323    | 808378    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 25063082  | 25064083  | -            | ...   |\n",
       "| chrY         | 25182276  | 25182587  | -            | ...   |\n",
       "| chrY         | 25378299  | 25378878  | -            | ...   |\n",
       "| chrY         | 57201142  | 57202020  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 30,643 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: le_id, gene_name, le_number, gene_id, transcript_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_gtf_out = pr.concat([merged_le_id_others, merged_le_id_3ext_les, merged_le_id_3ext_les_uniqs]).sort().drop_duplicate_positions()\n",
    "comb_gtf_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_gtf_path = os.path.join(outdir, \"papa_other_feature_counts.gtf\")\n",
    "comb_gtf_out.to_gtf(out_gtf_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
