{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from helpers import get_terminal_regions, _df_update_3p, _df_update_5p, add_region_number\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARHGAP32', 'CNPY3', 'RFNG', 'SIX3', 'STMN2', 'TLX1', 'SLIT3']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prioritised Zeng et al. target genes with PAPA annotations\n",
    "zeng_targets_papa = \"ARHGAP32 CNPY3 RFNG SIX3 STMN2 TLX1 SLIT3\".split(\" \")\n",
    "zeng_targets_papa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16097</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16098</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16099</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16100</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16101 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 16,101 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outdir = \"processed/isoform_specific/regions/\"\n",
    "zeng_bed = pr.PyRanges(pd.read_csv(os.path.join(outdir, \"zeng_pas_fig2c.updated.bed.tsv\"), sep=\"\\t\"))\n",
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of PAPA cryptics - 284\n",
      "{'ENSG00000139767.10_3', 'ENSG00000128487.19_1', 'ENSG00000162849.16_2', 'ENSG00000131375.10_2', 'ENSG00000164038.16_1', 'ENSG00000171094.18_1', 'ENSG00000138802.11_2', 'ENSG00000102678.7_2', 'ENSG00000101412.13_2', 'ENSG00000065457.11_1', 'ENSG00000126767.18_2', 'ENSG00000135315.12_2', 'ENSG00000213020.10_1', 'ENSG00000128708.13_3', 'ENSG00000197497.11_3', 'ENSG00000132394.11_3', 'ENSG00000253741.3_2', 'ENSG00000091009.8_1', 'ENSG00000247199.6_1', 'ENSG00000169856.9_3', 'ENSG00000188191.15_3', 'ENSG00000166135.14_3', 'ENSG00000100299.18_1', 'ENSG00000109534.17_4', 'ENSG00000265808.4_3', 'ENSG00000117151.13_1', 'ENSG00000198797.7_2', 'ENSG00000075035.10_3', 'ENSG00000135315.12_1', 'ENSG00000118922.18_1', 'ENSG00000204314.12_2', 'ENSG00000100557.10_4', 'ENSG00000166135.14_2', 'ENSG00000186908.15_2', 'ENSG00000163964.18_2', 'ENSG00000204764.14_9', 'ENSG00000026652.15_3', 'ENSG00000186615.12_2', 'ENSG00000169856.9_1', 'ENSG00000197563.11_3', 'ENSG00000075035.10_4', 'ENSG00000214562.15_2', 'ENSG00000105186.16_3', 'ENSG00000261221.4_2', 'ENSG00000066032.19_6', 'ENSG00000183715.14_3', 'ENSG00000032219.19_1', 'ENSG00000180182.11_3', 'ENSG00000124370.11_2', 'ENSG00000048649.14_1', 'ENSG00000134909.19_1', 'ENSG00000102678.7_1', 'ENSG00000153094.24_3', 'ENSG00000128805.15_4', 'ENSG00000263069.6_3', 'ENSG00000091009.8_2', 'ENSG00000028310.18_3', 'ENSG00000203546.7,ENSG00000129480.13_2', 'ENSG00000122254.7_2', 'ENSG00000179965.12_1', 'ENSG00000100299.18_2', 'ENSG00000156171.15_3', 'ENSG00000102290.23_3', 'ENSG00000246877.2_2', 'ENSG00000128708.13_2', 'ENSG00000105808.19_4', 'ENSG00000184347.15_7', 'ENSG00000216895.9_4', 'ENSG00000231312.8_4', 'ENSG00000155093.19_3', 'ENSG00000253741.3_1', 'ENSG00000135090.14_4', 'ENSG00000115392.13_1', 'ENSG00000155897.10_1', 'ENSG00000254602.2_1', 'ENSG00000261221.4_1', 'ENSG00000188026.13_1', 'ENSG00000162813.18_2', 'ENSG00000165633.13_1', 'ENSG00000107807.13_2', 'ENSG00000108599.15_4', 'ENSG00000169856.9_2', 'ENSG00000164609.10_5', 'ENSG00000186615.12_3', 'ENSG00000124370.11_1', 'ENSG00000153094.24_2', 'ENSG00000135953.11_2', 'ENSG00000204314.12_1', 'ENSG00000099381.19_3', 'ENSG00000171853.16_4', 'ENSG00000261879.8_2', 'ENSG00000107960.11_1', 'ENSG00000187951.11_3', 'ENSG00000137936.18_1', 'ENSG00000073605.19_5', 'ENSG00000138138.14_2', 'ENSG00000157500.12_4', 'ENSG00000028310.18_2', 'ENSG00000184178.17_1', 'ENSG00000153558.16_4', 'ENSG00000104435.14_1', 'ENSG00000138083.5_1', 'ENSG00000132394.11_2', 'ENSG00000242686.4_2', 'ENSG00000132849.22_1', 'ENSG00000132485.14_3', 'ENSG00000095203.15_3', 'ENSG00000198908.12_1', 'ENSG00000135090.14_5', 'ENSG00000103404.15_1', 'ENSG00000128805.15_2', 'ENSG00000183715.14_2', 'ENSG00000162944.11_1', 'ENSG00000101412.13_1', 'ENSG00000247708.8_2', 'ENSG00000198797.7_1', 'ENSG00000114541.15_3', 'ENSG00000078269.15_1', 'ENSG00000127995.17_2', 'ENSG00000100557.10_3', 'ENSG00000167264.18_1', 'ENSG00000164609.10_4', 'ENSG00000126767.18_1', 'ENSG00000139767.10_2', 'ENSG00000142794.19_3', 'ENSG00000065526.12_3', 'ENSG00000108587.16_3', 'ENSG00000183570.17_3', 'ENSG00000197565.17_1', 'ENSG00000114904.13_3', 'ENSG00000125676.20_3', 'ENSG00000179796.12_1', 'ENSG00000179796.12_2', 'ENSG00000149577.16_3', 'ENSG00000197724.11_1', 'ENSG00000173208.4_1', 'ENSG00000068383.19_2', 'ENSG00000173085.15_1', 'ENSG00000196275.16_6', 'ENSG00000255545.8_2', 'ENSG00000109534.17_3', 'ENSG00000132485.14_2', 'ENSG00000235173.7_2', 'ENSG00000136144.12_2', 'ENSG00000140750.17_5', 'ENSG00000149577.16_4', 'ENSG00000263956.8_2', 'ENSG00000254602.2_2', 'ENSG00000227110.7_3', 'ENSG00000228824.8_3', 'ENSG00000135953.11_3', 'ENSG00000105926.16_2', 'ENSG00000244115.1_1', 'ENSG00000127511.10_2', 'ENSG00000122783.17_3', 'ENSG00000092140.16_1', 'ENSG00000151692.15_7', 'ENSG00000186047.11_2', 'ENSG00000163964.18_3', 'ENSG00000162390.18_4', 'ENSG00000105926.16_1', 'ENSG00000130226.17_3', 'ENSG00000244115.1_2', 'ENSG00000185736.16_1', 'ENSG00000107807.13_1', 'ENSG00000105948.13_4', 'ENSG00000129158.11_4', 'ENSG00000109472.14_2', 'ENSG00000149050.10_1', 'ENSG00000162813.18_3', 'ENSG00000138083.5_2', 'ENSG00000004866.22_2', 'ENSG00000109118.14_1', 'ENSG00000128274.17_1', 'ENSG00000263956.8_3', 'ENSG00000281566.3_1', 'ENSG00000181090.21_5', 'ENSG00000007047.16_3', 'ENSG00000181090.21_6', 'ENSG00000171984.15_2', 'ENSG00000136144.12_3', 'ENSG00000066032.19_7', 'ENSG00000035141.8_3', 'ENSG00000119392.16_1', 'ENSG00000244558.6_1', 'ENSG00000187951.11_2', 'ENSG00000125676.20_2', 'ENSG00000026652.15_2', 'ENSG00000105948.13_3', 'ENSG00000173818.17_2', 'ENSG00000115665.9_1', 'ENSG00000196689.13_2', 'ENSG00000108599.15_3', 'ENSG00000164849.10_1', 'ENSG00000032219.19_2', 'ENSG00000101624.11_3', 'ENSG00000144460.13_2', 'ENSG00000122254.7_1', 'ENSG00000148814.18_3', 'ENSG00000122783.17_4', 'ENSG00000138802.11_1', 'ENSG00000146733.14_1', 'ENSG00000242686.4_1', 'ENSG00000072756.17_1', 'ENSG00000187260.16_1', 'ENSG00000068784.13_1', 'ENSG00000066379.15_2', 'ENSG00000118922.18_2', 'ENSG00000228824.8_4', 'ENSG00000105808.19_5', 'ENSG00000214562.15_1', 'ENSG00000277363.5_2', 'ENSG00000117020.19_2', 'ENSG00000035141.8_2', 'ENSG00000068784.13_2', 'ENSG00000121578.13_2', 'ENSG00000095203.15_2', 'ENSG00000128513.16_4', 'ENSG00000150760.13_2', 'ENSG00000121578.13_3', 'ENSG00000235173.7_1', 'ENSG00000145819.18_3', 'ENSG00000111450.14_1', 'ENSG00000139793.20_1', 'ENSG00000203485.14_3', 'ENSG00000181722.17_5', 'ENSG00000120594.17_1', 'ENSG00000203485.14_4', 'ENSG00000099381.19_2', 'ENSG00000265808.4_2', 'ENSG00000158604.15_1', 'ENSG00000078124.13_1', 'ENSG00000100060.18_2', 'ENSG00000128513.16_3', 'ENSG00000145349.18_3', 'ENSG00000213468.7_1', 'ENSG00000121904.19_3', 'ENSG00000107798.18_2', 'ENSG00000153558.16_3', 'ENSG00000117069.15_2', 'ENSG00000164741.15_3', 'ENSG00000158169.14_3', 'ENSG00000118257.17_5', 'ENSG00000109472.14_3', 'ENSG00000113448.20_2', 'ENSG00000119912.18_3', 'ENSG00000073605.19_4', 'ENSG00000170579.17_1', 'ENSG00000065526.12_4', 'ENSG00000115665.9_2', 'ENSG00000081760.17_1', 'ENSG00000196689.13_1', 'ENSG00000173085.15_2', 'ENSG00000232995.7_2', 'ENSG00000151692.15_6', 'ENSG00000066379.15_1', 'ENSG00000131375.10_3', 'ENSG00000156171.15_4', 'ENSG00000119509.13_2', 'ENSG00000119912.18_4', 'ENSG00000186908.15_3', 'ENSG00000065457.11_2', 'ENSG00000108587.16_4', 'ENSG00000263069.6_2', 'ENSG00000155093.19_2', 'ENSG00000173818.17_5', 'ENSG00000154319.16_1', 'ENSG00000107960.11_2', 'ENSG00000128274.17_2', 'ENSG00000140750.17_4', 'ENSG00000185352.9_1', 'ENSG00000138138.14_1', 'ENSG00000137161.18_1', 'ENSG00000152954.12_1', 'ENSG00000131831.18_1', 'ENSG00000148120.18_5', 'ENSG00000120868.14_2', 'ENSG00000115839.19_2', 'ENSG00000154319.16_2', 'ENSG00000007047.16_2', 'ENSG00000157500.12_3', 'ENSG00000174165.8_3', 'ENSG00000127995.17_1', 'ENSG00000244405.8_3'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177282422</td>\n",
       "      <td>ENSG00000198797.7_1|BRINP2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600573</td>\n",
       "      <td>123601338</td>\n",
       "      <td>ENSG00000125676.20_3|THOC2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940456  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 177280411 | 177282422 | ...   |\n",
       "| chr1         | 177280411 | 177284690 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 17835910  | 17837395  | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 40653641  | 40654363  | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 284 rows and 6 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load in BED file of PAPA cryptics\n",
    "papa_cryp_bed = pr.read_bed(\"data/2023-12-15_all.last_exons.cryptic.bed\")\n",
    "\n",
    "# get le_ids for cryptics (+ their annotated partners)\n",
    "# e.g. ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic\t\n",
    "papa_name_spl = papa_cryp_bed.Name.str.split(\"\\\\|\", expand=True)\n",
    "cryp_le_ids = set(papa_name_spl[0])\n",
    "cryp_gn = set(papa_name_spl[1])\n",
    "print(f\"Number of PAPA cryptics - {len(cryp_le_ids)}\")\n",
    "print(cryp_le_ids)\n",
    "papa_cryp_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RFNG', 'TLX1'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get Zeng IDs that are missing from iCLIP GTFs (+ add TLX1 which is misannotated)\n",
    "zeng_targets_papa_missing = set(zeng_targets_papa).difference(cryp_gn).union({\"TLX1\"})\n",
    "zeng_targets_papa_missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24419290</td>\n",
       "      <td>24420128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419290</td>\n",
       "      <td>24419640</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>488.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000001461.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24421813</td>\n",
       "      <td>24422110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24419640,24419640,24419640</td>\n",
       "      <td>24440171,24440171,24433103</td>\n",
       "      <td>ENST00000358028.8,ENST00000374399.9,ENST000000...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000001461.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24454053</td>\n",
       "      <td>24454824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24453504,24453504</td>\n",
       "      <td>24456137,24456137</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000001461.17_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24464025</td>\n",
       "      <td>24466378</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464025</td>\n",
       "      <td>24464120</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>2258.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000001461.17_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>24464120,24464120</td>\n",
       "      <td>24468985,24468985</td>\n",
       "      <td>ENST00000003912.7,ENST00000374399.9</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000001461.17_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124061</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124062</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880075</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124063</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21880307</td>\n",
       "      <td>21880652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21880652,21880652,21880652</td>\n",
       "      <td>21881063,21881063,21881063</td>\n",
       "      <td>ENST00000382653.6,ENST00000382680.5,ENST000004...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000244395.6</td>\n",
       "      <td>RBMY1D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000244395.6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124064</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21039044</td>\n",
       "      <td>21042268</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENSG00000254488</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000254488.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124065</th>\n",
       "      <td>chrY</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>23936726</td>\n",
       "      <td>23936831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>23936831</td>\n",
       "      <td>23939185</td>\n",
       "      <td>ENST00000631331.1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000280961.1</td>\n",
       "      <td>TTTY3B</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000280961.1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124066 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 24419290  | 24420128  | ...   |\n",
       "| chr1         | .          | exon         | 24421813  | 24422110  | ...   |\n",
       "| chr1         | .          | exon         | 24454053  | 24454824  | ...   |\n",
       "| chr1         | .          | exon         | 24464025  | 24466378  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | .          | exon         | 21880075  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21880307  | 21880652  | ...   |\n",
       "| chrY         | .          | exon         | 21038288  | 21039044  | ...   |\n",
       "| chrY         | .          | exon         | 23936726  | 23936831  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 124,066 rows and 22 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papa_gtf = pr.read_gtf(\"data/novel_ref_combined.last_exons.gtf\")\n",
    "papa_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Frame', 'gene_id', 'gene_name', 'transcript_id', 'exon_number',\n",
       "       'region_rank', 'Start_ref', 'End_ref', 'transcript_id_ref',\n",
       "       '3p_extension_length', 'event_type', 'ref_gene_id', 'ref_gene_name',\n",
       "       'le_number', 'le_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papa_gtf.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract partner, annotated/alt last exon for IPA and ALEs\n",
    "\n",
    "Bed file only has alt ALE for subset of cryptic ALE/IPAs. Extract them for quantification GTF using ofllowing criteria:\n",
    "- Cryptic ALE/IPA genes\n",
    "- Extract most distal non-cryptic le_id\n",
    "- Merge exons assigned to le_id, keep minimal features for gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21446036</td>\n",
       "      <td>21446482</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21445219,21445219,21445219,21445219,21445219</td>\n",
       "      <td>21470631,21468687,21468687,21466086,21466086</td>\n",
       "      <td>ENST00000454000.6,ENST00000318249.10,ENST00000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000142794.19_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21454306</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21445219,21445219,21445219,21445219,21445219</td>\n",
       "      <td>21470631,21468687,21468687,21466086,21466086</td>\n",
       "      <td>ENST00000454000.6,ENST00000318249.10,ENST00000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000142794.19_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21454855</td>\n",
       "      <td>21456901</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-2.310</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced,last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000142794.19_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689528</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-6.24702</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>792</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689804</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.27739</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155607</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155607</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>796 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 21446036  | 21446482  | ...   |\n",
       "| chr1         | .          | exon         | 21453372  | 21454306  | ...   |\n",
       "| chr1         | .          | exon         | 21454855  | 21456901  | ...   |\n",
       "| chr1         | .          | exon         | 21483142  | 21485005  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 131689804 | 131692636 | ...   |\n",
       "| chrX         | .          | exon         | 108155606 | 108157260 | ...   |\n",
       "| chrX         | .          | exon         | 108155607 | 108157260 | ...   |\n",
       "| chrX         | .          | exon         | 108155607 | 108157260 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 796 rows and 22 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract ids\n",
    "\n",
    "papa_cryp_bed_ale_ipa = papa_cryp_bed.subset(lambda df: df.Name.str.contains(\"spliced|bleedthrough\"))\n",
    "# ENSG00000162849.16_2|KIF26B|spliced|cryptic \n",
    "name_split_ale_ipa = papa_cryp_bed_ale_ipa.Name.str.split(\"\\\\|\", expand=True)\n",
    "cryp_le_ids_ale_ipa = set(name_split_ale_ipa[0])\n",
    "cryp_gn_ale_ipa = set(name_split_ale_ipa[1])\n",
    "\n",
    "papa_gtf_cryp_genes = papa_gtf.subset(lambda df: df.ref_gene_name.isin(cryp_gn_ale_ipa))\n",
    "\n",
    "# remove cryptic le_ids\n",
    "papa_gtf_cryp_genes_noncryp = papa_gtf_cryp_genes.subset(lambda df: ~df.le_id.isin(cryp_le_ids_ale_ipa))\n",
    "papa_gtf_cryp_genes_noncryp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21484900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21483467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21483412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>21482535,21482535,21482535,21482535,21482535</td>\n",
       "      <td>21483142,21483142,21483142,21483142,21483142</td>\n",
       "      <td>ENST00000318220.10,ENST00000318249.10,ENST0000...</td>\n",
       "      <td>NULL,NULL,NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689528</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.CTRL-6.24702</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131689804</td>\n",
       "      <td>131692636</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.ctrl_ctrl_2.27739</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155607</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108155607</td>\n",
       "      <td>108157260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108157260,108157260,108157260</td>\n",
       "      <td>108159461,108159461,108159461</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>349 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 21483142  | 21485005  | ...   |\n",
       "| chr1         | .          | exon         | 21483142  | 21484900  | ...   |\n",
       "| chr1         | .          | exon         | 21483142  | 21484900  | ...   |\n",
       "| chr1         | .          | exon         | 21483142  | 21483467  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 131689804 | 131692636 | ...   |\n",
       "| chrX         | .          | exon         | 108155606 | 108157260 | ...   |\n",
       "| chrX         | .          | exon         | 108155607 | 108157260 | ...   |\n",
       "| chrX         | .          | exon         | 108155607 | 108157260 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 349 rows and 22 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract all intervals associated with most distal ALE\n",
    "papa_gtf_cryp_genes_noncryp.le_number = papa_gtf_cryp_genes_noncryp.le_number.astype(float).astype(int)\n",
    "# pull out most distal isoform per gene\n",
    "le_ids_most_distal =  set(get_terminal_regions(papa_gtf_cryp_genes_noncryp, id_col=\"ref_gene_id\", region_number_col=\"le_number\").le_id)\n",
    "papa_gtf_ale_ipa_alt = papa_gtf_cryp_genes_noncryp.subset(lambda df: df.le_id.isin(le_ids_most_distal))\n",
    "papa_gtf_ale_ipa_alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>77062974</td>\n",
       "      <td>77067546</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117069.15_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62178422</td>\n",
       "      <td>62178675</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132849.22_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245702457</td>\n",
       "      <td>245709432</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162849.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98864610</td>\n",
       "      <td>98867720</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000281566.3_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17799730</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40648304</td>\n",
       "      <td>40651879</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+----------------------+\n",
       "| Chromosome   | Start     | End       | Strand       | le_id                |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)             |\n",
       "|--------------+-----------+-----------+--------------+----------------------|\n",
       "| chr1         | 77062974  | 77067546  | +            | ENSG00000117069.15_3 |\n",
       "| chr1         | 62178422  | 62178675  | +            | ENSG00000132849.22_8 |\n",
       "| chr1         | 21483142  | 21485005  | +            | ENSG00000142794.19_4 |\n",
       "| chr1         | 54608956  | 54610329  | +            | ENSG00000162390.18_3 |\n",
       "| ...          | ...       | ...       | ...          | ...                  |\n",
       "| chrX         | 17799730  | 17802034  | -            | ENSG00000131831.18_3 |\n",
       "| chrX         | 40648304  | 40651879  | -            | ENSG00000180182.11_4 |\n",
       "| chrX         | 108155606 | 108157260 | -            | ENSG00000197565.17_2 |\n",
       "| chrX         | 131688563 | 131692636 | -            | ENSG00000213468.7_4  |\n",
       "+--------------+-----------+-----------+--------------+----------------------+\n",
       "Stranded PyRanges object has 106 rows and 5 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge le_ids to sinlge interval\n",
    "papa_gtf_ale_ipa_alt_m = papa_gtf_ale_ipa_alt.merge(by=\"le_id\", strand=True)\n",
    "papa_gtf_ale_ipa_alt_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62178422</td>\n",
       "      <td>62178675</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000132849.22_8</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000132849.22</td>\n",
       "      <td>PATJ</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000132849.22_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>77062974</td>\n",
       "      <td>77067546</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000117069.15_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000117069.15</td>\n",
       "      <td>ST6GALNAC5</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000117069.15_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245702457</td>\n",
       "      <td>245709432</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162849.16_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000162849.16</td>\n",
       "      <td>KIF26B</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000162849.16_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>chrX</td>\n",
       "      <td>102748698</td>\n",
       "      <td>102753658</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198908.12_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000198908.12</td>\n",
       "      <td>BHLHB9</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000198908.12_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17799730</td>\n",
       "      <td>17802034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000131831.18_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000131831.18</td>\n",
       "      <td>RAI2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000131831.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40648304</td>\n",
       "      <td>40651879</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_4</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000180182.11_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 21483142  | 21485005  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610329  | +            | ...   |\n",
       "| chr1         | 62178422  | 62178675  | +            | ...   |\n",
       "| chr1         | 77062974  | 77067546  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 17799730  | 17802034  | -            | ...   |\n",
       "| chrX         | 40648304  | 40651879  | -            | ...   |\n",
       "| chrX         | 108155606 | 108157260 | -            | ...   |\n",
       "| chrX         | 131688563 | 131692636 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 106 rows and 10 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: le_id, Feature, gene_id, gene_name, pas_number, ... (+ 1 more.)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract minimal annotation per le_id + add to merged intervals\n",
    "meta_ale_ipa_alt = papa_gtf_ale_ipa_alt.as_df()[[\"Feature\", \"le_id\", \"ref_gene_id\", \"ref_gene_name\", \"le_number\"]].rename(columns={\"ref_gene_id\": \"gene_id\", \"ref_gene_name\": \"gene_name\", \"le_number\": \"pas_number\"}).drop_duplicates()\n",
    "papa_gtf_ale_ipa_alt_m = papa_gtf_ale_ipa_alt_m.apply(lambda df: df.merge(meta_ale_ipa_alt, on=\"le_id\", how=\"left\")).sort()\n",
    "# add transcript id (assign to le-id)\n",
    "papa_gtf_ale_ipa_alt_m.transcript_id = papa_gtf_ale_ipa_alt_m.le_id\n",
    "papa_gtf_ale_ipa_alt_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000142794.19_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54608956</td>\n",
       "      <td>54610329</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000162390.18_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108155606</td>\n",
       "      <td>108157260</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000197565.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>318 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 21483142  | 21485005  | +            | ...   |\n",
       "| chr1         | 21483142  | 21485005  | +            | ...   |\n",
       "| chr1         | 21483142  | 21485005  | +            | ...   |\n",
       "| chr1         | 54608956  | 54610329  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 108155606 | 108157260 | -            | ...   |\n",
       "| chrX         | 131688563 | 131692636 | -            | ...   |\n",
       "| chrX         | 131688563 | 131692636 | -            | ...   |\n",
       "| chrX         | 131688563 | 131692636 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 318 rows and 10 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: le_id, Feature, gene_id, gene_name, pas_number, ... (+ 1 more.)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get feature counts ready i.e. make copies with 'transcript' and 'gene' in the Feature col\n",
    "\n",
    "grs = [papa_gtf_ale_ipa_alt_m.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "papa_gtf_ale_ipa_alt_m\n",
    "# combine into single gr\n",
    "papa_gtf_ale_ipa_alt_m = pr.concat([papa_gtf_ale_ipa_alt_m, *grs]).sort()\n",
    "papa_gtf_ale_ipa_alt_m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean up TLX1 and RFNG\n",
    "\n",
    "xx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to novel predictions (never need annotated, so just overwrite existing object)\n",
    "papa_gtf_annot = papa_gtf.subset(lambda df: ~df.transcript_id.str.startswith(\"PAPA\"))\n",
    "papa_gtf_novel = papa_gtf.subset(lambda df: df.transcript_id.str.startswith(\"PAPA\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101138530</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.13599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000107807.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101134174</td>\n",
       "      <td>101138530</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.10958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>first_exon_spliced</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000107807.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138522</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.14398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.14536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045104</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_19065409_S29.22294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045147</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.NT_19074723_S36.21380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045423</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-4.21887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon         | 101134174 | 101138530 | ...   |\n",
       "| chr10        | .          | exon         | 101134174 | 101138530 | ...   |\n",
       "| chr10        | .          | exon         | 101136690 | 101138531 | ...   |\n",
       "| chr10        | .          | exon         | 101136690 | 101138522 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr17        | .          | exon         | 82045104  | 82048807  | ...   |\n",
       "| chr17        | .          | exon         | 82045147  | 82048807  | ...   |\n",
       "| chr17        | .          | exon         | 82045150  | 82048807  | ...   |\n",
       "| chr17        | .          | exon         | 82045423  | 82048807  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 9 rows and 22 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to Zeng et al. target genes (+ TLX1)\n",
    "papa_gtf_novel_sel = papa_gtf_novel.subset(lambda df: df.ref_gene_name.isin(zeng_targets_papa_missing))\n",
    "papa_gtf_novel_sel\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TLX1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101134376</td>\n",
       "      <td>101138530</td>\n",
       "      <td>ENSG00000107807.13_1|TLX1|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>ENSG00000107807.13_2|TLX1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +3    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr10        | 101134376 | 101138530 | ...   |\n",
       "| chr10        | 101136690 | 101138531 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Name, Score, Strand"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papa_cryp_bed.subset(lambda df: df.Name.str.contains(\"TLX1\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>101134376,101134376</td>\n",
       "      <td>101136690,101136667</td>\n",
       "      <td>ENST00000370196.11,ENST00000467928.2</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000107807.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | .          | exon         | 101136690 | 101138531 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papa_gtf_tlx1_prox = papa_gtf_annot.subset(lambda df: df.Start == 101136690)\n",
    "# distal PAS can be plucked straight as selected in BED file\n",
    "papa_gtf_tlx1_dist = papa_gtf_novel.subset(lambda df: df.End == 101138531) \n",
    "\n",
    "# now make proximal by overlapping with Zeng PAS and truncating\n",
    "tlx1 = pr.concat([papa_gtf_tlx1_prox, papa_gtf_tlx1_dist])\n",
    "tlx1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RFNG\n",
    "\n",
    "Problem here was a slight bug with PAPA grouping extensions together with the annot event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82049116</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82049030</td>\n",
       "      <td>82049116</td>\n",
       "      <td>ENST00000310496.9</td>\n",
       "      <td>1129.0</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82048444</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "| chr17        | .          | exon         |  82047901 |  82049116 | ...   |\n",
       "| chr17        | .          | exon         |  82048444 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 4 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfng_annot = papa_gtf_annot.subset(lambda df: df.ref_gene_name == \"RFNG\")\n",
    "rfng_novel = papa_gtf_novel.subset(lambda df: df.ref_gene_name == \"RFNG\")\n",
    "rfng_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pick longest annotated RFNG PAS\n",
    "rfng_prox = rfng_annot.subset(lambda df: (df.Start == 82047901 ) & (df.End == 82048807)).drop_duplicate_positions()\n",
    "rfng_prox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52583</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr17        | .          | exon         |  82045150 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# manual curation suggests x is one to pick (s)\n",
    "rfng_dist = rfng_novel.subset(lambda df: df.Start == 82045150)\n",
    "rfng_dist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr17        | .          | exon         |  82045150 |  82048807 | ...   |\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfng = pr.concat([rfng_prox, rfng_dist]).sort()\n",
    "rfng"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>101134376,101134376</td>\n",
       "      <td>101136690,101136667</td>\n",
       "      <td>ENST00000370196.11,ENST00000467928.2</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000107807.13_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | .          | exon         | 101136690 | 101138531 | ...   |\n",
       "| chr17        | .          | exon         |  82045150 |  82048807 | ...   |\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 4 rows and 22 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfng_tlx1 = pr.concat([rfng, tlx1]).sort()\n",
    "rfng_tlx1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101137789</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>...</td>\n",
       "      <td>101134376,101134376</td>\n",
       "      <td>101136690,101136667</td>\n",
       "      <td>ENST00000370196.11,ENST00000467928.2</td>\n",
       "      <td>NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000107807.13_1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82047901</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>...</td>\n",
       "      <td>82048807,82048807,82048807</td>\n",
       "      <td>82049030,82049030,82049030</td>\n",
       "      <td>ENST00000310496.9,ENST00000578676.5,ENST000005...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +18   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon         | 101136690 | 101137789 | ...   |\n",
       "| chr10        | .          | exon         | 101136690 | 101138531 | ...   |\n",
       "| chr17        | .          | exon         |  82045150 |  82048807 | ...   |\n",
       "| chr17        | .          | exon         |  82047901 |  82048807 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 4 rows and 23 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "18 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 11 more.)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign proximal + distal pas, generate bins\n",
    "rfng_tlx1 = add_region_number(rfng_tlx1,\n",
    "                  id_col=\"ref_gene_name\",\n",
    "                  feature_key=\"exon\",\n",
    "                  out_col=\"pas_number\",\n",
    "                  sort_col=\"End\",\n",
    "                  method=\"first\"\n",
    "                  ).sort()\n",
    "\n",
    "rfng_tlx1_prox = rfng_tlx1.subset(lambda df: df.pas_number == 1)\n",
    "rfng_tlx1_dist = rfng_tlx1.subset(lambda df: df.pas_number == 2)\n",
    "\n",
    "rfng_tlx1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +18   |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon         | 101137789 | 101138531 | ...   |\n",
       "| chr17        | .          | exon         |  82045150 |  82047901 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2 rows and 23 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "18 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 11 more.)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subtract prox from distal to get unique segemtn of distal cryptic\n",
    "rfng_tlx1_dist_uniq = rfng_tlx1_dist.subtract(rfng_tlx1_prox, strandedness=\"same\")\n",
    "rfng_tlx1_dist_uniq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating bins for all ALE events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177282422</td>\n",
       "      <td>ENSG00000198797.7_1|BRINP2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>76871267</td>\n",
       "      <td>76871821</td>\n",
       "      <td>ENSG00000117069.15_2|ST6GALNAC5|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ST6GALNAC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600573</td>\n",
       "      <td>123601338</td>\n",
       "      <td>ENSG00000125676.20_3|THOC2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17835910</td>\n",
       "      <td>17837395</td>\n",
       "      <td>ENSG00000131831.18_1|RAI2|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>RAI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>COL4A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653641</td>\n",
       "      <td>40654363</td>\n",
       "      <td>ENSG00000180182.11_3|MED14|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>MED14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940456  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 177280411 | 177282422 | ...   |\n",
       "| chr1         | 177280411 | 177284690 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 17835910  | 17837395  | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 40653641  | 40654363  | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 282 rows and 7 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Name, Score, Strand, gene_name"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove tlx1 from analysis \n",
    "papa_cryp_bed_upd = papa_cryp_bed.subset(lambda df: ~df.Name.str.contains(\"TLX1\"))\n",
    "# pull out gene name\n",
    "papa_cryp_bed_upd = papa_cryp_bed_upd.assign(\"gene_name\", lambda df: df.Name.str.split(\"\\\\|\", expand=True)[1])\n",
    "papa_cryp_bed_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>ENSG00000065526.12_3|SPEN|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61824444</td>\n",
       "      <td>61825501</td>\n",
       "      <td>ENSG00000132849.22_1|PATJ|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PATJ</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47635520</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_1|ELK1|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600573</td>\n",
       "      <td>123601338</td>\n",
       "      <td>ENSG00000125676.20_3|THOC2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +6    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940456  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | 54634687  | 54639192  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 123600550 | 123603748 | ...   |\n",
       "| chrX         | 123600573 | 123601338 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 282 rows and 9 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Name, Score, Strand, gene_name, Feature, ... (+ 1 more.)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate prox and distal \n",
    "papa_cryp_bed_upd = add_region_number(papa_cryp_bed_upd.assign(\"Feature\", lambda df: pd.Series([\"exon\"]*len(df), index=df.index)),\n",
    "                  id_col=\"gene_name\",\n",
    "                  feature_key=\"exon\",\n",
    "                  out_col=\"pas_number\",\n",
    "                  sort_col=\"End\",\n",
    "                  method=\"first\"\n",
    "                  ).sort()\n",
    "\n",
    "# extract to individual objects. Split by 3'UTRs to \n",
    "# papa_cryp_bed_upd_utr = papa_cryp_bed_upd.subset(lambda df: df.Name.str.contains(\"proximal|distal\"))\n",
    "\n",
    "\n",
    "papa_cryp_bed_upd_utr_prox = papa_cryp_bed_upd.subset(lambda df: df.pas_number == 1)\n",
    "papa_cryp_bed_upd_utr_dist = papa_cryp_bed_upd.subset(lambda df: df.pas_number == 2)\n",
    "\n",
    "papa_cryp_bed_upd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177282422</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>71063238</td>\n",
       "      <td>71063290</td>\n",
       "      <td>ENSG00000132485.14_3|ZRANB2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111117005</td>\n",
       "      <td>111117162</td>\n",
       "      <td>ENSG00000156171.15_4|DRAM2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>DRAM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120150756</td>\n",
       "      <td>120150897</td>\n",
       "      <td>ENSG00000265808.4_3|SEC22B|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>SEC22B</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50621309</td>\n",
       "      <td>50622753</td>\n",
       "      <td>ENSG00000100299.18_2|ARSA|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50623027</td>\n",
       "      <td>50625464</td>\n",
       "      <td>ENSG00000100299.18_2|ARSA|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47635520</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123600573</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123601338</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +6    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15940456  | 15940477  | ...   |\n",
       "| chr1         | 177282422 | 177284690 | ...   |\n",
       "| chr1         | 71063238  | 71063290  | ...   |\n",
       "| chr1         | 111117005 | 111117162 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr22        | 42681021  | 42692120  | ...   |\n",
       "| chr22        | 50621309  | 50622753  | ...   |\n",
       "| chr22        | 50623027  | 50625464  | ...   |\n",
       "| chrX         | 123601338 | 123603748 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 94 rows and 9 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Name, Score, Strand, gene_name, Feature, ... (+ 1 more.)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define distal specific region (ONLY for UTRs)\n",
    "papa_cryp_dist_uniq = papa_cryp_dist.subtract(papa_cryp_prox, strandedness=\"same\")\n",
    "papa_cryp_dist_uniq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FeatureCounts compliant\n",
    "\n",
    "featureCounts compliant GTF checklist:\n",
    "- each interval reported with 'exon' assigned to Feature column\n",
    "- Unique isoforms/intervals also represented as 'gene' & 'transcript' in Feature column\n",
    "- minimal attributes - gene_id, transcript_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>papa_bed_name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21453372</td>\n",
       "      <td>21457150</td>\n",
       "      <td>ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>gene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>ENSG00000162390.18_4</td>\n",
       "      <td>ENSG00000162390.18_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54634687</td>\n",
       "      <td>54639192</td>\n",
       "      <td>ENSG00000162390.18_4|ACOT11|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ACOT11</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162390.18</td>\n",
       "      <td>ENSG00000162390.18_4</td>\n",
       "      <td>ENSG00000162390.18_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108269152</td>\n",
       "      <td>ENSG00000197565.17_1|COL4A6|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>gene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>gene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>336 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | 21453372  | 21457150  | ...   |\n",
       "| chr1         | 54634687  | 54639192  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 108267641 | 108269152 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 336 rows and 12 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: papa_bed_name, Score, Strand, gene_name, ... (+ 5 more.)"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ALEs/IPAs dropped from papa_cryp_dist_uniq (because assigned as proximal)\n",
    "papa_cryp_ale_ipa = papa_cryp_bed_upd.subset(lambda df: ~df.Name.str.contains(\"proximal|distal\"))\n",
    "# assign gene_id, transcript_id\n",
    "# ENSG00000142794.19_3|NBPF3|bleedthrough|cryptic\n",
    "papa_cryp_ale_ipa.gene_id = papa_cryp_ale_ipa.Name.str.split(\"\\\\|\", expand=True)[0].str.split(\"_\", expand=True, regex=False)[0]\n",
    "papa_cryp_ale_ipa.transcript_id = papa_cryp_ale_ipa.Name.str.split(\"\\\\|\", expand=True)[0]\n",
    "papa_cryp_ale_ipa.le_id = papa_cryp_ale_ipa.Name.str.split(\"\\\\|\", expand=True)[0]\n",
    "papa_cryp_ale_ipa = papa_cryp_ale_ipa.apply(lambda df: df.rename(columns={\"Name\": \"papa_bed_name\"}))\n",
    "\n",
    "# \n",
    "grs = [papa_cryp_ale_ipa.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "papa_cryp_ale_ipa = pr.concat([papa_cryp_ale_ipa, *grs]).sort()\n",
    "papa_cryp_ale_ipa\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000198797.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>71063238</td>\n",
       "      <td>71065137</td>\n",
       "      <td>ENSG00000132485.14_3|ZRANB2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000132485.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111117005</td>\n",
       "      <td>111118267</td>\n",
       "      <td>ENSG00000156171.15_4|DRAM2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>DRAM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000156171.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120150756</td>\n",
       "      <td>120157192</td>\n",
       "      <td>ENSG00000265808.4_3|SEC22B|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>SEC22B</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000265808.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr20</td>\n",
       "      <td>33675132</td>\n",
       "      <td>33676979</td>\n",
       "      <td>ENSG00000101412.13_2|E2F1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000101412.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr22</td>\n",
       "      <td>42681021</td>\n",
       "      <td>42693997</td>\n",
       "      <td>ENSG00000128274.17_2|A4GALT|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000128274.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50621309</td>\n",
       "      <td>50625464</td>\n",
       "      <td>ENSG00000100299.18_2|ARSA|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000100299.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 177280411 | 177284690 | ...   |\n",
       "| chr1         | 71063238  | 71065137  | ...   |\n",
       "| chr1         | 111117005 | 111118267 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr22        | 42681021  | 42693997  | ...   |\n",
       "| chr22        | 50621309  | 50625464  | ...   |\n",
       "| chrX         | 47631618  | 47636927  | ...   |\n",
       "| chrX         | 123600550 | 123603748 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 89 rows and 10 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: Name, Score, Strand, gene_name, Feature, ... (+ 2 more.)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# currently annotated with exon already\n",
    "# extract gene_id from Name field\n",
    "# ENSG00000126767.18_2|ELK1|distal|cryptic -> ENSG00000126767.18\n",
    "\n",
    "# distal specific bins\n",
    "papa_cryp_dist_uniq.gene_id = papa_cryp_dist_uniq.Name.str.split(\"\\\\|\", expand=True)[0].str.split(\"_\", expand=True, regex=False)[0]\n",
    "# full length \n",
    "papa_cryp_dist.gene_id  = papa_cryp_dist.Name.str.split(\"\\\\|\", expand=True)[0].str.split(\"_\", expand=True, regex=False)[0]\n",
    "\n",
    "papa_cryp_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>177280411</td>\n",
       "      <td>177284690</td>\n",
       "      <td>ENSG00000198797.7_2|BRINP2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000198797.7</td>\n",
       "      <td>ENSG00000198797.7_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>71063238</td>\n",
       "      <td>71065137</td>\n",
       "      <td>ENSG00000132485.14_3|ZRANB2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000132485.14</td>\n",
       "      <td>ENSG00000132485.14_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111117005</td>\n",
       "      <td>111118267</td>\n",
       "      <td>ENSG00000156171.15_4|DRAM2|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>DRAM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000156171.15</td>\n",
       "      <td>ENSG00000156171.15_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120150756</td>\n",
       "      <td>120157192</td>\n",
       "      <td>ENSG00000265808.4_3|SEC22B|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>SEC22B</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000265808.4</td>\n",
       "      <td>ENSG00000265808.4_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr20</td>\n",
       "      <td>33675132</td>\n",
       "      <td>33676979</td>\n",
       "      <td>ENSG00000101412.13_2|E2F1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000101412.13</td>\n",
       "      <td>ENSG00000101412.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr22</td>\n",
       "      <td>42681021</td>\n",
       "      <td>42693997</td>\n",
       "      <td>ENSG00000128274.17_2|A4GALT|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000128274.17</td>\n",
       "      <td>ENSG00000128274.17_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50621309</td>\n",
       "      <td>50625464</td>\n",
       "      <td>ENSG00000100299.18_2|ARSA|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000100299.18</td>\n",
       "      <td>ENSG00000100299.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +8    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 177280411 | 177284690 | ...   |\n",
       "| chr1         | 71063238  | 71065137  | ...   |\n",
       "| chr1         | 111117005 | 111118267 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr22        | 42681021  | 42693997  | ...   |\n",
       "| chr22        | 50621309  | 50625464  | ...   |\n",
       "| chrX         | 47631618  | 47636927  | ...   |\n",
       "| chrX         | 123600550 | 123603748 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 89 rows and 11 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: Name, Score, Strand, gene_name, Feature, ... (+ 3 more.)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For PAPA defined events, am going to assign transcript ID as le_id \n",
    "# for distal unique regions, add suffix to transcript ID so represented separately\n",
    "papa_cryp_dist_uniq.transcript_id = papa_cryp_dist_uniq.Name.str.split(\"\\\\|\", expand=True)[0] + \"__distal\"\n",
    "papa_cryp_dist.transcript_id = papa_cryp_dist.Name.str.split(\"\\\\|\", expand=True)[0]\n",
    "\n",
    "papa_cryp_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Feature\n",
       "exon          89\n",
       "transcript    89\n",
       "gene          89\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need to create 'gene', 'transcript' and 'exon' entries\n",
    "# exon already present\n",
    "grs = [papa_cryp_dist.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "papa_cryp_dists = pr.concat([papa_cryp_dist, *grs])\n",
    "papa_cryp_dists.Feature.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Feature\n",
       "exon          94\n",
       "transcript    94\n",
       "gene          94\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# repeat for unique regions\n",
    "grs = [papa_cryp_dist_uniq.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "papa_cryp_dist_uniqs = pr.concat([papa_cryp_dist_uniq, *grs])\n",
    "papa_cryp_dist_uniqs.Feature.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>papa_bed_name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4__distal</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4__distal</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123600550</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123601338</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2__distal</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123601338</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2__distal</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123601338</td>\n",
       "      <td>123603748</td>\n",
       "      <td>ENSG00000125676.20_2|THOC2|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>THOC2</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000125676.20</td>\n",
       "      <td>ENSG00000125676.20_2__distal</td>\n",
       "      <td>ENSG00000125676.20_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>549 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15940456  | 15940477  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 123600550 | 123603748 | ...   |\n",
       "| chrX         | 123601338 | 123603748 | ...   |\n",
       "| chrX         | 123601338 | 123603748 | ...   |\n",
       "| chrX         | 123601338 | 123603748 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 549 rows and 12 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: papa_bed_name, Score, Strand, gene_name, ... (+ 5 more.)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine into single GTF\n",
    "comb_gtf = pr.concat([papa_cryp_dist_uniqs, papa_cryp_dists]).sort()\n",
    "comb_gtf = comb_gtf.apply(lambda df: df.rename(columns={\"Name\": \"papa_bed_name\"}))\n",
    "comb_gtf.le_id = comb_gtf.transcript_id.str.removesuffix(\"__distal\")\n",
    "comb_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>papa_bed_name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47635520</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2__distal</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47635520</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2__distal</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47635520</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2__distal</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631618</td>\n",
       "      <td>47636927</td>\n",
       "      <td>ENSG00000126767.18_2|ELK1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000126767.18</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "      <td>ENSG00000126767.18_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +9    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chrX         |  47631618 |  47635520 | ...   |\n",
       "| chrX         |  47631618 |  47635520 | ...   |\n",
       "| chrX         |  47631618 |  47635520 | ...   |\n",
       "| chrX         |  47631618 |  47636927 | ...   |\n",
       "| chrX         |  47631618 |  47636927 | ...   |\n",
       "| chrX         |  47631618 |  47636927 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 12 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: papa_bed_name, Score, Strand, gene_name, ... (+ 5 more.)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_gtf.subset(lambda df: df.gene_name == \"ELK1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>gene</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>.</td>\n",
       "      <td>gene</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.10465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>gene</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr17</td>\n",
       "      <td>.</td>\n",
       "      <td>gene</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.17048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_spliced</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +18   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr10        | .          | exon       | 101136690 | 101138531 | ...   |\n",
       "| chr10        | .          | transcript | 101136690 | 101138531 | ...   |\n",
       "| chr10        | .          | gene       | 101136690 | 101138531 | ...   |\n",
       "| chr10        | .          | exon       | 101137789 | 101138531 | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chr17        | .          | gene       | 82045150  | 82047901  | ...   |\n",
       "| chr17        | .          | exon       | 82045150  | 82048807  | ...   |\n",
       "| chr17        | .          | transcript | 82045150  | 82048807  | ...   |\n",
       "| chr17        | .          | gene       | 82045150  | 82048807  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 12 rows and 23 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "18 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 11 more.)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need to make copies where 'transcript' and 'gene' Feature columns are created for eacxh interval\n",
    "grs = [rfng_tlx1_dist.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "# combine into single gr\n",
    "rfng_tlx1_dists = pr.concat([rfng_tlx1_dist, *grs])\n",
    "grs = [rfng_tlx1_dist_uniq.assign(\"Feature\", lambda df: pd.Series([feat]*len(df), index=df.index)) for feat in [\"transcript\", \"gene\"]]\n",
    "rfng_tlx1_dist_uniqs = pr.concat([rfng_tlx1_dist_uniq, *grs])\n",
    "rfng_tlx1_dist_uniqs.le_id = rfng_tlx1_dist_uniqs.le_id + \"__distal\"\n",
    "\n",
    "rfng_tlx1_comb_gtf = pr.concat([rfng_tlx1_dists, rfng_tlx1_dist_uniqs]).sort()\n",
    "rfng_tlx1_comb_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>le_id</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>exon</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>transcript</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene</td>\n",
       "      <td>101136690</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>exon</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>transcript</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene</td>\n",
       "      <td>101137789</td>\n",
       "      <td>101138531</td>\n",
       "      <td>+</td>\n",
       "      <td>TLX1</td>\n",
       "      <td>ENSG00000107807.13</td>\n",
       "      <td>ENSG00000107807.13_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000107807.13_2__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>transcript</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>gene</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82047901</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1__distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr17</td>\n",
       "      <td>exon</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr17</td>\n",
       "      <td>transcript</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr17</td>\n",
       "      <td>gene</td>\n",
       "      <td>82045150</td>\n",
       "      <td>82048807</td>\n",
       "      <td>-</td>\n",
       "      <td>RFNG</td>\n",
       "      <td>ENSG00000169733.12</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000169733.12_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr10        | exon       | 101136690 | 101138531 | +            | ...   |\n",
       "| chr10        | transcript | 101136690 | 101138531 | +            | ...   |\n",
       "| chr10        | gene       | 101136690 | 101138531 | +            | ...   |\n",
       "| chr10        | exon       | 101137789 | 101138531 | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chr17        | gene       | 82045150  | 82047901  | -            | ...   |\n",
       "| chr17        | exon       | 82045150  | 82048807  | -            | ...   |\n",
       "| chr17        | transcript | 82045150  | 82048807  | -            | ...   |\n",
       "| chr17        | gene       | 82045150  | 82048807  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_name, gene_id, le_id, pas_number, transcript_id"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfng_tlx1_comb_gtf = rfng_tlx1_comb_gtf[[\"Feature\", \"ref_gene_name\", \"ref_gene_id\", \"le_id\", \"pas_number\"]]\n",
    "rfng_tlx1_comb_gtf = rfng_tlx1_comb_gtf.apply(lambda df: df.rename(columns={\"ref_gene_id\": \"gene_id\", \"ref_gene_name\": \"gene_name\"}))\n",
    "rfng_tlx1_comb_gtf.transcript_id = rfng_tlx1_comb_gtf.le_id\n",
    "rfng_tlx1_comb_gtf.le_id = rfng_tlx1_comb_gtf.le_id.str.removesuffix(\"__distal\")\n",
    "rfng_tlx1_comb_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>papa_bed_name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Feature</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4__distal</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940456</td>\n",
       "      <td>15940477</td>\n",
       "      <td>ENSG00000065526.12_4|SPEN|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>ENSG00000065526.12_4__distal</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1882</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>exon</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1883</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131688563</td>\n",
       "      <td>131692636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "      <td>ENSG00000213468.7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1884</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1885</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1886</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131825221</td>\n",
       "      <td>ENSG00000213468.7_1|FIRRE|bleedthrough|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>gene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1887 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +9    |\n",
       "| (object)     | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15939295  | 15940477  | ...   |\n",
       "| chr1         | 15940456  | 15940477  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 131688563 | 131692636 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "| chrX         | 131823775 | 131825221 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,887 rows and 12 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: papa_bed_name, Score, Strand, gene_name, ... (+ 5 more.)"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add \n",
    "comb_gtf = pr.concat([rfng_tlx1_comb_gtf, comb_gtf, papa_gtf_ale_ipa_alt_m, papa_cryp_ale_ipa]).sort()\n",
    "comb_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcript_id\n",
       "2    185\n",
       "4      5\n",
       "1      4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check that all cryptics have multiple PAS assigned\n",
    "comb_gtf.as_df().groupby(\"gene_id\")[\"transcript_id\"].nunique().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>87365805</td>\n",
       "      <td>87372223</td>\n",
       "      <td>ENSG00000214562.15_1|NUTM2D|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000214562.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>87366138</td>\n",
       "      <td>87372223</td>\n",
       "      <td>ENSG00000214562.15_2|NUTM2D|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000214562.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>134502943</td>\n",
       "      <td>134505661</td>\n",
       "      <td>ENSG00000255545.8_2|B3GAT1-DT|spliced|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000255545.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>31433709</td>\n",
       "      <td>31438140</td>\n",
       "      <td>ENSG00000203546.7,ENSG00000129480.13_2|ENSG000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000203546.7,ENSG00000129480.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>30520021</td>\n",
       "      <td>30528331</td>\n",
       "      <td>ENSG00000108587.16_3|GOSR1|proximal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000108587.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>30522253</td>\n",
       "      <td>30528331</td>\n",
       "      <td>ENSG00000108587.16_4|GOSR1|distal|cryptic</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000108587.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | +4    |\n",
       "| (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr10        |  87365805 |  87372223 | ...   |\n",
       "| chr10        |  87366138 |  87372223 | ...   |\n",
       "| chr11        | 134502943 | 134505661 | ...   |\n",
       "| chr14        |  31433709 |  31438140 | ...   |\n",
       "| chr17        |  30520021 |  30528331 | ...   |\n",
       "| chr17        |  30522253 |  30528331 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 6 rows and 7 columns from 4 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Name, Score, Strand, gene_id"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "single_gene_ids = set(comb_gtf.as_df().groupby(\"gene_id\")[\"transcript_id\"].nunique().loc[lambda x: x == 1].index)\n",
    "\"NUTM2D\"\n",
    "# comb_gtf.subset(lambda df: df.gene_id.isin(single_gene_ids))\n",
    "papa_cryp_bed.assign(\"gene_id\", lambda df: df.Name.str.split(\"\\\\|\", expand=True)[0].str.split(\"_\", expand=True)[0]).subset(lambda df: df.gene_id.isin(single_gene_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"NUTM2D\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_gtf.to_gtf(\"processed/isoform_specific/regions/papa_feature_counts.gtf\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
