{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ddd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from helpers import get_terminal_regions, _df_update_3p\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17902 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 17,902 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Strand"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BED file of proximal + distal PAS contained within genes (used to plot fig2c in Zeng et al. biorxiv)\n",
    "zeng_bed = pr.read_bed(\"data/zeng_2024/zeng_pas_fig2c.bed\")\n",
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17902 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 17,902 rows and 7 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Strand, pas_id"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate pas with unique id\n",
    "zeng_bed = zeng_bed.assign(\"pas_id\", lambda df: df.Chromosome.str.cat(df[[\"Start\", \"End\", \"Strand\", \"Name\"]].astype(str), sep=\":\"))\n",
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>3</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573257</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573258</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573259</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573260</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>4</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573261</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573262 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 11868     | 12227     | +            | ...   |\n",
       "| chr1         | exon       | 12612     | 12721     | +            | ...   |\n",
       "| chr1         | exon       | 13220     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 12009     | 12057     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57213879  | 57213964  | -            | ...   |\n",
       "| chrY         | exon       | 57213525  | 57213602  | -            | ...   |\n",
       "| chrY         | exon       | 57213203  | 57213357  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,573,262 rows and 10 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_id, gene_name, transcript_id, exon_number, transcript_type"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in Gencode GTF, extract exons\n",
    "exons = pr.read_gtf(\"data/gencode.v40.annotation.gtf\").subset(lambda df: df.Feature == \"exon\")\n",
    "exons = exons[[\"Feature\", \"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"transcript_type\"]]\n",
    "exons\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = exons.boundaries(group_by=\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066354:1066355:+:nan</td>\n",
       "      <td>1059686</td>\n",
       "      <td>1069355</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066455</td>\n",
       "      <td>1066456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066455:1066456:+:nan</td>\n",
       "      <td>1059686</td>\n",
       "      <td>1069355</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1575620</td>\n",
       "      <td>1575621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1575620:1575621:+:nan</td>\n",
       "      <td>1574101</td>\n",
       "      <td>1577075</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000215014.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1577070</td>\n",
       "      <td>1577071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1577070:1577071:+:nan</td>\n",
       "      <td>1574101</td>\n",
       "      <td>1577075</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000215014.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1728507</td>\n",
       "      <td>1728508</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1728507:1728508:+:nan</td>\n",
       "      <td>1702735</td>\n",
       "      <td>1737688</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000268575.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12326385</td>\n",
       "      <td>12326386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12326385:12326386:-:nan</td>\n",
       "      <td>12088009</td>\n",
       "      <td>12421587</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12378537:12378538:-:nan</td>\n",
       "      <td>12088009</td>\n",
       "      <td>12421587</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12378537:12378538:-:nan</td>\n",
       "      <td>12354095</td>\n",
       "      <td>12387328</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000225117.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-:nan</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1380</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1381 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +6    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 1066354   | 1066355   | nan        | .          | ...   |\n",
       "| chr1         | 1066455   | 1066456   | nan        | .          | ...   |\n",
       "| chr1         | 1575620   | 1575621   | nan        | .          | ...   |\n",
       "| chr1         | 1577070   | 1577071   | nan        | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 1,381 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Strand, pas_id, Start_b, End_b, Strand_b, gene_id"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do 'NA' genes overlap with gencode annotated genes\n",
    "pas_na_gene = zeng_bed.subset(lambda df: df.pas_id.str.contains(\"nan\")).join(genes, how = \"left\")\n",
    "pas_na_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_overlap\n",
       "True     1353\n",
       "False      28\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pas_na_gene.assign(\"gene_overlap\", lambda df: df.Start_b.ne(-1)).gene_overlap.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>3</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573257</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573258</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573259</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573260</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>4</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573261</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573262 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 11868     | 12227     | +            | ...   |\n",
       "| chr1         | exon       | 12612     | 12721     | +            | ...   |\n",
       "| chr1         | exon       | 13220     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 12009     | 12057     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57213879  | 57213964  | -            | ...   |\n",
       "| chrY         | exon       | 57213525  | 57213602  | -            | ...   |\n",
       "| chrY         | exon       | 57213203  | 57213357  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,573,262 rows and 10 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_id, gene_name, transcript_id, exon_number, transcript_type"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to last exons of all annotated transcripts\n",
    "exons = exons.assign(\"exon_number\", lambda df: df.exon_number.astype(float).astype(int))\n",
    "print(exons.exon_number.dtypes)\n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1:13220:14409:+:DDX11L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13452</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr1:13452:13670:+:DDX11L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>chr1:30365:30503:+:MIR1302-2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:30975:31097:+:MIR1302-2HG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:30975:31109:+:MIR1302-2HG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246619</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000237801.6_PAR_Y</td>\n",
       "      <td>AMD1P2</td>\n",
       "      <td>ENST00000412936.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57015104:57016096:-:AMD1P2</td>\n",
       "      <td>119470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246620</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228410.6_PAR_Y</td>\n",
       "      <td>ELOCP24</td>\n",
       "      <td>ENST00000456370.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57165511:57165845:-:ELOCP24</td>\n",
       "      <td>119471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246621</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223484.7_PAR_Y</td>\n",
       "      <td>TRPC6P</td>\n",
       "      <td>ENST00000421233.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57171889:57172769:-:TRPC6P</td>\n",
       "      <td>119472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246622</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>ENST00000399966.9_PAR_Y</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chrY:57201142:57202020:-:WASIR1</td>\n",
       "      <td>119473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246623</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>chrY:57212183:57213125:-:DDX11L16</td>\n",
       "      <td>119474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246624 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 13220     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 13452     | 13670     | +            | ...   |\n",
       "| chr1         | exon       | 30365     | 30503     | +            | ...   |\n",
       "| chr1         | exon       | 30975     | 31097     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57165511  | 57165845  | -            | ...   |\n",
       "| chrY         | exon       | 57171889  | 57172769  | -            | ...   |\n",
       "| chrY         | exon       | 57201142  | 57202020  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 246,624 rows and 11 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: gene_id, gene_name, transcript_id, transcript_type, exon_id, ... (+ 1 more.)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract last exons for all annotated transcripts\n",
    "last_exons = get_terminal_regions(exons, number_type=\"stranded\")\n",
    "\n",
    "# assign an 'exon id' - basically just the coordinates (so can handle identical LEs x transcripts of same gene)\n",
    "last_exons = last_exons.assign(\"exon_id\", lambda df: df.Chromosome.str.cat(df[[\"Start\", \"End\", \"Strand\", \"gene_name\"]].astype(str), sep=\":\"))\n",
    "\n",
    "# don't need it\n",
    "last_exons = last_exons.drop(\"exon_number\")\n",
    "\n",
    "# annotate overlapping last exons with a common ID\n",
    "last_exons = last_exons.cluster(strand=True)\n",
    "\n",
    "last_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17902"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# track input PAS\n",
    "pas_ids = set(zeng_bed.pas_id)\n",
    "len(pas_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854398</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:854398:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38045</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38046</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38047</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38048</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705416</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705416:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38049</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38050 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 853390    | 854398    | +            | ...   |\n",
       "| chr1         | 853390    | 859446    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705416  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 38,050 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 1 more.)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign PAS to last exons based on overlap\n",
    "pas_le = last_exons[[\"exon_id\", \"Cluster\"]].join(zeng_bed[[\"Score\", \"pas_id\"]], strandedness=\"same\",)\n",
    "pas_le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pas_id\n",
       "1    13557\n",
       "2     6923\n",
       "3        2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of unique PAS overlapping last exons \n",
    "# (2 = 3'UTR PAS, 1 = ALEs)\n",
    "num_olap_pas = pas_le.as_df().groupby(\"exon_id\")[\"pas_id\"].nunique().reset_index()\n",
    "\n",
    "# number of putative 3'UTR APA / ALE PAS\n",
    "num_olap_pas.pas_id.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1 overlap could arise due to a short annotated ALE (but have a longer annotated ALE of the same gene overlap it)\n",
    "\n",
    "At gene level, check if any last exon has two overlapping PAS, and recalculate the frequency distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pas_id\n",
       "2    4877\n",
       "1    3549\n",
       "3       2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract gene name from exon_id\n",
    "num_olap_pas.loc[:, \"gene_name\"] = num_olap_pas.exon_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "\n",
    "# for each gene, pick the max number of pas (pas_id column) and recount distribution (i.e. does gene have at least one tandem APA last exon)\n",
    "num_olap_pas.groupby(\"gene_name\")[\"pas_id\"].max().value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tandem APA last exons are now much more frequent, as expected\n",
    "\n",
    "To define tandem APA last exons, will need to subset for exon IDs with two overlapping PAS. Will keep last exons with distinct 5'ends (but still 2 overlapping) for now, as can't see a reasonable justification for removing/prioritising.\n",
    "\n",
    "1. Get list/set of exon IDs with two overlapping PAS, subset pas_le for these events\n",
    "2. Update the 3' coordinate of last exons (strand aware) to the reported PAS.\n",
    "3. Drop duplicate intervals (arise from identical last exons/extended 3'ends with same 5'end), get BED ready (i.e. define a cleaned 'Name' field)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066459</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1065829:1066459:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066354:1066355:+:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18689</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18690</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18691</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18692</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18694 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 853390    | 859446    | +            | ...   |\n",
       "| chr1         | 853390    | 859446    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 18,694 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 1 more.)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get last exons with two overlapping PAS & subset for joined last exon + pas coordinates\n",
    "tandem_exons_ids = set(num_olap_pas.loc[num_olap_pas[\"pas_id\"] == 2, \"exon_id\"])\n",
    "\n",
    "pas_le_3utr = pas_le.subset(lambda df: df.exon_id.isin(tandem_exons_ids))\n",
    "pas_le_3utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of negative or zero-length updated intervals to be dropped - 940\n",
      "Number of negative or zero-length updated intervals to be dropped - 1030\n",
      "Number of negative or zero-length updated intervals to be dropped - 728\n",
      "Number of negative or zero-length updated intervals to be dropped - 612\n",
      "Number of negative or zero-length updated intervals to be dropped - 552\n",
      "Number of negative or zero-length updated intervals to be dropped - 584\n",
      "Number of negative or zero-length updated intervals to be dropped - 384\n",
      "Number of negative or zero-length updated intervals to be dropped - 442\n",
      "Number of negative or zero-length updated intervals to be dropped - 452\n",
      "Number of negative or zero-length updated intervals to be dropped - 514\n",
      "Number of negative or zero-length updated intervals to be dropped - 408\n",
      "Number of negative or zero-length updated intervals to be dropped - 600\n",
      "Number of negative or zero-length updated intervals to be dropped - 408\n",
      "Number of negative or zero-length updated intervals to be dropped - 376\n",
      "Number of negative or zero-length updated intervals to be dropped - 358\n",
      "Number of negative or zero-length updated intervals to be dropped - 360\n",
      "Number of negative or zero-length updated intervals to be dropped - 310\n",
      "Number of negative or zero-length updated intervals to be dropped - 396\n",
      "Number of negative or zero-length updated intervals to be dropped - 400\n",
      "Number of negative or zero-length updated intervals to be dropped - 410\n",
      "Number of negative or zero-length updated intervals to be dropped - 384\n",
      "Number of negative or zero-length updated intervals to be dropped - 504\n",
      "Number of negative or zero-length updated intervals to be dropped - 534\n",
      "Number of negative or zero-length updated intervals to be dropped - 406\n",
      "Number of negative or zero-length updated intervals to be dropped - 158\n",
      "Number of negative or zero-length updated intervals to be dropped - 204\n",
      "Number of negative or zero-length updated intervals to be dropped - 328\n",
      "Number of negative or zero-length updated intervals to be dropped - 356\n",
      "Number of negative or zero-length updated intervals to be dropped - 268\n",
      "Number of negative or zero-length updated intervals to be dropped - 340\n",
      "Number of negative or zero-length updated intervals to be dropped - 384\n",
      "Number of negative or zero-length updated intervals to be dropped - 336\n",
      "Number of negative or zero-length updated intervals to be dropped - 492\n",
      "Number of negative or zero-length updated intervals to be dropped - 572\n",
      "Number of negative or zero-length updated intervals to be dropped - 158\n",
      "Number of negative or zero-length updated intervals to be dropped - 252\n",
      "Number of negative or zero-length updated intervals to be dropped - 432\n",
      "Number of negative or zero-length updated intervals to be dropped - 382\n",
      "Number of negative or zero-length updated intervals to be dropped - 290\n",
      "Number of negative or zero-length updated intervals to be dropped - 314\n",
      "Number of negative or zero-length updated intervals to be dropped - 104\n",
      "Number of negative or zero-length updated intervals to be dropped - 92\n",
      "Number of negative or zero-length updated intervals to be dropped - 190\n",
      "Number of negative or zero-length updated intervals to be dropped - 180\n",
      "Number of negative or zero-length updated intervals to be dropped - 432\n",
      "Number of negative or zero-length updated intervals to be dropped - 314\n",
      "Number of negative or zero-length updated intervals to be dropped - 16\n",
      "Number of negative or zero-length updated intervals to be dropped - 8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>629998</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>630683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630368</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854388</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859445</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>859444</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066355</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1065829:1066459:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066354:1066355:+:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18689</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19692491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18690</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19703867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18691</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18692</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19703867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19705420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18694 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 629998    | +            | ...   |\n",
       "| chr1         | 629639    | 630368    | +            | ...   |\n",
       "| chr1         | 853390    | 854388    | +            | ...   |\n",
       "| chr1         | 853390    | 859445    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703866  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703866  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 18,694 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 1 more.)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update 3' exon coordinates to the PAS coordinate\n",
    "pas_le_3utr_upd = pas_le_3utr.apply(lambda df: _df_update_3p(df))\n",
    "pas_le_3utr_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>629998</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630368</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854388</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859445</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066355</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1066354:1066355:+:ENSG00000217801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10373</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2934524</td>\n",
       "      <td>2934771</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:2934524:2934525:-:ENSG00000278847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10374</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10375</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10376</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10377</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10378 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Score      | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+-------|\n",
       "| chr1         | 629639    | 629998    | +            | .          | ...   |\n",
       "| chr1         | 629639    | 630368    | +            | .          | ...   |\n",
       "| chr1         | 853390    | 854388    | +            | .          | ...   |\n",
       "| chr1         | 853390    | 859445    | +            | .          | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...   |\n",
       "| chrY         | 19691944  | 19694606  | -            | .          | ...   |\n",
       "| chrY         | 19692490  | 19694606  | -            | .          | ...   |\n",
       "| chrY         | 19703866  | 19706345  | -            | .          | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | .          | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 10,378 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean up\n",
    "\n",
    "# drop duplicates\n",
    "pas_le_3utr_upd = pas_le_3utr_upd.drop_duplicate_positions(strand=True)\n",
    "\n",
    "# remake pas_id, using gene name from exon_id if required\n",
    "pas_le_3utr_upd = pas_le_3utr_upd.as_df()\n",
    "\n",
    "# generate standardised gene name column, preferring name from exon_id where provided PAS has no gene name\n",
    "pas_le_3utr_upd.loc[:, \"pas_gn\"] = pas_le_3utr_upd.pas_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "pas_le_3utr_upd.loc[:, \"exon_gn\"] = pas_le_3utr_upd.exon_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "pas_le_3utr_upd.loc[:, \"gene_name\"] = np.where(pas_le_3utr_upd.pas_gn == \"nan\", pas_le_3utr_upd[\"exon_gn\"], pas_le_3utr_upd[\"pas_gn\"])\n",
    "\n",
    "# Update pas_id to include the standardised gene name\n",
    "pas_le_3utr_upd.loc[:, \"pas_id\"] = pas_le_3utr_upd.pas_id.str.split(':').str[:4].str.join(':')\n",
    "pas_le_3utr_upd.loc[:, \"pas_id\"] = pas_le_3utr_upd.pas_id.str.cat(pas_le_3utr_upd.gene_name, sep=\":\")\n",
    "pas_le_3utr_upd.rename(columns={\"pas_id\": \"Name\"}, inplace=True)\n",
    "\n",
    "pas_le_3utr_upd = pr.PyRanges(pas_le_3utr_upd)\n",
    "\n",
    "pas_le_3utr_upd[[\"Score\", \"Name\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ALEs...\n",
    "\n",
    "Need to identify genes where have only 1 PAS overlapping each ALE.\n",
    "Then using those genes, check that the last exons themselves are not overlapping, and that the PAS are distinct (may only be one / 2 pas for gene overlapping an annotated exon - need to track those)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617898</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1619210</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1619210:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617323</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12049</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12050</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932006</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932006:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12051</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932316:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12052</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12053</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12054 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1013983   | 1014540   | +            | ...   |\n",
       "| chr1         | 1013983   | 1014540   | +            | ...   |\n",
       "| chr1         | 1616507   | 1617898   | +            | ...   |\n",
       "| chr1         | 1616507   | 1619210   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 18932006  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932316  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 12,054 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 1 more.)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each gene, get max overlapping PAS count of any of its last exons\n",
    "max_gene_olap_pas =  num_olap_pas.groupby(\"gene_name\")[\"pas_id\"].max()\n",
    "\n",
    "# subset to genes with overlap PAS count of 1, becasue these do not have any tandem APA\n",
    "put_ale_genes = set(max_gene_olap_pas[max_gene_olap_pas.eq(1)].index)\n",
    "# get exon ids\n",
    "put_ale_ids = set(num_olap_pas.loc[num_olap_pas.gene_name.isin(put_ale_genes), \"exon_id\"])\n",
    "\n",
    "pas_le_ale = pas_le.subset(lambda df: df.exon_id.isin(put_ale_ids))\n",
    "pas_le_ale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cluster  pas_id\n",
       "1        1         2621\n",
       "2        2          900\n",
       "1        2           27\n",
       "3        3            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ALEs should be non-overlapping. Each gene should therefore have two unique PAS, and two unique Cluster values\n",
    "pas_le_ale = pas_le_ale.assign(\"gene_name\", lambda df: df.exon_id.str.split(\":\", expand=True)[4])\n",
    "# count unqiue pas + non-overlapping ales\n",
    "ale_uniq_counts = pas_le_ale.as_df().groupby(\"gene_name\")[[\"Cluster\", \"pas_id\"]].nunique().reset_index()\n",
    "\n",
    "# quick check counts\n",
    "ale_uniq_counts[[\"Cluster\",\"pas_id\"]].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would anticipate 2 + 2 being the bonafide ALEs.\n",
    "1 + 1 likely mean 2nd PAS doesn't overlap with known last exon boundaries (could be intronic PAS, or just downstream of annotated ALEs?). Would have to better parse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42674847</td>\n",
       "      <td>42676446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42674847:42676446:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42675287</td>\n",
       "      <td>42676713</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42675287:42676713:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42675366</td>\n",
       "      <td>42676043</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42675366:42676043:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42675366</td>\n",
       "      <td>42676619</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42675366:42676619:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42675732</td>\n",
       "      <td>42676202</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42675732:42676202:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42675737</td>\n",
       "      <td>42676043</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42675737:42676043:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676040</td>\n",
       "      <td>42676041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676040:42676041:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42676583</td>\n",
       "      <td>42676758</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676583:42676758:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676757</td>\n",
       "      <td>42676758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676757:42676758:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42676583</td>\n",
       "      <td>42676758</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676583:42676758:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676757</td>\n",
       "      <td>42676758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676757:42676758:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42676583</td>\n",
       "      <td>42676758</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676583:42676758:+:PPIH</td>\n",
       "      <td>1320</td>\n",
       "      <td>42676757</td>\n",
       "      <td>42676758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:42676757:42676758:+:PPIH</td>\n",
       "      <td>PPIH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44777999</td>\n",
       "      <td>44778663</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44777999:44778663:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778053</td>\n",
       "      <td>44778054</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778053:44778054:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44777999</td>\n",
       "      <td>44778087</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44777999:44778087:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778053</td>\n",
       "      <td>44778054</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778053:44778054:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44778779</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778575:44778779:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778738</td>\n",
       "      <td>44778739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778738:44778739:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44778739</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778575:44778739:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778738</td>\n",
       "      <td>44778739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778738:44778739:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44778779</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778575:44778779:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778738</td>\n",
       "      <td>44778739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778738:44778739:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44778779</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778575:44778779:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778738</td>\n",
       "      <td>44778739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778738:44778739:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778575</td>\n",
       "      <td>44778779</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778575:44778779:+:RPS8</td>\n",
       "      <td>1468</td>\n",
       "      <td>44778738</td>\n",
       "      <td>44778739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778738:44778739:+:RPS8</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>231351303</td>\n",
       "      <td>231353132</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:231351303:231353132:+:SPRTN</td>\n",
       "      <td>5038</td>\n",
       "      <td>231352521</td>\n",
       "      <td>231352522</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:231352521:231352522:+:SPRTN</td>\n",
       "      <td>SPRTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>231352609</td>\n",
       "      <td>231355023</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:231352609:231355023:+:SPRTN</td>\n",
       "      <td>5038</td>\n",
       "      <td>231355022</td>\n",
       "      <td>231355023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:231355022:231355023:+:SPRTN</td>\n",
       "      <td>SPRTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89625961</td>\n",
       "      <td>89626826</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89625961:89626826:-:LRRC8C-DT</td>\n",
       "      <td>7583</td>\n",
       "      <td>89625974</td>\n",
       "      <td>89625975</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89625974:89625975:-:LRRC8C-DT</td>\n",
       "      <td>LRRC8C-DT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89625964</td>\n",
       "      <td>89626826</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89625964:89626826:-:LRRC8C-DT</td>\n",
       "      <td>7583</td>\n",
       "      <td>89625974</td>\n",
       "      <td>89625975</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89625974:89625975:-:LRRC8C-DT</td>\n",
       "      <td>LRRC8C-DT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89626019</td>\n",
       "      <td>89632209</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89626019:89632209:-:LRRC8C-DT</td>\n",
       "      <td>7583</td>\n",
       "      <td>89631878</td>\n",
       "      <td>89631879</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89631878:89631879:-:LRRC8C-DT</td>\n",
       "      <td>LRRC8C-DT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89631875</td>\n",
       "      <td>89632209</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89631875:89632209:-:LRRC8C-DT</td>\n",
       "      <td>7583</td>\n",
       "      <td>89631878</td>\n",
       "      <td>89631879</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:89631878:89631879:-:LRRC8C-DT</td>\n",
       "      <td>LRRC8C-DT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +8    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 42674847  | 42676446  | +            | ...   |\n",
       "| chr1         | 42675287  | 42676713  | +            | ...   |\n",
       "| chr1         | 42675366  | 42676043  | +            | ...   |\n",
       "| chr1         | 42675366  | 42676619  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chr1         | 89625961  | 89626826  | -            | ...   |\n",
       "| chr1         | 89625964  | 89626826  | -            | ...   |\n",
       "| chr1         | 89626019  | 89632209  | -            | ...   |\n",
       "| chr1         | 89631875  | 89632209  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 22 rows and 12 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the 1 ALE but two PAS events\n",
    "pas_le_ale.subset(lambda df: df.gene_name.isin(set(ale_uniq_counts.loc[(ale_uniq_counts[\"Cluster\"] == 1) & (ale_uniq_counts[\"pas_id\"] == 2), \"gene_name\"]))).head(22)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Partly overlapping ALEs with different 5'end coordinates. Probably better described as tandem ALEs, just need to merge the exons before updating the PAS?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2681</th>\n",
       "      <td>SFPQ</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  Cluster  pas_id\n",
       "2681      SFPQ        2       2"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# is SFPQ represented in putative ALEs?\n",
    "ale_uniq_counts.loc[(ale_uniq_counts[\"Cluster\"] == 2) & (ale_uniq_counts[\"pas_id\"] == 2) & (ale_uniq_counts[\"gene_name\"] == \"SFPQ\"),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617898</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1619210</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1619210:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617323</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630605</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630605:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630605</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630605</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630605:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630605</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4523</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4524</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932006</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932006:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4525</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932316:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4526</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4528 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +8    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1616507   | 1617898   | +            | ...   |\n",
       "| chr1         | 1616507   | 1619210   | +            | ...   |\n",
       "| chr1         | 1617077   | 1617323   | +            | ...   |\n",
       "| chr1         | 1630291   | 1630605   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 18932006  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932316  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 4,528 rows and 12 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get last exons with two overlapping PAS & subset for joined last exon + pas coordinates\n",
    "# get IDs for 'bonafide' ALEs\n",
    "ale_gene_names = ale_uniq_counts.loc[(ale_uniq_counts[\"Cluster\"] == 2) & (ale_uniq_counts[\"pas_id\"] == 2), \"gene_name\"]\n",
    "pas_le_ale_bf = pas_le_ale.subset(lambda df: df.gene_name.isin(ale_gene_names))\n",
    "pas_le_ale_bf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of negative or zero-length updated intervals to be dropped - 177\n",
      "Number of negative or zero-length updated intervals to be dropped - 217\n",
      "Number of negative or zero-length updated intervals to be dropped - 237\n",
      "Number of negative or zero-length updated intervals to be dropped - 99\n",
      "Number of negative or zero-length updated intervals to be dropped - 125\n",
      "Number of negative or zero-length updated intervals to be dropped - 82\n",
      "Number of negative or zero-length updated intervals to be dropped - 66\n",
      "Number of negative or zero-length updated intervals to be dropped - 107\n",
      "Number of negative or zero-length updated intervals to be dropped - 83\n",
      "Number of negative or zero-length updated intervals to be dropped - 48\n",
      "Number of negative or zero-length updated intervals to be dropped - 104\n",
      "Number of negative or zero-length updated intervals to be dropped - 136\n",
      "Number of negative or zero-length updated intervals to be dropped - 119\n",
      "Number of negative or zero-length updated intervals to be dropped - 168\n",
      "Number of negative or zero-length updated intervals to be dropped - 158\n",
      "Number of negative or zero-length updated intervals to be dropped - 119\n",
      "Number of negative or zero-length updated intervals to be dropped - 98\n",
      "Number of negative or zero-length updated intervals to be dropped - 98\n",
      "Number of negative or zero-length updated intervals to be dropped - 58\n",
      "Number of negative or zero-length updated intervals to be dropped - 93\n",
      "Number of negative or zero-length updated intervals to be dropped - 90\n",
      "Number of negative or zero-length updated intervals to be dropped - 155\n",
      "Number of negative or zero-length updated intervals to be dropped - 80\n",
      "Number of negative or zero-length updated intervals to be dropped - 106\n",
      "Number of negative or zero-length updated intervals to be dropped - 27\n",
      "Number of negative or zero-length updated intervals to be dropped - 61\n",
      "Number of negative or zero-length updated intervals to be dropped - 62\n",
      "Number of negative or zero-length updated intervals to be dropped - 114\n",
      "Number of negative or zero-length updated intervals to be dropped - 110\n",
      "Number of negative or zero-length updated intervals to be dropped - 91\n",
      "Number of negative or zero-length updated intervals to be dropped - 167\n",
      "Number of negative or zero-length updated intervals to be dropped - 159\n",
      "Number of negative or zero-length updated intervals to be dropped - 88\n",
      "Number of negative or zero-length updated intervals to be dropped - 86\n",
      "Number of negative or zero-length updated intervals to be dropped - 60\n",
      "Number of negative or zero-length updated intervals to be dropped - 45\n",
      "Number of negative or zero-length updated intervals to be dropped - 74\n",
      "Number of negative or zero-length updated intervals to be dropped - 113\n",
      "Number of negative or zero-length updated intervals to be dropped - 127\n",
      "Number of negative or zero-length updated intervals to be dropped - 31\n",
      "Number of negative or zero-length updated intervals to be dropped - 17\n",
      "Number of negative or zero-length updated intervals to be dropped - 43\n",
      "Number of negative or zero-length updated intervals to be dropped - 45\n",
      "Number of negative or zero-length updated intervals to be dropped - 27\n",
      "Number of negative or zero-length updated intervals to be dropped - 61\n",
      "Number of negative or zero-length updated intervals to be dropped - 74\n",
      "Number of negative or zero-length updated intervals to be dropped - 4\n",
      "Number of negative or zero-length updated intervals to be dropped - 19\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617898</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630605</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630605:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630605</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "      <td>MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6632795</td>\n",
       "      <td>6633562</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6632795:6633562:+:THAP3</td>\n",
       "      <td>248</td>\n",
       "      <td>6633561</td>\n",
       "      <td>6633562</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6633561:6633562:+:THAP3</td>\n",
       "      <td>THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6634019</td>\n",
       "      <td>6635586</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6634019:6635586:+:THAP3</td>\n",
       "      <td>249</td>\n",
       "      <td>6635585</td>\n",
       "      <td>6635586</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6635585:6635586:+:THAP3</td>\n",
       "      <td>THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19593087</td>\n",
       "      <td>19594158</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19593087:19594161:+:TXLNGY</td>\n",
       "      <td>118984</td>\n",
       "      <td>19594157</td>\n",
       "      <td>19594161</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19594157:19594158:+:nan</td>\n",
       "      <td>TXLNGY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2047</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234577</td>\n",
       "      <td>13234826</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234576:13234826:-:UTY</td>\n",
       "      <td>119277</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234578</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234577:13234578:-:UTY</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2048</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13248385</td>\n",
       "      <td>13249882</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13248378:13249882:-:UTY</td>\n",
       "      <td>119278</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13248386</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13248385:13248386:-:UTY</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2049</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872593</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18872500:18872834:-:TTTY14</td>\n",
       "      <td>119337</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18872593:18872594:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2050</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932450</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2051 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +8    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1616507   | 1617321   | +            | ...   |\n",
       "| chr1         | 1617077   | 1617321   | +            | ...   |\n",
       "| chr1         | 1630291   | 1630605   | +            | ...   |\n",
       "| chr1         | 6632795   | 6633562   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 13234577  | 13234826  | -            | ...   |\n",
       "| chrY         | 13248385  | 13249882  | -            | ...   |\n",
       "| chrY         | 18872593  | 18872834  | -            | ...   |\n",
       "| chrY         | 18932449  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 2,051 rows and 12 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: exon_id, Cluster, Start_b, End_b, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update 3' exon coordinates to the PAS coordinate\n",
    "pas_le_ale_bf_upd = pas_le_ale_bf.apply(lambda df: _df_update_3p(df))\n",
    "pas_le_ale_bf_upd = pas_le_ale_bf_upd.drop_duplicate_positions(strand=True)\n",
    "pas_le_ale_bf_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630605</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6632795</td>\n",
       "      <td>6633562</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:6633561:6633562:+:THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6634019</td>\n",
       "      <td>6635586</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:6635585:6635586:+:THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19593087</td>\n",
       "      <td>19594158</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:19594157:19594158:+:TXLNGY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2047</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234577</td>\n",
       "      <td>13234826</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:13234577:13234578:-:UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2048</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13248385</td>\n",
       "      <td>13249882</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:13248385:13248386:-:UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2049</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872593</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:18872593:18872594:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2050</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2051 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Score      | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+-------|\n",
       "| chr1         | 1616507   | 1617321   | +            | .          | ...   |\n",
       "| chr1         | 1617077   | 1617321   | +            | .          | ...   |\n",
       "| chr1         | 1630291   | 1630605   | +            | .          | ...   |\n",
       "| chr1         | 6632795   | 6633562   | +            | .          | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...   |\n",
       "| chrY         | 13234577  | 13234826  | -            | .          | ...   |\n",
       "| chrY         | 13248385  | 13249882  | -            | .          | ...   |\n",
       "| chrY         | 18872593  | 18872834  | -            | .          | ...   |\n",
       "| chrY         | 18932449  | 18932841  | -            | .          | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 2,051 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remake pas_id, using gene name from exon_id if required\n",
    "pas_le_ale_bf_upd = pas_le_ale_bf_upd.drop(\"gene_name\").as_df()\n",
    "\n",
    "# generate standardised gene name column, preferring name from exon_id where provided PAS has no gene name\n",
    "pas_le_ale_bf_upd.loc[:, \"pas_gn\"] = pas_le_ale_bf_upd.pas_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "pas_le_ale_bf_upd.loc[:, \"exon_gn\"] = pas_le_ale_bf_upd.exon_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "pas_le_ale_bf_upd.loc[:, \"gene_name\"] = np.where(pas_le_ale_bf_upd.pas_gn == \"nan\", pas_le_ale_bf_upd[\"exon_gn\"], pas_le_ale_bf_upd[\"pas_gn\"])\n",
    "\n",
    "# Update pas_id to include the standardised gene name\n",
    "pas_le_ale_bf_upd.loc[:, \"pas_id\"] = pas_le_ale_bf_upd.pas_id.str.split(':').str[:4].str.join(':')\n",
    "pas_le_ale_bf_upd.loc[:, \"pas_id\"] = pas_le_ale_bf_upd.pas_id.str.cat(pas_le_ale_bf_upd.gene_name, sep=\":\")\n",
    "pas_le_ale_bf_upd.rename(columns={\"pas_id\": \"Name\"}, inplace=True)\n",
    "\n",
    "pas_le_ale_bf_upd = pr.PyRanges(pas_le_ale_bf_upd)\n",
    "\n",
    "pas_le_ale_bf_upd[[\"Score\", \"Name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35176380</td>\n",
       "      <td>35176474</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:35176380:35176381:-:SFPQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35176380</td>\n",
       "      <td>35176471</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:35176380:35176381:-:SFPQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35183601</td>\n",
       "      <td>35184593</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:35183601:35183602:-:SFPQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Score      | +1    |\n",
       "| (category)   |   (int64) |   (int64) | (category)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+-------|\n",
       "| chr1         |  35176380 |  35176474 | -            | .          | ...   |\n",
       "| chr1         |  35176380 |  35176471 | -            | .          | ...   |\n",
       "| chr1         |  35183601 |  35184593 | -            | .          | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pas_le_ale_bf_upd[[\"Score\", \"Name\"]].subset(lambda df: df.Name.str.contains(\"SFPQ\", regex=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TMED10', 'GOLGA7B', 'UBE2R2', 'SLC6A8', 'SLC24A3', 'PAK4', 'TTBK1', 'SMC1A', 'ELP3', 'JPT1', 'TFDP2', 'TMEM106B', 'F11R', 'BRD9', 'MARK3', 'EIF4E2', 'KLHL42', 'ARHGAP32', 'SLC19A1', 'NDRG4', 'EMC10', 'FOXK2', 'SSR1', 'CCNL2', 'SLIT3', 'STMN2', 'NEFL', 'GSTO2', 'DPYSL5', 'ELAVL4', 'ZNF462', 'NUFIP2', 'TLX1', 'GPR173', 'SEC14L1', 'DYNC1LI1', 'MDGA1', 'ZMAT2', 'TARDBP', 'SMARCA4', 'PGRMC2', 'RAB11A', 'AZIN1', 'AGPAT4', 'CNPY3', 'SREK1', 'CSNK2A1', 'RFNG', 'G3BP1', 'HNRNPC', 'SSU72', 'LRRC3', 'EGFR', 'RPN1', 'ARNT', 'H3-3B', 'CORO1C', 'TAFA5', 'NUCKS1', 'SYP', 'PTPN9', 'ELP6', 'NDUFA9', 'NAV1', 'PARD6G', 'NSMAF', 'ELP1', 'OLA1', 'SIX3', 'CHRNB4', 'LSM14A', 'KPNA4', 'TEF', 'SFPQ', 'PPM1A', 'SMG7', 'DYRK2', 'NFE2L1', 'GREB1', 'ABCC5', 'HIF1AN', 'TFDP1', 'DIDO1', 'CADM1', 'CLSTN3', 'GGA2', 'SERF2', 'ARMC10', 'UBE2H'}\n"
     ]
    }
   ],
   "source": [
    "# How many of Zeng et al. targets are missing?\n",
    "with open(\"data/zeng_2024/zeng_target_genes.txt\", \"r\") as infile:\n",
    "    zeng_target_genes = set([line.rstrip(\"\\n\").replace(\"'\",\"\") for line in infile])\n",
    "\n",
    "print(zeng_target_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ARHGAP32',\n",
       " 'BRD9',\n",
       " 'CHRNB4',\n",
       " 'CLSTN3',\n",
       " 'CNPY3',\n",
       " 'EGFR',\n",
       " 'ELP6',\n",
       " 'GSTO2',\n",
       " 'HNRNPC',\n",
       " 'LSM14A',\n",
       " 'NAV1',\n",
       " 'NSMAF',\n",
       " 'PAK4',\n",
       " 'PARD6G',\n",
       " 'RFNG',\n",
       " 'SEC14L1',\n",
       " 'SIX3',\n",
       " 'SLC24A3',\n",
       " 'SLIT3',\n",
       " 'SMG7',\n",
       " 'SREK1',\n",
       " 'STMN2',\n",
       " 'TAFA5',\n",
       " 'TLX1',\n",
       " 'TTBK1',\n",
       " 'UBE2R2'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get events where have defined regions\n",
    "defined_genes = set(pas_le_3utr_upd.gene_name).union(set(pas_le_ale_bf_upd.gene_name))\n",
    "missing_genes = zeng_target_genes.difference(defined_genes)\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Missing - class of 3'UTR extensions - TLX1, RFNG, SIX3, ELP6 (?). Others are cryptic ALEs/IPAs that I have in my hits (STMN2, CNPY3, ARHGAP32). Would wager remainin are intronic / 3'UTR extensions I miss (check in IGV)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39007689</td>\n",
       "      <td>39007690</td>\n",
       "      <td>chr1:39007689:+|3'UTR extension|AKIRIN1|0.0461...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>AKIRIN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39766107</td>\n",
       "      <td>39766108</td>\n",
       "      <td>chr1:39766107:+|3'UTR extension|PPIE|0.04583|0...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PPIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>52349654</td>\n",
       "      <td>52349655</td>\n",
       "      <td>chr1:52349654:+|3'UTR extension|ZFYVE9|0.02687...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ZFYVE9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110203695</td>\n",
       "      <td>110203696</td>\n",
       "      <td>chr1:110203695:+|3'UTR extension|SLC6A17|0.008...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC6A17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150353337</td>\n",
       "      <td>150353338</td>\n",
       "      <td>chr1:150353337:+|3'UTR extension|PRPF3|0.01155...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PRPF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47217646</td>\n",
       "      <td>47217647</td>\n",
       "      <td>chrX:47217646:+|3'UTR extension|UBA1|0.01779|0...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>UBA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>chrX</td>\n",
       "      <td>77909737</td>\n",
       "      <td>77909738</td>\n",
       "      <td>chrX:77909737:+|3'UTR extension|COX7B|0.04784|...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>COX7B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47631616</td>\n",
       "      <td>47631617</td>\n",
       "      <td>chrX:47631616:-|3'UTR extension|ELK1|0.01899|0...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ELK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54441090</td>\n",
       "      <td>54441091</td>\n",
       "      <td>chrX:54441090:-|3'UTR extension|FGD1|0.05151|0...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FGD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155255938</td>\n",
       "      <td>155255939</td>\n",
       "      <td>chrX:155255938:-|3'UTR extension|RAB39B|0.0365...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB39B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +4    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 39007689  | 39007690  | ...   |\n",
       "| chr1         | 39766107  | 39766108  | ...   |\n",
       "| chr1         | 52349654  | 52349655  | ...   |\n",
       "| chr1         | 110203695 | 110203696 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 77909737  | 77909738  | ...   |\n",
       "| chrX         | 47631616  | 47631617  | ...   |\n",
       "| chrX         | 54441090  | 54441091  | ...   |\n",
       "| chrX         | 155255938 | 155255939 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 162 rows and 7 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Name, Score, Strand, gene_name"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zeng_ext = pr.read_bed(\"data/zeng_2024/supplementary_s5.cryptic_pas.3UTR_extension.bed\")\n",
    "zeng_intron = pr.read_bed(\"data/zeng_2024/supplementary_s5.cryptic_pas.intron.bed\")\n",
    "zeng_ext = zeng_ext.assign(\"gene_name\", lambda df: df.Name.str.split(\"|\", expand=True)[2])\n",
    "zeng_intron = zeng_intron.assign(\"gene_name\", lambda df: df.Name.str.split(\"|\", expand=True)[2])\n",
    "zeng_ext_gn = set(zeng_ext.gene_name)\n",
    "zeng_intron_gn = set(zeng_intron.gene_name)\n",
    "zeng_ext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RFNG', 'SIX3'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3'exts\n",
    "missing_genes.intersection(zeng_ext_gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EGFR', 'GSTO2', 'STMN2'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# intronic PAS\n",
    "missing_genes.intersection(zeng_intron_gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RFNG', 'SIX3', 'TLX1'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# missing from provided PAS\n",
    "missing_genes.difference(set(zeng_bed.Name))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So basically pretty much all provided, just need to rescue from current approach...\n",
    "1. How many are just downstream of known last exons? (regardless of intron vs 3'UTR)\n",
    "2. How many don't have a known ALE?\n",
    "\n",
    "first, need to just know how many of these missing PAS don't overlap known last exons. Some genes will have at least 1 PAS overlapping (2621 genes based on ALE analysis),"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6694\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066354:1066355:+:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066455</td>\n",
       "      <td>1066456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066455:1066456:+:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1575620</td>\n",
       "      <td>1575621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1575620:1575621:+:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6689</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2934524</td>\n",
       "      <td>2934525</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2934524:2934525:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6690</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12326385</td>\n",
       "      <td>12326386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12326385:12326386:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6691</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12378537:12378538:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6692</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6694 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 1011462   | 1011463   | ISG15      | .          | ...   |\n",
       "| chr1         | 1014537   | 1014538   | ISG15      | .          | ...   |\n",
       "| chr1         | 1066354   | 1066355   | nan        | .          | ...   |\n",
       "| chr1         | 1066455   | 1066456   | nan        | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 12326385  | 12326386  | nan        | .          | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 6,694 rows and 7 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Strand, pas_id"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get pas IDs that haven't formed part of tandem 3'UTR-APA / bonafide ALEs\n",
    "missing_pas_ids = pas_ids.difference(set(pas_le_3utr_upd.Name).union(set(pas_le_ale_bf_upd.Name)))\n",
    "print(len(missing_pas_ids))\n",
    "\n",
    "# subset to missing PAS\n",
    "zeng_bed_missing = zeng_bed.subset(lambda df: df.pas_id.isin(missing_pas_ids))\n",
    "zeng_bed_missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    6.694000e+03\n",
      "mean     1.525460e+04\n",
      "std      5.585132e+04\n",
      "min      0.000000e+00\n",
      "0%       0.000000e+00\n",
      "20%      0.000000e+00\n",
      "40%      0.000000e+00\n",
      "50%      4.000000e+00\n",
      "60%      1.108600e+03\n",
      "80%      9.649800e+03\n",
      "100%     1.123544e+06\n",
      "max      1.123544e+06\n",
      "Name: Distance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "      <td>exon</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187608.10</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>ENST00000624697.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066354:1066355:+:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801.11</td>\n",
       "      <td>ENSG00000217801</td>\n",
       "      <td>ENST00000394517.7</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1:1065829:1066459:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1066455</td>\n",
       "      <td>1066456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1066455:1066456:+:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801.11</td>\n",
       "      <td>ENSG00000217801</td>\n",
       "      <td>ENST00000394517.7</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1:1065829:1066459:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1575620</td>\n",
       "      <td>1575621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1575620:1575621:+:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>1574974</td>\n",
       "      <td>1577075</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000215014.5</td>\n",
       "      <td>ENSG00000215014</td>\n",
       "      <td>ENST00000366221.3</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>chr1:1574974:1577075:+:ENSG00000215014</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1577070</td>\n",
       "      <td>1577071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1577070:1577071:+:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>1574974</td>\n",
       "      <td>1577075</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000215014.5</td>\n",
       "      <td>ENSG00000215014</td>\n",
       "      <td>ENST00000366221.3</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>chr1:1574974:1577075:+:ENSG00000215014</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6689</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2934524</td>\n",
       "      <td>2934525</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2934524:2934525:-:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>2934405</td>\n",
       "      <td>2934771</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278847.1</td>\n",
       "      <td>ENSG00000278847</td>\n",
       "      <td>ENST00000611750.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chrY:2934405:2934771:-:ENSG00000278847</td>\n",
       "      <td>119157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6690</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12378537:12378538:-:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>12378531</td>\n",
       "      <td>12378749</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>ENST00000382965.3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chrY:12378531:12378749:-:GYG2P1</td>\n",
       "      <td>119266</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6691</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>ENST00000566193.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6692</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>ENST00000566193.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12326385</td>\n",
       "      <td>12326386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:12326385:12326386:-:nan</td>\n",
       "      <td>exon</td>\n",
       "      <td>12354415</td>\n",
       "      <td>12354771</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>ENST00000382966.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chrY:12354415:12354771:-:GYG2P1</td>\n",
       "      <td>119265</td>\n",
       "      <td>28030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6694 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +13   |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 1014537   | 1014538   | ISG15      | .          | ...   |\n",
       "| chr1         | 1066354   | 1066355   | nan        | .          | ...   |\n",
       "| chr1         | 1066455   | 1066456   | nan        | .          | ...   |\n",
       "| chr1         | 1575620   | 1575621   | nan        | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 12326385  | 12326386  | nan        | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 6,694 rows and 18 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Strand, pas_id, Feature, Start_b, End_b, Strand_b, gene_id, ... (+ 6 more.)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate distance to nearest (upstream) last exon (allowing overlaps)\n",
    "missing_pas_le_nr = zeng_bed_missing.nearest(last_exons, strandedness=\"same\", overlap=True,how=\"upstream\",)\n",
    "print(missing_pas_le_nr.Distance.describe(percentiles=[i * 0.1 for i in range(0,12,2)]))\n",
    "missing_pas_le_nr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "min    max  \n",
       "True   False    1713\n",
       "False  False    1507\n",
       "True   True     1055\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each gene, report the min and max distance (i.e. how many genes are just missing annotated last exon for 1 PAS?)\n",
    "dist_min_max = missing_pas_le_nr.as_df().groupby(\"gene_name\")[\"Distance\"].agg([min, max]).reset_index().sort_values(by=[\"min\", \"max\"])\n",
    "# gene counts for both overlap last exons (True, True), 1 overlaps last exon (True, False) and neither overlaps (False, False)\n",
    "dist_min_max.drop(columns=\"gene_name\").eq(0,axis=\"columns\").value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1055 should've been captured by previous categories..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>overlap_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ABCA10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ABCA3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ACBD5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ACD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2174</th>\n",
       "      <td>LINC01435</td>\n",
       "      <td>731685</td>\n",
       "      <td>731685</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1642</th>\n",
       "      <td>FECHP1</td>\n",
       "      <td>769519</td>\n",
       "      <td>769519</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3190</th>\n",
       "      <td>RNU6-974P</td>\n",
       "      <td>799441</td>\n",
       "      <td>799441</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>SCP2D1</td>\n",
       "      <td>867610</td>\n",
       "      <td>867610</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>MFSD4BP1</td>\n",
       "      <td>1104855</td>\n",
       "      <td>1123544</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4275 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name      min      max  overlap_group\n",
       "3         AADAT        0        0              2\n",
       "9        ABCA10        0        0              2\n",
       "11        ABCA3        0        0              2\n",
       "25        ACBD5        0        0              2\n",
       "26          ACD        0        0              2\n",
       "...         ...      ...      ...            ...\n",
       "2174  LINC01435   731685   731685              1\n",
       "1642     FECHP1   769519   769519              1\n",
       "3190  RNU6-974P   799441   799441              1\n",
       "3310     SCP2D1   867610   867610              1\n",
       "2368   MFSD4BP1  1104855  1123544              1\n",
       "\n",
       "[4275 rows x 4 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign genes to categories based on overlaps\n",
    "dist_min_max.loc[:, \"overlap_group\"] = dist_min_max.eq(0,axis=\"columns\").sum(axis=\"columns\")\n",
    "dist_min_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3.220000e+03\n",
       "mean     2.843969e+04\n",
       "std      7.187899e+04\n",
       "min      1.000000e+00\n",
       "0%       1.000000e+00\n",
       "20%      4.440000e+02\n",
       "40%      3.001800e+03\n",
       "50%      5.400500e+03\n",
       "60%      8.860200e+03\n",
       "80%      3.098040e+04\n",
       "100%     1.123544e+06\n",
       "max      1.123544e+06\n",
       "Name: max, dtype: float64"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes with 1 missing PAS - how far away are the non-distal PAS?\n",
    "dist_min_max_1 = dist_min_max[dist_min_max[\"overlap_group\"].eq(1)]\n",
    "dist_min_max_1[\"max\"].describe(percentiles=[i * 0.1 for i in range(0,12,2)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0': 0,\n",
       " '25': 420,\n",
       " '50': 462,\n",
       " '100': 506,\n",
       " '200': 555,\n",
       " '500': 667,\n",
       " '1000': 817,\n",
       " '2500': 1183,\n",
       " '5000': 1555}"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# try different max distance thresholds (how many genes rescue?)\n",
    "{str(cutoff): len(dist_min_max_1[dist_min_max_1[\"max\"] <= cutoff]) for cutoff in [0,25,50,100,200,500,1000, 2500, 5000]}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suspect most of the provided ones here are 3'UTR extensions (1555 / 1713). Remaining possibly unannotated ALEs\n",
    "\n",
    "How are Yi's missing target genes distributed among the overlap categories\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overlap_group\n",
      "1    13\n",
      "2     9\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>overlap_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>CHRNB4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1809</th>\n",
       "      <td>GSTO2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2645</th>\n",
       "      <td>NSMAF</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2718</th>\n",
       "      <td>PAK4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3386</th>\n",
       "      <td>SLC24A3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3435</th>\n",
       "      <td>SLIT3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3569</th>\n",
       "      <td>STMN2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>TAFA5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2541</th>\n",
       "      <td>NAV1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3527</th>\n",
       "      <td>SREK1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3907</th>\n",
       "      <td>UBE2R2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>CLSTN3</td>\n",
       "      <td>0</td>\n",
       "      <td>690</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>ELP6</td>\n",
       "      <td>0</td>\n",
       "      <td>989</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1900</th>\n",
       "      <td>HNRNPC</td>\n",
       "      <td>0</td>\n",
       "      <td>1109</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>SMG7</td>\n",
       "      <td>0</td>\n",
       "      <td>1902</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>BRD9</td>\n",
       "      <td>0</td>\n",
       "      <td>2406</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>TTBK1</td>\n",
       "      <td>0</td>\n",
       "      <td>4036</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>CNPY3</td>\n",
       "      <td>0</td>\n",
       "      <td>4134</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2727</th>\n",
       "      <td>PARD6G</td>\n",
       "      <td>0</td>\n",
       "      <td>7662</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>SEC14L1</td>\n",
       "      <td>0</td>\n",
       "      <td>10276</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2264</th>\n",
       "      <td>LSM14A</td>\n",
       "      <td>0</td>\n",
       "      <td>22898</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  min    max  overlap_group\n",
       "549     CHRNB4    0      0              2\n",
       "853       EGFR    0      0              2\n",
       "1809     GSTO2    0      0              2\n",
       "2645     NSMAF    0      0              2\n",
       "2718      PAK4    0      0              2\n",
       "3386   SLC24A3    0      0              2\n",
       "3435     SLIT3    0      0              2\n",
       "3569     STMN2    0      0              2\n",
       "3627     TAFA5    0      0              2\n",
       "2541      NAV1    0      1              1\n",
       "3527     SREK1    0      1              1\n",
       "3907    UBE2R2    0      3              1\n",
       "576     CLSTN3    0    690              1\n",
       "882       ELP6    0    989              1\n",
       "1900    HNRNPC    0   1109              1\n",
       "3446      SMG7    0   1902              1\n",
       "309       BRD9    0   2406              1\n",
       "3850     TTBK1    0   4036              1\n",
       "588      CNPY3    0   4134              1\n",
       "2727    PARD6G    0   7662              1\n",
       "3320   SEC14L1    0  10276              1\n",
       "2264    LSM14A    0  22898              1"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(dist_min_max[dist_min_max.gene_name.isin(missing_genes)][\"overlap_group\"].value_counts())\n",
    "dist_min_max[dist_min_max.gene_name.isin(missing_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>201823342</td>\n",
       "      <td>201823343</td>\n",
       "      <td>NAV1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:201823342:201823343:+:NAV1</td>\n",
       "      <td>exon</td>\n",
       "      <td>201819836</td>\n",
       "      <td>201826969</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134369.16</td>\n",
       "      <td>NAV1</td>\n",
       "      <td>ENST00000367295.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:201819836:201826969:+:NAV1</td>\n",
       "      <td>4359</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>201826969</td>\n",
       "      <td>201826970</td>\n",
       "      <td>NAV1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:201826969:201826970:+:NAV1</td>\n",
       "      <td>exon</td>\n",
       "      <td>201819836</td>\n",
       "      <td>201826969</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134369.16</td>\n",
       "      <td>NAV1</td>\n",
       "      <td>ENST00000685211.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:201819836:201826969:+:NAV1</td>\n",
       "      <td>4359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>35643372</td>\n",
       "      <td>35643373</td>\n",
       "      <td>ARPP21</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:35643372:35643373:+:ARPP21</td>\n",
       "      <td>exon</td>\n",
       "      <td>34872640</td>\n",
       "      <td>34873854</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235534.1</td>\n",
       "      <td>FECHP1</td>\n",
       "      <td>ENST00000431896.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr3:34872640:34873854:+:FECHP1</td>\n",
       "      <td>19881</td>\n",
       "      <td>769519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>47495640</td>\n",
       "      <td>47495641</td>\n",
       "      <td>ELP6</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:47495640:47495641:-:ELP6</td>\n",
       "      <td>exon</td>\n",
       "      <td>47495639</td>\n",
       "      <td>47496197</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000163832.16</td>\n",
       "      <td>ELP6</td>\n",
       "      <td>ENST00000296149.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr3:47495639:47496197:-:ELP6</td>\n",
       "      <td>23737</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>47500537</td>\n",
       "      <td>47500538</td>\n",
       "      <td>ELP6</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:47500537:47500538:-:ELP6</td>\n",
       "      <td>exon</td>\n",
       "      <td>47501526</td>\n",
       "      <td>47501851</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000163832.16</td>\n",
       "      <td>ELP6</td>\n",
       "      <td>ENST00000461208.5</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>chr3:47501526:47501851:-:ELP6</td>\n",
       "      <td>23739</td>\n",
       "      <td>989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79666156</td>\n",
       "      <td>79666157</td>\n",
       "      <td>STMN2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr8:79666156:79666157:+:STMN2</td>\n",
       "      <td>exon</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>STMN2</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr8:79664814:79666158:+:STMN2</td>\n",
       "      <td>50603</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name       | Score      | +13   |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 201823342 | 201823343 | NAV1       | .          | ...   |\n",
       "| chr1         | 201826969 | 201826970 | NAV1       | .          | ...   |\n",
       "| chr3         |  35643372 |  35643373 | ARPP21     | .          | ...   |\n",
       "| chr3         |  47495640 |  47495641 | ELP6       | .          | ...   |\n",
       "| chr3         |  47500537 |  47500538 | ELP6       | .          | ...   |\n",
       "| chr8         |  79666156 |  79666157 | STMN2      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 6 rows and 18 columns from 3 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Strand, pas_id, Feature, Start_b, End_b, Strand_b, gene_id, ... (+ 6 more.)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_pas_le_nr.subset(lambda df: df.gene_name.isin([\"STMN2\", \"ELP6\", \"FECHP1\", \"NAV1\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692490:19692491:-:nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17902 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 17,902 rows and 7 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Strand, pas_id"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79617048</td>\n",
       "      <td>79617049</td>\n",
       "      <td>STMN2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr8:79617048:79617049:+:STMN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79666156</td>\n",
       "      <td>79666157</td>\n",
       "      <td>STMN2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr8:79666156:79666157:+:STMN2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name       | Score      | +2    |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr8         |  79617048 |  79617049 | STMN2      | .          | ...   |\n",
       "| chr8         |  79666156 |  79666157 | STMN2      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Strand, pas_id"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zeng_bed.subset(lambda df: df.Name == \"STMN2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output BED \n",
    "le_3utr_bed = pas_le_3utr_upd[[\"Score\", \"Name\"]].sort()\n",
    "le_ale_bed = pas_le_ale_bf_upd[[\"Score\", \"Name\"]].sort()\n",
    "le_3utr_bed.to_bed(\"processed/zeng_2024_3utr_pas_last_exons.bed\")\n",
    "le_ale_bed.to_bed(\"processed/zeng_2024_ale_pas_last_exons.bed\")\n",
    "pr.concat([le_3utr_bed, le_ale_bed]).sort().to_bed(\"processed/zeng_2024_combined_pas_last_exons.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
