{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from helpers import get_terminal_regions, _df_update_3p\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17902 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 17,902 rows and 6 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Strand"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BED file of proximal + distal PAS contained within genes (used to plot fig2c in Zeng et al. biorxiv)\n",
    "zeng_bed = pr.read_bed(\"data/zeng_2024/zeng_pas_fig2c.bed\")\n",
    "zeng_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ABCC5',\n",
       " 'AGPAT4',\n",
       " 'ARHGAP32',\n",
       " 'ARMC10',\n",
       " 'ARNT',\n",
       " 'AZIN1',\n",
       " 'BRD9',\n",
       " 'CADM1',\n",
       " 'CCNL2',\n",
       " 'CHRNB4',\n",
       " 'CLSTN3',\n",
       " 'CNPY3',\n",
       " 'CORO1C',\n",
       " 'CSNK2A1',\n",
       " 'DIDO1',\n",
       " 'DPYSL5',\n",
       " 'DYNC1LI1',\n",
       " 'DYRK2',\n",
       " 'EGFR',\n",
       " 'EIF4E2',\n",
       " 'ELAVL4',\n",
       " 'ELP1',\n",
       " 'ELP3',\n",
       " 'ELP6',\n",
       " 'EMC10',\n",
       " 'F11R',\n",
       " 'FOXK2',\n",
       " 'G3BP1',\n",
       " 'GGA2',\n",
       " 'GOLGA7B',\n",
       " 'GPR173',\n",
       " 'GREB1',\n",
       " 'GSTO2',\n",
       " 'H3-3B',\n",
       " 'HIF1AN',\n",
       " 'HNRNPC',\n",
       " 'JPT1',\n",
       " 'KLHL42',\n",
       " 'KPNA4',\n",
       " 'LRRC3',\n",
       " 'LSM14A',\n",
       " 'MARK3',\n",
       " 'MDGA1',\n",
       " 'NAV1',\n",
       " 'NDRG4',\n",
       " 'NDUFA9',\n",
       " 'NEFL',\n",
       " 'NFE2L1',\n",
       " 'NSMAF',\n",
       " 'NUCKS1',\n",
       " 'NUFIP2',\n",
       " 'OLA1',\n",
       " 'PAK4',\n",
       " 'PARD6G',\n",
       " 'PGRMC2',\n",
       " 'PPM1A',\n",
       " 'PTPN9',\n",
       " 'RAB11A',\n",
       " 'RFNG',\n",
       " 'RPN1',\n",
       " 'SEC14L1',\n",
       " 'SERF2',\n",
       " 'SFPQ',\n",
       " 'SIX3',\n",
       " 'SLC19A1',\n",
       " 'SLC24A3',\n",
       " 'SLC6A8',\n",
       " 'SLIT3',\n",
       " 'SMARCA4',\n",
       " 'SMC1A',\n",
       " 'SMG7',\n",
       " 'SREK1',\n",
       " 'SSR1',\n",
       " 'SSU72',\n",
       " 'STMN2',\n",
       " 'SYP',\n",
       " 'TAFA5',\n",
       " 'TARDBP',\n",
       " 'TEF',\n",
       " 'TFDP1',\n",
       " 'TFDP2',\n",
       " 'TLX1',\n",
       " 'TMED10',\n",
       " 'TMEM106B',\n",
       " 'TTBK1',\n",
       " 'UBE2H',\n",
       " 'UBE2R2',\n",
       " 'ZMAT2',\n",
       " 'ZNF462'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prioritised Zeng et al. target genes\n",
    "with open(\"data/zeng_2024/zeng_target_genes.txt\", \"r\") as infile:\n",
    "    zeng_target_genes = [line.rstrip(\"\\n\").replace(\"'\",\"\") for line in infile]\n",
    "\n",
    "zeng_target_genes_df = pd.DataFrame({\"gene_name\": zeng_target_genes})\n",
    "zeng_target_genes = set(zeng_target_genes)\n",
    "zeng_target_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>3</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573257</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573258</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573259</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573260</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>4</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573261</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573262 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 11868     | 12227     | +            | ...   |\n",
       "| chr1         | exon       | 12612     | 12721     | +            | ...   |\n",
       "| chr1         | exon       | 13220     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 12009     | 12057     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57213879  | 57213964  | -            | ...   |\n",
       "| chrY         | exon       | 57213525  | 57213602  | -            | ...   |\n",
       "| chrY         | exon       | 57213203  | 57213357  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,573,262 rows and 10 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_id, gene_name, transcript_id, exon_number, transcript_type"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in Gencode GTF, extract exons\n",
    "exons = pr.read_gtf(\"data/gencode.v40.annotation.gtf\").subset(lambda df: df.Feature == \"exon\")\n",
    "exons = exons[[\"Feature\", \"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"transcript_type\"]]\n",
    "# define genic regions\n",
    "genes = exons.boundaries(group_by=\"gene_name\")\n",
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing step 1 - Consolidate gene name assignment & generate unique PAS ID\n",
    "\n",
    "Some Name fields weren't filled in BED, need to use gencode annotation to update\n",
    "Also need to make sure that PAS fall within annotated boundaries, otherwise will never get counting regions for SLAM-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>629061</td>\n",
       "      <td>629433</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND1P23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>631073</td>\n",
       "      <td>632616</td>\n",
       "      <td>+</td>\n",
       "      <td>MTCO1P12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>632756</td>\n",
       "      <td>633438</td>\n",
       "      <td>+</td>\n",
       "      <td>MTCO2P12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>633534</td>\n",
       "      <td>633741</td>\n",
       "      <td>+</td>\n",
       "      <td>MTATP8P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56639</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56640</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56641</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56642</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19744939</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56643</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19744939</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56644 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +5    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 56,644 rows and 10 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Strand, Start_b, End_b, Strand_b, gene_name"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reassign gene name for pas\n",
    "zeng_bed_gene = zeng_bed.join(genes, strandedness=\"same\", how = \"left\", slack=10000)\n",
    "zeng_bed_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_overlap\n",
       "True     56641\n",
       "False        3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many PAS do not overlap with any gencode annotated genes (/ within 10kb)\n",
    "zeng_bed_gene.assign(\"gene_overlap\", lambda df: df.Start_b.ne(-1)).gene_overlap.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>134790979</td>\n",
       "      <td>134790980</td>\n",
       "      <td>LINC02714</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>21332430</td>\n",
       "      <td>21332431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>26420990</td>\n",
       "      <td>26420991</td>\n",
       "      <td>CDK8</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name       | Score      | +5    |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr11        | 134790979 | 134790980 | LINC02714  | .          | ...   |\n",
       "| chr13        |  21332430 |  21332431 | nan        | .          | ...   |\n",
       "| chr13        |  26420990 |  26420991 | CDK8       | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Strand, Start_b, End_b, Strand_b, gene_name"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at nnon-overlaps\n",
    "zeng_bed_gene.subset(lambda df: df.Start_b == -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>134790979</td>\n",
       "      <td>134790980</td>\n",
       "      <td>LINC02714</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>134763810</td>\n",
       "      <td>134763811</td>\n",
       "      <td>LINC02714</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>134735595</td>\n",
       "      <td>134763810</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC02714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>21332430</td>\n",
       "      <td>21332431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>26420990</td>\n",
       "      <td>26420991</td>\n",
       "      <td>CDK8</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>26405233</td>\n",
       "      <td>26405234</td>\n",
       "      <td>CDK8</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>26254103</td>\n",
       "      <td>26405238</td>\n",
       "      <td>+</td>\n",
       "      <td>CDK8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   |     Start |       End | Name       | Score      | +5    |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr11        | 134790979 | 134790980 | LINC02714  | .          | ...   |\n",
       "| chr11        | 134763810 | 134763811 | LINC02714  | .          | ...   |\n",
       "| chr13        |  21332430 |  21332431 | nan        | .          | ...   |\n",
       "| chr13        |  26420990 |  26420991 | CDK8       | .          | ...   |\n",
       "| chr13        |  26405233 |  26405234 | CDK8       | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 5 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Strand, Start_b, End_b, Strand_b, gene_name"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pull out alt PAS for non overlaps, track excluded pas\n",
    "no_gene_overlap_names = set(zeng_bed_gene.subset(lambda df: df.Start_b == -1).as_df().dropna(subset=\"Name\").Name)\n",
    "no_gene_overlap_bed = pr.concat([zeng_bed_gene.subset(lambda df: df.Start_b == -1), zeng_bed_gene.subset(lambda df: df.Name.isin(no_gene_overlap_names))]).drop_duplicate_positions()\n",
    "# exclude from next steps\n",
    "zeng_bed_gene = zeng_bed_gene.subset(lambda df: (df.Start_b != -1) & ~(df.gene_name.isin(no_gene_overlap_names)))\n",
    "\n",
    "no_gene_overlap_bed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_name_common</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>629061</td>\n",
       "      <td>629433</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>631073</td>\n",
       "      <td>632616</td>\n",
       "      <td>+</td>\n",
       "      <td>MTCO1P12</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>632756</td>\n",
       "      <td>633438</td>\n",
       "      <td>+</td>\n",
       "      <td>MTCO2P12</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>633534</td>\n",
       "      <td>633741</td>\n",
       "      <td>+</td>\n",
       "      <td>MTATP8P1</td>\n",
       "      <td>MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56634</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56635</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56636</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56637</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19744939</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56638</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19744939</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56639 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +6    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 56,639 rows and 11 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Strand, Start_b, End_b, Strand_b, gene_name, gene_name_common"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# consolidate gene name - trust original where not NA, otherwise use from gencode\n",
    "zeng_bed_gene = zeng_bed_gene.assign(\"gene_name_common\", lambda df: pd.Series(np.where(df[\"Name\"].isna(), df[\"gene_name\"], df[\"Name\"])))\n",
    "zeng_bed_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp\n",
    "zeng_bed_gene = zeng_bed_gene.as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gn_match\n",
       "True     16804\n",
       "False     1093\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check that gene names from annotation + names for PAS in BED file match\n",
    "zeng_bed_gene.loc[:, \"gn_match\"] = zeng_bed_gene.Name == zeng_bed_gene.gene_name\n",
    "\n",
    "# may be multiple genic overlaps for each pas, so prefer match where applicable\n",
    "# Sort by name so matches appear as first row\n",
    "(zeng_bed_gene.sort_values(by=[\"Name\", \"gn_match\"], ascending=False)\n",
    " # subset to one overlap per PAS, prioritising matches\n",
    " .drop_duplicates(subset=[\"Chromosome\", \"Start\", \"End\", \"Strand\"])\n",
    " .gn_match.value_counts()\n",
    " )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Name\n",
       "True     927\n",
       "False    166\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1093 PAS where no match with assigned gene name\n",
    "# how many are NaNs from the BED?\n",
    "pas_no_match = (zeng_bed_gene.sort_values(by=[\"Name\", \"gn_match\"], ascending=False)\n",
    " # subset to one overlap per PAS, prioritising matches\n",
    " .drop_duplicates(subset=[\"Chromosome\", \"Start\", \"End\", \"Strand\"])\n",
    " .loc[lambda x: ~x[\"gn_match\"], :]\n",
    " )\n",
    "\n",
    "pas_no_match.Name.isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ZFTRAF1', 'IFT25', 'ZNF892', 'TCF12-DT', 'FRMD6-AS2', 'CDHR18P', 'BLTP1', 'RPSA2', 'ENTREP3', 'H2BC26', 'LINC01013', 'ENTREP2', 'ZNG1B', 'ATOSA', 'CASTOR3P', 'LINC03051', 'CSTPP1', 'ERCC6L2-AS1', 'KCNA3', 'FAM200C', 'ADAM7-AS1', 'POLR1HASP', 'ABTB3', 'SKIC8', 'ZNG1A', 'C2orf74-DT', 'SLC38A4-AS1', 'ZNG1E', 'GATD1-DT', 'TTC23-AS1', 'LINC03077', 'UQCC6', 'LINC03072', 'LINC02976', 'FLRT2-AS1', 'RAB35-AS1', 'PRPF18', 'UBE2CP5', 'C22orf46P', 'CTDP1-DT', 'HYCC1', 'RNF32-DT', 'ZNG1F', 'LINC02977', 'ZNF496-DT', 'LINC00951', 'TOX-DT', 'MFSD14CP', 'LIPT2-AS1', 'LINC03009', 'EEIG1', 'LINGO2', 'LINC00938', 'BLTP3B', 'TMA7B', 'SLC38A6', 'CFAP90', 'SKIC3', 'MIR3171HG', 'UQCC5', 'HAPSTR1', 'LCDR', 'HYCC2', 'BMAL2', 'LINC02981', 'BLTP3A', 'H2AC25', 'LINC02987', 'IRX2-DT', 'MRFAP1P1', 'PCGF3-AS1', 'MATCAP1', 'MSANTD7', 'LYSET', 'ZNF778-DT', 'ZNF225-AS1', 'PHB1', 'PIERCE2', 'MATCAP2', 'LINC03006', 'PTGR3', 'LASP1NB', 'LINC03026', 'ALG1L1P', 'RIMOC1', 'SCAND3'}\n",
      "total gene names - 86\n",
      "gene name found in annotation - 7\n",
      "{'PRPF18', 'SLC38A6', 'LINC00951', 'FRMD6-AS2', 'LINC01013', 'KCNA3', 'LINGO2'}\n"
     ]
    }
   ],
   "source": [
    "# 927 just have no assigned gene from BED file (trust gencode)\n",
    "# what's happening with the 166?\n",
    "pas_no_match_nna = pas_no_match[~pas_no_match.Name.isna()]\n",
    "names_no_match_nna = set(pas_no_match_nna.Name)\n",
    "print(names_no_match_nna)\n",
    "\n",
    "# are their gene names found in annotation?\n",
    "print(f\"total gene names - {len(names_no_match_nna)}\")\n",
    "print(f\"gene name found in annotation - {len(names_no_match_nna.intersection(set(genes.gene_name)))}\")\n",
    "print(names_no_match_nna.intersection(set(genes.gene_name)))\n",
    "\n",
    "# set(genes.gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Are any of Yi's prioritised targets found in this list?\n",
    "names_no_match_nna.intersection(zeng_target_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No intersection, so will just drop these 166 for now... (but report)\n",
    "\n",
    "Priority for updating names:\n",
    "zeng_bed_gene - keep the gn_match and NAs + match, then generate pas_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247334576</td>\n",
       "      <td>247334577</td>\n",
       "      <td>ZNF496-DT</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247377231</td>\n",
       "      <td>247377232</td>\n",
       "      <td>ZNF496-DT</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>228460764</td>\n",
       "      <td>228460765</td>\n",
       "      <td>H2BC26</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>228463065</td>\n",
       "      <td>228463066</td>\n",
       "      <td>H2BC26</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110653567</td>\n",
       "      <td>110653568</td>\n",
       "      <td>KCNA3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>chr22</td>\n",
       "      <td>41698135</td>\n",
       "      <td>41698136</td>\n",
       "      <td>C22orf46P</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153599891</td>\n",
       "      <td>153599892</td>\n",
       "      <td>MRFAP1P1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153609261</td>\n",
       "      <td>153609262</td>\n",
       "      <td>MRFAP1P1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98865549</td>\n",
       "      <td>98865550</td>\n",
       "      <td>LINC03077</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>chrX</td>\n",
       "      <td>98867708</td>\n",
       "      <td>98867709</td>\n",
       "      <td>LINC03077</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +1    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 247334576 | 247334577 | ZNF496-DT  | .          | ...   |\n",
       "| chr1         | 247377231 | 247377232 | ZNF496-DT  | .          | ...   |\n",
       "| chr1         | 228460764 | 228460765 | H2BC26     | .          | ...   |\n",
       "| chr1         | 228463065 | 228463066 | H2BC26     | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrX         | 153599891 | 153599892 | MRFAP1P1   | .          | ...   |\n",
       "| chrX         | 153609261 | 153609262 | MRFAP1P1   | .          | ...   |\n",
       "| chrX         | 98865549  | 98865550  | LINC03077  | .          | ...   |\n",
       "| chrX         | 98867708  | 98867709  | LINC03077  | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 166 rows and 6 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: Strand"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# track excluded 166 non na and no match PAS\n",
    "no_genename_match_bed = pr.PyRanges(pas_no_match[~pas_no_match.Name.isna()].drop(columns=[\"Start_b\", \"End_b\", \"Strand_b\", \"gene_name\", \"gene_name_common\", \"gn_match\"]))\n",
    "no_genename_match_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17726</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17727</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17728</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17729</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17730</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17731 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 17,731 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop the 166, return to pyranges object\n",
    "zeng_bed_gene = zeng_bed_gene[zeng_bed_gene.gn_match | zeng_bed_gene.Name.isna()]\n",
    "zeng_bed_gene.loc[:, \"pas_id\"] = zeng_bed_gene.Chromosome.str.cat(zeng_bed_gene[[\"Start\", \"End\", \"Strand\", \"gene_name_common\"]].astype(str), sep=\":\")\n",
    "zeng_bed_upd = pr.PyRanges(zeng_bed_gene.drop(columns=[\"Start_b\", \"End_b\", \"Strand_b\", \"gene_name\", \"gn_match\"]))\n",
    "zeng_bed_upd = zeng_bed_upd.drop_duplicate_positions()\n",
    "zeng_bed_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16101"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# track total input PAS\n",
    "pas_ids = set(zeng_bed_upd.pas_id)\n",
    "len(pas_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing step 2 - Check for (non-terminal) exon overlapping PAS\n",
    "\n",
    "Slam-seq counting performed using featurecounts. To do isoform-specific properly, it is essential that regions for each event are non-ambiguous. If PAS overlaps an internal exon, it's impossible to differentiate between txs using region as an internal exon vs a last exon. Such PAS will need to be removed from further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1:13220:14409:+:DDX11L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13452</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr1:13452:13670:+:DDX11L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>chr1:30365:30503:+:MIR1302-2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:30975:31097:+:MIR1302-2HG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:30975:31109:+:MIR1302-2HG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246619</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000237801.6_PAR_Y</td>\n",
       "      <td>AMD1P2</td>\n",
       "      <td>ENST00000412936.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57015104:57016096:-:AMD1P2</td>\n",
       "      <td>119470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246620</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228410.6_PAR_Y</td>\n",
       "      <td>ELOCP24</td>\n",
       "      <td>ENST00000456370.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57165511:57165845:-:ELOCP24</td>\n",
       "      <td>119471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246621</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223484.7_PAR_Y</td>\n",
       "      <td>TRPC6P</td>\n",
       "      <td>ENST00000421233.6_PAR_Y</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chrY:57171889:57172769:-:TRPC6P</td>\n",
       "      <td>119472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246622</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57202020</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>ENST00000399966.9_PAR_Y</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chrY:57201142:57202020:-:WASIR1</td>\n",
       "      <td>119473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246623</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>chrY:57212183:57213125:-:DDX11L16</td>\n",
       "      <td>119474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246624 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 13220     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 13452     | 13670     | +            | ...   |\n",
       "| chr1         | exon       | 30365     | 30503     | +            | ...   |\n",
       "| chr1         | exon       | 30975     | 31097     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57165511  | 57165845  | -            | ...   |\n",
       "| chrY         | exon       | 57171889  | 57172769  | -            | ...   |\n",
       "| chrY         | exon       | 57201142  | 57202020  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 246,624 rows and 11 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: gene_id, gene_name, transcript_id, transcript_type, exon_id, ... (+ 1 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get objects containing last exons and non-last exons\n",
    "\n",
    "# Extract last exons from annotation\n",
    "exons = exons.assign(\"exon_number\", lambda df: df.exon_number.astype(float).astype(int))\n",
    "last_exons = get_terminal_regions(exons, number_type=\"stranded\")\n",
    "# assign an 'exon id' - basically just the coordinates (so can handle identical LEs x transcripts of same gene)\n",
    "last_exons = last_exons.assign(\"exon_id\", lambda df: df.Chromosome.str.cat(df[[\"Start\", \"End\", \"Strand\", \"gene_name\"]].astype(str), sep=\":\"))\n",
    "# don't need it\n",
    "last_exons = last_exons.drop(\"exon_number\")\n",
    "# annotate overlapping last exons with a common ID\n",
    "last_exons = last_exons.cluster(strand=True)\n",
    "\n",
    "last_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12697</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935310</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203181</td>\n",
       "      <td>57203357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>ENST00000399966.9_PAR_Y</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935311</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935312</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935313</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935314</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>4</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>935315 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +5    |\n",
       "| (category)   | (object)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 11868     | 12227     | +            | ...   |\n",
       "| chr1         | exon       | 12612     | 12721     | +            | ...   |\n",
       "| chr1         | exon       | 12009     | 12057     | +            | ...   |\n",
       "| chr1         | exon       | 12178     | 12227     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57214349  | 57214397  | -            | ...   |\n",
       "| chrY         | exon       | 57213879  | 57213964  | -            | ...   |\n",
       "| chrY         | exon       | 57213525  | 57213602  | -            | ...   |\n",
       "| chrY         | exon       | 57213203  | 57213357  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 935,315 rows and 10 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: gene_id, gene_name, transcript_id, exon_number, transcript_type"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# anti-overlap to get non-last exons\n",
    "nonlast_exons = exons.overlap(last_exons, strandedness=\"same\", invert=True)\n",
    "nonlast_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15727029</td>\n",
       "      <td>15727030</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>chr1:15727029:15727030:+:PLEKHM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>15727013</td>\n",
       "      <td>15727832</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116786.13</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>ENST00000375799.8</td>\n",
       "      <td>9</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15727029</td>\n",
       "      <td>15727030</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>chr1:15727029:15727030:+:PLEKHM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>15727013</td>\n",
       "      <td>15727832</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116786.13</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>ENST00000375793.2</td>\n",
       "      <td>8</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15727029</td>\n",
       "      <td>15727030</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>chr1:15727029:15727030:+:PLEKHM2</td>\n",
       "      <td>exon</td>\n",
       "      <td>15727013</td>\n",
       "      <td>15727068</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116786.13</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>ENST00000642363.1</td>\n",
       "      <td>9</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19904972</td>\n",
       "      <td>19904973</td>\n",
       "      <td>OTUD3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>OTUD3</td>\n",
       "      <td>chr1:19904972:19904973:+:OTUD3</td>\n",
       "      <td>exon</td>\n",
       "      <td>19904890</td>\n",
       "      <td>19904987</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000169914.6</td>\n",
       "      <td>OTUD3</td>\n",
       "      <td>ENST00000375120.4</td>\n",
       "      <td>6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26268506</td>\n",
       "      <td>26268507</td>\n",
       "      <td>CEP85</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>CEP85</td>\n",
       "      <td>chr1:26268506:26268507:+:CEP85</td>\n",
       "      <td>exon</td>\n",
       "      <td>26268482</td>\n",
       "      <td>26268635</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000130695.16</td>\n",
       "      <td>CEP85</td>\n",
       "      <td>ENST00000640292.2</td>\n",
       "      <td>7</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19864</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19865</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19866</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19867</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19868</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19869 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +12   |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 15727029  | 15727030  | PLEKHM2    | .          | ...   |\n",
       "| chr1         | 15727029  | 15727030  | PLEKHM2    | .          | ...   |\n",
       "| chr1         | 15727029  | 15727030  | PLEKHM2    | .          | ...   |\n",
       "| chr1         | 19904972  | 19904973  | OTUD3      | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 19,869 rows and 17 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "12 hidden columns: Strand, gene_name_common, pas_id, Feature, Start_b, End_b, ... (+ 6 more.)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find PAS overlapping with non-terminal exons\n",
    "pas_nonlast_olap = zeng_bed_upd.join(nonlast_exons, strandedness=\"same\",how=\"left\")\n",
    "pas_nonlast_olap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "olap\n",
       "False    17242\n",
       "True       489\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count PAS overlapping with non-last exons\n",
    "(pas_nonlast_olap.as_df()\n",
    " .assign(olap=lambda df: df[\"Start_b\"].ne(-1))\n",
    " .drop_duplicates(subset=[\"pas_id\"])\n",
    " .olap.value_counts()\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15727029</td>\n",
       "      <td>15727030</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>chr1:15727029:15727030:+:PLEKHM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19904972</td>\n",
       "      <td>19904973</td>\n",
       "      <td>OTUD3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>OTUD3</td>\n",
       "      <td>chr1:19904972:19904973:+:OTUD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26268506</td>\n",
       "      <td>26268507</td>\n",
       "      <td>CEP85</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>CEP85</td>\n",
       "      <td>chr1:26268506:26268507:+:CEP85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32031568</td>\n",
       "      <td>32031569</td>\n",
       "      <td>KHDRBS1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>KHDRBS1</td>\n",
       "      <td>chr1:32031568:32031569:+:KHDRBS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40059359</td>\n",
       "      <td>40059360</td>\n",
       "      <td>CAP1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>CAP1</td>\n",
       "      <td>chr1:40059359:40059360:+:CAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>chrX</td>\n",
       "      <td>119841736</td>\n",
       "      <td>119841737</td>\n",
       "      <td>UPF3B</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>UPF3B</td>\n",
       "      <td>chrX:119841736:119841737:-:UPF3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131768620</td>\n",
       "      <td>131768621</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>chrX:131768620:131768621:-:FIRRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chrX</td>\n",
       "      <td>136225931</td>\n",
       "      <td>136225932</td>\n",
       "      <td>MAP7D3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP7D3</td>\n",
       "      <td>chrX:136225931:136225932:-:MAP7D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139588359</td>\n",
       "      <td>139588360</td>\n",
       "      <td>MCF2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>MCF2</td>\n",
       "      <td>chrX:139588359:139588360:-:MCF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139814911</td>\n",
       "      <td>139814912</td>\n",
       "      <td>ATP11C</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ATP11C</td>\n",
       "      <td>chrX:139814911:139814912:-:ATP11C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 15727029  | 15727030  | PLEKHM2    | .          | ...   |\n",
       "| chr1         | 19904972  | 19904973  | OTUD3      | .          | ...   |\n",
       "| chr1         | 26268506  | 26268507  | CEP85      | .          | ...   |\n",
       "| chr1         | 32031568  | 32031569  | KHDRBS1    | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrX         | 131768620 | 131768621 | FIRRE      | .          | ...   |\n",
       "| chrX         | 136225931 | 136225932 | MAP7D3     | .          | ...   |\n",
       "| chrX         | 139588359 | 139588360 | MCF2       | .          | ...   |\n",
       "| chrX         | 139814911 | 139814912 | ATP11C     | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 489 rows and 8 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# isolate exon-overlapping, track gene names to remvoe their counterpart PAS\n",
    "nonlast_olap_bed = (pas_nonlast_olap.subset(lambda df: df.Start_b.ne(-1))\n",
    " .drop_duplicate_positions()\n",
    " .drop(like=\"_b$\")\n",
    " .drop([\"Feature\", \"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"transcript_type\"])\n",
    " )\n",
    "\n",
    "# gene names\n",
    "nonlast_olap_names = set(nonlast_olap_bed.gene_name_common)\n",
    "\n",
    "nonlast_olap_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16097</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19691944:19691945:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16098</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16099</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16100</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16101 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 629997    | 629998    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 630367    | 630368    | MTND2P28   | .          | ...   |\n",
       "| chr1         | 854387    | 854388    | LINC01128  | .          | ...   |\n",
       "| chr1         | 859444    | 859445    | LINC01128  | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 19691944  | 19691945  | nan        | .          | ...   |\n",
       "| chrY         | 19692490  | 19692491  | nan        | .          | ...   |\n",
       "| chrY         | 19703866  | 19703867  | KDM5D      | .          | ...   |\n",
       "| chrY         | 19705419  | 19705420  | KDM5D      | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 16,101 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove the exon overlapping PAS (+ their counterparts)\n",
    "zeng_bed_upd = (pas_nonlast_olap.subset(lambda df: df.Start_b.eq(-1))\n",
    "                .subset(lambda df: ~df.gene_name_common.isin(nonlast_olap_names))\n",
    "                .drop(like=\"_b$\")\n",
    "                .drop([\"Feature\", \"gene_id\", \"gene_name\", \"transcript_id\", \"exon_number\", \"transcript_type\"])\n",
    "                )\n",
    "\n",
    "zeng_bed_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BRD9', 'CLSTN3', 'HNRNPC', 'LSM14A', 'SEC14L1', 'SLC24A3'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check which of provided targets are lost to exon overlap\n",
    "zeng_targets_nonlast = zeng_target_genes.intersection(nonlast_olap_names)\n",
    "zeng_targets_nonlast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define tandem 3'UTR regions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854398</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:854398:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854387</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:854387:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854385</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:854385:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46186</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46187</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46188</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705416</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705416:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46189</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46190</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705424</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705424:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46191 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 853390    | 854398    | +            | ...   |\n",
       "| chr1         | 853390    | 854387    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705416  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705424  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 46,191 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign PAS to last exons based on overlap\n",
    "# allow small slack of 100nt, so PAS close to annotated bou3'ends are considered overlapping\n",
    "pas_le = last_exons[[\"exon_id\", \"Cluster\"]].join(zeng_bed_upd, strandedness=\"same\", slack=100)\n",
    "pas_le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pas_id\n",
       "2    5000\n",
       "1    2909\n",
       "3       5\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of unique PAS overlapping last exons.\n",
    "# do this at gene level, so interpretation = at least one annot last exon has x overlapping PAS\n",
    "# (could have problems with shorter annotated last exons of same gene only overlapping with proximal PAS)\n",
    "# (2 = 3'UTR PAS, 1 = ALEs)\n",
    "num_olap_pas = pas_le.as_df().groupby(\"exon_id\")[\"pas_id\"].nunique().reset_index()\n",
    "num_olap_pas.loc[:, \"gene_name\"] = num_olap_pas.exon_id.str.split(\":\", expand=True, regex=False)[4]\n",
    "\n",
    "# number of putative 3'UTR APA / ALE PAS\n",
    "num_olap_pas_genemax = num_olap_pas.groupby(\"gene_name\")[\"pas_id\"].max()\n",
    "num_olap_pas_genemax.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>629998</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630368</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>854388</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859446</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>859444</td>\n",
       "      <td>859445</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066449</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1065829:1066449:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801</td>\n",
       "      <td>chr1:1066354:1066355:+:ENSG00000217801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24793</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24794</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24795</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24796</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24797</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24798 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | 853390    | 859446    | +            | ...   |\n",
       "| chr1         | 853390    | 859446    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703864  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 24,798 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get last exons with two overlapping PAS & subset for joined last exon + pas coordinates\n",
    "tandem_exons_ids = set(num_olap_pas.loc[num_olap_pas[\"pas_id\"] == 2, \"exon_id\"])\n",
    "\n",
    "\n",
    "pas_le_3utr = pas_le.subset(lambda df: df.exon_id.isin(tandem_exons_ids))\n",
    "pas_le_3utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of negative or zero-length updated intervals to be dropped - 1312\n",
      "Number of negative or zero-length updated intervals to be dropped - 1304\n",
      "Number of negative or zero-length updated intervals to be dropped - 1040\n",
      "Number of negative or zero-length updated intervals to be dropped - 822\n",
      "Number of negative or zero-length updated intervals to be dropped - 776\n",
      "Number of negative or zero-length updated intervals to be dropped - 790\n",
      "Number of negative or zero-length updated intervals to be dropped - 568\n",
      "Number of negative or zero-length updated intervals to be dropped - 542\n",
      "Number of negative or zero-length updated intervals to be dropped - 680\n",
      "Number of negative or zero-length updated intervals to be dropped - 668\n",
      "Number of negative or zero-length updated intervals to be dropped - 472\n",
      "Number of negative or zero-length updated intervals to be dropped - 720\n",
      "Number of negative or zero-length updated intervals to be dropped - 596\n",
      "Number of negative or zero-length updated intervals to be dropped - 534\n",
      "Number of negative or zero-length updated intervals to be dropped - 526\n",
      "Number of negative or zero-length updated intervals to be dropped - 406\n",
      "Number of negative or zero-length updated intervals to be dropped - 450\n",
      "Number of negative or zero-length updated intervals to be dropped - 536\n",
      "Number of negative or zero-length updated intervals to be dropped - 592\n",
      "Number of negative or zero-length updated intervals to be dropped - 518\n",
      "Number of negative or zero-length updated intervals to be dropped - 508\n",
      "Number of negative or zero-length updated intervals to be dropped - 688\n",
      "Number of negative or zero-length updated intervals to be dropped - 720\n",
      "Number of negative or zero-length updated intervals to be dropped - 470\n",
      "Number of negative or zero-length updated intervals to be dropped - 224\n",
      "Number of negative or zero-length updated intervals to be dropped - 254\n",
      "Number of negative or zero-length updated intervals to be dropped - 534\n",
      "Number of negative or zero-length updated intervals to be dropped - 486\n",
      "Number of negative or zero-length updated intervals to be dropped - 380\n",
      "Number of negative or zero-length updated intervals to be dropped - 376\n",
      "Number of negative or zero-length updated intervals to be dropped - 564\n",
      "Number of negative or zero-length updated intervals to be dropped - 374\n",
      "Number of negative or zero-length updated intervals to be dropped - 738\n",
      "Number of negative or zero-length updated intervals to be dropped - 664\n",
      "Number of negative or zero-length updated intervals to be dropped - 204\n",
      "Number of negative or zero-length updated intervals to be dropped - 276\n",
      "Number of negative or zero-length updated intervals to be dropped - 598\n",
      "Number of negative or zero-length updated intervals to be dropped - 476\n",
      "Number of negative or zero-length updated intervals to be dropped - 340\n",
      "Number of negative or zero-length updated intervals to be dropped - 376\n",
      "Number of negative or zero-length updated intervals to be dropped - 132\n",
      "Number of negative or zero-length updated intervals to be dropped - 114\n",
      "Number of negative or zero-length updated intervals to be dropped - 228\n",
      "Number of negative or zero-length updated intervals to be dropped - 214\n",
      "Number of negative or zero-length updated intervals to be dropped - 610\n",
      "Number of negative or zero-length updated intervals to be dropped - 372\n",
      "Number of negative or zero-length updated intervals to be dropped - 18\n",
      "Number of negative or zero-length updated intervals to be dropped - 8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>629998</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>629997</td>\n",
       "      <td>630683</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:629997:629998:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629639</td>\n",
       "      <td>630368</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629639:630683:+:MTND2P28</td>\n",
       "      <td>16</td>\n",
       "      <td>630367</td>\n",
       "      <td>630683</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:630367:630368:+:MTND2P28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>854388</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>854387</td>\n",
       "      <td>859446</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:854387:854388:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>853390</td>\n",
       "      <td>859445</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:853390:859446:+:LINC01128</td>\n",
       "      <td>35</td>\n",
       "      <td>859444</td>\n",
       "      <td>859446</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>chr1:859444:859445:+:LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066355</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1065829:1066449:+:ENSG00000217801</td>\n",
       "      <td>57</td>\n",
       "      <td>1066354</td>\n",
       "      <td>1066449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000217801</td>\n",
       "      <td>chr1:1066354:1066355:+:ENSG00000217801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24671</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692490</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197</td>\n",
       "      <td>119348</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19692491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197</td>\n",
       "      <td>chrY:19692490:19692491:-:ENSG00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24672</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24673</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703866</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19703867</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19703866:19703867:-:KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19703864:19706345:-:KDM5D</td>\n",
       "      <td>119349</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19705420</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>chrY:19705419:19705420:-:KDM5D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24676 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 629639    | 629998    | +            | ...   |\n",
       "| chr1         | 629639    | 630368    | +            | ...   |\n",
       "| chr1         | 853390    | 854388    | +            | ...   |\n",
       "| chr1         | 853390    | 859445    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 19703866  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "| chrY         | 19703866  | 19706345  | -            | ...   |\n",
       "| chrY         | 19705419  | 19706345  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 24,676 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update 3' exon coordinates to the PAS coordinate\n",
    "pas_le_3utr_upd = pas_le_3utr.apply(lambda df: _df_update_3p(df))\n",
    "pas_le_3utr_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# target genes fitting tandem criteria\n",
    "tandem_genes = set(pas_le_3utr_upd.gene_name_common)\n",
    "zeng_targets_tandem = zeng_target_genes.intersection(tandem_genes)\n",
    "len(zeng_targets_tandem)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define ALE regions\n",
    "\n",
    "In theory these are genes with only 1 pas overlapping each of two distinct ALEs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617898</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1619210</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1619210:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617323</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13590</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13591</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932006</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932006:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13592</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932316:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13593</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13594</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13595 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1013983   | 1014540   | +            | ...   |\n",
       "| chr1         | 1013983   | 1014540   | +            | ...   |\n",
       "| chr1         | 1616507   | 1617898   | +            | ...   |\n",
       "| chr1         | 1616507   | 1619210   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 18932006  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932316  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 13,595 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get putative ALEs/ALE-overlapping IDs\n",
    "# i.e. genes with max exon-PAS overlap count of 1, becasue these do not have any tandem APA\n",
    "put_ale_genes = set(num_olap_pas_genemax[num_olap_pas_genemax.eq(1)].index)\n",
    "put_ale_ids = set(num_olap_pas.loc[num_olap_pas.gene_name.isin(put_ale_genes), \"exon_id\"])\n",
    "\n",
    "pas_le_ale = pas_le.subset(lambda df: df.exon_id.isin(put_ale_ids))\n",
    "pas_le_ale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cluster  pas_id\n",
       "1        1         1704\n",
       "2        2         1116\n",
       "1        2           14\n",
       "3        2           12\n",
       "2        1           10\n",
       "3        3            1\n",
       "4        2            1\n",
       "         4            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# true ALEs should be non-overlapping.\n",
    "# Each gene should therefore have two unique PAS, and two unique Cluster values\n",
    "\n",
    "# count unqiue pas + non-overlapping ales\n",
    "ale_uniq_counts = pas_le_ale.as_df().groupby(\"gene_name_common\")[[\"Cluster\", \"pas_id\"]].nunique().reset_index()\n",
    "\n",
    "# quick check counts\n",
    "ale_uniq_counts[[\"Cluster\",\"pas_id\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617898</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1619210</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1619210:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617323</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617321</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630604</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630604:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630605</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630604</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630604:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630605</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7341</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7342</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932006</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932006:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7343</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932316</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932316:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7344</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7345</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932435</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18932435:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7346 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1616507   | 1617898   | +            | ...   |\n",
       "| chr1         | 1616507   | 1619210   | +            | ...   |\n",
       "| chr1         | 1617077   | 1617323   | +            | ...   |\n",
       "| chr1         | 1630291   | 1630604   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 18932006  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932316  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "| chrY         | 18932435  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 7,346 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2 + 2 represent 'bonafide' ALEs - extract + report\n",
    "# get last exons with two overlapping PAS & subset for joined last exon + pas coordinates\n",
    "# get IDs for 'bonafide' ALEs\n",
    "ale_bf_gene_names = ale_uniq_counts.loc[(ale_uniq_counts[\"Cluster\"] == 2) & (ale_uniq_counts[\"pas_id\"] == 2), \"gene_name_common\"]\n",
    "pas_le_ale_bf = pas_le_ale.subset(lambda df: df.gene_name_common.isin(ale_bf_gene_names))\n",
    "pas_le_ale_bf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of negative or zero-length updated intervals to be dropped - 411\n",
      "Number of negative or zero-length updated intervals to be dropped - 282\n",
      "Number of negative or zero-length updated intervals to be dropped - 396\n",
      "Number of negative or zero-length updated intervals to be dropped - 264\n",
      "Number of negative or zero-length updated intervals to be dropped - 212\n",
      "Number of negative or zero-length updated intervals to be dropped - 125\n",
      "Number of negative or zero-length updated intervals to be dropped - 187\n",
      "Number of negative or zero-length updated intervals to be dropped - 170\n",
      "Number of negative or zero-length updated intervals to be dropped - 128\n",
      "Number of negative or zero-length updated intervals to be dropped - 61\n",
      "Number of negative or zero-length updated intervals to be dropped - 171\n",
      "Number of negative or zero-length updated intervals to be dropped - 220\n",
      "Number of negative or zero-length updated intervals to be dropped - 210\n",
      "Number of negative or zero-length updated intervals to be dropped - 152\n",
      "Number of negative or zero-length updated intervals to be dropped - 215\n",
      "Number of negative or zero-length updated intervals to be dropped - 171\n",
      "Number of negative or zero-length updated intervals to be dropped - 193\n",
      "Number of negative or zero-length updated intervals to be dropped - 120\n",
      "Number of negative or zero-length updated intervals to be dropped - 140\n",
      "Number of negative or zero-length updated intervals to be dropped - 146\n",
      "Number of negative or zero-length updated intervals to be dropped - 192\n",
      "Number of negative or zero-length updated intervals to be dropped - 225\n",
      "Number of negative or zero-length updated intervals to be dropped - 136\n",
      "Number of negative or zero-length updated intervals to be dropped - 181\n",
      "Number of negative or zero-length updated intervals to be dropped - 72\n",
      "Number of negative or zero-length updated intervals to be dropped - 80\n",
      "Number of negative or zero-length updated intervals to be dropped - 103\n",
      "Number of negative or zero-length updated intervals to be dropped - 127\n",
      "Number of negative or zero-length updated intervals to be dropped - 158\n",
      "Number of negative or zero-length updated intervals to be dropped - 107\n",
      "Number of negative or zero-length updated intervals to be dropped - 229\n",
      "Number of negative or zero-length updated intervals to be dropped - 199\n",
      "Number of negative or zero-length updated intervals to be dropped - 189\n",
      "Number of negative or zero-length updated intervals to be dropped - 166\n",
      "Number of negative or zero-length updated intervals to be dropped - 100\n",
      "Number of negative or zero-length updated intervals to be dropped - 62\n",
      "Number of negative or zero-length updated intervals to be dropped - 161\n",
      "Number of negative or zero-length updated intervals to be dropped - 164\n",
      "Number of negative or zero-length updated intervals to be dropped - 211\n",
      "Number of negative or zero-length updated intervals to be dropped - 41\n",
      "Number of negative or zero-length updated intervals to be dropped - 39\n",
      "Number of negative or zero-length updated intervals to be dropped - 52\n",
      "Number of negative or zero-length updated intervals to be dropped - 63\n",
      "Number of negative or zero-length updated intervals to be dropped - 41\n",
      "Number of negative or zero-length updated intervals to be dropped - 98\n",
      "Number of negative or zero-length updated intervals to be dropped - 148\n",
      "Number of negative or zero-length updated intervals to be dropped - 5\n",
      "Number of negative or zero-length updated intervals to be dropped - 23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1616507</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1616507:1617898:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617898</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1617077</td>\n",
       "      <td>1617321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1617077:1617323:+:MIB2</td>\n",
       "      <td>100</td>\n",
       "      <td>1617320</td>\n",
       "      <td>1617323</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1617320:1617321:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1630291</td>\n",
       "      <td>1630605</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1630291:1630604:+:MIB2</td>\n",
       "      <td>109</td>\n",
       "      <td>1630604</td>\n",
       "      <td>1630604</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>chr1:1630604:1630605:+:MIB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6632795</td>\n",
       "      <td>6633562</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6632795:6633562:+:THAP3</td>\n",
       "      <td>248</td>\n",
       "      <td>6633561</td>\n",
       "      <td>6633562</td>\n",
       "      <td>THAP3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>THAP3</td>\n",
       "      <td>chr1:6633561:6633562:+:THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6634019</td>\n",
       "      <td>6635586</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6634019:6635586:+:THAP3</td>\n",
       "      <td>249</td>\n",
       "      <td>6635585</td>\n",
       "      <td>6635586</td>\n",
       "      <td>THAP3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>THAP3</td>\n",
       "      <td>chr1:6635585:6635586:+:THAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2643</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19593087</td>\n",
       "      <td>19594158</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19593087:19594161:+:TXLNGY</td>\n",
       "      <td>118984</td>\n",
       "      <td>19594157</td>\n",
       "      <td>19594161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>chrY:19594157:19594158:+:TXLNGY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2644</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234577</td>\n",
       "      <td>13234826</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234576:13234826:-:UTY</td>\n",
       "      <td>119277</td>\n",
       "      <td>13234576</td>\n",
       "      <td>13234578</td>\n",
       "      <td>UTY</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>chrY:13234577:13234578:-:UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2645</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13248385</td>\n",
       "      <td>13249882</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13248378:13249882:-:UTY</td>\n",
       "      <td>119278</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13248386</td>\n",
       "      <td>UTY</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>chrY:13248385:13248386:-:UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2646</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18872593</td>\n",
       "      <td>18872834</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18872500:18872834:-:TTTY14</td>\n",
       "      <td>119337</td>\n",
       "      <td>18872500</td>\n",
       "      <td>18872594</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18872593:18872594:-:TTTY14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2647</th>\n",
       "      <td>chrY</td>\n",
       "      <td>18932449</td>\n",
       "      <td>18932841</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:18930470:18932841:-:TTTY14</td>\n",
       "      <td>119340</td>\n",
       "      <td>18930470</td>\n",
       "      <td>18932450</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTY14</td>\n",
       "      <td>chrY:18932449:18932450:-:TTTY14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2648 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 1616507   | 1617321   | +            | ...   |\n",
       "| chr1         | 1617077   | 1617321   | +            | ...   |\n",
       "| chr1         | 1630291   | 1630605   | +            | ...   |\n",
       "| chr1         | 6632795   | 6633562   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 13234577  | 13234826  | -            | ...   |\n",
       "| chrY         | 13248385  | 13249882  | -            | ...   |\n",
       "| chrY         | 18872593  | 18872834  | -            | ...   |\n",
       "| chrY         | 18932449  | 18932841  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 2,648 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: exon_id, Cluster, Start_b, End_b, Name, Score, Strand_b, ... (+ 2 more.)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update 3' exon coordinates to the PAS coordinate\n",
    "pas_le_ale_bf_upd = pas_le_ale_bf.apply(lambda df: _df_update_3p(df))\n",
    "pas_le_ale_bf_upd = pas_le_ale_bf_upd.drop_duplicate_positions(strand=True)\n",
    "pas_le_ale_bf_upd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ABCC5',\n",
       " 'CCNL2',\n",
       " 'DIDO1',\n",
       " 'EIF4E2',\n",
       " 'ELAVL4',\n",
       " 'GREB1',\n",
       " 'SFPQ',\n",
       " 'SLC19A1',\n",
       " 'SSU72',\n",
       " 'SYP',\n",
       " 'TFDP1'}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# track target genes classified as bonafide ALEs\n",
    "# target genes fitting tandem criteria\n",
    "zeng_targets_ale_bf = zeng_target_genes.intersection(ale_bf_gene_names)\n",
    "print(len(zeng_targets_ale_bf))\n",
    "zeng_targets_ale_bf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ARHGAP32',\n",
       " 'CHRNB4',\n",
       " 'CNPY3',\n",
       " 'EGFR',\n",
       " 'ELP6',\n",
       " 'GSTO2',\n",
       " 'NSMAF',\n",
       " 'PAK4',\n",
       " 'PARD6G',\n",
       " 'RFNG',\n",
       " 'SIX3',\n",
       " 'SLIT3',\n",
       " 'SMG7',\n",
       " 'STMN2',\n",
       " 'TAFA5',\n",
       " 'TLX1',\n",
       " 'TTBK1'}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what are the unnacounted for target genes?\n",
    "zeng_int_unaccounted = zeng_target_genes.difference(zeng_targets_tandem.union(zeng_targets_nonlast).union(zeng_targets_ale_bf))\n",
    "print(len(zeng_int_unaccounted))\n",
    "zeng_int_unaccounted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ARHGAP32, CNPY3, RFNG, SIX3, STMN2, TLX1 are known events I've captured with my pipeline, so I'll add them in via those models. For the remaining, need to dig deeper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CHRNB4',\n",
       " 'EGFR',\n",
       " 'ELP6',\n",
       " 'GSTO2',\n",
       " 'NSMAF',\n",
       " 'PAK4',\n",
       " 'PARD6G',\n",
       " 'SLIT3',\n",
       " 'SMG7',\n",
       " 'TAFA5',\n",
       " 'TTBK1'}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papa_hits = \"ARHGAP32, CNPY3, RFNG, SIX3, STMN2, TLX1\".split(\", \")\n",
    "zeng_int_unaccounted = zeng_int_unaccounted.difference(papa_hits)\n",
    "zeng_int_unaccounted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rescuing missing targets - novel extensions of annotated last exons? Or just novel events\n",
    "\n",
    "Find nearest upstream ALEs for the missing targets. With nearest approach, overlapping PAS will be assigned 0, so can assess whether both PAS are accounted for.\n",
    "\n",
    "Realistically assessing two points:\n",
    "1. Are PAS just downstream of annotated last exons? In which case, can rescue these events by attaching to and extending annotated event\n",
    "2. Do they represent unannotated last exons? In which case, quite tricky to use with bulk RNA-seq. If deep intronic, could just extend a specified distance upstream of the PAS to define a counting window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1011462</td>\n",
       "      <td>1011463</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1011462:1011463:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1728507</td>\n",
       "      <td>1728508</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227775</td>\n",
       "      <td>chr1:1728507:1728508:+:ENSG00000227775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1735571</td>\n",
       "      <td>1735572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227775</td>\n",
       "      <td>chr1:1735571:1735572:+:ENSG00000227775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10275251</td>\n",
       "      <td>10275252</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>chr1:10275251:10275252:+:KIF1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3940</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13711999</td>\n",
       "      <td>13712000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>chrY:13711999:13712000:+:TMSB4Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3941</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13889010</td>\n",
       "      <td>13889011</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>chrY:13889010:13889011:+:ANOS2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3942</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13899472</td>\n",
       "      <td>13899473</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>chrY:13899472:13899473:+:ANOS2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3943</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12326385</td>\n",
       "      <td>12326386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>chrY:12326385:12326386:-:GYG2P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3944</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>chrY:12378537:12378538:-:GYG2P1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3945 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 1011462   | 1011463   | ISG15      | .          | ...   |\n",
       "| chr1         | 1014537   | 1014538   | ISG15      | .          | ...   |\n",
       "| chr1         | 1728507   | 1728508   | nan        | .          | ...   |\n",
       "| chr1         | 1735571   | 1735572   | nan        | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 13889010  | 13889011  | ANOS2P     | .          | ...   |\n",
       "| chrY         | 13899472  | 13899473  | ANOS2P     | .          | ...   |\n",
       "| chrY         | 12326385  | 12326386  | nan        | .          | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 3,945 rows and 8 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Strand, gene_name_common, pas_id"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to unnacounted PAS\n",
    "missing_pas_ids = pas_ids.difference(set(pas_le_3utr_upd.pas_id)).difference(set(pas_le_ale_bf_upd.pas_id))\n",
    "len(missing_pas_ids)\n",
    "zeng_bed_upd_missing = zeng_bed_upd.subset(lambda df: df.pas_id.isin(missing_pas_ids))\n",
    "zeng_bed_upd_missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1014538</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1:1014537:1014538:+:ISG15</td>\n",
       "      <td>exon</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014540</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187608.10</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>ENST00000624697.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:1013983:1014540:+:ISG15</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10381599</td>\n",
       "      <td>10381600</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>chr1:10381599:10381600:+:KIF1B</td>\n",
       "      <td>exon</td>\n",
       "      <td>10376544</td>\n",
       "      <td>10381603</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000054523.20</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>ENST00000377086.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:10376544:10381603:+:KIF1B</td>\n",
       "      <td>334</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10630756</td>\n",
       "      <td>10630757</td>\n",
       "      <td>PEX14</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>PEX14</td>\n",
       "      <td>chr1:10630756:10630757:+:PEX14</td>\n",
       "      <td>exon</td>\n",
       "      <td>10629530</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142655.13</td>\n",
       "      <td>PEX14</td>\n",
       "      <td>ENST00000356607.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:10629530:10630758:+:PEX14</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940455</td>\n",
       "      <td>15940456</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>chr1:15940455:15940456:+:SPEN</td>\n",
       "      <td>exon</td>\n",
       "      <td>15939295</td>\n",
       "      <td>15940456</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>ENST00000375759.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:15939295:15940456:+:SPEN</td>\n",
       "      <td>517</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21484899</td>\n",
       "      <td>21484900</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>chr1:21484899:21484900:+:NBPF3</td>\n",
       "      <td>exon</td>\n",
       "      <td>21483142</td>\n",
       "      <td>21485005</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENST00000318220.10</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>chr1:21483142:21485005:+:NBPF3</td>\n",
       "      <td>653</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3940</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13711999</td>\n",
       "      <td>13712000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>chrY:13711999:13712000:+:TMSB4Y</td>\n",
       "      <td>exon</td>\n",
       "      <td>13705224</td>\n",
       "      <td>13706024</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000154620.6</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>ENST00000284856.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:13705224:13706024:+:TMSB4Y</td>\n",
       "      <td>118922</td>\n",
       "      <td>5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3941</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13889010</td>\n",
       "      <td>13889011</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>chrY:13889010:13889011:+:ANOS2P</td>\n",
       "      <td>exon</td>\n",
       "      <td>13887257</td>\n",
       "      <td>13888980</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000241859.7</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>ENST00000652544.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chrY:13887257:13888980:+:ANOS2P</td>\n",
       "      <td>118925</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3942</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13899472</td>\n",
       "      <td>13899473</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>chrY:13899472:13899473:+:ANOS2P</td>\n",
       "      <td>exon</td>\n",
       "      <td>13887257</td>\n",
       "      <td>13888980</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000241859.7</td>\n",
       "      <td>ANOS2P</td>\n",
       "      <td>ENST00000652544.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chrY:13887257:13888980:+:ANOS2P</td>\n",
       "      <td>118925</td>\n",
       "      <td>10493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3943</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12378537</td>\n",
       "      <td>12378538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>chrY:12378537:12378538:-:GYG2P1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12378531</td>\n",
       "      <td>12378749</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>ENST00000382965.3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chrY:12378531:12378749:-:GYG2P1</td>\n",
       "      <td>119266</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3944</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12326385</td>\n",
       "      <td>12326386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>chrY:12326385:12326386:-:GYG2P1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12354415</td>\n",
       "      <td>12354771</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000206159.12</td>\n",
       "      <td>GYG2P1</td>\n",
       "      <td>ENST00000382966.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chrY:12354415:12354771:-:GYG2P1</td>\n",
       "      <td>119265</td>\n",
       "      <td>28030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3945 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Name       | Score      | +14   |\n",
       "| (category)   | (int64)   | (int64)   | (object)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+------------+-------|\n",
       "| chr1         | 1014537   | 1014538   | ISG15      | .          | ...   |\n",
       "| chr1         | 10381599  | 10381600  | KIF1B      | .          | ...   |\n",
       "| chr1         | 10630756  | 10630757  | PEX14      | .          | ...   |\n",
       "| chr1         | 15940455  | 15940456  | SPEN       | .          | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...   |\n",
       "| chrY         | 13889010  | 13889011  | ANOS2P     | .          | ...   |\n",
       "| chrY         | 13899472  | 13899473  | ANOS2P     | .          | ...   |\n",
       "| chrY         | 12378537  | 12378538  | nan        | .          | ...   |\n",
       "| chrY         | 12326385  | 12326386  | nan        | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+------------+-------+\n",
       "Stranded PyRanges object has 3,945 rows and 19 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "14 hidden columns: Strand, gene_name_common, pas_id, Feature, Start_b, End_b, ... (+ 8 more.)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate distance to nearest (upstream) last exon (allowing overlaps)\n",
    "pas_le_missing_nr = zeng_bed_upd_missing.nearest(last_exons, strandedness=\"same\", overlap=True,how=\"upstream\",)\n",
    "pas_le_missing_nr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "min    max  \n",
       "True   False    1592\n",
       "False  False     356\n",
       "True   True       46\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each gene, report the min and max distance (i.e. how many genes are just missing annotated last exon for 1 PAS?)\n",
    "dist_min_max = pas_le_missing_nr.as_df().groupby(\"gene_name_common\")[\"Distance\"].agg([min, max]).reset_index().sort_values(by=[\"min\", \"max\"])\n",
    "\n",
    "# gene counts for \n",
    "# both overlap last exons (True, True),\n",
    "# 1 overlaps last exon (True, False) \n",
    "# and neither overlaps (False, False)\n",
    "dist_min_max.drop(columns=\"gene_name_common\").eq(0,axis=\"columns\").value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>SLIT3</td>\n",
       "      <td>0</td>\n",
       "      <td>743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>ELP6</td>\n",
       "      <td>0</td>\n",
       "      <td>989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>SMG7</td>\n",
       "      <td>0</td>\n",
       "      <td>1902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1753</th>\n",
       "      <td>TTBK1</td>\n",
       "      <td>0</td>\n",
       "      <td>4036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>GSTO2</td>\n",
       "      <td>0</td>\n",
       "      <td>4802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1175</th>\n",
       "      <td>PARD6G</td>\n",
       "      <td>0</td>\n",
       "      <td>7662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>CHRNB4</td>\n",
       "      <td>0</td>\n",
       "      <td>28792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1168</th>\n",
       "      <td>PAK4</td>\n",
       "      <td>0</td>\n",
       "      <td>38634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>0</td>\n",
       "      <td>83003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>TAFA5</td>\n",
       "      <td>0</td>\n",
       "      <td>144633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132</th>\n",
       "      <td>NSMAF</td>\n",
       "      <td>0</td>\n",
       "      <td>169042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name_common  min     max\n",
       "1528            SLIT3    0     743\n",
       "487              ELP6    0     989\n",
       "1534             SMG7    0    1902\n",
       "1753            TTBK1    0    4036\n",
       "683             GSTO2    0    4802\n",
       "1175           PARD6G    0    7662\n",
       "307            CHRNB4    0   28792\n",
       "1168             PAK4    0   38634\n",
       "475              EGFR    0   83003\n",
       "1632            TAFA5    0  144633\n",
       "1132            NSMAF    0  169042"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dist_min_max[dist_min_max.gene_name_common.isin(zeng_int_unaccounted)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_common</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>overlap_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>APEX1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>ATXN2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>B4GALNT1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>CPT1B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>CTSL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>XKR4</td>\n",
       "      <td>202640</td>\n",
       "      <td>233429</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>TEX10</td>\n",
       "      <td>239536</td>\n",
       "      <td>241106</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1872</th>\n",
       "      <td>ZCCHC17</td>\n",
       "      <td>259686</td>\n",
       "      <td>274242</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>NR4A2</td>\n",
       "      <td>436250</td>\n",
       "      <td>441767</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>NDST4</td>\n",
       "      <td>692094</td>\n",
       "      <td>746823</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1994 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name_common     min     max  overlap_group\n",
       "80              APEX1       0       0              2\n",
       "135             ATXN2       0       0              2\n",
       "143          B4GALNT1       0       0              2\n",
       "358             CPT1B       0       0              2\n",
       "377              CTSL       0       0              2\n",
       "...               ...     ...     ...            ...\n",
       "1853             XKR4  202640  233429              0\n",
       "1658            TEX10  239536  241106              0\n",
       "1872          ZCCHC17  259686  274242              0\n",
       "1121            NR4A2  436250  441767              0\n",
       "1084            NDST4  692094  746823              0\n",
       "\n",
       "[1994 rows x 4 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign to overlap group based on both overlap last exons (2), 1 overlaps last exon (1) or neither (0) \n",
    "dist_min_max.loc[:, \"overlap_group\"] = dist_min_max.drop(columns=\"gene_name_common\").eq(0,axis=\"columns\").sum(axis=\"columns\")\n",
    "dist_min_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes - 1638\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'0': 46,\n",
       " '25': 55,\n",
       " '50': 55,\n",
       " '100': 58,\n",
       " '200': 92,\n",
       " '500': 162,\n",
       " '1000': 251,\n",
       " '2500': 487,\n",
       " '5000': 705}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# try different max distance thresholds (how many genes rescued?)\n",
    "print(f\"Number of genes - {dist_min_max.loc[dist_min_max.overlap_group.ne(0), 'gene_name_common'].nunique()}\")\n",
    "{str(cutoff): len(dist_min_max[(dist_min_max[\"overlap_group\"].ne(0)) & (dist_min_max[\"max\"] <= cutoff)]) for cutoff in [0,25,50,100,200,500,1000, 2500, 5000]}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
