### Generating decoy index

Commands used:

`python scripts/add_decoys_to_gtf.py -i data/PAPA/2023-06-20_cryptic_last_exons.gtf -r data/reference_filtered.gtf -d full -o processed/2023-06-21_cryptics_plus_decoys.decoys_full`

`python scripts/add_decoys_to_gtf.py -i processed/2023-06-20_cryptic_last_exons.gtf -r data/reference_filtered.gtf -d full -o processed/2023-06-22_cryptics_plus_decoys.decoys_full_fix_tx2le`

add_decoys_to_gtf requirements:

- `event_type_simple` attribute, which is generated using `preproccesing/scripts/clean_papa_tbls.R`:

  - spliced - last exon with novel SJ
  - bleedthrough - last exon that extends an annotated internal exon (or annotated last exon that bleeds through )
  - distal_3utr_extension - xx

- input last exons contain a 'ref_gene_id` attribute, which corresponds to the reference gene_id to which the last exon overlaps/belongs. This is satisfied if you use 'novel_ref_combined.last_exons.gtf' generated by PAPA pipeline.


### PAPA to splice junctions


Commands used:

`python scripts/last_exons_to_sj.py data/2023-09-06_papa_cryptic_spliced.last_exons.cryptic_only.blacklist_filtered.bed data/reference_filtered.gtf processed/2023-09-12_papa_as_ale_cryptic &> processed/2023-09-12_papa_as_ale_cryptic.last_exons_to_sj.log.txt`

1 missing event - `ENSG00000002746.15_3|HECW1|spliced|cryptic` - which is better described as a novel bleedthrough event that an AS-ALE, but has been categorised as annotated by this scriptin this case. Why classified as a spliced event?
- 5'end is shifted 1 nucleotide upstream relative to annotated internal exon
- No annotated intron/SJ that contains this exon
- Probably initially classified as spliced because of novel exon 5'end (+ 3'end being contained within intron), but 5'end of last intron matches a known intron