{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How many cryptic PAS are supported by polyA tail containing reads?\n",
    "\n",
    "Strategy for analysis:\n",
    "- Predicted PAS from GTF file\n",
    "- Subset quant GTF to cryptics\n",
    "- pr.k_nearest on each interval (pick relaxed, allow overlaps + either direction)\n",
    "\n",
    "Q1: What %age of events are supported by reads at given distance threshold?\n",
    "- e.g. 0, 10, 25, 50, 100\n",
    "- Each le_id can contain multiple 3'ends per event - first collapse to min distance as representative for each le_id\n",
    "- Then apply range of cut-offs - number of events in each window\n",
    "\n",
    "\n",
    "Q2: Are distinct isoforms supported for each event?\n",
    "- e.g. STMN2s and ARHGAP32s of the cryptic world\n",
    "- perform naive clustering/collapsing of pas by le_id e.g. 25/50nt \n",
    "  - How many PAS have multiple unique predicted isoforms?\n",
    "  - How many have junction read support?\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "TO CONSIDER:\n",
    "\n",
    "3-5 overhang could be missing some lower confidence/shorter supporting reads. Repeat counting of lower confidence PAS by running count_pas with BED file containing cryptic PAS from PAPA (all PAS for each le_id) AND/OR (e.g. separate runs) PAS clusters defined using the 3-5/6+ criteria (i.e. for the same clusters, is there additional support from short overhangs?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from typing import Union, List, Optional, Tuple, Dict\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nearest_threshold_wrapper(gr: pr.PyRanges,\n",
    "                              gr2: pr.PyRanges,\n",
    "                              thresholds: Union[int, List[int]],\n",
    "                              id_col: str = \"le_id\",\n",
    "                              nearest_only: bool = True,\n",
    "                              return_grs: bool = False,\n",
    "                              nearest_kwargs: dict = {\"k\": 10,\n",
    "                                                      \"strandedness\": \"same\",\n",
    "                                                      \"overlap\": True,\n",
    "                                                      \"how\": None},\n",
    "                              ) -> Tuple[Dict[str, set], Optional[Dict[str, pr.PyRanges]]]:\n",
    "    '''Get events with nearest overlap passing a maximum distance threshold(s)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    gr : pr.PyRanges\n",
    "        _description_\n",
    "    gr2 : pr.PyRanges\n",
    "        _description_\n",
    "    thresholds : Union[int, List[int]]\n",
    "        _description_\n",
    "    id_col : str, optional\n",
    "        _description_, by default \"le_id\"\n",
    "    nearest_only : bool, optional\n",
    "        _description_, by default True\n",
    "    return_grs : bool, optional\n",
    "        _description_, by default False\n",
    "    nearest_kwargs : _type_, optional\n",
    "        _description_, by default {\"k\": 10, \"strandedness\": \"same\", \"overlap\": True, \"how\": None}\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Tuple[Dict[str, set], Optional[Dict[str, pr.PyRanges]]]\n",
    "        return {threshold: set(id_col values)} and optionally {threshold: pr.PyRanges} (if return_grs=True)\n",
    "    '''\n",
    "    \n",
    "\n",
    "    if nearest_only:\n",
    "        if \"k\" in nearest_kwargs.keys():\n",
    "            kwargs = {key: v for key, v in nearest_kwargs.items()if key != \"k\"}\n",
    "        gr_nr = gr.nearest(gr2, **kwargs)\n",
    "\n",
    "    else:\n",
    "        assert \"k\" in nearest_kwargs.keys()\n",
    "        gr_nr = gr.k_nearest(gr2, **nearest_kwargs)\n",
    "\n",
    "\n",
    "    # convert to absolute values (don't care if up/downstream rn)\n",
    "    gr_nr = gr_nr.assign(\"DistanceAbs\", lambda df: df.Distance.abs())\n",
    "    \n",
    "    # if nearest_only:\n",
    "    #     # # for each le_id, select the smallest distance\n",
    "    #     gr_nr = gr_nr.apply(lambda df: df.sort_values(by=[id_col, \"DistanceAbs\"]).drop_duplicates(subset=[id_col])).sort()\n",
    "\n",
    "    if isinstance(thresholds, int):\n",
    "        # subset to those pasing distance threshold\n",
    "        pass_grs = {str(thresholds): gr_nr.subset(lambda df: df.DistanceAbs.le(thresholds))}\n",
    "\n",
    "    else:\n",
    "        pass_grs = {str(threshold): gr_nr.subset(lambda df: df.DistanceAbs.le(threshold)) for threshold in thresholds}\n",
    "\n",
    "    # return set of ids at each threshold\n",
    "    pass_ids = {threshold: set(gr.as_df()[id_col]) for threshold, gr in pass_grs.items()}\n",
    "\n",
    "    if return_grs:\n",
    "        return pass_ids, pass_grs\n",
    "    else:\n",
    "        return pass_ids\n",
    "\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BED files of polyA-tail containing reads - 3-5 nt overhang + 100 % As | 6+nt & >80 % As\n",
    "# Pooling across all KD samples from all experiments\n",
    "pas_all = pr.read_bed(\"../data/bulk_polya_reads/tdp_ko_collection/pas_clusters/condition__TDP43KD/two_class_simple/polya_clusters.bed\")\n",
    "# pooling x KD samples from i3Neuron datasets\n",
    "pas_i3 = pr.read_bed(\"../data/bulk_polya_reads/tdp_ko_collection/pas_clusters/cell_type__i3_cortical___condition__TDP43KD/two_class_simple/polya_clusters.bed\")\n",
    "# pooling x KD samples from SH-SY5Y datasets\n",
    "pas_shsy5y = pr.read_bed(\"../data/bulk_polya_reads/tdp_ko_collection/pas_clusters/cell_type__shsy5y___condition__TDP43KD/two_class_simple/polya_clusters.bed\")\n",
    "# poolin x KD samples from SK-N-BE2 datasets\n",
    "pas_sknbe2 = pr.read_bed(\"../data/bulk_polya_reads/tdp_ko_collection/pas_clusters/cell_type__sknbe2___condition__TDP43KD/two_class_simple/polya_clusters.bed\")\n",
    "\n",
    "# GTF file of predicted last exons\n",
    "quant_le = pr.read_gtf(\"../data/novel_ref_combined.last_exons.gtf\")\n",
    "\n",
    "# Table with cryptic le_ids (prior to mv filtering)\n",
    "cryptics_df = pd.read_csv(\"../processed/2023-12-10_cryptics_summary_all_events.tsv\", sep=\"\\t\")\n",
    "# after manual validation (just to get some idea)\n",
    "cryptics_mv_df = pd.read_csv(\"../processed/2023-12-10_cryptics_summary_all_events_bleedthrough_manual_validation.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>region_rank</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>3p_extension_length</th>\n",
       "      <th>event_type</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>le_number</th>\n",
       "      <th>le_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>15940476</td>\n",
       "      <td>15940477</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>147623355</td>\n",
       "      <td>147623356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_2.1123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>147622692</td>\n",
       "      <td>147622693</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.chx_tdp_DOX_ctrl_2.1186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>147623700</td>\n",
       "      <td>147623701</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.doxconc_DOX_0075_1.1139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>21457149</td>\n",
       "      <td>21457150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP-1.345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>40653676</td>\n",
       "      <td>40653677</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_3.26410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>last_exon_extension</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "      <td>MED14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131823776</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43-F_S6.20505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131823778</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.TDP43_ctrl_2.26896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>internal_exon_extension</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108267642</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108221374,108221374,108221374</td>\n",
       "      <td>108310747,108310747,108310747</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>chrX</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108267642</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>...</td>\n",
       "      <td>last</td>\n",
       "      <td>108221374,108221374,108221374</td>\n",
       "      <td>108310747,108310747,108310747</td>\n",
       "      <td>ENST00000372216.8,ENST00000538570.5,ENST000006...</td>\n",
       "      <td>NULL,NULL,NULL</td>\n",
       "      <td>internal_exon_spliced</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>975 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +17   |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | .          | exon         | 15940476  | 15940477  | ...   |\n",
       "| chr1         | .          | exon         | 147623355 | 147623356 | ...   |\n",
       "| chr1         | .          | exon         | 147622692 | 147622693 | ...   |\n",
       "| chr1         | .          | exon         | 147623700 | 147623701 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrX         | .          | exon         | 131823775 | 131823776 | ...   |\n",
       "| chrX         | .          | exon         | 131823777 | 131823778 | ...   |\n",
       "| chrX         | .          | exon         | 108267641 | 108267642 | ...   |\n",
       "| chrX         | .          | exon         | 108267641 | 108267642 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 975 rows and 22 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Score, Strand, Frame, gene_id, gene_name, transcript_id, exon_number, ... (+ 10 more.)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cryptic IDs\n",
    "cryp_le_ids = set(cryptics_df.le_id)\n",
    "cryp_le_ids_mv = set(cryptics_mv_df.le_id)\n",
    "quant_le_cryp = quant_le.subset(lambda df: df.le_id.isin(cryp_le_ids))\n",
    "quant_le_cryp_mv = quant_le.subset(lambda df: df.le_id.isin(cryp_le_ids_mv))\n",
    "quant_le_cryp_pas = quant_le_cryp.three_end()\n",
    "quant_le_cryp_mv_pas = quant_le_cryp_mv.three_end()\n",
    "quant_le_cryp_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>le_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "      <th>ref_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15940476</td>\n",
       "      <td>15940477</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000065526.12_4</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>ENSG00000065526.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147623355</td>\n",
       "      <td>147623356</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147622692</td>\n",
       "      <td>147622693</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147623700</td>\n",
       "      <td>147623701</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116128.12_2</td>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENSG00000116128.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21457149</td>\n",
       "      <td>21457150</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142794.19_3</td>\n",
       "      <td>NBPF3</td>\n",
       "      <td>ENSG00000142794.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40653676</td>\n",
       "      <td>40653677</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000180182.11_3</td>\n",
       "      <td>MED14</td>\n",
       "      <td>ENSG00000180182.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823775</td>\n",
       "      <td>131823776</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131823777</td>\n",
       "      <td>131823778</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000213468.7_1</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>ENSG00000213468.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108267642</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108267641</td>\n",
       "      <td>108267642</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000197565.17_1</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>ENSG00000197565.17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>975 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 15940476  | 15940477  | +            | ...   |\n",
       "| chr1         | 147623355 | 147623356 | +            | ...   |\n",
       "| chr1         | 147622692 | 147622693 | +            | ...   |\n",
       "| chr1         | 147623700 | 147623701 | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 131823775 | 131823776 | -            | ...   |\n",
       "| chrX         | 131823777 | 131823778 | -            | ...   |\n",
       "| chrX         | 108267641 | 108267642 | -            | ...   |\n",
       "| chrX         | 108267641 | 108267642 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 975 rows and 7 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: le_id, ref_gene_name, ref_gene_id"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove junky columns\n",
    "quant_le_cryp_pas = quant_le_cryp_pas[[\"le_id\", \"ref_gene_name\", \"ref_gene_id\", \"event_id\"]]\n",
    "quant_le_cryp_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'0': 54, '10': 112, '25': 122, '50': 131, '100': 138, '200': 151, '500': 176}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'0': 0.20930232558139536,\n",
       " '10': 0.43410852713178294,\n",
       " '25': 0.4728682170542636,\n",
       " '50': 0.5077519379844961,\n",
       " '100': 0.5348837209302325,\n",
       " '200': 0.5852713178294574,\n",
       " '500': 0.6821705426356589}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overlap with jnc reads from all PAS\n",
    "cryp_pas_kn = quant_le_cryp_pas.k_nearest(pas_all, k=10, strandedness=\"same\", overlap=True, how=None)\n",
    "# convert to absolute values (don't care if up/downstream rn)\n",
    "cryp_pas_kn = cryp_pas_kn.assign(\"DistanceAbs\", lambda df: df.Distance.abs())\n",
    "\n",
    "# for each le_id, select the smallest distance\n",
    "cryp_pas_kn_nr = cryp_pas_kn.apply(lambda df: df.sort_values(by=[\"le_id\", \"DistanceAbs\"]).drop_duplicates(subset=[\"le_id\"])).sort()\n",
    "\n",
    "# how \n",
    "n_cryp_supported = {str(d): cryp_pas_kn_nr.subset(lambda df: df.DistanceAbs.le(d)).le_id.nunique() for d in [0,10,25,50,100,200,500]}\n",
    "print(n_cryp_supported)\n",
    "{d: n / len(cryp_le_ids) for d,n in n_cryp_supported.items()}\n",
    "# cryp_pas_kn_nr.as_df()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'0': 54, '10': 112, '25': 122, '50': 131, '100': 138, '200': 151, '500': 176}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'0': 0.23788546255506607,\n",
       " '10': 0.4933920704845815,\n",
       " '25': 0.5374449339207048,\n",
       " '50': 0.5770925110132159,\n",
       " '100': 0.6079295154185022,\n",
       " '200': 0.6651982378854625,\n",
       " '500': 0.775330396475771}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overlap with jnc reads from all PAS\n",
    "cryp_pas_mv_kn = quant_le_cryp_pas.k_nearest(pas_all, k=10, strandedness=\"same\", overlap=True, how=None)\n",
    "# convert to absolute values (don't care if up/downstream rn)\n",
    "cryp_pas_mv_kn = cryp_pas_mv_kn.assign(\"DistanceAbs\", lambda df: df.Distance.abs())\n",
    "\n",
    "# for each le_id, select the smallest distance\n",
    "cryp_pas_mv_kn_nr = cryp_pas_mv_kn.apply(lambda df: df.sort_values(by=[\"le_id\", \"DistanceAbs\"]).drop_duplicates(subset=[\"le_id\"])).sort()\n",
    "\n",
    "# how \n",
    "n_cryp_supported_mv = {str(d): cryp_pas_mv_kn_nr.subset(lambda df: df.DistanceAbs.le(d)).le_id.nunique() for d in [0,10,25,50,100,200,500]}\n",
    "print(n_cryp_supported_mv)\n",
    "{d: n / len(cryp_le_ids_mv) for d,n in n_cryp_supported_mv.items()}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For manually validated (i.e. bleedthrough IR artefacts removed):\n",
    "~ 49 % +/- 10nt, rising ~ 61 % +/- 100nt\n",
    "\n",
    "vs all initial cryptics:\n",
    "~ 43 % +/- 10nt, rising ~ 53 % +/- 100nt\n",
    "\n",
    "Haven't checked read support..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pybioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
